{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import anndata as ad\n",
    "import torch\n",
    "import pandas as pd\n",
    "import requests\n",
    "import sys\n",
    "import os\n",
    "import scanpy as sc\n",
    "#import pertpy as pt      # comment-out when using nen_env\n",
    "import re\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import ast\n",
    "import pickle\n",
    "from scipy.stats import wasserstein_distance, permutation_test\n",
    "from scipy.spatial.distance import cdist, pdist, squareform\n",
    "\n",
    "from dotenv import find_dotenv, load_dotenv\n",
    "\n",
    "load_dotenv(find_dotenv())\n",
    "\n",
    "sys.path.append('../src/null-effect-net')\n",
    "\n",
    "import utils            # comment-out when using pertpy_env"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load perturbation screens"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "perturbation_screen_name = 'GasperiniShendure2019_atscale'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 207324 × 13135\n",
       "    obs: 'sample', 'total_umis', 'Size_Factor', 'gene', 'all_gene', 'barcode', 'read_count', 'umi_count', 'proportion', 'guide_count', 'sample_directory', 'ko_barcode_file', 'id', 'prep_batch', 'within_batch_chip', 'within_chip_lane', 'percent.mito', 'tissue_type', 'cell_line', 'cancer', 'disease', 'perturbation_type', 'celltype', 'organism', 'perturbation', 'nperts', 'ngenes', 'ncounts', 'percent_mito', 'percent_ribo'\n",
       "    var: 'ensembl_id', 'ncounts', 'ncells'"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata = sc.read_h5ad(f'../data/perturbation_screens/{perturbation_screen_name}.h5ad')\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 207324 × 13135\n",
       "    obs: 'sample', 'total_umis', 'Size_Factor', 'gene', 'all_gene', 'barcode', 'read_count', 'umi_count', 'proportion', 'guide_count', 'sample_directory', 'ko_barcode_file', 'id', 'prep_batch', 'within_batch_chip', 'within_chip_lane', 'percent.mito', 'tissue_type', 'cell_line', 'cancer', 'disease', 'perturbation_type', 'celltype', 'organism', 'perturbation', 'nperts', 'ngenes', 'ncounts', 'percent_mito', 'percent_ribo', 'n_genes'\n",
       "    var: 'ensembl_id', 'ncounts', 'ncells', 'n_cells'"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sc.pp.filter_cells(adata, min_genes=200)\n",
    "sc.pp.filter_genes(adata, min_cells=3)\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>total_umis</th>\n",
       "      <th>Size_Factor</th>\n",
       "      <th>gene</th>\n",
       "      <th>all_gene</th>\n",
       "      <th>barcode</th>\n",
       "      <th>read_count</th>\n",
       "      <th>umi_count</th>\n",
       "      <th>proportion</th>\n",
       "      <th>guide_count</th>\n",
       "      <th>...</th>\n",
       "      <th>perturbation_type</th>\n",
       "      <th>celltype</th>\n",
       "      <th>organism</th>\n",
       "      <th>perturbation</th>\n",
       "      <th>nperts</th>\n",
       "      <th>ngenes</th>\n",
       "      <th>ncounts</th>\n",
       "      <th>percent_mito</th>\n",
       "      <th>percent_ribo</th>\n",
       "      <th>n_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAGGTACC</th>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>17572</td>\n",
       "      <td>1.009682</td>\n",
       "      <td>chr10.845_top_two_chr1.11183_top_two_chr1.1129...</td>\n",
       "      <td>chr10.845_top_two_chr1.11183_top_two_chr1.1129...</td>\n",
       "      <td>AGAAAGCTCCTCCAGTTCAC_TGATCGCTTTGACTGTGACA_ACAA...</td>\n",
       "      <td>14135.0</td>\n",
       "      <td>964.0</td>\n",
       "      <td>0.969819</td>\n",
       "      <td>67.0</td>\n",
       "      <td>...</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>lymphoblasts</td>\n",
       "      <td>human</td>\n",
       "      <td>chr10.845_top_two_chr1.11183_top_two_chr1.1129...</td>\n",
       "      <td>195</td>\n",
       "      <td>3549</td>\n",
       "      <td>17566.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>30.507799</td>\n",
       "      <td>3549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGTCAATAG</th>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>8923</td>\n",
       "      <td>0.939677</td>\n",
       "      <td>chr1.12695_top_two_chr11.3294_top_two_chr1.679...</td>\n",
       "      <td>chr1.12695_top_two_chr11.3294_top_two_chr1.679...</td>\n",
       "      <td>GTAGAGCCTCCAGAACTGTG_AGGTTTATCCAGATGAACTG_CATC...</td>\n",
       "      <td>4329.0</td>\n",
       "      <td>293.0</td>\n",
       "      <td>0.844380</td>\n",
       "      <td>26.0</td>\n",
       "      <td>...</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>lymphoblasts</td>\n",
       "      <td>human</td>\n",
       "      <td>chr1.12695_top_two_chr11.3294_top_two_chr1.679...</td>\n",
       "      <td>68</td>\n",
       "      <td>2543</td>\n",
       "      <td>8917.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>28.955927</td>\n",
       "      <td>2543</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCAAACAACA</th>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>14637</td>\n",
       "      <td>0.990803</td>\n",
       "      <td>ALDH1A2_TSS_BRI3_TSS_chr10.1918_top_two_chr10....</td>\n",
       "      <td>ALDH1A2_TSS_BRI3_TSS_chr10.1918_top_two_chr10....</td>\n",
       "      <td>CCAAGGCGTCCTCAGACCAG_AGCTCCAGGAAGGACCCCCG_TCAC...</td>\n",
       "      <td>12362.0</td>\n",
       "      <td>884.0</td>\n",
       "      <td>0.950538</td>\n",
       "      <td>61.0</td>\n",
       "      <td>...</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>lymphoblasts</td>\n",
       "      <td>human</td>\n",
       "      <td>ALDH1A2_BRI3_chr10.1918_top_two_chr10.350_top_...</td>\n",
       "      <td>167</td>\n",
       "      <td>3191</td>\n",
       "      <td>14626.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>26.548612</td>\n",
       "      <td>3191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCACTTCTGC</th>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>22798</td>\n",
       "      <td>1.036578</td>\n",
       "      <td>C16orf91_TSS_chr1.11332_top_two_chr1.1933_top_...</td>\n",
       "      <td>C16orf91_TSS_chr1.11332_top_two_chr1.1933_top_...</td>\n",
       "      <td>GGCGTCAGTCGAGGAGTCAG_GCCAGCACTTCAGCTCACCG_GCTG...</td>\n",
       "      <td>7459.0</td>\n",
       "      <td>544.0</td>\n",
       "      <td>0.939551</td>\n",
       "      <td>39.0</td>\n",
       "      <td>...</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>lymphoblasts</td>\n",
       "      <td>human</td>\n",
       "      <td>C16orf91_chr1.11332_top_two_chr1.1933_top_two_...</td>\n",
       "      <td>109</td>\n",
       "      <td>4539</td>\n",
       "      <td>22783.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>20.616249</td>\n",
       "      <td>4539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCATGTAGTC</th>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>10136</td>\n",
       "      <td>0.952844</td>\n",
       "      <td>chr10.185_top_two_chr10.484_top_two_chr11.4167...</td>\n",
       "      <td>chr10.185_top_two_chr10.484_top_two_chr11.4167...</td>\n",
       "      <td>ATAAGGCACTCACATCCACC_GCTTGTCCCTAACACTCAGA_GGGC...</td>\n",
       "      <td>14831.0</td>\n",
       "      <td>1054.0</td>\n",
       "      <td>0.959927</td>\n",
       "      <td>37.0</td>\n",
       "      <td>...</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>lymphoblasts</td>\n",
       "      <td>human</td>\n",
       "      <td>chr10.185_top_two_chr10.484_top_two_chr11.4167...</td>\n",
       "      <td>107</td>\n",
       "      <td>2605</td>\n",
       "      <td>10124.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>32.418017</td>\n",
       "      <td>2605</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAGTACCTACA</th>\n",
       "      <td>2B_8_SI-GA-H9</td>\n",
       "      <td>17938</td>\n",
       "      <td>1.011811</td>\n",
       "      <td>BRIX1_TSS_chr10.2059_top_two_chr10.350_top_two...</td>\n",
       "      <td>BRIX1_TSS_chr10.2059_top_two_chr10.350_top_two...</td>\n",
       "      <td>AGGAGGCCAAGAGCGCGGGG_TGATTGAAGGAGGCTCCCCA_GGTA...</td>\n",
       "      <td>4570.0</td>\n",
       "      <td>394.0</td>\n",
       "      <td>0.856522</td>\n",
       "      <td>31.0</td>\n",
       "      <td>...</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>lymphoblasts</td>\n",
       "      <td>human</td>\n",
       "      <td>BRIX1_chr10.2059_top_two_chr10.350_top_two_chr...</td>\n",
       "      <td>85</td>\n",
       "      <td>3451</td>\n",
       "      <td>17934.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>31.733021</td>\n",
       "      <td>3451</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAGTATCACCA-1</th>\n",
       "      <td>2B_8_SI-GA-H9</td>\n",
       "      <td>16543</td>\n",
       "      <td>1.003448</td>\n",
       "      <td>chr10.221_top_two_chr11.3853_top_two_chr11.498...</td>\n",
       "      <td>chr10.221_top_two_chr11.3853_top_two_chr11.498...</td>\n",
       "      <td>CCTGCAACTGTCTATGGCCT_TCAGTGGGTGAGTCTTCAGG_TGAG...</td>\n",
       "      <td>4311.0</td>\n",
       "      <td>387.0</td>\n",
       "      <td>0.861915</td>\n",
       "      <td>33.0</td>\n",
       "      <td>...</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>lymphoblasts</td>\n",
       "      <td>human</td>\n",
       "      <td>chr10.221_top_two_chr11.3853_top_two_chr11.498...</td>\n",
       "      <td>92</td>\n",
       "      <td>3693</td>\n",
       "      <td>16542.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>24.694716</td>\n",
       "      <td>3693</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAGTTCAGACT-1</th>\n",
       "      <td>2B_8_SI-GA-H9</td>\n",
       "      <td>15009</td>\n",
       "      <td>0.993396</td>\n",
       "      <td>chr12.3400_top_two_chr4.1281_second_two_ID3_TSS</td>\n",
       "      <td>chr12.3400_top_two_chr4.1281_second_two_ID3_TSS</td>\n",
       "      <td>CCAGTTGCTAGGCAGGACAG_CTTTCAGGCAGGAATCTGAG_TGGT...</td>\n",
       "      <td>669.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>0.876923</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>lymphoblasts</td>\n",
       "      <td>human</td>\n",
       "      <td>chr12.3400_top_two_chr4.1281_second_two_ID3</td>\n",
       "      <td>7</td>\n",
       "      <td>3162</td>\n",
       "      <td>14992.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>32.123799</td>\n",
       "      <td>3162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAGTTCTGTTT</th>\n",
       "      <td>2B_8_SI-GA-H9</td>\n",
       "      <td>5149</td>\n",
       "      <td>0.882878</td>\n",
       "      <td>chr7.1014_top_two</td>\n",
       "      <td>chr7.1014_top_two</td>\n",
       "      <td>CACCCCCCAGAGATAAGAGA</td>\n",
       "      <td>77.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.086420</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>lymphoblasts</td>\n",
       "      <td>human</td>\n",
       "      <td>chr7.1014_top_two</td>\n",
       "      <td>3</td>\n",
       "      <td>1392</td>\n",
       "      <td>5149.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>35.074772</td>\n",
       "      <td>1392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCAAAGTAG-1</th>\n",
       "      <td>2B_8_SI-GA-H9</td>\n",
       "      <td>9503</td>\n",
       "      <td>0.946182</td>\n",
       "      <td>chr11.2790_top_two_chr12.306_top_two_chr12.474...</td>\n",
       "      <td>chr11.2790_top_two_chr12.306_top_two_chr12.474...</td>\n",
       "      <td>AGTATGGGCCTGGAACACAG_AGATCTCCCCTGTCACCACG_CGTT...</td>\n",
       "      <td>4630.0</td>\n",
       "      <td>410.0</td>\n",
       "      <td>0.813492</td>\n",
       "      <td>35.0</td>\n",
       "      <td>...</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>lymphoblasts</td>\n",
       "      <td>human</td>\n",
       "      <td>chr11.2790_top_two_chr12.306_top_two_chr12.474...</td>\n",
       "      <td>99</td>\n",
       "      <td>2429</td>\n",
       "      <td>9499.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>36.624908</td>\n",
       "      <td>2429</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>207324 rows × 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           sample  total_umis  Size_Factor  \\\n",
       "AAACCTGAGAGGTACC    1A_1_SI-GA-E2       17572     1.009682   \n",
       "AAACCTGAGTCAATAG    1A_1_SI-GA-E2        8923     0.939677   \n",
       "AAACCTGCAAACAACA    1A_1_SI-GA-E2       14637     0.990803   \n",
       "AAACCTGCACTTCTGC    1A_1_SI-GA-E2       22798     1.036578   \n",
       "AAACCTGCATGTAGTC    1A_1_SI-GA-E2       10136     0.952844   \n",
       "...                           ...         ...          ...   \n",
       "TTTGTCAGTACCTACA    2B_8_SI-GA-H9       17938     1.011811   \n",
       "TTTGTCAGTATCACCA-1  2B_8_SI-GA-H9       16543     1.003448   \n",
       "TTTGTCAGTTCAGACT-1  2B_8_SI-GA-H9       15009     0.993396   \n",
       "TTTGTCAGTTCTGTTT    2B_8_SI-GA-H9        5149     0.882878   \n",
       "TTTGTCATCAAAGTAG-1  2B_8_SI-GA-H9        9503     0.946182   \n",
       "\n",
       "                                                                 gene  \\\n",
       "AAACCTGAGAGGTACC    chr10.845_top_two_chr1.11183_top_two_chr1.1129...   \n",
       "AAACCTGAGTCAATAG    chr1.12695_top_two_chr11.3294_top_two_chr1.679...   \n",
       "AAACCTGCAAACAACA    ALDH1A2_TSS_BRI3_TSS_chr10.1918_top_two_chr10....   \n",
       "AAACCTGCACTTCTGC    C16orf91_TSS_chr1.11332_top_two_chr1.1933_top_...   \n",
       "AAACCTGCATGTAGTC    chr10.185_top_two_chr10.484_top_two_chr11.4167...   \n",
       "...                                                               ...   \n",
       "TTTGTCAGTACCTACA    BRIX1_TSS_chr10.2059_top_two_chr10.350_top_two...   \n",
       "TTTGTCAGTATCACCA-1  chr10.221_top_two_chr11.3853_top_two_chr11.498...   \n",
       "TTTGTCAGTTCAGACT-1    chr12.3400_top_two_chr4.1281_second_two_ID3_TSS   \n",
       "TTTGTCAGTTCTGTTT                                    chr7.1014_top_two   \n",
       "TTTGTCATCAAAGTAG-1  chr11.2790_top_two_chr12.306_top_two_chr12.474...   \n",
       "\n",
       "                                                             all_gene  \\\n",
       "AAACCTGAGAGGTACC    chr10.845_top_two_chr1.11183_top_two_chr1.1129...   \n",
       "AAACCTGAGTCAATAG    chr1.12695_top_two_chr11.3294_top_two_chr1.679...   \n",
       "AAACCTGCAAACAACA    ALDH1A2_TSS_BRI3_TSS_chr10.1918_top_two_chr10....   \n",
       "AAACCTGCACTTCTGC    C16orf91_TSS_chr1.11332_top_two_chr1.1933_top_...   \n",
       "AAACCTGCATGTAGTC    chr10.185_top_two_chr10.484_top_two_chr11.4167...   \n",
       "...                                                               ...   \n",
       "TTTGTCAGTACCTACA    BRIX1_TSS_chr10.2059_top_two_chr10.350_top_two...   \n",
       "TTTGTCAGTATCACCA-1  chr10.221_top_two_chr11.3853_top_two_chr11.498...   \n",
       "TTTGTCAGTTCAGACT-1    chr12.3400_top_two_chr4.1281_second_two_ID3_TSS   \n",
       "TTTGTCAGTTCTGTTT                                    chr7.1014_top_two   \n",
       "TTTGTCATCAAAGTAG-1  chr11.2790_top_two_chr12.306_top_two_chr12.474...   \n",
       "\n",
       "                                                              barcode  \\\n",
       "AAACCTGAGAGGTACC    AGAAAGCTCCTCCAGTTCAC_TGATCGCTTTGACTGTGACA_ACAA...   \n",
       "AAACCTGAGTCAATAG    GTAGAGCCTCCAGAACTGTG_AGGTTTATCCAGATGAACTG_CATC...   \n",
       "AAACCTGCAAACAACA    CCAAGGCGTCCTCAGACCAG_AGCTCCAGGAAGGACCCCCG_TCAC...   \n",
       "AAACCTGCACTTCTGC    GGCGTCAGTCGAGGAGTCAG_GCCAGCACTTCAGCTCACCG_GCTG...   \n",
       "AAACCTGCATGTAGTC    ATAAGGCACTCACATCCACC_GCTTGTCCCTAACACTCAGA_GGGC...   \n",
       "...                                                               ...   \n",
       "TTTGTCAGTACCTACA    AGGAGGCCAAGAGCGCGGGG_TGATTGAAGGAGGCTCCCCA_GGTA...   \n",
       "TTTGTCAGTATCACCA-1  CCTGCAACTGTCTATGGCCT_TCAGTGGGTGAGTCTTCAGG_TGAG...   \n",
       "TTTGTCAGTTCAGACT-1  CCAGTTGCTAGGCAGGACAG_CTTTCAGGCAGGAATCTGAG_TGGT...   \n",
       "TTTGTCAGTTCTGTTT                                 CACCCCCCAGAGATAAGAGA   \n",
       "TTTGTCATCAAAGTAG-1  AGTATGGGCCTGGAACACAG_AGATCTCCCCTGTCACCACG_CGTT...   \n",
       "\n",
       "                    read_count  umi_count  proportion  guide_count  ...  \\\n",
       "AAACCTGAGAGGTACC       14135.0      964.0    0.969819         67.0  ...   \n",
       "AAACCTGAGTCAATAG        4329.0      293.0    0.844380         26.0  ...   \n",
       "AAACCTGCAAACAACA       12362.0      884.0    0.950538         61.0  ...   \n",
       "AAACCTGCACTTCTGC        7459.0      544.0    0.939551         39.0  ...   \n",
       "AAACCTGCATGTAGTC       14831.0     1054.0    0.959927         37.0  ...   \n",
       "...                        ...        ...         ...          ...  ...   \n",
       "TTTGTCAGTACCTACA        4570.0      394.0    0.856522         31.0  ...   \n",
       "TTTGTCAGTATCACCA-1      4311.0      387.0    0.861915         33.0  ...   \n",
       "TTTGTCAGTTCAGACT-1       669.0       57.0    0.876923          3.0  ...   \n",
       "TTTGTCAGTTCTGTTT          77.0        7.0    0.086420          1.0  ...   \n",
       "TTTGTCATCAAAGTAG-1      4630.0      410.0    0.813492         35.0  ...   \n",
       "\n",
       "                   perturbation_type      celltype organism  \\\n",
       "AAACCTGAGAGGTACC              CRISPR  lymphoblasts    human   \n",
       "AAACCTGAGTCAATAG              CRISPR  lymphoblasts    human   \n",
       "AAACCTGCAAACAACA              CRISPR  lymphoblasts    human   \n",
       "AAACCTGCACTTCTGC              CRISPR  lymphoblasts    human   \n",
       "AAACCTGCATGTAGTC              CRISPR  lymphoblasts    human   \n",
       "...                              ...           ...      ...   \n",
       "TTTGTCAGTACCTACA              CRISPR  lymphoblasts    human   \n",
       "TTTGTCAGTATCACCA-1            CRISPR  lymphoblasts    human   \n",
       "TTTGTCAGTTCAGACT-1            CRISPR  lymphoblasts    human   \n",
       "TTTGTCAGTTCTGTTT              CRISPR  lymphoblasts    human   \n",
       "TTTGTCATCAAAGTAG-1            CRISPR  lymphoblasts    human   \n",
       "\n",
       "                                                         perturbation nperts  \\\n",
       "AAACCTGAGAGGTACC    chr10.845_top_two_chr1.11183_top_two_chr1.1129...    195   \n",
       "AAACCTGAGTCAATAG    chr1.12695_top_two_chr11.3294_top_two_chr1.679...     68   \n",
       "AAACCTGCAAACAACA    ALDH1A2_BRI3_chr10.1918_top_two_chr10.350_top_...    167   \n",
       "AAACCTGCACTTCTGC    C16orf91_chr1.11332_top_two_chr1.1933_top_two_...    109   \n",
       "AAACCTGCATGTAGTC    chr10.185_top_two_chr10.484_top_two_chr11.4167...    107   \n",
       "...                                                               ...    ...   \n",
       "TTTGTCAGTACCTACA    BRIX1_chr10.2059_top_two_chr10.350_top_two_chr...     85   \n",
       "TTTGTCAGTATCACCA-1  chr10.221_top_two_chr11.3853_top_two_chr11.498...     92   \n",
       "TTTGTCAGTTCAGACT-1        chr12.3400_top_two_chr4.1281_second_two_ID3      7   \n",
       "TTTGTCAGTTCTGTTT                                    chr7.1014_top_two      3   \n",
       "TTTGTCATCAAAGTAG-1  chr11.2790_top_two_chr12.306_top_two_chr12.474...     99   \n",
       "\n",
       "                   ngenes  ncounts percent_mito percent_ribo  n_genes  \n",
       "AAACCTGAGAGGTACC     3549  17566.0          0.0    30.507799     3549  \n",
       "AAACCTGAGTCAATAG     2543   8917.0          0.0    28.955927     2543  \n",
       "AAACCTGCAAACAACA     3191  14626.0          0.0    26.548612     3191  \n",
       "AAACCTGCACTTCTGC     4539  22783.0          0.0    20.616249     4539  \n",
       "AAACCTGCATGTAGTC     2605  10124.0          0.0    32.418017     2605  \n",
       "...                   ...      ...          ...          ...      ...  \n",
       "TTTGTCAGTACCTACA     3451  17934.0          0.0    31.733021     3451  \n",
       "TTTGTCAGTATCACCA-1   3693  16542.0          0.0    24.694716     3693  \n",
       "TTTGTCAGTTCAGACT-1   3162  14992.0          0.0    32.123799     3162  \n",
       "TTTGTCAGTTCTGTTT     1392   5149.0          0.0    35.074772     1392  \n",
       "TTTGTCATCAAAGTAG-1   2429   9499.0          0.0    36.624908     2429  \n",
       "\n",
       "[207324 rows x 31 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 310385 × 8563\n",
       "    obs: 'batch', 'gene', 'gene_id', 'transcript', 'gene_transcript', 'guide_id', 'percent_mito', 'UMI_count', 'z_gemgroup_UMI', 'core_scale_factor', 'core_adjusted_UMI_count', 'disease', 'cancer', 'cell_line', 'sex', 'age', 'perturbation', 'organism', 'perturbation_type', 'tissue_type', 'ncounts', 'ngenes', 'nperts', 'percent_ribo', 'n_genes'\n",
       "    var: 'chr', 'start', 'end', 'class', 'strand', 'length', 'in_matrix', 'mean', 'std', 'cv', 'fano', 'ensembl_id', 'ncounts', 'ncells', 'n_cells'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# only single perturbation\n",
    "# no nan\n",
    "\n",
    "valid_indices = adata.obs[(~adata.obs['perturbation'].str.contains('_')) & (~adata.obs['perturbation'].str.contains('nan'))].index\n",
    "adata = adata[valid_indices, :]\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/preprocessing/_normalization.py:216: UserWarning: Received a view of an AnnData. Making a copy.\n",
      "  view_to_actual(adata)\n"
     ]
    }
   ],
   "source": [
    "sc.pp.normalize_total(adata, target_sum=1e4)\n",
    "sc.pp.log1p(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.highly_variable_genes(adata, min_mean=0.0125, max_mean=3, min_disp=0.5)\n",
    "adata = adata[:, adata.var.highly_variable]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>batch</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript</th>\n",
       "      <th>gene_transcript</th>\n",
       "      <th>guide_id</th>\n",
       "      <th>percent_mito</th>\n",
       "      <th>UMI_count</th>\n",
       "      <th>z_gemgroup_UMI</th>\n",
       "      <th>core_scale_factor</th>\n",
       "      <th>...</th>\n",
       "      <th>age</th>\n",
       "      <th>perturbation</th>\n",
       "      <th>organism</th>\n",
       "      <th>perturbation_type</th>\n",
       "      <th>tissue_type</th>\n",
       "      <th>ncounts</th>\n",
       "      <th>ngenes</th>\n",
       "      <th>nperts</th>\n",
       "      <th>percent_ribo</th>\n",
       "      <th>n_genes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGAAATCCA-27</th>\n",
       "      <td>27</td>\n",
       "      <td>NAF1</td>\n",
       "      <td>ENSG00000145414</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>5449_NAF1_P1P2_ENSG00000145414</td>\n",
       "      <td>NAF1_+_164087918.23-P1P2|NAF1_-_164087674.23-P1P2</td>\n",
       "      <td>0.112083</td>\n",
       "      <td>11438.0</td>\n",
       "      <td>0.013047</td>\n",
       "      <td>0.813253</td>\n",
       "      <td>...</td>\n",
       "      <td>53</td>\n",
       "      <td>NAF1</td>\n",
       "      <td>human</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>11324.0</td>\n",
       "      <td>3332</td>\n",
       "      <td>1</td>\n",
       "      <td>0.225362</td>\n",
       "      <td>3332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGAACTTCC-31</th>\n",
       "      <td>31</td>\n",
       "      <td>BUB1</td>\n",
       "      <td>ENSG00000169679</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>935_BUB1_P1P2_ENSG00000169679</td>\n",
       "      <td>BUB1_-_111435363.23-P1P2|BUB1_-_111435372.23-P1P2</td>\n",
       "      <td>0.179895</td>\n",
       "      <td>5342.0</td>\n",
       "      <td>-1.522247</td>\n",
       "      <td>0.844107</td>\n",
       "      <td>...</td>\n",
       "      <td>53</td>\n",
       "      <td>BUB1</td>\n",
       "      <td>human</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>5257.0</td>\n",
       "      <td>2192</td>\n",
       "      <td>1</td>\n",
       "      <td>0.129732</td>\n",
       "      <td>2192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGAAGCCAC-34</th>\n",
       "      <td>34</td>\n",
       "      <td>UBL5</td>\n",
       "      <td>ENSG00000198258</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>9534_UBL5_P1P2_ENSG00000198258</td>\n",
       "      <td>UBL5_-_9938639.23-P1P2|UBL5_+_9938801.23-P1P2</td>\n",
       "      <td>0.105287</td>\n",
       "      <td>17305.0</td>\n",
       "      <td>0.384157</td>\n",
       "      <td>1.091537</td>\n",
       "      <td>...</td>\n",
       "      <td>53</td>\n",
       "      <td>UBL5</td>\n",
       "      <td>human</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>17135.0</td>\n",
       "      <td>4002</td>\n",
       "      <td>1</td>\n",
       "      <td>0.236825</td>\n",
       "      <td>4002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGAATAGTC-43</th>\n",
       "      <td>43</td>\n",
       "      <td>C9orf16</td>\n",
       "      <td>ENSG00000171159</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>1131_C9orf16_P1P2_ENSG00000171159</td>\n",
       "      <td>C9orf16_+_130922603.23-P1P2|C9orf16_+_13092264...</td>\n",
       "      <td>0.099359</td>\n",
       "      <td>30244.0</td>\n",
       "      <td>3.721912</td>\n",
       "      <td>0.948277</td>\n",
       "      <td>...</td>\n",
       "      <td>53</td>\n",
       "      <td>C9orf16</td>\n",
       "      <td>human</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>29717.0</td>\n",
       "      <td>5358</td>\n",
       "      <td>1</td>\n",
       "      <td>0.246828</td>\n",
       "      <td>5358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGACAGCGT-28</th>\n",
       "      <td>28</td>\n",
       "      <td>TIMM9</td>\n",
       "      <td>ENSG00000100575</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>8927_TIMM9_P1P2_ENSG00000100575</td>\n",
       "      <td>TIMM9_-_58893843.23-P1P2|TIMM9_-_58893848.23-P1P2</td>\n",
       "      <td>0.137623</td>\n",
       "      <td>8407.0</td>\n",
       "      <td>-0.975371</td>\n",
       "      <td>0.868942</td>\n",
       "      <td>...</td>\n",
       "      <td>53</td>\n",
       "      <td>TIMM9</td>\n",
       "      <td>human</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>8261.0</td>\n",
       "      <td>2944</td>\n",
       "      <td>1</td>\n",
       "      <td>0.183392</td>\n",
       "      <td>2944</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGTCTGTCGTC-45</th>\n",
       "      <td>45</td>\n",
       "      <td>ATP6V1D</td>\n",
       "      <td>ENSG00000100554</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>682_ATP6V1D_P1P2_ENSG00000100554</td>\n",
       "      <td>ATP6V1D_+_67826485.23-P1P2|ATP6V1D_+_67826497....</td>\n",
       "      <td>0.100272</td>\n",
       "      <td>18350.0</td>\n",
       "      <td>0.428227</td>\n",
       "      <td>1.115052</td>\n",
       "      <td>...</td>\n",
       "      <td>53</td>\n",
       "      <td>ATP6V1D</td>\n",
       "      <td>human</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>18044.0</td>\n",
       "      <td>4300</td>\n",
       "      <td>1</td>\n",
       "      <td>0.253547</td>\n",
       "      <td>4300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGTCTGTCTCG-27</th>\n",
       "      <td>27</td>\n",
       "      <td>CNOT3</td>\n",
       "      <td>ENSG00000088038</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>1718_CNOT3_P1P2_ENSG00000088038</td>\n",
       "      <td>CNOT3_+_54641532.23-P1P2|CNOT3_-_54641691.23-P1P2</td>\n",
       "      <td>0.093876</td>\n",
       "      <td>8671.0</td>\n",
       "      <td>-0.633593</td>\n",
       "      <td>0.813253</td>\n",
       "      <td>...</td>\n",
       "      <td>53</td>\n",
       "      <td>CNOT3</td>\n",
       "      <td>human</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>8510.0</td>\n",
       "      <td>3158</td>\n",
       "      <td>1</td>\n",
       "      <td>0.183196</td>\n",
       "      <td>3158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGTCTGTGCGG-44</th>\n",
       "      <td>44</td>\n",
       "      <td>METTL3</td>\n",
       "      <td>ENSG00000165819</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>5004_METTL3_P1P2_ENSG00000165819</td>\n",
       "      <td>METTL3_+_21979431.23-P1P2|METTL3_-_21979084.23...</td>\n",
       "      <td>0.107983</td>\n",
       "      <td>20568.0</td>\n",
       "      <td>1.054624</td>\n",
       "      <td>0.973352</td>\n",
       "      <td>...</td>\n",
       "      <td>53</td>\n",
       "      <td>METTL3</td>\n",
       "      <td>human</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>20355.0</td>\n",
       "      <td>4247</td>\n",
       "      <td>1</td>\n",
       "      <td>0.256595</td>\n",
       "      <td>4247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGTCTTGCAGA-14</th>\n",
       "      <td>14</td>\n",
       "      <td>RPL5</td>\n",
       "      <td>ENSG00000122406</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>7475_RPL5_P1P2_ENSG00000122406</td>\n",
       "      <td>RPL5_+_93297664.23-P1P2|RPL5_-_93297968.23-P1P2</td>\n",
       "      <td>0.128225</td>\n",
       "      <td>23568.0</td>\n",
       "      <td>1.676254</td>\n",
       "      <td>1.050055</td>\n",
       "      <td>...</td>\n",
       "      <td>53</td>\n",
       "      <td>RPL5</td>\n",
       "      <td>human</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>23169.0</td>\n",
       "      <td>4690</td>\n",
       "      <td>1</td>\n",
       "      <td>0.210713</td>\n",
       "      <td>4690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGTCTTTACAC-25</th>\n",
       "      <td>25</td>\n",
       "      <td>SEC61B</td>\n",
       "      <td>ENSG00000106803</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>7752_SEC61B_P1P2_ENSG00000106803</td>\n",
       "      <td>SEC61B_+_101984577.23-P1P2|SEC61B_-_101984591....</td>\n",
       "      <td>0.095009</td>\n",
       "      <td>13504.0</td>\n",
       "      <td>0.341625</td>\n",
       "      <td>0.855594</td>\n",
       "      <td>...</td>\n",
       "      <td>53</td>\n",
       "      <td>SEC61B</td>\n",
       "      <td>human</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>13314.0</td>\n",
       "      <td>3794</td>\n",
       "      <td>1</td>\n",
       "      <td>0.211432</td>\n",
       "      <td>3794</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>299694 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     batch     gene          gene_id transcript  \\\n",
       "cell_barcode                                                      \n",
       "AAACCCAAGAAATCCA-27     27     NAF1  ENSG00000145414       P1P2   \n",
       "AAACCCAAGAACTTCC-31     31     BUB1  ENSG00000169679       P1P2   \n",
       "AAACCCAAGAAGCCAC-34     34     UBL5  ENSG00000198258       P1P2   \n",
       "AAACCCAAGAATAGTC-43     43  C9orf16  ENSG00000171159       P1P2   \n",
       "AAACCCAAGACAGCGT-28     28    TIMM9  ENSG00000100575       P1P2   \n",
       "...                    ...      ...              ...        ...   \n",
       "TTTGTTGTCTGTCGTC-45     45  ATP6V1D  ENSG00000100554       P1P2   \n",
       "TTTGTTGTCTGTCTCG-27     27    CNOT3  ENSG00000088038       P1P2   \n",
       "TTTGTTGTCTGTGCGG-44     44   METTL3  ENSG00000165819       P1P2   \n",
       "TTTGTTGTCTTGCAGA-14     14     RPL5  ENSG00000122406       P1P2   \n",
       "TTTGTTGTCTTTACAC-25     25   SEC61B  ENSG00000106803       P1P2   \n",
       "\n",
       "                                       gene_transcript  \\\n",
       "cell_barcode                                             \n",
       "AAACCCAAGAAATCCA-27     5449_NAF1_P1P2_ENSG00000145414   \n",
       "AAACCCAAGAACTTCC-31      935_BUB1_P1P2_ENSG00000169679   \n",
       "AAACCCAAGAAGCCAC-34     9534_UBL5_P1P2_ENSG00000198258   \n",
       "AAACCCAAGAATAGTC-43  1131_C9orf16_P1P2_ENSG00000171159   \n",
       "AAACCCAAGACAGCGT-28    8927_TIMM9_P1P2_ENSG00000100575   \n",
       "...                                                ...   \n",
       "TTTGTTGTCTGTCGTC-45   682_ATP6V1D_P1P2_ENSG00000100554   \n",
       "TTTGTTGTCTGTCTCG-27    1718_CNOT3_P1P2_ENSG00000088038   \n",
       "TTTGTTGTCTGTGCGG-44   5004_METTL3_P1P2_ENSG00000165819   \n",
       "TTTGTTGTCTTGCAGA-14     7475_RPL5_P1P2_ENSG00000122406   \n",
       "TTTGTTGTCTTTACAC-25   7752_SEC61B_P1P2_ENSG00000106803   \n",
       "\n",
       "                                                              guide_id  \\\n",
       "cell_barcode                                                             \n",
       "AAACCCAAGAAATCCA-27  NAF1_+_164087918.23-P1P2|NAF1_-_164087674.23-P1P2   \n",
       "AAACCCAAGAACTTCC-31  BUB1_-_111435363.23-P1P2|BUB1_-_111435372.23-P1P2   \n",
       "AAACCCAAGAAGCCAC-34      UBL5_-_9938639.23-P1P2|UBL5_+_9938801.23-P1P2   \n",
       "AAACCCAAGAATAGTC-43  C9orf16_+_130922603.23-P1P2|C9orf16_+_13092264...   \n",
       "AAACCCAAGACAGCGT-28  TIMM9_-_58893843.23-P1P2|TIMM9_-_58893848.23-P1P2   \n",
       "...                                                                ...   \n",
       "TTTGTTGTCTGTCGTC-45  ATP6V1D_+_67826485.23-P1P2|ATP6V1D_+_67826497....   \n",
       "TTTGTTGTCTGTCTCG-27  CNOT3_+_54641532.23-P1P2|CNOT3_-_54641691.23-P1P2   \n",
       "TTTGTTGTCTGTGCGG-44  METTL3_+_21979431.23-P1P2|METTL3_-_21979084.23...   \n",
       "TTTGTTGTCTTGCAGA-14    RPL5_+_93297664.23-P1P2|RPL5_-_93297968.23-P1P2   \n",
       "TTTGTTGTCTTTACAC-25  SEC61B_+_101984577.23-P1P2|SEC61B_-_101984591....   \n",
       "\n",
       "                     percent_mito  UMI_count  z_gemgroup_UMI  \\\n",
       "cell_barcode                                                   \n",
       "AAACCCAAGAAATCCA-27      0.112083    11438.0        0.013047   \n",
       "AAACCCAAGAACTTCC-31      0.179895     5342.0       -1.522247   \n",
       "AAACCCAAGAAGCCAC-34      0.105287    17305.0        0.384157   \n",
       "AAACCCAAGAATAGTC-43      0.099359    30244.0        3.721912   \n",
       "AAACCCAAGACAGCGT-28      0.137623     8407.0       -0.975371   \n",
       "...                           ...        ...             ...   \n",
       "TTTGTTGTCTGTCGTC-45      0.100272    18350.0        0.428227   \n",
       "TTTGTTGTCTGTCTCG-27      0.093876     8671.0       -0.633593   \n",
       "TTTGTTGTCTGTGCGG-44      0.107983    20568.0        1.054624   \n",
       "TTTGTTGTCTTGCAGA-14      0.128225    23568.0        1.676254   \n",
       "TTTGTTGTCTTTACAC-25      0.095009    13504.0        0.341625   \n",
       "\n",
       "                     core_scale_factor  ...  age perturbation  organism  \\\n",
       "cell_barcode                            ...                               \n",
       "AAACCCAAGAAATCCA-27           0.813253  ...   53         NAF1     human   \n",
       "AAACCCAAGAACTTCC-31           0.844107  ...   53         BUB1     human   \n",
       "AAACCCAAGAAGCCAC-34           1.091537  ...   53         UBL5     human   \n",
       "AAACCCAAGAATAGTC-43           0.948277  ...   53      C9orf16     human   \n",
       "AAACCCAAGACAGCGT-28           0.868942  ...   53        TIMM9     human   \n",
       "...                                ...  ...  ...          ...       ...   \n",
       "TTTGTTGTCTGTCGTC-45           1.115052  ...   53      ATP6V1D     human   \n",
       "TTTGTTGTCTGTCTCG-27           0.813253  ...   53        CNOT3     human   \n",
       "TTTGTTGTCTGTGCGG-44           0.973352  ...   53       METTL3     human   \n",
       "TTTGTTGTCTTGCAGA-14           1.050055  ...   53         RPL5     human   \n",
       "TTTGTTGTCTTTACAC-25           0.855594  ...   53       SEC61B     human   \n",
       "\n",
       "                    perturbation_type tissue_type  ncounts ngenes nperts  \\\n",
       "cell_barcode                                                               \n",
       "AAACCCAAGAAATCCA-27            CRISPR   cell_line  11324.0   3332      1   \n",
       "AAACCCAAGAACTTCC-31            CRISPR   cell_line   5257.0   2192      1   \n",
       "AAACCCAAGAAGCCAC-34            CRISPR   cell_line  17135.0   4002      1   \n",
       "AAACCCAAGAATAGTC-43            CRISPR   cell_line  29717.0   5358      1   \n",
       "AAACCCAAGACAGCGT-28            CRISPR   cell_line   8261.0   2944      1   \n",
       "...                               ...         ...      ...    ...    ...   \n",
       "TTTGTTGTCTGTCGTC-45            CRISPR   cell_line  18044.0   4300      1   \n",
       "TTTGTTGTCTGTCTCG-27            CRISPR   cell_line   8510.0   3158      1   \n",
       "TTTGTTGTCTGTGCGG-44            CRISPR   cell_line  20355.0   4247      1   \n",
       "TTTGTTGTCTTGCAGA-14            CRISPR   cell_line  23169.0   4690      1   \n",
       "TTTGTTGTCTTTACAC-25            CRISPR   cell_line  13314.0   3794      1   \n",
       "\n",
       "                    percent_ribo n_genes  \n",
       "cell_barcode                              \n",
       "AAACCCAAGAAATCCA-27     0.225362    3332  \n",
       "AAACCCAAGAACTTCC-31     0.129732    2192  \n",
       "AAACCCAAGAAGCCAC-34     0.236825    4002  \n",
       "AAACCCAAGAATAGTC-43     0.246828    5358  \n",
       "AAACCCAAGACAGCGT-28     0.183392    2944  \n",
       "...                          ...     ...  \n",
       "TTTGTTGTCTGTCGTC-45     0.253547    4300  \n",
       "TTTGTTGTCTGTCTCG-27     0.183196    3158  \n",
       "TTTGTTGTCTGTGCGG-44     0.256595    4247  \n",
       "TTTGTTGTCTTGCAGA-14     0.210713    4690  \n",
       "TTTGTTGTCTTTACAC-25     0.211432    3794  \n",
       "\n",
       "[299694 rows x 25 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs[adata.obs['perturbation'] != 'control']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene_string='AAAS'\n",
      "gene_list=['AAAS']\n",
      "match=None\n",
      "['AAAS']:     ENSG00000094914\n",
      "gene_string='AAMP'\n",
      "gene_list=['AAMP']\n",
      "match=None\n",
      "['AAMP']:     ENSG00000127837\n",
      "gene_string='AARS'\n",
      "gene_list=['AARS']\n",
      "match=None\n",
      "['AARS']:     ENSG00000090861\n",
      "gene_string='AARS2'\n",
      "gene_list=['AARS2']\n",
      "match=None\n",
      "['AARS2']:     ENSG00000124608\n",
      "gene_string='AASDHPPT'\n",
      "gene_list=['AASDHPPT']\n",
      "match=None\n",
      "['AASDHPPT']:     ENSG00000149313\n",
      "gene_string='AATF'\n",
      "gene_list=['AATF']\n",
      "match=None\n",
      "['AATF']:     ENSG00000275700\n",
      "gene_string='ABCB7'\n",
      "gene_list=['ABCB7']\n",
      "match=None\n",
      "['ABCB7']:     ENSG00000131269\n",
      "gene_string='ABCB10'\n",
      "gene_list=['ABCB10']\n",
      "match=None\n",
      "['ABCB10']:     ENSG00000135776\n",
      "gene_string='ABCE1'\n",
      "gene_list=['ABCE1']\n",
      "match=None\n",
      "['ABCE1']:     ENSG00000164163\n",
      "gene_string='ABCF1'\n",
      "gene_list=['ABCF1']\n",
      "match=None\n",
      "['ABCF1']:     ENSG00000204574\n",
      "gene_string='ABCG1'\n",
      "gene_list=['ABCG1']\n",
      "match=None\n",
      "['ABCG1']:     ENSG00000160179\n",
      "gene_string='ABHD11'\n",
      "gene_list=['ABHD11']\n",
      "match=None\n",
      "['ABHD11']:     ENSG00000106077\n",
      "gene_string='ABHD17A'\n",
      "gene_list=['ABHD17A']\n",
      "match=None\n",
      "['ABHD17A']:     ENSG00000129968\n",
      "gene_string='ABT1'\n",
      "gene_list=['ABT1']\n",
      "match=None\n",
      "['ABT1']:     ENSG00000146109\n",
      "gene_string='ACD'\n",
      "gene_list=['ACD']\n",
      "match=None\n",
      "['ACD']:     ENSG00000102977\n",
      "gene_string='ACIN1'\n",
      "gene_list=['ACIN1']\n",
      "match=None\n",
      "['ACIN1']:     ENSG00000100813\n",
      "gene_string='ACOT12'\n",
      "gene_list=['ACOT12']\n",
      "match=None\n",
      "['ACOT12']:     ENSG00000172497\n",
      "gene_string='ACTB'\n",
      "gene_list=['ACTB']\n",
      "match=None\n",
      "['ACTB']:     ENSG00000075624\n",
      "gene_string='ACTL6A'\n",
      "gene_list=['ACTL6A']\n",
      "match=None\n",
      "['ACTL6A']:     ENSG00000136518\n",
      "gene_string='ACTR1A'\n",
      "gene_list=['ACTR1A']\n",
      "match=None\n",
      "['ACTR1A']:     ENSG00000138107\n",
      "gene_string='ACTR1B'\n",
      "gene_list=['ACTR1B']\n",
      "match=None\n",
      "['ACTR1B']:     ENSG00000115073\n",
      "gene_string='ACTR2'\n",
      "gene_list=['ACTR2']\n",
      "match=None\n",
      "['ACTR2']:     ENSG00000138071\n",
      "gene_string='ACTR3'\n",
      "gene_list=['ACTR3']\n",
      "match=None\n",
      "['ACTR3']:     ENSG00000115091\n",
      "gene_string='ACTR6'\n",
      "gene_list=['ACTR6']\n",
      "match=None\n",
      "['ACTR6']:     ENSG00000075089\n",
      "gene_string='ACTR8'\n",
      "gene_list=['ACTR8']\n",
      "match=None\n",
      "['ACTR8']:     ENSG00000113812\n",
      "gene_string='ACTR10'\n",
      "gene_list=['ACTR10']\n",
      "match=None\n",
      "['ACTR10']:     ENSG00000131966\n",
      "gene_string='ADAT2'\n",
      "gene_list=['ADAT2']\n",
      "match=None\n",
      "['ADAT2']:     ENSG00000189007\n",
      "gene_string='ADAT3'\n",
      "gene_list=['ADAT3']\n",
      "match=None\n",
      "['ADAT3']:     ENSG00000213638\n",
      "gene_string='ADRM1'\n",
      "gene_list=['ADRM1']\n",
      "match=None\n",
      "['ADRM1']:     ENSG00000130706\n",
      "gene_string='ADSL'\n",
      "gene_list=['ADSL']\n",
      "match=None\n",
      "['ADSL']:     ENSG00000239900\n",
      "gene_string='AFG3L2'\n",
      "gene_list=['AFG3L2']\n",
      "match=None\n",
      "['AFG3L2']:     ENSG00000141385\n",
      "gene_string='AHCTF1'\n",
      "gene_list=['AHCTF1']\n",
      "match=None\n",
      "['AHCTF1']:     ENSG00000153207\n",
      "gene_string='AHCY'\n",
      "gene_list=['AHCY']\n",
      "match=None\n",
      "['AHCY']:     ENSG00000101444\n",
      "gene_string='AIFM1'\n",
      "gene_list=['AIFM1']\n",
      "match=None\n",
      "['AIFM1']:     ENSG00000156709\n",
      "gene_string='AKIRIN2'\n",
      "gene_list=['AKIRIN2']\n",
      "match=None\n",
      "['AKIRIN2']:     ENSG00000135334\n",
      "gene_string='ALDOA'\n",
      "gene_list=['ALDOA']\n",
      "match=None\n",
      "['ALDOA']:     ENSG00000149925\n",
      "gene_string='ALG1'\n",
      "gene_list=['ALG1']\n",
      "match=None\n",
      "['ALG1']:     ENSG00000033011\n",
      "gene_string='ALG1L'\n",
      "gene_list=['ALG1L']\n",
      "match=None\n",
      "['ALG1L']:     ENSG00000189366\n",
      "gene_string='ALG2'\n",
      "gene_list=['ALG2']\n",
      "match=None\n",
      "['ALG2']:     ENSG00000119523\n",
      "gene_string='ALG11'\n",
      "gene_list=['ALG11']\n",
      "match=None\n",
      "['ALG11']:     ENSG00000253710\n",
      "gene_string='ALG13'\n",
      "gene_list=['ALG13']\n",
      "match=None\n",
      "['ALG13']:     ENSG00000101901\n",
      "gene_string='ALG14'\n",
      "gene_list=['ALG14']\n",
      "match=None\n",
      "['ALG14']:     ENSG00000172339\n",
      "gene_string='ALYREF'\n",
      "gene_list=['ALYREF']\n",
      "match=None\n",
      "['ALYREF']:     ENSG00000183684\n",
      "gene_string='ANAPC1'\n",
      "gene_list=['ANAPC1']\n",
      "match=None\n",
      "['ANAPC1']:     ENSG00000153107\n",
      "gene_string='ANAPC2'\n",
      "gene_list=['ANAPC2']\n",
      "match=None\n",
      "['ANAPC2']:     ENSG00000176248\n",
      "gene_string='ANAPC4'\n",
      "gene_list=['ANAPC4']\n",
      "match=None\n",
      "['ANAPC4']:     ENSG00000053900\n",
      "gene_string='ANAPC5'\n",
      "gene_list=['ANAPC5']\n",
      "match=None\n",
      "['ANAPC5']:     ENSG00000089053\n",
      "gene_string='ANAPC10'\n",
      "gene_list=['ANAPC10']\n",
      "match=None\n",
      "['ANAPC10']:     ENSG00000164162\n",
      "gene_string='ANAPC11'\n",
      "gene_list=['ANAPC11']\n",
      "match=None\n",
      "['ANAPC11']:     ENSG00000141552\n",
      "gene_string='ANAPC13'\n",
      "gene_list=['ANAPC13']\n",
      "match=None\n",
      "['ANAPC13']:     ENSG00000129055\n",
      "gene_string='ANAPC15'\n",
      "gene_list=['ANAPC15']\n",
      "match=None\n",
      "['ANAPC15']:     ENSG00000110200\n",
      "gene_string='ANKLE2'\n",
      "gene_list=['ANKLE2']\n",
      "match=None\n",
      "['ANKLE2']:     ENSG00000176915\n",
      "gene_string='ANKRD11'\n",
      "gene_list=['ANKRD11']\n",
      "match=None\n",
      "['ANKRD11']:     ENSG00000167522\n",
      "gene_string='ANKRD49'\n",
      "gene_list=['ANKRD49']\n",
      "match=None\n",
      "['ANKRD49']:     ENSG00000168876\n",
      "gene_string='ANKS6'\n",
      "gene_list=['ANKS6']\n",
      "match=None\n",
      "['ANKS6']:     ENSG00000165138\n",
      "gene_string='ANLN'\n",
      "gene_list=['ANLN']\n",
      "match=None\n",
      "['ANLN']:     ENSG00000011426\n",
      "gene_string='AP2M1'\n",
      "gene_list=['AP2M1']\n",
      "match=None\n",
      "['AP2M1']:     ENSG00000161203\n",
      "gene_string='AP2S1'\n",
      "gene_list=['AP2S1']\n",
      "match=None\n",
      "['AP2S1']:     ENSG00000042753\n",
      "gene_string='AQP7'\n",
      "gene_list=['AQP7']\n",
      "match=None\n",
      "['AQP7']:     ENSG00000165269\n",
      "gene_string='AQR'\n",
      "gene_list=['AQR']\n",
      "match=None\n",
      "['AQR']:     ENSG00000021776\n",
      "gene_string='ARCN1'\n",
      "gene_list=['ARCN1']\n",
      "match=None\n",
      "['ARCN1']:     ENSG00000095139\n",
      "gene_string='ARF4'\n",
      "gene_list=['ARF4']\n",
      "match=None\n",
      "['ARF4']:     ENSG00000168374\n",
      "gene_string='ARFRP1'\n",
      "gene_list=['ARFRP1']\n",
      "match=None\n",
      "['ARFRP1']:     ENSG00000101246\n",
      "gene_string='ARGFX'\n",
      "gene_list=['ARGFX']\n",
      "match=None\n",
      "['ARGFX']:     ENSG00000186103\n",
      "gene_string='ARGLU1'\n",
      "gene_list=['ARGLU1']\n",
      "match=None\n",
      "['ARGLU1']:     ENSG00000134884\n",
      "gene_string='ARHGAP11B'\n",
      "gene_list=['ARHGAP11B']\n",
      "match=None\n",
      "['ARHGAP11B']:     ENSG00000285077\n",
      "gene_string='ARID3A'\n",
      "gene_list=['ARID3A']\n",
      "match=None\n",
      "['ARID3A']:     ENSG00000116017\n",
      "gene_string='ARIH1'\n",
      "gene_list=['ARIH1']\n",
      "match=None\n",
      "['ARIH1']:     ENSG00000166233\n",
      "gene_string='ARL2'\n",
      "gene_list=['ARL2']\n",
      "match=None\n",
      "['ARL2']:     ENSG00000213465\n",
      "gene_string='ARL4D'\n",
      "gene_list=['ARL4D']\n",
      "match=None\n",
      "['ARL4D']:     ENSG00000175906\n",
      "gene_string='ARMC6'\n",
      "gene_list=['ARMC6']\n",
      "match=None\n",
      "['ARMC6']:     ENSG00000105676\n",
      "gene_string='ARMC7'\n",
      "gene_list=['ARMC7']\n",
      "match=None\n",
      "['ARMC7']:     ENSG00000125449\n",
      "gene_string='ARPC2'\n",
      "gene_list=['ARPC2']\n",
      "match=None\n",
      "['ARPC2']:     ENSG00000163466\n",
      "gene_string='ARPC3'\n",
      "gene_list=['ARPC3']\n",
      "match=None\n",
      "['ARPC3']:     ENSG00000111229\n",
      "gene_string='ARPC4'\n",
      "gene_list=['ARPC4']\n",
      "match=None\n",
      "['ARPC4']:     ENSG00000241553\n",
      "gene_string='ASF1B'\n",
      "gene_list=['ASF1B']\n",
      "match=None\n",
      "['ASF1B']:     ENSG00000105011\n",
      "gene_string='ATF4'\n",
      "gene_list=['ATF4']\n",
      "match=None\n",
      "['ATF4']:     ENSG00000128272\n",
      "gene_string='ATF5'\n",
      "gene_list=['ATF5']\n",
      "match=None\n",
      "['ATF5']:     ENSG00000169136\n",
      "gene_string='ATL2'\n",
      "gene_list=['ATL2']\n",
      "match=None\n",
      "['ATL2']:     ENSG00000119787\n",
      "gene_string='ATP1A1'\n",
      "gene_list=['ATP1A1']\n",
      "match=None\n",
      "['ATP1A1']:     ENSG00000163399\n",
      "gene_string='ATP2A2'\n",
      "gene_list=['ATP2A2']\n",
      "match=None\n",
      "['ATP2A2']:     ENSG00000174437\n",
      "gene_string='ATP6AP1'\n",
      "gene_list=['ATP6AP1']\n",
      "match=None\n",
      "['ATP6AP1']:     ENSG00000071553\n",
      "gene_string='ATP6AP2'\n",
      "gene_list=['ATP6AP2']\n",
      "match=None\n",
      "['ATP6AP2']:     ENSG00000182220\n",
      "gene_string='ATP6V0B'\n",
      "gene_list=['ATP6V0B']\n",
      "match=None\n",
      "['ATP6V0B']:     ENSG00000117410\n",
      "gene_string='ATP6V0C'\n",
      "gene_list=['ATP6V0C']\n",
      "match=None\n",
      "['ATP6V0C']:     ENSG00000185883\n",
      "gene_string='ATP6V0D1'\n",
      "gene_list=['ATP6V0D1']\n",
      "match=None\n",
      "['ATP6V0D1']:     ENSG00000159720\n",
      "gene_string='ATP6V1A'\n",
      "gene_list=['ATP6V1A']\n",
      "match=None\n",
      "['ATP6V1A']:     ENSG00000114573\n",
      "gene_string='ATP6V1B2'\n",
      "gene_list=['ATP6V1B2']\n",
      "match=None\n",
      "['ATP6V1B2']:     ENSG00000147416\n",
      "gene_string='ATP6V1C1'\n",
      "gene_list=['ATP6V1C1']\n",
      "match=None\n",
      "['ATP6V1C1']:     ENSG00000155097\n",
      "gene_string='ATP6V1D'\n",
      "gene_list=['ATP6V1D']\n",
      "match=None\n",
      "['ATP6V1D']:     ENSG00000100554\n",
      "gene_string='ATP6V1E1'\n",
      "gene_list=['ATP6V1E1']\n",
      "match=None\n",
      "['ATP6V1E1']:     ENSG00000131100\n",
      "gene_string='ATP6V1F'\n",
      "gene_list=['ATP6V1F']\n",
      "match=None\n",
      "['ATP6V1F']:     ENSG00000128524\n",
      "gene_string='ATP6V1G1'\n",
      "gene_list=['ATP6V1G1']\n",
      "match=None\n",
      "['ATP6V1G1']:     ENSG00000136888\n",
      "gene_string='ATP6V1H'\n",
      "gene_list=['ATP6V1H']\n",
      "match=None\n",
      "['ATP6V1H']:     ENSG00000047249\n",
      "gene_string='ATR'\n",
      "gene_list=['ATR']\n",
      "match=None\n",
      "['ATR']:     ENSG00000175054\n",
      "gene_string='ATRIP'\n",
      "gene_list=['ATRIP']\n",
      "match=None\n",
      "['ATRIP']:     ENSG00000164053\n",
      "gene_string='AURKA'\n",
      "gene_list=['AURKA']\n",
      "match=None\n",
      "['AURKA']:     ENSG00000087586\n",
      "gene_string='AURKAIP1'\n",
      "gene_list=['AURKAIP1']\n",
      "match=None\n",
      "['AURKAIP1']:     ENSG00000175756\n",
      "gene_string='AURKB'\n",
      "gene_list=['AURKB']\n",
      "match=None\n",
      "['AURKB']:     ENSG00000178999\n",
      "gene_string='BANF1'\n",
      "gene_list=['BANF1']\n",
      "match=None\n",
      "['BANF1']:     ENSG00000175334\n",
      "gene_string='BANP'\n",
      "gene_list=['BANP']\n",
      "match=None\n",
      "['BANP']:     ENSG00000172530\n",
      "gene_string='BAP1'\n",
      "gene_list=['BAP1']\n",
      "match=None\n",
      "['BAP1']:     ENSG00000163930\n",
      "gene_string='BARD1'\n",
      "gene_list=['BARD1']\n",
      "match=None\n",
      "['BARD1']:     ENSG00000138376\n",
      "gene_string='BCAR1'\n",
      "gene_list=['BCAR1']\n",
      "match=None\n",
      "['BCAR1']:     ENSG00000050820\n",
      "gene_string='BCAS2'\n",
      "gene_list=['BCAS2']\n",
      "match=None\n",
      "['BCAS2']:     ENSG00000116752\n",
      "gene_string='BCL2L1'\n",
      "gene_list=['BCL2L1']\n",
      "match=None\n",
      "['BCL2L1']:     ENSG00000171552\n",
      "gene_string='BCLAF1'\n",
      "gene_list=['BCLAF1']\n",
      "match=None\n",
      "['BCLAF1']:     ENSG00000029363\n",
      "gene_string='BCR'\n",
      "gene_list=['BCR']\n",
      "match=None\n",
      "['BCR']:     ENSG00000186716\n",
      "gene_string='BCS1L'\n",
      "gene_list=['BCS1L']\n",
      "match=None\n",
      "['BCS1L']:     ENSG00000074582\n",
      "gene_string='BDP1'\n",
      "gene_list=['BDP1']\n",
      "match=None\n",
      "['BDP1']:     ENSG00000072135\n",
      "gene_string='BET1'\n",
      "gene_list=['BET1']\n",
      "match=None\n",
      "['BET1']:     ENSG00000105829\n",
      "gene_string='BGLAP'\n",
      "gene_list=['BGLAP']\n",
      "match=None\n",
      "['BGLAP']:     ENSG00000242252\n",
      "gene_string='BIRC5'\n",
      "gene_list=['BIRC5']\n",
      "match=None\n",
      "['BIRC5']:     ENSG00000089685\n",
      "gene_string='BMS1'\n",
      "gene_list=['BMS1']\n",
      "match=None\n",
      "['BMS1']:     ENSG00000165733\n",
      "gene_string='BNIP1'\n",
      "gene_list=['BNIP1']\n",
      "match=None\n",
      "['BNIP1']:     ENSG00000113734\n",
      "gene_string='BOP1'\n",
      "gene_list=['BOP1']\n",
      "match=None\n",
      "['BOP1']:     ENSG00000261236\n",
      "gene_string='BORA'\n",
      "gene_list=['BORA']\n",
      "match=None\n",
      "['BORA']:     ENSG00000136122\n",
      "gene_string='BPTF'\n",
      "gene_list=['BPTF']\n",
      "match=None\n",
      "['BPTF']:     ENSG00000171634\n",
      "gene_string='BRCA1'\n",
      "gene_list=['BRCA1']\n",
      "match=None\n",
      "['BRCA1']:     ENSG00000012048\n",
      "gene_string='BRCA2'\n",
      "gene_list=['BRCA2']\n",
      "match=None\n",
      "['BRCA2']:     ENSG00000139618\n",
      "gene_string='BRD4'\n",
      "gene_list=['BRD4']\n",
      "match=None\n",
      "['BRD4']:     ENSG00000141867\n",
      "gene_string='BRD8'\n",
      "gene_list=['BRD8']\n",
      "match=None\n",
      "['BRD8']:     ENSG00000112983\n",
      "gene_string='BRF1'\n",
      "gene_list=['BRF1']\n",
      "match=None\n",
      "['BRF1']:     ENSG00000185650\n",
      "gene_string='BRF2'\n",
      "gene_list=['BRF2']\n",
      "match=None\n",
      "['BRF2']:     ENSG00000104221\n",
      "gene_string='BRIP1'\n",
      "gene_list=['BRIP1']\n",
      "match=None\n",
      "['BRIP1']:     ENSG00000136492\n",
      "gene_string='BRIX1'\n",
      "gene_list=['BRIX1']\n",
      "match=None\n",
      "['BRIX1']:     ENSG00000113460\n",
      "gene_string='BRK1'\n",
      "gene_list=['BRK1']\n",
      "match=None\n",
      "['BRK1']:     ENSG00000254999\n",
      "gene_string='BTF3'\n",
      "gene_list=['BTF3']\n",
      "match=None\n",
      "['BTF3']:     ENSG00000111801\n",
      "gene_string='BTF3L4'\n",
      "gene_list=['BTF3L4']\n",
      "match=None\n",
      "['BTF3L4']:     ENSG00000134717\n",
      "gene_string='BUB1'\n",
      "gene_list=['BUB1']\n",
      "match=None\n",
      "['BUB1']:     ENSG00000169679\n",
      "gene_string='BUB1B'\n",
      "gene_list=['BUB1B']\n",
      "match=None\n",
      "['BUB1B']:     ENSG00000156970\n",
      "gene_string='BUB3'\n",
      "gene_list=['BUB3']\n",
      "match=None\n",
      "['BUB3']:     ENSG00000154473\n",
      "gene_string='BUD13'\n",
      "gene_list=['BUD13']\n",
      "match=None\n",
      "['BUD13']:     ENSG00000137656\n",
      "gene_string='BUD31'\n",
      "gene_list=['BUD31']\n",
      "match=None\n",
      "['BUD31']:     ENSG00000106245\n",
      "gene_string='BYSL'\n",
      "gene_list=['BYSL']\n",
      "match=None\n",
      "['BYSL']:     ENSG00000112578\n",
      "gene_string='C1D'\n",
      "gene_list=['C1D']\n",
      "match=None\n",
      "['C1D']:     ENSG00000197223\n",
      "gene_string='C1QBP'\n",
      "gene_list=['C1QBP']\n",
      "match=None\n",
      "['C1QBP']:     ENSG00000108561\n",
      "gene_string='C1QTNF4'\n",
      "gene_list=['C1QTNF4']\n",
      "match=None\n",
      "['C1QTNF4']:     ENSG00000172247\n",
      "gene_string='C1orf109'\n",
      "gene_list=['C1orf109']\n",
      "match=None\n",
      "['C1orf109']:     ENSG00000116922\n",
      "gene_string='C6orf15'\n",
      "gene_list=['C6orf15']\n",
      "match=None\n",
      "['C6orf15']:     ENSG00000204542\n",
      "gene_string='C7orf26'\n",
      "gene_list=['C7orf26']\n",
      "match=None\n",
      "['C7orf26']:     ENSG00000146576\n",
      "gene_string='C7orf50'\n",
      "gene_list=['C7orf50']\n",
      "match=None\n",
      "['C7orf50']:     ENSG00000146540\n",
      "gene_string='C9orf16'\n",
      "gene_list=['C9orf16']\n",
      "match=None\n",
      "['C9orf16']:     ENSG00000171159\n",
      "gene_string='C9orf78'\n",
      "gene_list=['C9orf78']\n",
      "match=None\n",
      "['C9orf78']:     ENSG00000136819\n",
      "gene_string='C12orf45'\n",
      "gene_list=['C12orf45']\n",
      "match=None\n",
      "['C12orf45']:     ENSG00000151131\n",
      "gene_string='C12orf60'\n",
      "gene_list=['C12orf60']\n",
      "match=None\n",
      "['C12orf60']:     ENSG00000182993\n",
      "gene_string='C14orf178'\n",
      "gene_list=['C14orf178']\n",
      "match=None\n",
      "['C14orf178']:     ENSG00000197734\n",
      "gene_string='C16orf86'\n",
      "gene_list=['C16orf86']\n",
      "match=None\n",
      "['C16orf86']:     ENSG00000159761\n",
      "gene_string='C17orf58'\n",
      "gene_list=['C17orf58']\n",
      "match=None\n",
      "['C17orf58']:     ENSG00000186665\n",
      "gene_string='C19orf25'\n",
      "gene_list=['C19orf25']\n",
      "match=None\n",
      "['C19orf25']:     ENSG00000119559\n",
      "gene_string='C19orf53'\n",
      "gene_list=['C19orf53']\n",
      "match=None\n",
      "['C19orf53']:     ENSG00000104979\n",
      "gene_string='CA5A'\n",
      "gene_list=['CA5A']\n",
      "match=None\n",
      "['CA5A']:     ENSG00000174990\n",
      "gene_string='CACNB3'\n",
      "gene_list=['CACNB3']\n",
      "match=None\n",
      "['CACNB3']:     ENSG00000167535\n",
      "gene_string='CACTIN'\n",
      "gene_list=['CACTIN']\n",
      "match=None\n",
      "['CACTIN']:     ENSG00000105298\n",
      "gene_string='CADM4'\n",
      "gene_list=['CADM4']\n",
      "match=None\n",
      "['CADM4']:     ENSG00000105767\n",
      "gene_string='CAMLG'\n",
      "gene_list=['CAMLG']\n",
      "match=None\n",
      "['CAMLG']:     ENSG00000164615\n",
      "gene_string='CAP1'\n",
      "gene_list=['CAP1']\n",
      "match=None\n",
      "['CAP1']:     ENSG00000052344\n",
      "gene_string='CAPZB'\n",
      "gene_list=['CAPZB']\n",
      "match=None\n",
      "['CAPZB']:     ENSG00000077549\n",
      "gene_string='CARS'\n",
      "gene_list=['CARS']\n",
      "match=None\n",
      "['CARS']:     ENSG00000110619\n",
      "gene_string='CASP8AP2'\n",
      "gene_list=['CASP8AP2']\n",
      "match=None\n",
      "['CASP8AP2']:     ENSG00000118412\n",
      "gene_string='CBLL1'\n",
      "gene_list=['CBLL1']\n",
      "match=None\n",
      "['CBLL1']:     ENSG00000105879\n",
      "gene_string='CBLN1'\n",
      "gene_list=['CBLN1']\n",
      "match=None\n",
      "['CBLN1']:     ENSG00000102924\n",
      "gene_string='CBX1'\n",
      "gene_list=['CBX1']\n",
      "match=None\n",
      "['CBX1']:     ENSG00000108468\n",
      "gene_string='CCAR1'\n",
      "gene_list=['CCAR1']\n",
      "match=None\n",
      "['CCAR1']:     ENSG00000060339\n",
      "gene_string='CCDC59'\n",
      "gene_list=['CCDC59']\n",
      "match=None\n",
      "['CCDC59']:     ENSG00000133773\n",
      "gene_string='CCDC78'\n",
      "gene_list=['CCDC78']\n",
      "match=None\n",
      "['CCDC78']:     ENSG00000162004\n",
      "gene_string='CCDC84'\n",
      "gene_list=['CCDC84']\n",
      "match=None\n",
      "['CCDC84']:     ENSG00000186166\n",
      "gene_string='CCDC86'\n",
      "gene_list=['CCDC86']\n",
      "match=None\n",
      "['CCDC86']:     ENSG00000110104\n",
      "gene_string='CCDC115'\n",
      "gene_list=['CCDC115']\n",
      "match=None\n",
      "['CCDC115']:     ENSG00000136710\n",
      "gene_string='CCDC130'\n",
      "gene_list=['CCDC130']\n",
      "match=None\n",
      "['CCDC130']:     ENSG00000104957\n",
      "gene_string='CCDC137'\n",
      "gene_list=['CCDC137']\n",
      "match=None\n",
      "['CCDC137']:     ENSG00000185298\n",
      "gene_string='CCDC144A'\n",
      "gene_list=['CCDC144A']\n",
      "match=None\n",
      "['CCDC144A']:     ENSG00000266302\n",
      "gene_string='CCDC144NL'\n",
      "gene_list=['CCDC144NL']\n",
      "match=None\n",
      "['CCDC144NL']:     ENSG00000291074\n",
      "gene_string='CCDC174'\n",
      "gene_list=['CCDC174']\n",
      "match=None\n",
      "['CCDC174']:     ENSG00000154781\n",
      "gene_string='CCNA2'\n",
      "gene_list=['CCNA2']\n",
      "match=None\n",
      "['CCNA2']:     ENSG00000145386\n",
      "gene_string='CCNB1'\n",
      "gene_list=['CCNB1']\n",
      "match=None\n",
      "['CCNB1']:     ENSG00000134057\n",
      "gene_string='CCND1'\n",
      "gene_list=['CCND1']\n",
      "match=None\n",
      "['CCND1']:     ENSG00000110092\n",
      "gene_string='CCNH'\n",
      "gene_list=['CCNH']\n",
      "match=None\n",
      "['CCNH']:     ENSG00000134480\n",
      "gene_string='CCNK'\n",
      "gene_list=['CCNK']\n",
      "match=None\n",
      "['CCNK']:     ENSG00000090061\n",
      "gene_string='CCNL1'\n",
      "gene_list=['CCNL1']\n",
      "match=None\n",
      "['CCNL1']:     ENSG00000073111\n",
      "gene_string='CCP110'\n",
      "gene_list=['CCP110']\n",
      "match=None\n",
      "['CCP110']:     ENSG00000103540\n",
      "gene_string='CCT2'\n",
      "gene_list=['CCT2']\n",
      "match=None\n",
      "['CCT2']:     ENSG00000166226\n",
      "gene_string='CCT3'\n",
      "gene_list=['CCT3']\n",
      "match=None\n",
      "['CCT3']:     ENSG00000163468\n",
      "gene_string='CCT4'\n",
      "gene_list=['CCT4']\n",
      "match=None\n",
      "['CCT4']:     ENSG00000115484\n",
      "gene_string='CCT5'\n",
      "gene_list=['CCT5']\n",
      "match=None\n",
      "['CCT5']:     ENSG00000150753\n",
      "gene_string='CCT6A'\n",
      "gene_list=['CCT6A']\n",
      "match=None\n",
      "['CCT6A']:     ENSG00000146731\n",
      "gene_string='CCT7'\n",
      "gene_list=['CCT7']\n",
      "match=None\n",
      "['CCT7']:     ENSG00000135624\n",
      "gene_string='CCT8'\n",
      "gene_list=['CCT8']\n",
      "match=None\n",
      "['CCT8']:     ENSG00000156261\n",
      "gene_string='CD2BP2'\n",
      "gene_list=['CD2BP2']\n",
      "match=None\n",
      "['CD2BP2']:     ENSG00000169217\n",
      "gene_string='CD3EAP'\n",
      "gene_list=['CD3EAP']\n",
      "match=None\n",
      "['CD3EAP']:     ENSG00000117877\n",
      "gene_string='CD8B'\n",
      "gene_list=['CD8B']\n",
      "match=None\n",
      "['CD8B']:     ENSG00000172116\n",
      "gene_string='CDAN1'\n",
      "gene_list=['CDAN1']\n",
      "match=None\n",
      "['CDAN1']:     ENSG00000140326\n",
      "gene_string='CDC5L'\n",
      "gene_list=['CDC5L']\n",
      "match=None\n",
      "['CDC5L']:     ENSG00000096401\n",
      "gene_string='CDC6'\n",
      "gene_list=['CDC6']\n",
      "match=None\n",
      "['CDC6']:     ENSG00000094804\n",
      "gene_string='CDC7'\n",
      "gene_list=['CDC7']\n",
      "match=None\n",
      "['CDC7']:     ENSG00000097046\n",
      "gene_string='CDC16'\n",
      "gene_list=['CDC16']\n",
      "match=None\n",
      "['CDC16']:     ENSG00000130177\n",
      "gene_string='CDC20'\n",
      "gene_list=['CDC20']\n",
      "match=None\n",
      "['CDC20']:     ENSG00000117399\n",
      "gene_string='CDC23'\n",
      "gene_list=['CDC23']\n",
      "match=None\n",
      "['CDC23']:     ENSG00000094880\n",
      "gene_string='CDC26'\n",
      "gene_list=['CDC26']\n",
      "match=None\n",
      "['CDC26']:     ENSG00000176386\n",
      "gene_string='CDC27'\n",
      "gene_list=['CDC27']\n",
      "match=None\n",
      "['CDC27']:     ENSG00000004897\n",
      "gene_string='CDC37'\n",
      "gene_list=['CDC37']\n",
      "match=None\n",
      "['CDC37']:     ENSG00000105401\n",
      "gene_string='CDC40'\n",
      "gene_list=['CDC40']\n",
      "match=None\n",
      "['CDC40']:     ENSG00000168438\n",
      "gene_string='CDC42'\n",
      "gene_list=['CDC42']\n",
      "match=None\n",
      "['CDC42']:     ENSG00000070831\n",
      "gene_string='CDC45'\n",
      "gene_list=['CDC45']\n",
      "match=None\n",
      "['CDC45']:     ENSG00000093009\n",
      "gene_string='CDC73'\n",
      "gene_list=['CDC73']\n",
      "match=None\n",
      "['CDC73']:     ENSG00000134371\n",
      "gene_string='CDC123'\n",
      "gene_list=['CDC123']\n",
      "match=None\n",
      "['CDC123']:     ENSG00000151465\n",
      "gene_string='CDCA5'\n",
      "gene_list=['CDCA5']\n",
      "match=None\n",
      "['CDCA5']:     ENSG00000146670\n",
      "gene_string='CDCA8'\n",
      "gene_list=['CDCA8']\n",
      "match=None\n",
      "['CDCA8']:     ENSG00000134690\n",
      "gene_string='CDIPT'\n",
      "gene_list=['CDIPT']\n",
      "match=None\n",
      "['CDIPT']:     ENSG00000103502\n",
      "gene_string='CDK1'\n",
      "gene_list=['CDK1']\n",
      "match=None\n",
      "['CDK1']:     ENSG00000170312\n",
      "gene_string='CDK2'\n",
      "gene_list=['CDK2']\n",
      "match=None\n",
      "['CDK2']:     ENSG00000123374\n",
      "gene_string='CDK7'\n",
      "gene_list=['CDK7']\n",
      "match=None\n",
      "['CDK7']:     ENSG00000134058\n",
      "gene_string='CDK9'\n",
      "gene_list=['CDK9']\n",
      "match=None\n",
      "['CDK9']:     ENSG00000136807\n",
      "gene_string='CDK11A'\n",
      "gene_list=['CDK11A']\n",
      "match=None\n",
      "['CDK11A']:     ENSG00000008128\n",
      "gene_string='CDT1'\n",
      "gene_list=['CDT1']\n",
      "match=None\n",
      "['CDT1']:     ENSG00000167513\n",
      "gene_string='CEBPZ'\n",
      "gene_list=['CEBPZ']\n",
      "match=None\n",
      "['CEBPZ']:     ENSG00000115816\n",
      "gene_string='CENPA'\n",
      "gene_list=['CENPA']\n",
      "match=None\n",
      "['CENPA']:     ENSG00000115163\n",
      "gene_string='CENPC'\n",
      "gene_list=['CENPC']\n",
      "match=None\n",
      "['CENPC']:     ENSG00000145241\n",
      "gene_string='CENPE'\n",
      "gene_list=['CENPE']\n",
      "match=None\n",
      "['CENPE']:     ENSG00000138778\n",
      "gene_string='CENPH'\n",
      "gene_list=['CENPH']\n",
      "match=None\n",
      "['CENPH']:     ENSG00000153044\n",
      "gene_string='CENPI'\n",
      "gene_list=['CENPI']\n",
      "match=None\n",
      "['CENPI']:     ENSG00000102384\n",
      "gene_string='CENPJ'\n",
      "gene_list=['CENPJ']\n",
      "match=None\n",
      "['CENPJ']:     ENSG00000151849\n",
      "gene_string='CENPK'\n",
      "gene_list=['CENPK']\n",
      "match=None\n",
      "['CENPK']:     ENSG00000123219\n",
      "gene_string='CENPN'\n",
      "gene_list=['CENPN']\n",
      "match=None\n",
      "['CENPN']:     ENSG00000166451\n",
      "gene_string='CENPT'\n",
      "gene_list=['CENPT']\n",
      "match=None\n",
      "['CENPT']:     ENSG00000102901\n",
      "gene_string='CENPW'\n",
      "gene_list=['CENPW']\n",
      "match=None\n",
      "['CENPW']:     ENSG00000203760\n",
      "gene_string='CEP68'\n",
      "gene_list=['CEP68']\n",
      "match=None\n",
      "['CEP68']:     ENSG00000011523\n",
      "gene_string='CEP85'\n",
      "gene_list=['CEP85']\n",
      "match=None\n",
      "['CEP85']:     ENSG00000130695\n",
      "gene_string='CEP97'\n",
      "gene_list=['CEP97']\n",
      "match=None\n",
      "['CEP97']:     ENSG00000182504\n",
      "gene_string='CEP152'\n",
      "gene_list=['CEP152']\n",
      "match=None\n",
      "['CEP152']:     ENSG00000103995\n",
      "gene_string='CEP192'\n",
      "gene_list=['CEP192']\n",
      "match=None\n",
      "['CEP192']:     ENSG00000101639\n",
      "gene_string='CFDP1'\n",
      "gene_list=['CFDP1']\n",
      "match=None\n",
      "['CFDP1']:     ENSG00000153774\n",
      "gene_string='CFL1'\n",
      "gene_list=['CFL1']\n",
      "match=None\n",
      "['CFL1']:     ENSG00000172757\n",
      "gene_string='CHAF1A'\n",
      "gene_list=['CHAF1A']\n",
      "match=None\n",
      "['CHAF1A']:     ENSG00000167670\n",
      "gene_string='CHAF1B'\n",
      "gene_list=['CHAF1B']\n",
      "match=None\n",
      "['CHAF1B']:     ENSG00000159259\n",
      "gene_string='CHCHD1'\n",
      "gene_list=['CHCHD1']\n",
      "match=None\n",
      "['CHCHD1']:     ENSG00000172586\n",
      "gene_string='CHCHD2'\n",
      "gene_list=['CHCHD2']\n",
      "match=None\n",
      "['CHCHD2']:     ENSG00000106153\n",
      "gene_string='CHCHD4'\n",
      "gene_list=['CHCHD4']\n",
      "match=None\n",
      "['CHCHD4']:     ENSG00000163528\n",
      "gene_string='CHD4'\n",
      "gene_list=['CHD4']\n",
      "match=None\n",
      "['CHD4']:     ENSG00000111642\n",
      "gene_string='CHEK1'\n",
      "gene_list=['CHEK1']\n",
      "match=None\n",
      "['CHEK1']:     ENSG00000149554\n",
      "gene_string='CHERP'\n",
      "gene_list=['CHERP']\n",
      "match=None\n",
      "['CHERP']:     ENSG00000085872\n",
      "gene_string='CHMP1A'\n",
      "gene_list=['CHMP1A']\n",
      "match=None\n",
      "['CHMP1A']:     ENSG00000131165\n",
      "gene_string='CHMP2A'\n",
      "gene_list=['CHMP2A']\n",
      "match=None\n",
      "['CHMP2A']:     ENSG00000130724\n",
      "gene_string='CHMP3'\n",
      "gene_list=['CHMP3']\n",
      "match=None\n",
      "['CHMP3']:     ENSG00000115561\n",
      "gene_string='CHMP4B'\n",
      "gene_list=['CHMP4B']\n",
      "match=None\n",
      "['CHMP4B']:     ENSG00000101421\n",
      "gene_string='CHMP5'\n",
      "gene_list=['CHMP5']\n",
      "match=None\n",
      "['CHMP5']:     ENSG00000086065\n",
      "gene_string='CHMP6'\n",
      "gene_list=['CHMP6']\n",
      "match=None\n",
      "['CHMP6']:     ENSG00000176108\n",
      "gene_string='CHMP7'\n",
      "gene_list=['CHMP7']\n",
      "match=None\n",
      "['CHMP7']:     ENSG00000147457\n",
      "gene_string='CHORDC1'\n",
      "gene_list=['CHORDC1']\n",
      "match=None\n",
      "['CHORDC1']:     ENSG00000110172\n",
      "gene_string='CHTF8'\n",
      "gene_list=['CHTF8']\n",
      "match=None\n",
      "['CHTF8']:     ENSG00000286140\n",
      "gene_string='CHTF18'\n",
      "gene_list=['CHTF18']\n",
      "match=None\n",
      "['CHTF18']:     ENSG00000127586\n",
      "gene_string='CHTOP'\n",
      "gene_list=['CHTOP']\n",
      "match=None\n",
      "['CHTOP']:     ENSG00000160679\n",
      "gene_string='CIAO1'\n",
      "gene_list=['CIAO1']\n",
      "match=None\n",
      "['CIAO1']:     ENSG00000144021\n",
      "gene_string='CIAPIN1'\n",
      "gene_list=['CIAPIN1']\n",
      "match=None\n",
      "['CIAPIN1']:     ENSG00000005194\n",
      "gene_string='CINP'\n",
      "gene_list=['CINP']\n",
      "match=None\n",
      "['CINP']:     ENSG00000100865\n",
      "gene_string='CKAP5'\n",
      "gene_list=['CKAP5']\n",
      "match=None\n",
      "['CKAP5']:     ENSG00000175216\n",
      "gene_string='CLASRP'\n",
      "gene_list=['CLASRP']\n",
      "match=None\n",
      "['CLASRP']:     ENSG00000104859\n",
      "gene_string='CLCC1'\n",
      "gene_list=['CLCC1']\n",
      "match=None\n",
      "['CLCC1']:     ENSG00000121940\n",
      "gene_string='CLK2'\n",
      "gene_list=['CLK2']\n",
      "match=None\n",
      "['CLK2']:     ENSG00000176444\n",
      "gene_string='CLNS1A'\n",
      "gene_list=['CLNS1A']\n",
      "match=None\n",
      "['CLNS1A']:     ENSG00000074201\n",
      "gene_string='CLP1'\n",
      "gene_list=['CLP1']\n",
      "match=None\n",
      "['CLP1']:     ENSG00000172409\n",
      "gene_string='CLSPN'\n",
      "gene_list=['CLSPN']\n",
      "match=None\n",
      "['CLSPN']:     ENSG00000092853\n",
      "gene_string='CLTC'\n",
      "gene_list=['CLTC']\n",
      "match=None\n",
      "['CLTC']:     ENSG00000141367\n",
      "gene_string='CMPK1'\n",
      "gene_list=['CMPK1']\n",
      "match=None\n",
      "['CMPK1']:     ENSG00000162368\n",
      "gene_string='CMTR1'\n",
      "gene_list=['CMTR1']\n",
      "match=None\n",
      "['CMTR1']:     ENSG00000137200\n",
      "gene_string='CNIH4'\n",
      "gene_list=['CNIH4']\n",
      "match=None\n",
      "['CNIH4']:     ENSG00000143771\n",
      "gene_string='CNN2'\n",
      "gene_list=['CNN2']\n",
      "match=None\n",
      "['CNN2']:     ENSG00000064666\n",
      "gene_string='CNOT1'\n",
      "gene_list=['CNOT1']\n",
      "match=None\n",
      "['CNOT1']:     ENSG00000125107\n",
      "gene_string='CNOT2'\n",
      "gene_list=['CNOT2']\n",
      "match=None\n",
      "['CNOT2']:     ENSG00000111596\n",
      "gene_string='CNOT3'\n",
      "gene_list=['CNOT3']\n",
      "match=None\n",
      "['CNOT3']:     ENSG00000088038\n",
      "gene_string='COA5'\n",
      "gene_list=['COA5']\n",
      "match=None\n",
      "['COA5']:     ENSG00000183513\n",
      "gene_string='COASY'\n",
      "gene_list=['COASY']\n",
      "match=None\n",
      "['COASY']:     ENSG00000068120\n",
      "gene_string='COG1'\n",
      "gene_list=['COG1']\n",
      "match=None\n",
      "['COG1']:     ENSG00000166685\n",
      "gene_string='COG2'\n",
      "gene_list=['COG2']\n",
      "match=None\n",
      "['COG2']:     ENSG00000135775\n",
      "gene_string='COG3'\n",
      "gene_list=['COG3']\n",
      "match=None\n",
      "['COG3']:     ENSG00000136152\n",
      "gene_string='COG4'\n",
      "gene_list=['COG4']\n",
      "match=None\n",
      "['COG4']:     ENSG00000103051\n",
      "gene_string='COG6'\n",
      "gene_list=['COG6']\n",
      "match=None\n",
      "['COG6']:     ENSG00000133103\n",
      "gene_string='COG8'\n",
      "gene_list=['COG8']\n",
      "match=None\n",
      "['COG8']:     ENSG00000213380\n",
      "gene_string='COPA'\n",
      "gene_list=['COPA']\n",
      "match=None\n",
      "['COPA']:     ENSG00000122218\n",
      "gene_string='COPB1'\n",
      "gene_list=['COPB1']\n",
      "match=None\n",
      "['COPB1']:     ENSG00000129083\n",
      "gene_string='COPB2'\n",
      "gene_list=['COPB2']\n",
      "match=None\n",
      "['COPB2']:     ENSG00000184432\n",
      "gene_string='COPE'\n",
      "gene_list=['COPE']\n",
      "match=None\n",
      "['COPE']:     ENSG00000105669\n",
      "gene_string='COPG1'\n",
      "gene_list=['COPG1']\n",
      "match=None\n",
      "['COPG1']:     ENSG00000181789\n",
      "gene_string='COPS2'\n",
      "gene_list=['COPS2']\n",
      "match=None\n",
      "['COPS2']:     ENSG00000166200\n",
      "gene_string='COPS3'\n",
      "gene_list=['COPS3']\n",
      "match=None\n",
      "['COPS3']:     ENSG00000141030\n",
      "gene_string='COPS4'\n",
      "gene_list=['COPS4']\n",
      "match=None\n",
      "['COPS4']:     ENSG00000138663\n",
      "gene_string='COPS5'\n",
      "gene_list=['COPS5']\n",
      "match=None\n",
      "['COPS5']:     ENSG00000121022\n",
      "gene_string='COPS6'\n",
      "gene_list=['COPS6']\n",
      "match=None\n",
      "['COPS6']:     ENSG00000168090\n",
      "gene_string='COPS8'\n",
      "gene_list=['COPS8']\n",
      "match=None\n",
      "['COPS8']:     ENSG00000198612\n",
      "gene_string='COPZ1'\n",
      "gene_list=['COPZ1']\n",
      "match=None\n",
      "['COPZ1']:     ENSG00000111481\n",
      "gene_string='COQ2'\n",
      "gene_list=['COQ2']\n",
      "match=None\n",
      "['COQ2']:     ENSG00000173085\n",
      "gene_string='COQ4'\n",
      "gene_list=['COQ4']\n",
      "match=None\n",
      "['COQ4']:     ENSG00000167113\n",
      "gene_string='COQ5'\n",
      "gene_list=['COQ5']\n",
      "match=None\n",
      "['COQ5']:     ENSG00000110871\n",
      "gene_string='COX5A'\n",
      "gene_list=['COX5A']\n",
      "match=None\n",
      "['COX5A']:     ENSG00000178741\n",
      "gene_string='COX5B'\n",
      "gene_list=['COX5B']\n",
      "match=None\n",
      "['COX5B']:     ENSG00000135940\n",
      "gene_string='COX6C'\n",
      "gene_list=['COX6C']\n",
      "match=None\n",
      "['COX6C']:     ENSG00000164919\n",
      "gene_string='COX7B'\n",
      "gene_list=['COX7B']\n",
      "match=None\n",
      "['COX7B']:     ENSG00000131174\n",
      "gene_string='COX7C'\n",
      "gene_list=['COX7C']\n",
      "match=None\n",
      "['COX7C']:     ENSG00000127184\n",
      "gene_string='COX10'\n",
      "gene_list=['COX10']\n",
      "match=None\n",
      "['COX10']:     ENSG00000006695\n",
      "gene_string='COX11'\n",
      "gene_list=['COX11']\n",
      "match=None\n",
      "['COX11']:     ENSG00000166260\n",
      "gene_string='COX15'\n",
      "gene_list=['COX15']\n",
      "match=None\n",
      "['COX15']:     ENSG00000014919\n",
      "gene_string='COX17'\n",
      "gene_list=['COX17']\n",
      "match=None\n",
      "['COX17']:     ENSG00000138495\n",
      "gene_string='CPAMD8'\n",
      "gene_list=['CPAMD8']\n",
      "match=None\n",
      "['CPAMD8']:     ENSG00000160111\n",
      "gene_string='CPNE7'\n",
      "gene_list=['CPNE7']\n",
      "match=None\n",
      "['CPNE7']:     ENSG00000178773\n",
      "gene_string='CPOX'\n",
      "gene_list=['CPOX']\n",
      "match=None\n",
      "['CPOX']:     ENSG00000080819\n",
      "gene_string='CPSF1'\n",
      "gene_list=['CPSF1']\n",
      "match=None\n",
      "['CPSF1']:     ENSG00000071894\n",
      "gene_string='CPSF2'\n",
      "gene_list=['CPSF2']\n",
      "match=None\n",
      "['CPSF2']:     ENSG00000165934\n",
      "gene_string='CPSF3'\n",
      "gene_list=['CPSF3']\n",
      "match=None\n",
      "['CPSF3']:     ENSG00000119203\n",
      "gene_string='CPSF4'\n",
      "gene_list=['CPSF4']\n",
      "match=None\n",
      "['CPSF4']:     ENSG00000160917\n",
      "gene_string='CPSF6'\n",
      "gene_list=['CPSF6']\n",
      "match=None\n",
      "['CPSF6']:     ENSG00000111605\n",
      "gene_string='CRCP'\n",
      "gene_list=['CRCP']\n",
      "match=None\n",
      "['CRCP']:     ENSG00000241258\n",
      "gene_string='CRKL'\n",
      "gene_list=['CRKL']\n",
      "match=None\n",
      "['CRKL']:     ENSG00000099942\n",
      "gene_string='CRLS1'\n",
      "gene_list=['CRLS1']\n",
      "match=None\n",
      "['CRLS1']:     ENSG00000088766\n",
      "gene_string='CRNKL1'\n",
      "gene_list=['CRNKL1']\n",
      "match=None\n",
      "['CRNKL1']:     ENSG00000101343\n",
      "gene_string='CS'\n",
      "gene_list=['CS']\n",
      "match=None\n",
      "['CS']:     ENSG00000062485\n",
      "gene_string='CSE1L'\n",
      "gene_list=['CSE1L']\n",
      "match=None\n",
      "['CSE1L']:     ENSG00000124207\n",
      "gene_string='CSH2'\n",
      "gene_list=['CSH2']\n",
      "match=None\n",
      "['CSH2']:     ENSG00000213218\n",
      "gene_string='CSNK1A1'\n",
      "gene_list=['CSNK1A1']\n",
      "match=None\n",
      "['CSNK1A1']:     ENSG00000113712\n",
      "gene_string='CSNK2B'\n",
      "gene_list=['CSNK2B']\n",
      "match=None\n",
      "['CSNK2B']:     ENSG00000263020\n",
      "gene_string='CSTF1'\n",
      "gene_list=['CSTF1']\n",
      "match=None\n",
      "['CSTF1']:     ENSG00000101138\n",
      "gene_string='CSTF3'\n",
      "gene_list=['CSTF3']\n",
      "match=None\n",
      "['CSTF3']:     ENSG00000176102\n",
      "gene_string='CTBP2'\n",
      "gene_list=['CTBP2']\n",
      "match=None\n",
      "['CTBP2']:     ENSG00000175029\n",
      "gene_string='CTCF'\n",
      "gene_list=['CTCF']\n",
      "match=None\n",
      "['CTCF']:     ENSG00000102974\n",
      "gene_string='CTDP1'\n",
      "gene_list=['CTDP1']\n",
      "match=None\n",
      "['CTDP1']:     ENSG00000060069\n",
      "gene_string='CTNNBL1'\n",
      "gene_list=['CTNNBL1']\n",
      "match=None\n",
      "['CTNNBL1']:     ENSG00000132792\n",
      "gene_string='CTPS1'\n",
      "gene_list=['CTPS1']\n",
      "match=None\n",
      "['CTPS1']:     ENSG00000171793\n",
      "gene_string='CTR9'\n",
      "gene_list=['CTR9']\n",
      "match=None\n",
      "['CTR9']:     ENSG00000198730\n",
      "gene_string='CTU2'\n",
      "gene_list=['CTU2']\n",
      "match=None\n",
      "['CTU2']:     ENSG00000174177\n",
      "gene_string='CUL1'\n",
      "gene_list=['CUL1']\n",
      "match=None\n",
      "['CUL1']:     ENSG00000055130\n",
      "gene_string='CUL2'\n",
      "gene_list=['CUL2']\n",
      "match=None\n",
      "['CUL2']:     ENSG00000108094\n",
      "gene_string='CUL3'\n",
      "gene_list=['CUL3']\n",
      "match=None\n",
      "['CUL3']:     ENSG00000036257\n",
      "gene_string='CUL7'\n",
      "gene_list=['CUL7']\n",
      "match=None\n",
      "['CUL7']:     ENSG00000044090\n",
      "gene_string='CWC15'\n",
      "gene_list=['CWC15']\n",
      "match=None\n",
      "['CWC15']:     ENSG00000150316\n",
      "gene_string='CWC22'\n",
      "gene_list=['CWC22']\n",
      "match=None\n",
      "['CWC22']:     ENSG00000163510\n",
      "gene_string='CWC25'\n",
      "gene_list=['CWC25']\n",
      "match=None\n",
      "['CWC25']:     ENSG00000273559\n",
      "gene_string='CWF19L2'\n",
      "gene_list=['CWF19L2']\n",
      "match=None\n",
      "['CWF19L2']:     ENSG00000152404\n",
      "gene_string='CXXC1'\n",
      "gene_list=['CXXC1']\n",
      "match=None\n",
      "['CXXC1']:     ENSG00000154832\n",
      "gene_string='CYC1'\n",
      "gene_list=['CYC1']\n",
      "match=None\n",
      "['CYC1']:     ENSG00000179091\n",
      "gene_string='CYCS'\n",
      "gene_list=['CYCS']\n",
      "match=None\n",
      "['CYCS']:     ENSG00000172115\n",
      "gene_string='CYFIP1'\n",
      "gene_list=['CYFIP1']\n",
      "match=None\n",
      "['CYFIP1']:     ENSG00000273749\n",
      "gene_string='CYP2A13'\n",
      "gene_list=['CYP2A13']\n",
      "match=None\n",
      "['CYP2A13']:     ENSG00000197838\n",
      "gene_string='CYP4F11'\n",
      "gene_list=['CYP4F11']\n",
      "match=None\n",
      "['CYP4F11']:     ENSG00000171903\n",
      "gene_string='CYS1'\n",
      "gene_list=['CYS1']\n",
      "match=None\n",
      "['CYS1']:     ENSG00000205795\n",
      "gene_string='DAD1'\n",
      "gene_list=['DAD1']\n",
      "match=None\n",
      "['DAD1']:     ENSG00000129562\n",
      "gene_string='DAP3'\n",
      "gene_list=['DAP3']\n",
      "match=None\n",
      "['DAP3']:     ENSG00000116544\n",
      "gene_string='DARS'\n",
      "gene_list=['DARS']\n",
      "match=None\n",
      "['DARS']:     ENSG00000115866\n",
      "gene_string='DBF4'\n",
      "gene_list=['DBF4']\n",
      "match=None\n",
      "['DBF4']:     ENSG00000006634\n",
      "gene_string='DBR1'\n",
      "gene_list=['DBR1']\n",
      "match=None\n",
      "['DBR1']:     ENSG00000138231\n",
      "gene_string='DCAF6'\n",
      "gene_list=['DCAF6']\n",
      "match=None\n",
      "['DCAF6']:     ENSG00000143164\n",
      "gene_string='DCLRE1B'\n",
      "gene_list=['DCLRE1B']\n",
      "match=None\n",
      "['DCLRE1B']:     ENSG00000118655\n",
      "gene_string='DCTN1'\n",
      "gene_list=['DCTN1']\n",
      "match=None\n",
      "['DCTN1']:     ENSG00000204843\n",
      "gene_string='DCTN2'\n",
      "gene_list=['DCTN2']\n",
      "match=None\n",
      "['DCTN2']:     ENSG00000175203\n",
      "gene_string='DCTN3'\n",
      "gene_list=['DCTN3']\n",
      "match=None\n",
      "['DCTN3']:     ENSG00000137100\n",
      "gene_string='DCTN4'\n",
      "gene_list=['DCTN4']\n",
      "match=None\n",
      "['DCTN4']:     ENSG00000132912\n",
      "gene_string='DCTN5'\n",
      "gene_list=['DCTN5']\n",
      "match=None\n",
      "['DCTN5']:     ENSG00000166847\n",
      "gene_string='DCTN6'\n",
      "gene_list=['DCTN6']\n",
      "match=None\n",
      "['DCTN6']:     ENSG00000104671\n",
      "gene_string='DCUN1D5'\n",
      "gene_list=['DCUN1D5']\n",
      "match=None\n",
      "['DCUN1D5']:     ENSG00000137692\n",
      "gene_string='DDB1'\n",
      "gene_list=['DDB1']\n",
      "match=None\n",
      "['DDB1']:     ENSG00000167986\n",
      "gene_string='DDN'\n",
      "gene_list=['DDN']\n",
      "match=None\n",
      "['DDN']:     ENSG00000181418\n",
      "gene_string='DDOST'\n",
      "gene_list=['DDOST']\n",
      "match=None\n",
      "['DDOST']:     ENSG00000244038\n",
      "gene_string='DDX1'\n",
      "gene_list=['DDX1']\n",
      "match=None\n",
      "['DDX1']:     ENSG00000079785\n",
      "gene_string='DDX3X'\n",
      "gene_list=['DDX3X']\n",
      "match=None\n",
      "['DDX3X']:     ENSG00000215301\n",
      "gene_string='DDX5'\n",
      "gene_list=['DDX5']\n",
      "match=None\n",
      "['DDX5']:     ENSG00000108654\n",
      "gene_string='DDX6'\n",
      "gene_list=['DDX6']\n",
      "match=None\n",
      "['DDX6']:     ENSG00000110367\n",
      "gene_string='DDX10'\n",
      "gene_list=['DDX10']\n",
      "match=None\n",
      "['DDX10']:     ENSG00000178105\n",
      "gene_string='DDX11'\n",
      "gene_list=['DDX11']\n",
      "match=None\n",
      "['DDX11']:     ENSG00000013573\n",
      "gene_string='DDX18'\n",
      "gene_list=['DDX18']\n",
      "match=None\n",
      "['DDX18']:     ENSG00000088205\n",
      "gene_string='DDX19A'\n",
      "gene_list=['DDX19A']\n",
      "match=None\n",
      "['DDX19A']:     ENSG00000168872\n",
      "gene_string='DDX19B'\n",
      "gene_list=['DDX19B']\n",
      "match=None\n",
      "['DDX19B']:     ENSG00000260537\n",
      "gene_string='DDX20'\n",
      "gene_list=['DDX20']\n",
      "match=None\n",
      "['DDX20']:     ENSG00000064703\n",
      "gene_string='DDX21'\n",
      "gene_list=['DDX21']\n",
      "match=None\n",
      "['DDX21']:     ENSG00000165732\n",
      "gene_string='DDX23'\n",
      "gene_list=['DDX23']\n",
      "match=None\n",
      "['DDX23']:     ENSG00000174243\n",
      "gene_string='DDX24'\n",
      "gene_list=['DDX24']\n",
      "match=None\n",
      "['DDX24']:     ENSG00000089737\n",
      "gene_string='DDX27'\n",
      "gene_list=['DDX27']\n",
      "match=None\n",
      "['DDX27']:     ENSG00000124228\n",
      "gene_string='DDX41'\n",
      "gene_list=['DDX41']\n",
      "match=None\n",
      "['DDX41']:     ENSG00000183258\n",
      "gene_string='DDX42'\n",
      "gene_list=['DDX42']\n",
      "match=None\n",
      "['DDX42']:     ENSG00000198231\n",
      "gene_string='DDX46'\n",
      "gene_list=['DDX46']\n",
      "match=None\n",
      "['DDX46']:     ENSG00000145833\n",
      "gene_string='DDX47'\n",
      "gene_list=['DDX47']\n",
      "match=None\n",
      "['DDX47']:     ENSG00000213782\n",
      "gene_string='DDX49'\n",
      "gene_list=['DDX49']\n",
      "match=None\n",
      "['DDX49']:     ENSG00000105671\n",
      "gene_string='DDX51'\n",
      "gene_list=['DDX51']\n",
      "match=None\n",
      "['DDX51']:     ENSG00000185163\n",
      "gene_string='DDX52'\n",
      "gene_list=['DDX52']\n",
      "match=None\n",
      "['DDX52']:     ENSG00000278053\n",
      "gene_string='DDX54'\n",
      "gene_list=['DDX54']\n",
      "match=None\n",
      "['DDX54']:     ENSG00000123064\n",
      "gene_string='DDX55'\n",
      "gene_list=['DDX55']\n",
      "match=None\n",
      "['DDX55']:     ENSG00000111364\n",
      "gene_string='DDX56'\n",
      "gene_list=['DDX56']\n",
      "match=None\n",
      "['DDX56']:     ENSG00000136271\n",
      "gene_string='DDX59'\n",
      "gene_list=['DDX59']\n",
      "match=None\n",
      "['DDX59']:     ENSG00000118197\n",
      "gene_string='DENR'\n",
      "gene_list=['DENR']\n",
      "match=None\n",
      "['DENR']:     ENSG00000139726\n",
      "gene_string='DERL2'\n",
      "gene_list=['DERL2']\n",
      "match=None\n",
      "['DERL2']:     ENSG00000072849\n",
      "gene_string='DHDDS'\n",
      "gene_list=['DHDDS']\n",
      "match=None\n",
      "['DHDDS']:     ENSG00000117682\n",
      "gene_string='DHFR'\n",
      "gene_list=['DHFR']\n",
      "match=None\n",
      "['DHFR']:     ENSG00000228716\n",
      "gene_string='DHODH'\n",
      "gene_list=['DHODH']\n",
      "match=None\n",
      "['DHODH']:     ENSG00000102967\n",
      "gene_string='DHPS'\n",
      "gene_list=['DHPS']\n",
      "match=None\n",
      "['DHPS']:     ENSG00000095059\n",
      "gene_string='DHX8'\n",
      "gene_list=['DHX8']\n",
      "match=None\n",
      "['DHX8']:     ENSG00000067596\n",
      "gene_string='DHX9'\n",
      "gene_list=['DHX9']\n",
      "match=None\n",
      "['DHX9']:     ENSG00000135829\n",
      "gene_string='DHX15'\n",
      "gene_list=['DHX15']\n",
      "match=None\n",
      "['DHX15']:     ENSG00000109606\n",
      "gene_string='DHX16'\n",
      "gene_list=['DHX16']\n",
      "match=None\n",
      "['DHX16']:     ENSG00000204560\n",
      "gene_string='DHX33'\n",
      "gene_list=['DHX33']\n",
      "match=None\n",
      "['DHX33']:     ENSG00000005100\n",
      "gene_string='DHX36'\n",
      "gene_list=['DHX36']\n",
      "match=None\n",
      "['DHX36']:     ENSG00000174953\n",
      "gene_string='DHX37'\n",
      "gene_list=['DHX37']\n",
      "match=None\n",
      "['DHX37']:     ENSG00000150990\n",
      "gene_string='DICER1'\n",
      "gene_list=['DICER1']\n",
      "match=None\n",
      "['DICER1']:     ENSG00000100697\n",
      "gene_string='DIMT1'\n",
      "gene_list=['DIMT1']\n",
      "match=None\n",
      "['DIMT1']:     ENSG00000086189\n",
      "gene_string='DIS3'\n",
      "gene_list=['DIS3']\n",
      "match=None\n",
      "['DIS3']:     ENSG00000083520\n",
      "gene_string='DKC1'\n",
      "gene_list=['DKC1']\n",
      "match=None\n",
      "['DKC1']:     ENSG00000130826\n",
      "gene_string='DLD'\n",
      "gene_list=['DLD']\n",
      "match=None\n",
      "['DLD']:     ENSG00000091140\n",
      "gene_string='DMAP1'\n",
      "gene_list=['DMAP1']\n",
      "match=None\n",
      "['DMAP1']:     ENSG00000178028\n",
      "gene_string='DMRTA2'\n",
      "gene_list=['DMRTA2']\n",
      "match=None\n",
      "['DMRTA2']:     ENSG00000142700\n",
      "gene_string='DNAJA1'\n",
      "gene_list=['DNAJA1']\n",
      "match=None\n",
      "['DNAJA1']:     ENSG00000086061\n",
      "gene_string='DNAJA3'\n",
      "gene_list=['DNAJA3']\n",
      "match=None\n",
      "['DNAJA3']:     ENSG00000103423\n",
      "gene_string='DNAJC8'\n",
      "gene_list=['DNAJC8']\n",
      "match=None\n",
      "['DNAJC8']:     ENSG00000126698\n",
      "gene_string='DNAJC9'\n",
      "gene_list=['DNAJC9']\n",
      "match=None\n",
      "['DNAJC9']:     ENSG00000213551\n",
      "gene_string='DNAJC11'\n",
      "gene_list=['DNAJC11']\n",
      "match=None\n",
      "['DNAJC11']:     ENSG00000007923\n",
      "gene_string='DNAJC17'\n",
      "gene_list=['DNAJC17']\n",
      "match=None\n",
      "['DNAJC17']:     ENSG00000104129\n",
      "gene_string='DNAJC19'\n",
      "gene_list=['DNAJC19']\n",
      "match=None\n",
      "['DNAJC19']:     ENSG00000205981\n",
      "gene_string='DNLZ'\n",
      "gene_list=['DNLZ']\n",
      "match=None\n",
      "['DNLZ']:     ENSG00000213221\n",
      "gene_string='DNM1'\n",
      "gene_list=['DNM1']\n",
      "match=None\n",
      "['DNM1']:     ENSG00000106976\n",
      "gene_string='DNM1L'\n",
      "gene_list=['DNM1L']\n",
      "match=None\n",
      "['DNM1L']:     ENSG00000087470\n",
      "gene_string='DNM2'\n",
      "gene_list=['DNM2']\n",
      "match=None\n",
      "['DNM2']:     ENSG00000079805\n",
      "gene_string='DNMT1'\n",
      "gene_list=['DNMT1']\n",
      "match=None\n",
      "['DNMT1']:     ENSG00000130816\n",
      "gene_string='DNTTIP2'\n",
      "gene_list=['DNTTIP2']\n",
      "match=None\n",
      "['DNTTIP2']:     ENSG00000067334\n",
      "gene_string='DOHH'\n",
      "gene_list=['DOHH']\n",
      "match=None\n",
      "['DOHH']:     ENSG00000129932\n",
      "gene_string='DOLK'\n",
      "gene_list=['DOLK']\n",
      "match=None\n",
      "['DOLK']:     ENSG00000175283\n",
      "gene_string='DONSON'\n",
      "gene_list=['DONSON']\n",
      "match=None\n",
      "['DONSON']:     ENSG00000159147\n",
      "gene_string='DPAGT1'\n",
      "gene_list=['DPAGT1']\n",
      "match=None\n",
      "['DPAGT1']:     ENSG00000172269\n",
      "gene_string='DPH1'\n",
      "gene_list=['DPH1']\n",
      "match=None\n",
      "['DPH1']:     ENSG00000108963\n",
      "gene_string='DPH2'\n",
      "gene_list=['DPH2']\n",
      "match=None\n",
      "['DPH2']:     ENSG00000132768\n",
      "gene_string='DPH3'\n",
      "gene_list=['DPH3']\n",
      "match=None\n",
      "['DPH3']:     ENSG00000154813\n",
      "gene_string='DPH6'\n",
      "gene_list=['DPH6']\n",
      "match=None\n",
      "['DPH6']:     ENSG00000134146\n",
      "gene_string='DPM2'\n",
      "gene_list=['DPM2']\n",
      "match=None\n",
      "['DPM2']:     ENSG00000136908\n",
      "gene_string='DPPA2'\n",
      "gene_list=['DPPA2']\n",
      "match=None\n",
      "['DPPA2']:     ENSG00000163530\n",
      "gene_string='DPY19L2'\n",
      "gene_list=['DPY19L2']\n",
      "match=None\n",
      "['DPY19L2']:     ENSG00000177990\n",
      "gene_string='DR1'\n",
      "gene_list=['DR1']\n",
      "match=None\n",
      "['DR1']:     ENSG00000117505\n",
      "gene_string='DRAP1'\n",
      "gene_list=['DRAP1']\n",
      "match=None\n",
      "['DRAP1']:     ENSG00000175550\n",
      "gene_string='DRG1'\n",
      "gene_list=['DRG1']\n",
      "match=None\n",
      "['DRG1']:     ENSG00000185721\n",
      "gene_string='DSN1'\n",
      "gene_list=['DSN1']\n",
      "match=None\n",
      "['DSN1']:     ENSG00000149636\n",
      "gene_string='DTL'\n",
      "gene_list=['DTL']\n",
      "match=None\n",
      "['DTL']:     ENSG00000143476\n",
      "gene_string='DTYMK'\n",
      "gene_list=['DTYMK']\n",
      "match=None\n",
      "['DTYMK']:     ENSG00000168393\n",
      "gene_string='DUT'\n",
      "gene_list=['DUT']\n",
      "match=None\n",
      "['DUT']:     ENSG00000128951\n",
      "gene_string='DYNC1H1'\n",
      "gene_list=['DYNC1H1']\n",
      "match=None\n",
      "['DYNC1H1']:     ENSG00000197102\n",
      "gene_string='DYNC1I2'\n",
      "gene_list=['DYNC1I2']\n",
      "match=None\n",
      "['DYNC1I2']:     ENSG00000077380\n",
      "gene_string='DYNLL1'\n",
      "gene_list=['DYNLL1']\n",
      "match=None\n",
      "['DYNLL1']:     ENSG00000088986\n",
      "gene_string='DYNLRB1'\n",
      "gene_list=['DYNLRB1']\n",
      "match=None\n",
      "['DYNLRB1']:     ENSG00000125971\n",
      "gene_string='E2F6'\n",
      "gene_list=['E2F6']\n",
      "match=None\n",
      "['E2F6']:     ENSG00000169016\n",
      "gene_string='E4F1'\n",
      "gene_list=['E4F1']\n",
      "match=None\n",
      "['E4F1']:     ENSG00000167967\n",
      "gene_string='EARS2'\n",
      "gene_list=['EARS2']\n",
      "match=None\n",
      "['EARS2']:     ENSG00000103356\n",
      "gene_string='EBNA1BP2'\n",
      "gene_list=['EBNA1BP2']\n",
      "match=None\n",
      "['EBNA1BP2']:     ENSG00000117395\n",
      "gene_string='ECD'\n",
      "gene_list=['ECD']\n",
      "match=None\n",
      "['ECD']:     ENSG00000127922\n",
      "gene_string='ECT2'\n",
      "gene_list=['ECT2']\n",
      "match=None\n",
      "['ECT2']:     ENSG00000114346\n",
      "gene_string='EDC4'\n",
      "gene_list=['EDC4']\n",
      "match=None\n",
      "['EDC4']:     ENSG00000038358\n",
      "gene_string='EEF1A1'\n",
      "gene_list=['EEF1A1']\n",
      "match=None\n",
      "['EEF1A1']:     ENSG00000156508\n",
      "gene_string='EEF1B2'\n",
      "gene_list=['EEF1B2']\n",
      "match=None\n",
      "['EEF1B2']:     ENSG00000114942\n",
      "gene_string='EEF1G'\n",
      "gene_list=['EEF1G']\n",
      "match=None\n",
      "['EEF1G']:     ENSG00000254772\n",
      "gene_string='EEF2'\n",
      "gene_list=['EEF2']\n",
      "match=None\n",
      "['EEF2']:     ENSG00000167658\n",
      "gene_string='EFTUD2'\n",
      "gene_list=['EFTUD2']\n",
      "match=None\n",
      "['EFTUD2']:     ENSG00000108883\n",
      "gene_string='EGLN2'\n",
      "gene_list=['EGLN2']\n",
      "match=None\n",
      "['EGLN2']:     ENSG00000269858\n",
      "gene_string='EIF1'\n",
      "gene_list=['EIF1']\n",
      "match=None\n",
      "['EIF1']:     ENSG00000173812\n",
      "gene_string='EIF1AD'\n",
      "gene_list=['EIF1AD']\n",
      "match=None\n",
      "['EIF1AD']:     ENSG00000175376\n",
      "gene_string='EIF1AX'\n",
      "gene_list=['EIF1AX']\n",
      "match=None\n",
      "['EIF1AX']:     ENSG00000173674\n",
      "gene_string='EIF2B1'\n",
      "gene_list=['EIF2B1']\n",
      "match=None\n",
      "['EIF2B1']:     ENSG00000111361\n",
      "gene_string='EIF2B2'\n",
      "gene_list=['EIF2B2']\n",
      "match=None\n",
      "['EIF2B2']:     ENSG00000119718\n",
      "gene_string='EIF2B3'\n",
      "gene_list=['EIF2B3']\n",
      "match=None\n",
      "['EIF2B3']:     ENSG00000070785\n",
      "gene_string='EIF2B4'\n",
      "gene_list=['EIF2B4']\n",
      "match=None\n",
      "['EIF2B4']:     ENSG00000115211\n",
      "gene_string='EIF2B5'\n",
      "gene_list=['EIF2B5']\n",
      "match=None\n",
      "['EIF2B5']:     ENSG00000145191\n",
      "gene_string='EIF2S1'\n",
      "gene_list=['EIF2S1']\n",
      "match=None\n",
      "['EIF2S1']:     ENSG00000134001\n",
      "gene_string='EIF2S2'\n",
      "gene_list=['EIF2S2']\n",
      "match=None\n",
      "['EIF2S2']:     ENSG00000125977\n",
      "gene_string='EIF2S3'\n",
      "gene_list=['EIF2S3']\n",
      "match=None\n",
      "['EIF2S3']:     ENSG00000130741\n",
      "gene_string='EIF3A'\n",
      "gene_list=['EIF3A']\n",
      "match=None\n",
      "['EIF3A']:     ENSG00000107581\n",
      "gene_string='EIF3B'\n",
      "gene_list=['EIF3B']\n",
      "match=None\n",
      "['EIF3B']:     ENSG00000106263\n",
      "gene_string='EIF3CL'\n",
      "gene_list=['EIF3CL']\n",
      "match=None\n",
      "['EIF3CL']:     ENSG00000184110\n",
      "gene_string='EIF3D'\n",
      "gene_list=['EIF3D']\n",
      "match=None\n",
      "['EIF3D']:     ENSG00000100353\n",
      "gene_string='EIF3E'\n",
      "gene_list=['EIF3E']\n",
      "match=None\n",
      "['EIF3E']:     ENSG00000104408\n",
      "gene_string='EIF3F'\n",
      "gene_list=['EIF3F']\n",
      "match=None\n",
      "['EIF3F']:     ENSG00000175390\n",
      "gene_string='EIF3G'\n",
      "gene_list=['EIF3G']\n",
      "match=None\n",
      "['EIF3G']:     ENSG00000130811\n",
      "gene_string='EIF3H'\n",
      "gene_list=['EIF3H']\n",
      "match=None\n",
      "['EIF3H']:     ENSG00000147677\n",
      "gene_string='EIF3I'\n",
      "gene_list=['EIF3I']\n",
      "match=None\n",
      "['EIF3I']:     ENSG00000084623\n",
      "gene_string='EIF3J'\n",
      "gene_list=['EIF3J']\n",
      "match=None\n",
      "['EIF3J']:     ENSG00000104131\n",
      "gene_string='EIF3L'\n",
      "gene_list=['EIF3L']\n",
      "match=None\n",
      "['EIF3L']:     ENSG00000100129\n",
      "gene_string='EIF3M'\n",
      "gene_list=['EIF3M']\n",
      "match=None\n",
      "['EIF3M']:     ENSG00000149100\n",
      "gene_string='EIF4A1'\n",
      "gene_list=['EIF4A1']\n",
      "match=None\n",
      "['EIF4A1']:     ENSG00000161960\n",
      "gene_string='EIF4A3'\n",
      "gene_list=['EIF4A3']\n",
      "match=None\n",
      "['EIF4A3']:     ENSG00000141543\n",
      "gene_string='EIF4B'\n",
      "gene_list=['EIF4B']\n",
      "match=None\n",
      "['EIF4B']:     ENSG00000063046\n",
      "gene_string='EIF4E'\n",
      "gene_list=['EIF4E']\n",
      "match=None\n",
      "['EIF4E']:     ENSG00000151247\n",
      "gene_string='EIF4G1'\n",
      "gene_list=['EIF4G1']\n",
      "match=None\n",
      "['EIF4G1']:     ENSG00000114867\n",
      "gene_string='EIF4G2'\n",
      "gene_list=['EIF4G2']\n",
      "match=None\n",
      "['EIF4G2']:     ENSG00000110321\n",
      "gene_string='EIF5'\n",
      "gene_list=['EIF5']\n",
      "match=None\n",
      "['EIF5']:     ENSG00000100664\n",
      "gene_string='EIF5A'\n",
      "gene_list=['EIF5A']\n",
      "match=None\n",
      "['EIF5A']:     ENSG00000132507\n",
      "gene_string='EIF6'\n",
      "gene_list=['EIF6']\n",
      "match=None\n",
      "['EIF6']:     ENSG00000242372\n",
      "gene_string='ELAC2'\n",
      "gene_list=['ELAC2']\n",
      "match=None\n",
      "['ELAC2']:     ENSG00000006744\n",
      "gene_string='ELL'\n",
      "gene_list=['ELL']\n",
      "match=None\n",
      "['ELL']:     ENSG00000105656\n",
      "gene_string='ELOVL1'\n",
      "gene_list=['ELOVL1']\n",
      "match=None\n",
      "['ELOVL1']:     ENSG00000066322\n",
      "gene_string='ELP2'\n",
      "gene_list=['ELP2']\n",
      "match=None\n",
      "['ELP2']:     ENSG00000134759\n",
      "gene_string='ELP3'\n",
      "gene_list=['ELP3']\n",
      "match=None\n",
      "['ELP3']:     ENSG00000134014\n",
      "gene_string='ELP5'\n",
      "gene_list=['ELP5']\n",
      "match=None\n",
      "['ELP5']:     ENSG00000170291\n",
      "gene_string='ELP6'\n",
      "gene_list=['ELP6']\n",
      "match=None\n",
      "['ELP6']:     ENSG00000163832\n",
      "gene_string='EMC1'\n",
      "gene_list=['EMC1']\n",
      "match=None\n",
      "['EMC1']:     ENSG00000127463\n",
      "gene_string='EMC3'\n",
      "gene_list=['EMC3']\n",
      "match=None\n",
      "['EMC3']:     ENSG00000125037\n",
      "gene_string='EMC4'\n",
      "gene_list=['EMC4']\n",
      "match=None\n",
      "['EMC4']:     ENSG00000128463\n",
      "gene_string='EMC7'\n",
      "gene_list=['EMC7']\n",
      "match=None\n",
      "['EMC7']:     ENSG00000134153\n",
      "gene_string='ENO1'\n",
      "gene_list=['ENO1']\n",
      "match=None\n",
      "['ENO1']:     ENSG00000074800\n",
      "gene_string='EP400'\n",
      "gene_list=['EP400']\n",
      "match=None\n",
      "['EP400']:     ENSG00000183495\n",
      "gene_string='EPRS'\n",
      "gene_list=['EPRS']\n",
      "match=None\n",
      "['EPRS']:     ENSG00000136628\n",
      "gene_string='ERAL1'\n",
      "gene_list=['ERAL1']\n",
      "match=None\n",
      "['ERAL1']:     ENSG00000132591\n",
      "gene_string='ERCC2'\n",
      "gene_list=['ERCC2']\n",
      "match=None\n",
      "['ERCC2']:     ENSG00000104884\n",
      "gene_string='ERCC3'\n",
      "gene_list=['ERCC3']\n",
      "match=None\n",
      "['ERCC3']:     ENSG00000163161\n",
      "gene_string='ERH'\n",
      "gene_list=['ERH']\n",
      "match=None\n",
      "['ERH']:     ENSG00000100632\n",
      "gene_string='ERVW-1'\n",
      "gene_list=['ERVW-1']\n",
      "match=None\n",
      "['ERVW-1']:     ENSG00000242950\n",
      "gene_string='ESF1'\n",
      "gene_list=['ESF1']\n",
      "match=None\n",
      "['ESF1']:     ENSG00000089048\n",
      "gene_string='ESPL1'\n",
      "gene_list=['ESPL1']\n",
      "match=None\n",
      "['ESPL1']:     ENSG00000135476\n",
      "gene_string='ESPN'\n",
      "gene_list=['ESPN']\n",
      "match=None\n",
      "['ESPN']:     ENSG00000187017\n",
      "gene_string='ETF1'\n",
      "gene_list=['ETF1']\n",
      "match=None\n",
      "['ETF1']:     ENSG00000120705\n",
      "gene_string='EWSR1'\n",
      "gene_list=['EWSR1']\n",
      "match=None\n",
      "['EWSR1']:     ENSG00000182944\n",
      "gene_string='EXOC1'\n",
      "gene_list=['EXOC1']\n",
      "match=None\n",
      "['EXOC1']:     ENSG00000090989\n",
      "gene_string='EXOC2'\n",
      "gene_list=['EXOC2']\n",
      "match=None\n",
      "['EXOC2']:     ENSG00000112685\n",
      "gene_string='EXOC3'\n",
      "gene_list=['EXOC3']\n",
      "match=None\n",
      "['EXOC3']:     ENSG00000180104\n",
      "gene_string='EXOC4'\n",
      "gene_list=['EXOC4']\n",
      "match=None\n",
      "['EXOC4']:     ENSG00000131558\n",
      "gene_string='EXOC5'\n",
      "gene_list=['EXOC5']\n",
      "match=None\n",
      "['EXOC5']:     ENSG00000070367\n",
      "gene_string='EXOC7'\n",
      "gene_list=['EXOC7']\n",
      "match=None\n",
      "['EXOC7']:     ENSG00000182473\n",
      "gene_string='EXOC8'\n",
      "gene_list=['EXOC8']\n",
      "match=None\n",
      "['EXOC8']:     ENSG00000116903\n",
      "gene_string='EXOSC1'\n",
      "gene_list=['EXOSC1']\n",
      "match=None\n",
      "['EXOSC1']:     ENSG00000171311\n",
      "gene_string='EXOSC2'\n",
      "gene_list=['EXOSC2']\n",
      "match=None\n",
      "['EXOSC2']:     ENSG00000130713\n",
      "gene_string='EXOSC3'\n",
      "gene_list=['EXOSC3']\n",
      "match=None\n",
      "['EXOSC3']:     ENSG00000107371\n",
      "gene_string='EXOSC4'\n",
      "gene_list=['EXOSC4']\n",
      "match=None\n",
      "['EXOSC4']:     ENSG00000178896\n",
      "gene_string='EXOSC5'\n",
      "gene_list=['EXOSC5']\n",
      "match=None\n",
      "['EXOSC5']:     ENSG00000077348\n",
      "gene_string='EXOSC6'\n",
      "gene_list=['EXOSC6']\n",
      "match=None\n",
      "['EXOSC6']:     ENSG00000223496\n",
      "gene_string='EXOSC7'\n",
      "gene_list=['EXOSC7']\n",
      "match=None\n",
      "['EXOSC7']:     ENSG00000075914\n",
      "gene_string='EXOSC8'\n",
      "gene_list=['EXOSC8']\n",
      "match=None\n",
      "['EXOSC8']:     ENSG00000120699\n",
      "gene_string='EXOSC9'\n",
      "gene_list=['EXOSC9']\n",
      "match=None\n",
      "['EXOSC9']:     ENSG00000123737\n",
      "gene_string='EXOSC10'\n",
      "gene_list=['EXOSC10']\n",
      "match=None\n",
      "['EXOSC10']:     ENSG00000171824\n",
      "gene_string='F8A1'\n",
      "gene_list=['F8A1']\n",
      "match=None\n",
      "['F8A1']:     ENSG00000277150\n",
      "gene_string='FAF2'\n",
      "gene_list=['FAF2']\n",
      "match=None\n",
      "['FAF2']:     ENSG00000113194\n",
      "gene_string='FAM32A'\n",
      "gene_list=['FAM32A']\n",
      "match=None\n",
      "['FAM32A']:     ENSG00000105058\n",
      "gene_string='FAM50A'\n",
      "gene_list=['FAM50A']\n",
      "match=None\n",
      "['FAM50A']:     ENSG00000071859\n",
      "gene_string='FAM72D'\n",
      "gene_list=['FAM72D']\n",
      "match=None\n",
      "['FAM72D']:     ENSG00000263513\n",
      "gene_string='FAM133B'\n",
      "gene_list=['FAM133B']\n",
      "match=None\n",
      "['FAM133B']:     ENSG00000234545\n",
      "gene_string='FAM136A'\n",
      "gene_list=['FAM136A']\n",
      "match=None\n",
      "['FAM136A']:     ENSG00000035141\n",
      "gene_string='FAM207A'\n",
      "gene_list=['FAM207A']\n",
      "match=None\n",
      "['FAM207A']:     ENSG00000160256\n",
      "gene_string='FAM229A'\n",
      "gene_list=['FAM229A']\n",
      "match=None\n",
      "['FAM229A']:     ENSG00000225828\n",
      "gene_string='FARS2'\n",
      "gene_list=['FARS2']\n",
      "match=None\n",
      "['FARS2']:     ENSG00000145982\n",
      "gene_string='FARSA'\n",
      "gene_list=['FARSA']\n",
      "match=None\n",
      "['FARSA']:     ENSG00000179115\n",
      "gene_string='FARSB'\n",
      "gene_list=['FARSB']\n",
      "match=None\n",
      "['FARSB']:     ENSG00000116120\n",
      "gene_string='FASTKD5'\n",
      "gene_list=['FASTKD5']\n",
      "match=None\n",
      "['FASTKD5']:     ENSG00000215251\n",
      "gene_string='FAU'\n",
      "gene_list=['FAU']\n",
      "match=None\n",
      "['FAU']:     ENSG00000149806\n",
      "gene_string='FBL'\n",
      "gene_list=['FBL']\n",
      "match=None\n",
      "['FBL']:     ENSG00000105202\n",
      "gene_string='FBLIM1'\n",
      "gene_list=['FBLIM1']\n",
      "match=None\n",
      "['FBLIM1']:     ENSG00000162458\n",
      "gene_string='FBXO5'\n",
      "gene_list=['FBXO5']\n",
      "match=None\n",
      "['FBXO5']:     ENSG00000112029\n",
      "gene_string='FCF1'\n",
      "gene_list=['FCF1']\n",
      "match=None\n",
      "['FCF1']:     ENSG00000119616\n",
      "gene_string='FDPS'\n",
      "gene_list=['FDPS']\n",
      "match=None\n",
      "['FDPS']:     ENSG00000160752\n",
      "gene_string='FDXR'\n",
      "gene_list=['FDXR']\n",
      "match=None\n",
      "['FDXR']:     ENSG00000161513\n",
      "gene_string='FEN1'\n",
      "gene_list=['FEN1']\n",
      "match=None\n",
      "['FEN1']:     ENSG00000168496\n",
      "gene_string='FGFR1OP'\n",
      "gene_list=['FGFR1OP']\n",
      "match=None\n",
      "['FGFR1OP']:     ENSG00000213066\n",
      "gene_string='FIP1L1'\n",
      "gene_list=['FIP1L1']\n",
      "match=None\n",
      "['FIP1L1']:     ENSG00000145216\n",
      "gene_string='FNBP4'\n",
      "gene_list=['FNBP4']\n",
      "match=None\n",
      "['FNBP4']:     ENSG00000109920\n",
      "gene_string='FNTA'\n",
      "gene_list=['FNTA']\n",
      "match=None\n",
      "['FNTA']:     ENSG00000168522\n",
      "gene_string='FNTB'\n",
      "gene_list=['FNTB']\n",
      "match=None\n",
      "['FNTB']:     ENSG00000257365\n",
      "gene_string='FOLR3'\n",
      "gene_list=['FOLR3']\n",
      "match=None\n",
      "['FOLR3']:     ENSG00000110203\n",
      "gene_string='FOXD4'\n",
      "gene_list=['FOXD4']\n",
      "match=None\n",
      "['FOXD4']:     ENSG00000170122\n",
      "gene_string='FOXL2'\n",
      "gene_list=['FOXL2']\n",
      "match=None\n",
      "['FOXL2']:     ENSG00000183770\n",
      "gene_string='FRG2'\n",
      "gene_list=['FRG2']\n",
      "match=None\n",
      "['FRG2']:     ENSG00000205097\n",
      "gene_string='FTSJ3'\n",
      "gene_list=['FTSJ3']\n",
      "match=None\n",
      "['FTSJ3']:     ENSG00000108592\n",
      "gene_string='FXN'\n",
      "gene_list=['FXN']\n",
      "match=None\n",
      "['FXN']:     ENSG00000285130\n",
      "gene_string='GAB2'\n",
      "gene_list=['GAB2']\n",
      "match=None\n",
      "['GAB2']:     ENSG00000033327\n",
      "gene_string='GABPA'\n",
      "gene_list=['GABPA']\n",
      "match=None\n",
      "['GABPA']:     ENSG00000154727\n",
      "gene_string='GAK'\n",
      "gene_list=['GAK']\n",
      "match=None\n",
      "['GAK']:     ENSG00000178950\n",
      "gene_string='GAPDH'\n",
      "gene_list=['GAPDH']\n",
      "match=None\n",
      "['GAPDH']:     ENSG00000111640\n",
      "gene_string='GAR1'\n",
      "gene_list=['GAR1']\n",
      "match=None\n",
      "['GAR1']:     ENSG00000109534\n",
      "gene_string='GARS'\n",
      "gene_list=['GARS']\n",
      "match=None\n",
      "['GARS']:     ENSG00000106105\n",
      "gene_string='GATA1'\n",
      "gene_list=['GATA1']\n",
      "match=None\n",
      "['GATA1']:     ENSG00000102145\n",
      "gene_string='GBF1'\n",
      "gene_list=['GBF1']\n",
      "match=None\n",
      "['GBF1']:     ENSG00000107862\n",
      "gene_string='GCOM1'\n",
      "gene_list=['GCOM1']\n",
      "match=None\n",
      "['GCOM1']:     ENSG00000137878\n",
      "gene_string='GEMIN4'\n",
      "gene_list=['GEMIN4']\n",
      "match=None\n",
      "['GEMIN4']:     ENSG00000179409\n",
      "gene_string='GEMIN5'\n",
      "gene_list=['GEMIN5']\n",
      "match=None\n",
      "['GEMIN5']:     ENSG00000082516\n",
      "gene_string='GEMIN6'\n",
      "gene_list=['GEMIN6']\n",
      "match=None\n",
      "['GEMIN6']:     ENSG00000152147\n",
      "gene_string='GEMIN7'\n",
      "gene_list=['GEMIN7']\n",
      "match=None\n",
      "['GEMIN7']:     ENSG00000142252\n",
      "gene_string='GEMIN8'\n",
      "gene_list=['GEMIN8']\n",
      "match=None\n",
      "['GEMIN8']:     ENSG00000046647\n",
      "gene_string='GET1'\n",
      "gene_list=['GET1']\n",
      "match=None\n",
      "['GET1']:     ENSG00000182093\n",
      "gene_string='GET3'\n",
      "gene_list=['GET3']\n",
      "match=None\n",
      "['GET3']:     ENSG00000198356\n",
      "gene_string='GFER'\n",
      "gene_list=['GFER']\n",
      "match=None\n",
      "['GFER']:     ENSG00000127554\n",
      "gene_string='GFM1'\n",
      "gene_list=['GFM1']\n",
      "match=None\n",
      "['GFM1']:     ENSG00000168827\n",
      "gene_string='GGPS1'\n",
      "gene_list=['GGPS1']\n",
      "match=None\n",
      "['GGPS1']:     ENSG00000152904\n",
      "gene_string='GINS1'\n",
      "gene_list=['GINS1']\n",
      "match=None\n",
      "['GINS1']:     ENSG00000101003\n",
      "gene_string='GINS2'\n",
      "gene_list=['GINS2']\n",
      "match=None\n",
      "['GINS2']:     ENSG00000131153\n",
      "gene_string='GINS3'\n",
      "gene_list=['GINS3']\n",
      "match=None\n",
      "['GINS3']:     ENSG00000181938\n",
      "gene_string='GINS4'\n",
      "gene_list=['GINS4']\n",
      "match=None\n",
      "['GINS4']:     ENSG00000147536\n",
      "gene_string='GIT2'\n",
      "gene_list=['GIT2']\n",
      "match=None\n",
      "['GIT2']:     ENSG00000139436\n",
      "gene_string='GJA3'\n",
      "gene_list=['GJA3']\n",
      "match=None\n",
      "['GJA3']:     ENSG00000121743\n",
      "gene_string='GLE1'\n",
      "gene_list=['GLE1']\n",
      "match=None\n",
      "['GLE1']:     ENSG00000119392\n",
      "gene_string='GLI4'\n",
      "gene_list=['GLI4']\n",
      "match=None\n",
      "['GLI4']:     ENSG00000250571\n",
      "gene_string='GLMN'\n",
      "gene_list=['GLMN']\n",
      "match=None\n",
      "['GLMN']:     ENSG00000174842\n",
      "gene_string='GLRX3'\n",
      "gene_list=['GLRX3']\n",
      "match=None\n",
      "['GLRX3']:     ENSG00000108010\n",
      "gene_string='GLRX5'\n",
      "gene_list=['GLRX5']\n",
      "match=None\n",
      "['GLRX5']:     ENSG00000182512\n",
      "gene_string='GMPPB'\n",
      "gene_list=['GMPPB']\n",
      "match=None\n",
      "['GMPPB']:     ENSG00000173540\n",
      "gene_string='GMPS'\n",
      "gene_list=['GMPS']\n",
      "match=None\n",
      "['GMPS']:     ENSG00000163655\n",
      "gene_string='GNB1L'\n",
      "gene_list=['GNB1L']\n",
      "match=None\n",
      "['GNB1L']:     ENSG00000185838\n",
      "gene_string='GNL2'\n",
      "gene_list=['GNL2']\n",
      "match=None\n",
      "['GNL2']:     ENSG00000134697\n",
      "gene_string='GNL3'\n",
      "gene_list=['GNL3']\n",
      "match=None\n",
      "['GNL3']:     ENSG00000163938\n",
      "gene_string='GNL3L'\n",
      "gene_list=['GNL3L']\n",
      "match=None\n",
      "['GNL3L']:     ENSG00000130119\n",
      "gene_string='GNPNAT1'\n",
      "gene_list=['GNPNAT1']\n",
      "match=None\n",
      "['GNPNAT1']:     ENSG00000100522\n",
      "gene_string='GOLGA6L1'\n",
      "gene_list=['GOLGA6L1']\n",
      "match=None\n",
      "['GOLGA6L1']:     ENSG00000273976\n",
      "gene_string='GOLT1B'\n",
      "gene_list=['GOLT1B']\n",
      "match=None\n",
      "['GOLT1B']:     ENSG00000111711\n",
      "gene_string='GON4L'\n",
      "gene_list=['GON4L']\n",
      "match=None\n",
      "['GON4L']:     ENSG00000116580\n",
      "gene_string='GOSR2'\n",
      "gene_list=['GOSR2']\n",
      "match=None\n",
      "['GOSR2']:     ENSG00000108433\n",
      "gene_string='GPKOW'\n",
      "gene_list=['GPKOW']\n",
      "match=None\n",
      "['GPKOW']:     ENSG00000068394\n",
      "gene_string='GPN1'\n",
      "gene_list=['GPN1']\n",
      "match=None\n",
      "['GPN1']:     ENSG00000198522\n",
      "gene_string='GPN2'\n",
      "gene_list=['GPN2']\n",
      "match=None\n",
      "['GPN2']:     ENSG00000142751\n",
      "gene_string='GPN3'\n",
      "gene_list=['GPN3']\n",
      "match=None\n",
      "['GPN3']:     ENSG00000111231\n",
      "gene_string='GPR61'\n",
      "gene_list=['GPR61']\n",
      "match=None\n",
      "['GPR61']:     ENSG00000156097\n",
      "gene_string='GPS1'\n",
      "gene_list=['GPS1']\n",
      "match=None\n",
      "['GPS1']:     ENSG00000169727\n",
      "gene_string='GRB2'\n",
      "gene_list=['GRB2']\n",
      "match=None\n",
      "['GRB2']:     ENSG00000177885\n",
      "gene_string='GRPEL1'\n",
      "gene_list=['GRPEL1']\n",
      "match=None\n",
      "['GRPEL1']:     ENSG00000109519\n",
      "gene_string='GRSF1'\n",
      "gene_list=['GRSF1']\n",
      "match=None\n",
      "['GRSF1']:     ENSG00000132463\n",
      "gene_string='GRWD1'\n",
      "gene_list=['GRWD1']\n",
      "match=None\n",
      "['GRWD1']:     ENSG00000105447\n",
      "gene_string='GSDMA'\n",
      "gene_list=['GSDMA']\n",
      "match=None\n",
      "['GSDMA']:     ENSG00000167914\n",
      "gene_string='GSPT1'\n",
      "gene_list=['GSPT1']\n",
      "match=None\n",
      "['GSPT1']:     ENSG00000103342\n",
      "gene_string='GTF2A1'\n",
      "gene_list=['GTF2A1']\n",
      "match=None\n",
      "['GTF2A1']:     ENSG00000165417\n",
      "gene_string='GTF2A2'\n",
      "gene_list=['GTF2A2']\n",
      "match=None\n",
      "['GTF2A2']:     ENSG00000140307\n",
      "gene_string='GTF2B'\n",
      "gene_list=['GTF2B']\n",
      "match=None\n",
      "['GTF2B']:     ENSG00000137947\n",
      "gene_string='GTF2E1'\n",
      "gene_list=['GTF2E1']\n",
      "match=None\n",
      "['GTF2E1']:     ENSG00000153767\n",
      "gene_string='GTF2E2'\n",
      "gene_list=['GTF2E2']\n",
      "match=None\n",
      "['GTF2E2']:     ENSG00000197265\n",
      "gene_string='GTF2F1'\n",
      "gene_list=['GTF2F1']\n",
      "match=None\n",
      "['GTF2F1']:     ENSG00000125651\n",
      "gene_string='GTF2F2'\n",
      "gene_list=['GTF2F2']\n",
      "match=None\n",
      "['GTF2F2']:     ENSG00000188342\n",
      "gene_string='GTF2H1'\n",
      "gene_list=['GTF2H1']\n",
      "match=None\n",
      "['GTF2H1']:     ENSG00000110768\n",
      "gene_string='GTF2H2'\n",
      "gene_list=['GTF2H2']\n",
      "match=None\n",
      "['GTF2H2']:     ENSG00000145736\n",
      "gene_string='GTF2H2C'\n",
      "gene_list=['GTF2H2C']\n",
      "match=None\n",
      "['GTF2H2C']:     ENSG00000183474\n",
      "gene_string='GTF2H3'\n",
      "gene_list=['GTF2H3']\n",
      "match=None\n",
      "['GTF2H3']:     ENSG00000111358\n",
      "gene_string='GTF2H4'\n",
      "gene_list=['GTF2H4']\n",
      "match=None\n",
      "['GTF2H4']:     ENSG00000213780\n",
      "gene_string='GTF3A'\n",
      "gene_list=['GTF3A']\n",
      "match=None\n",
      "['GTF3A']:     ENSG00000122034\n",
      "gene_string='GTF3C1'\n",
      "gene_list=['GTF3C1']\n",
      "match=None\n",
      "['GTF3C1']:     ENSG00000077235\n",
      "gene_string='GTF3C2'\n",
      "gene_list=['GTF3C2']\n",
      "match=None\n",
      "['GTF3C2']:     ENSG00000115207\n",
      "gene_string='GTF3C3'\n",
      "gene_list=['GTF3C3']\n",
      "match=None\n",
      "['GTF3C3']:     ENSG00000119041\n",
      "gene_string='GTF3C4'\n",
      "gene_list=['GTF3C4']\n",
      "match=None\n",
      "['GTF3C4']:     ENSG00000125484\n",
      "gene_string='GTF3C5'\n",
      "gene_list=['GTF3C5']\n",
      "match=None\n",
      "['GTF3C5']:     ENSG00000148308\n",
      "gene_string='GTF3C6'\n",
      "gene_list=['GTF3C6']\n",
      "match=None\n",
      "['GTF3C6']:     ENSG00000155115\n",
      "gene_string='GTPBP4'\n",
      "gene_list=['GTPBP4']\n",
      "match=None\n",
      "['GTPBP4']:     ENSG00000107937\n",
      "gene_string='GUCD1'\n",
      "gene_list=['GUCD1']\n",
      "match=None\n",
      "['GUCD1']:     ENSG00000138867\n",
      "gene_string='GUK1'\n",
      "gene_list=['GUK1']\n",
      "match=None\n",
      "['GUK1']:     ENSG00000143774\n",
      "gene_string='H2AFX'\n",
      "gene_list=['H2AFX']\n",
      "match=None\n",
      "['H2AFX']:     ENSG00000188486\n",
      "gene_string='H2AFZ'\n",
      "gene_list=['H2AFZ']\n",
      "match=None\n",
      "['H2AFZ']:     ENSG00000164032\n",
      "gene_string='H3F3A'\n",
      "gene_list=['H3F3A']\n",
      "match=None\n",
      "['H3F3A']:     ENSG00000163041\n",
      "gene_string='HAMP'\n",
      "gene_list=['HAMP']\n",
      "match=None\n",
      "['HAMP']:     ENSG00000105697\n",
      "gene_string='HAPLN2'\n",
      "gene_list=['HAPLN2']\n",
      "match=None\n",
      "['HAPLN2']:     ENSG00000132702\n",
      "gene_string='HARS'\n",
      "gene_list=['HARS']\n",
      "match=None\n",
      "['HARS']:     ENSG00000170445\n",
      "gene_string='HARS2'\n",
      "gene_list=['HARS2']\n",
      "match=None\n",
      "['HARS2']:     ENSG00000125821\n",
      "gene_string='HAUS1'\n",
      "gene_list=['HAUS1']\n",
      "match=None\n",
      "['HAUS1']:     ENSG00000152240\n",
      "gene_string='HAUS3'\n",
      "gene_list=['HAUS3']\n",
      "match=None\n",
      "['HAUS3']:     ENSG00000214367\n",
      "gene_string='HAUS4'\n",
      "gene_list=['HAUS4']\n",
      "match=None\n",
      "['HAUS4']:     ENSG00000092036\n",
      "gene_string='HAUS5'\n",
      "gene_list=['HAUS5']\n",
      "match=None\n",
      "['HAUS5']:     ENSG00000249115\n",
      "gene_string='HAUS6'\n",
      "gene_list=['HAUS6']\n",
      "match=None\n",
      "['HAUS6']:     ENSG00000147874\n",
      "gene_string='HAUS7'\n",
      "gene_list=['HAUS7']\n",
      "match=None\n",
      "['HAUS7']:     ENSG00000213397\n",
      "gene_string='HAUS8'\n",
      "gene_list=['HAUS8']\n",
      "match=None\n",
      "['HAUS8']:     ENSG00000131351\n",
      "gene_string='HBS1L'\n",
      "gene_list=['HBS1L']\n",
      "match=None\n",
      "['HBS1L']:     ENSG00000112339\n",
      "gene_string='HCFC1'\n",
      "gene_list=['HCFC1']\n",
      "match=None\n",
      "['HCFC1']:     ENSG00000172534\n",
      "gene_string='HCRTR1'\n",
      "gene_list=['HCRTR1']\n",
      "match=None\n",
      "['HCRTR1']:     ENSG00000121764\n",
      "gene_string='HDAC3'\n",
      "gene_list=['HDAC3']\n",
      "match=None\n",
      "['HDAC3']:     ENSG00000171720\n",
      "gene_string='HDAC7'\n",
      "gene_list=['HDAC7']\n",
      "match=None\n",
      "['HDAC7']:     ENSG00000061273\n",
      "gene_string='HEATR1'\n",
      "gene_list=['HEATR1']\n",
      "match=None\n",
      "['HEATR1']:     ENSG00000119285\n",
      "gene_string='HECTD1'\n",
      "gene_list=['HECTD1']\n",
      "match=None\n",
      "['HECTD1']:     ENSG00000092148\n",
      "gene_string='HGS'\n",
      "gene_list=['HGS']\n",
      "match=None\n",
      "['HGS']:     ENSG00000185359\n",
      "gene_string='HINFP'\n",
      "gene_list=['HINFP']\n",
      "match=None\n",
      "['HINFP']:     ENSG00000172273\n",
      "gene_string='HIRA'\n",
      "gene_list=['HIRA']\n",
      "match=None\n",
      "['HIRA']:     ENSG00000100084\n",
      "gene_string='HIST1H2AB'\n",
      "gene_list=['HIST1H2AB']\n",
      "match=None\n",
      "['HIST1H2AB']:     ENSG00000277075\n",
      "gene_string='HIST1H2AE'\n",
      "gene_list=['HIST1H2AE']\n",
      "match=None\n",
      "['HIST1H2AE']:     ENSG00000277075\n",
      "gene_string='HIST1H2AI'\n",
      "gene_list=['HIST1H2AI']\n",
      "match=None\n",
      "['HIST1H2AI']:     ENSG00000196747\n",
      "gene_string='HIST1H2BB'\n",
      "gene_list=['HIST1H2BB']\n",
      "match=None\n",
      "['HIST1H2BB']:     ENSG00000276410\n",
      "gene_string='HIST1H2BC'\n",
      "gene_list=['HIST1H2BC']\n",
      "match=None\n",
      "['HIST1H2BC']:     ENSG00000180596\n",
      "gene_string='HIST1H2BE'\n",
      "gene_list=['HIST1H2BE']\n",
      "match=None\n",
      "['HIST1H2BE']:     ENSG00000273802\n",
      "gene_string='HIST1H2BJ'\n",
      "gene_list=['HIST1H2BJ']\n",
      "match=None\n",
      "['HIST1H2BJ']:     ENSG00000124635\n",
      "gene_string='HIST1H2BL'\n",
      "gene_list=['HIST1H2BL']\n",
      "match=None\n",
      "['HIST1H2BL']:     ENSG00000185130\n",
      "gene_string='HIST1H2BM'\n",
      "gene_list=['HIST1H2BM']\n",
      "match=None\n",
      "['HIST1H2BM']:     ENSG00000273703\n",
      "gene_string='HIST1H2BN'\n",
      "gene_list=['HIST1H2BN']\n",
      "match=None\n",
      "['HIST1H2BN']:     ENSG00000233822\n",
      "gene_string='HIST2H2AA3'\n",
      "gene_list=['HIST2H2AA3']\n",
      "match=None\n",
      "['HIST2H2AA3']:     ENSG00000288825\n",
      "gene_string='HIST2H2AC'\n",
      "gene_list=['HIST2H2AC']\n",
      "match=None\n",
      "['HIST2H2AC']:     ENSG00000184260\n",
      "gene_string='HIST2H2BE'\n",
      "gene_list=['HIST2H2BE']\n",
      "match=None\n",
      "['HIST2H2BE']:     ENSG00000184678\n",
      "gene_string='HIST2H2BF'\n",
      "gene_list=['HIST2H2BF']\n",
      "match=None\n",
      "['HIST2H2BF']:     ENSG00000203814\n",
      "gene_string='HIST2H3A'\n",
      "gene_list=['HIST2H3A']\n",
      "match=None\n",
      "['HIST2H3A']:     ENSG00000203852\n",
      "gene_string='HIST2H3D'\n",
      "gene_list=['HIST2H3D']\n",
      "match=None\n",
      "['HIST2H3D']:     ENSG00000183598\n",
      "gene_string='HJURP'\n",
      "gene_list=['HJURP']\n",
      "match=None\n",
      "['HJURP']:     ENSG00000123485\n",
      "gene_string='HLA-C'\n",
      "gene_list=['HLA-C']\n",
      "match=None\n",
      "['HLA-C']:     ENSG00000204525\n",
      "gene_string='HMGA1'\n",
      "gene_list=['HMGA1']\n",
      "match=None\n",
      "['HMGA1']:     ENSG00000137309\n",
      "gene_string='HMGB1'\n",
      "gene_list=['HMGB1']\n",
      "match=None\n",
      "['HMGB1']:     ENSG00000189403\n",
      "gene_string='HMGCR'\n",
      "gene_list=['HMGCR']\n",
      "match=None\n",
      "['HMGCR']:     ENSG00000113161\n",
      "gene_string='HMGCS1'\n",
      "gene_list=['HMGCS1']\n",
      "match=None\n",
      "['HMGCS1']:     ENSG00000112972\n",
      "gene_string='HMGN2'\n",
      "gene_list=['HMGN2']\n",
      "match=None\n",
      "['HMGN2']:     ENSG00000198830\n",
      "gene_string='HNRNPC'\n",
      "gene_list=['HNRNPC']\n",
      "match=None\n",
      "['HNRNPC']:     ENSG00000092199\n",
      "gene_string='HNRNPH1'\n",
      "gene_list=['HNRNPH1']\n",
      "match=None\n",
      "['HNRNPH1']:     ENSG00000169045\n",
      "gene_string='HNRNPK'\n",
      "gene_list=['HNRNPK']\n",
      "match=None\n",
      "['HNRNPK']:     ENSG00000165119\n",
      "gene_string='HNRNPL'\n",
      "gene_list=['HNRNPL']\n",
      "match=None\n",
      "['HNRNPL']:     ENSG00000104824\n",
      "gene_string='HNRNPM'\n",
      "gene_list=['HNRNPM']\n",
      "match=None\n",
      "['HNRNPM']:     ENSG00000099783\n",
      "gene_string='HNRNPR'\n",
      "gene_list=['HNRNPR']\n",
      "match=None\n",
      "['HNRNPR']:     ENSG00000125944\n",
      "gene_string='HNRNPU'\n",
      "gene_list=['HNRNPU']\n",
      "match=None\n",
      "['HNRNPU']:     ENSG00000153187\n",
      "gene_string='HOXC10'\n",
      "gene_list=['HOXC10']\n",
      "match=None\n",
      "['HOXC10']:     ENSG00000180818\n",
      "gene_string='HS6ST1'\n",
      "gene_list=['HS6ST1']\n",
      "match=None\n",
      "['HS6ST1']:     ENSG00000136720\n",
      "gene_string='HSCB'\n",
      "gene_list=['HSCB']\n",
      "match=None\n",
      "['HSCB']:     ENSG00000100209\n",
      "gene_string='HSD17B10'\n",
      "gene_list=['HSD17B10']\n",
      "match=None\n",
      "['HSD17B10']:     ENSG00000072506\n",
      "gene_string='HSD17B12'\n",
      "gene_list=['HSD17B12']\n",
      "match=None\n",
      "['HSD17B12']:     ENSG00000149084\n",
      "gene_string='HSF1'\n",
      "gene_list=['HSF1']\n",
      "match=None\n",
      "['HSF1']:     ENSG00000185122\n",
      "gene_string='HSP90B1'\n",
      "gene_list=['HSP90B1']\n",
      "match=None\n",
      "['HSP90B1']:     ENSG00000166598\n",
      "gene_string='HSPA5'\n",
      "gene_list=['HSPA5']\n",
      "match=None\n",
      "['HSPA5']:     ENSG00000044574\n",
      "gene_string='HSPA8'\n",
      "gene_list=['HSPA8']\n",
      "match=None\n",
      "['HSPA8']:     ENSG00000109971\n",
      "gene_string='HSPA9'\n",
      "gene_list=['HSPA9']\n",
      "match=None\n",
      "['HSPA9']:     ENSG00000113013\n",
      "gene_string='HSPA14'\n",
      "gene_list=['HSPA14']\n",
      "match=None\n",
      "['HSPA14']:     ENSG00000187522\n",
      "gene_string='HSPD1'\n",
      "gene_list=['HSPD1']\n",
      "match=None\n",
      "['HSPD1']:     ENSG00000144381\n",
      "gene_string='HSPE1'\n",
      "gene_list=['HSPE1']\n",
      "match=None\n",
      "['HSPE1']:     ENSG00000115541\n",
      "gene_string='HTATSF1'\n",
      "gene_list=['HTATSF1']\n",
      "match=None\n",
      "['HTATSF1']:     ENSG00000102241\n",
      "gene_string='HUS1'\n",
      "gene_list=['HUS1']\n",
      "match=None\n",
      "['HUS1']:     ENSG00000136273\n",
      "gene_string='HUWE1'\n",
      "gene_list=['HUWE1']\n",
      "match=None\n",
      "['HUWE1']:     ENSG00000086758\n",
      "gene_string='HYOU1'\n",
      "gene_list=['HYOU1']\n",
      "match=None\n",
      "['HYOU1']:     ENSG00000149428\n",
      "gene_string='HYPK'\n",
      "gene_list=['HYPK']\n",
      "match=None\n",
      "['HYPK']:     ENSG00000242028\n",
      "gene_string='IARS'\n",
      "gene_list=['IARS']\n",
      "match=None\n",
      "['IARS']:     ENSG00000196305\n",
      "gene_string='IARS2'\n",
      "gene_list=['IARS2']\n",
      "match=None\n",
      "['IARS2']:     ENSG00000067704\n",
      "gene_string='ICE1'\n",
      "gene_list=['ICE1']\n",
      "match=None\n",
      "['ICE1']:     ENSG00000164151\n",
      "gene_string='IFITM2'\n",
      "gene_list=['IFITM2']\n",
      "match=None\n",
      "['IFITM2']:     ENSG00000185201\n",
      "gene_string='IFITM3'\n",
      "gene_list=['IFITM3']\n",
      "match=None\n",
      "['IFITM3']:     ENSG00000142089\n",
      "gene_string='IGBP1'\n",
      "gene_list=['IGBP1']\n",
      "match=None\n",
      "['IGBP1']:     ENSG00000089289\n",
      "gene_string='IK'\n",
      "gene_list=['IK']\n",
      "match=None\n",
      "['IK']:     ENSG00000104783\n",
      "gene_string='ILF2'\n",
      "gene_list=['ILF2']\n",
      "match=None\n",
      "['ILF2']:     ENSG00000143621\n",
      "gene_string='ILF3'\n",
      "gene_list=['ILF3']\n",
      "match=None\n",
      "['ILF3']:     ENSG00000129351\n",
      "gene_string='IMP3'\n",
      "gene_list=['IMP3']\n",
      "match=None\n",
      "['IMP3']:     ENSG00000138600\n",
      "gene_string='IMP4'\n",
      "gene_list=['IMP4']\n",
      "match=None\n",
      "['IMP4']:     ENSG00000005206\n",
      "gene_string='IMPDH2'\n",
      "gene_list=['IMPDH2']\n",
      "match=None\n",
      "['IMPDH2']:     ENSG00000178035\n",
      "gene_string='INCENP'\n",
      "gene_list=['INCENP']\n",
      "match=None\n",
      "['INCENP']:     ENSG00000149503\n",
      "gene_string='ING3'\n",
      "gene_list=['ING3']\n",
      "match=None\n",
      "['ING3']:     ENSG00000071243\n",
      "gene_string='INO80'\n",
      "gene_list=['INO80']\n",
      "match=None\n",
      "['INO80']:     ENSG00000128908\n",
      "gene_string='INO80B'\n",
      "gene_list=['INO80B']\n",
      "match=None\n",
      "['INO80B']:     ENSG00000115274\n",
      "gene_string='INTS1'\n",
      "gene_list=['INTS1']\n",
      "match=None\n",
      "['INTS1']:     ENSG00000164880\n",
      "gene_string='INTS2'\n",
      "gene_list=['INTS2']\n",
      "match=None\n",
      "['INTS2']:     ENSG00000108506\n",
      "gene_string='INTS3'\n",
      "gene_list=['INTS3']\n",
      "match=None\n",
      "['INTS3']:     ENSG00000143624\n",
      "gene_string='INTS4'\n",
      "gene_list=['INTS4']\n",
      "match=None\n",
      "['INTS4']:     ENSG00000149262\n",
      "gene_string='INTS5'\n",
      "gene_list=['INTS5']\n",
      "match=None\n",
      "['INTS5']:     ENSG00000185085\n",
      "gene_string='INTS6'\n",
      "gene_list=['INTS6']\n",
      "match=None\n",
      "['INTS6']:     ENSG00000102786\n",
      "gene_string='INTS7'\n",
      "gene_list=['INTS7']\n",
      "match=None\n",
      "['INTS7']:     ENSG00000143493\n",
      "gene_string='INTS8'\n",
      "gene_list=['INTS8']\n",
      "match=None\n",
      "['INTS8']:     ENSG00000164941\n",
      "gene_string='INTS9'\n",
      "gene_list=['INTS9']\n",
      "match=None\n",
      "['INTS9']:     ENSG00000104299\n",
      "gene_string='IPO7'\n",
      "gene_list=['IPO7']\n",
      "match=None\n",
      "['IPO7']:     ENSG00000205339\n",
      "gene_string='IPO9'\n",
      "gene_list=['IPO9']\n",
      "match=None\n",
      "['IPO9']:     ENSG00000198700\n",
      "gene_string='IPO11'\n",
      "gene_list=['IPO11']\n",
      "match=None\n",
      "['IPO11']:     ENSG00000086200\n",
      "gene_string='IPO13'\n",
      "gene_list=['IPO13']\n",
      "match=None\n",
      "['IPO13']:     ENSG00000117408\n",
      "gene_string='ISCA2'\n",
      "gene_list=['ISCA2']\n",
      "match=None\n",
      "['ISCA2']:     ENSG00000165898\n",
      "gene_string='ISCU'\n",
      "gene_list=['ISCU']\n",
      "match=None\n",
      "['ISCU']:     ENSG00000136003\n",
      "gene_string='ISG20L2'\n",
      "gene_list=['ISG20L2']\n",
      "match=None\n",
      "['ISG20L2']:     ENSG00000143319\n",
      "gene_string='ISY1'\n",
      "gene_list=['ISY1']\n",
      "match=None\n",
      "['ISY1']:     ENSG00000240682\n",
      "gene_string='JAZF1'\n",
      "gene_list=['JAZF1']\n",
      "match=None\n",
      "['JAZF1']:     ENSG00000153814\n",
      "gene_string='JMJD6'\n",
      "gene_list=['JMJD6']\n",
      "match=None\n",
      "['JMJD6']:     ENSG00000070495\n",
      "gene_string='KANSL1'\n",
      "gene_list=['KANSL1']\n",
      "match=None\n",
      "['KANSL1']:     ENSG00000120071\n",
      "gene_string='KANSL2'\n",
      "gene_list=['KANSL2']\n",
      "match=None\n",
      "['KANSL2']:     ENSG00000139620\n",
      "gene_string='KANSL3'\n",
      "gene_list=['KANSL3']\n",
      "match=None\n",
      "['KANSL3']:     ENSG00000114982\n",
      "gene_string='KARS'\n",
      "gene_list=['KARS']\n",
      "match=None\n",
      "['KARS']:     ENSG00000065427\n",
      "gene_string='KAT5'\n",
      "gene_list=['KAT5']\n",
      "match=None\n",
      "['KAT5']:     ENSG00000172977\n",
      "gene_string='KAT8'\n",
      "gene_list=['KAT8']\n",
      "match=None\n",
      "['KAT8']:     ENSG00000103510\n",
      "gene_string='KATNB1'\n",
      "gene_list=['KATNB1']\n",
      "match=None\n",
      "['KATNB1']:     ENSG00000140854\n",
      "gene_string='KCMF1'\n",
      "gene_list=['KCMF1']\n",
      "match=None\n",
      "['KCMF1']:     ENSG00000176407\n",
      "gene_string='KCNA10'\n",
      "gene_list=['KCNA10']\n",
      "match=None\n",
      "['KCNA10']:     ENSG00000143105\n",
      "gene_string='KCTD10'\n",
      "gene_list=['KCTD10']\n",
      "match=None\n",
      "['KCTD10']:     ENSG00000110906\n",
      "gene_string='KDM2A'\n",
      "gene_list=['KDM2A']\n",
      "match=None\n",
      "['KDM2A']:     ENSG00000173120\n",
      "gene_string='KDM8'\n",
      "gene_list=['KDM8']\n",
      "match=None\n",
      "['KDM8']:     ENSG00000155666\n",
      "gene_string='KIAA1143'\n",
      "gene_list=['KIAA1143']\n",
      "match=None\n",
      "['KIAA1143']:     ENSG00000163807\n",
      "gene_string='KIF4A'\n",
      "gene_list=['KIF4A']\n",
      "match=None\n",
      "['KIF4A']:     ENSG00000090889\n",
      "gene_string='KIF11'\n",
      "gene_list=['KIF11']\n",
      "match=None\n",
      "['KIF11']:     ENSG00000138160\n",
      "gene_string='KIF14'\n",
      "gene_list=['KIF14']\n",
      "match=None\n",
      "['KIF14']:     ENSG00000118193\n",
      "gene_string='KIF18A'\n",
      "gene_list=['KIF18A']\n",
      "match=None\n",
      "['KIF18A']:     ENSG00000121621\n",
      "gene_string='KIF18B'\n",
      "gene_list=['KIF18B']\n",
      "match=None\n",
      "['KIF18B']:     ENSG00000186185\n",
      "gene_string='KIF20A'\n",
      "gene_list=['KIF20A']\n",
      "match=None\n",
      "['KIF20A']:     ENSG00000112984\n",
      "gene_string='KIF23'\n",
      "gene_list=['KIF23']\n",
      "match=None\n",
      "['KIF23']:     ENSG00000137807\n",
      "gene_string='KIN'\n",
      "gene_list=['KIN']\n",
      "match=None\n",
      "['KIN']:     ENSG00000151657\n",
      "gene_string='KLC2'\n",
      "gene_list=['KLC2']\n",
      "match=None\n",
      "['KLC2']:     ENSG00000174996\n",
      "gene_string='KLF7'\n",
      "gene_list=['KLF7']\n",
      "match=None\n",
      "['KLF7']:     ENSG00000118263\n",
      "gene_string='KLHL11'\n",
      "gene_list=['KLHL11']\n",
      "match=None\n",
      "['KLHL11']:     ENSG00000178502\n",
      "gene_string='KPNA6'\n",
      "gene_list=['KPNA6']\n",
      "match=None\n",
      "['KPNA6']:     ENSG00000025800\n",
      "gene_string='KPNB1'\n",
      "gene_list=['KPNB1']\n",
      "match=None\n",
      "['KPNB1']:     ENSG00000108424\n",
      "gene_string='KRI1'\n",
      "gene_list=['KRI1']\n",
      "match=None\n",
      "['KRI1']:     ENSG00000129347\n",
      "gene_string='KRR1'\n",
      "gene_list=['KRR1']\n",
      "match=None\n",
      "['KRR1']:     ENSG00000111615\n",
      "gene_string='KRT8'\n",
      "gene_list=['KRT8']\n",
      "match=None\n",
      "['KRT8']:     ENSG00000170421\n",
      "gene_string='KRT10'\n",
      "gene_list=['KRT10']\n",
      "match=None\n",
      "['KRT10']:     ENSG00000186395\n",
      "gene_string='KRT17'\n",
      "gene_list=['KRT17']\n",
      "match=None\n",
      "['KRT17']:     ENSG00000128422\n",
      "gene_string='KRTAP4-2'\n",
      "gene_list=['KRTAP4-2']\n",
      "match=None\n",
      "['KRTAP4-2']:     ENSG00000244537\n",
      "gene_string='KRTAP4-7'\n",
      "gene_list=['KRTAP4-7']\n",
      "match=None\n",
      "['KRTAP4-7']:     ENSG00000240871\n",
      "gene_string='KTI12'\n",
      "gene_list=['KTI12']\n",
      "match=None\n",
      "['KTI12']:     ENSG00000198841\n",
      "gene_string='KXD1'\n",
      "gene_list=['KXD1']\n",
      "match=None\n",
      "['KXD1']:     ENSG00000105700\n",
      "gene_string='LAMTOR2'\n",
      "gene_list=['LAMTOR2']\n",
      "match=None\n",
      "['LAMTOR2']:     ENSG00000116586\n",
      "gene_string='LAMTOR3'\n",
      "gene_list=['LAMTOR3']\n",
      "match=None\n",
      "['LAMTOR3']:     ENSG00000109270\n",
      "gene_string='LAMTOR4'\n",
      "gene_list=['LAMTOR4']\n",
      "match=None\n",
      "['LAMTOR4']:     ENSG00000188186\n",
      "gene_string='LARS'\n",
      "gene_list=['LARS']\n",
      "match=None\n",
      "['LARS']:     ENSG00000133706\n",
      "gene_string='LARS2'\n",
      "gene_list=['LARS2']\n",
      "match=None\n",
      "['LARS2']:     ENSG00000011376\n",
      "gene_string='LAS1L'\n",
      "gene_list=['LAS1L']\n",
      "match=None\n",
      "['LAS1L']:     ENSG00000001497\n",
      "gene_string='LCE1C'\n",
      "gene_list=['LCE1C']\n",
      "match=None\n",
      "['LCE1C']:     ENSG00000197084\n",
      "gene_string='LCE1E'\n",
      "gene_list=['LCE1E']\n",
      "match=None\n",
      "['LCE1E']:     ENSG00000186226\n",
      "gene_string='LCN10'\n",
      "gene_list=['LCN10']\n",
      "match=None\n",
      "['LCN10']:     ENSG00000187922\n",
      "gene_string='LENG8'\n",
      "gene_list=['LENG8']\n",
      "match=None\n",
      "['LENG8']:     ENSG00000167615\n",
      "gene_string='LETM1'\n",
      "gene_list=['LETM1']\n",
      "match=None\n",
      "['LETM1']:     ENSG00000168924\n",
      "gene_string='LIMS1'\n",
      "gene_list=['LIMS1']\n",
      "match=None\n",
      "['LIMS1']:     ENSG00000169756\n",
      "gene_string='LIN52'\n",
      "gene_list=['LIN52']\n",
      "match=None\n",
      "['LIN52']:     ENSG00000205659\n",
      "gene_string='LIN54'\n",
      "gene_list=['LIN54']\n",
      "match=None\n",
      "['LIN54']:     ENSG00000189308\n",
      "gene_string='LONP1'\n",
      "gene_list=['LONP1']\n",
      "match=None\n",
      "['LONP1']:     ENSG00000196365\n",
      "gene_string='LPA'\n",
      "gene_list=['LPA']\n",
      "match=None\n",
      "['LPA']:     ENSG00000198670\n",
      "gene_string='LRPPRC'\n",
      "gene_list=['LRPPRC']\n",
      "match=None\n",
      "['LRPPRC']:     ENSG00000138095\n",
      "gene_string='LRR1'\n",
      "gene_list=['LRR1']\n",
      "match=None\n",
      "['LRR1']:     ENSG00000165501\n",
      "gene_string='LRRC37A3'\n",
      "gene_list=['LRRC37A3']\n",
      "match=None\n",
      "['LRRC37A3']:     ENSG00000176809\n",
      "gene_string='LRRC37B'\n",
      "gene_list=['LRRC37B']\n",
      "match=None\n",
      "['LRRC37B']:     ENSG00000185158\n",
      "gene_string='LSG1'\n",
      "gene_list=['LSG1']\n",
      "match=None\n",
      "['LSG1']:     ENSG00000041802\n",
      "gene_string='LSM2'\n",
      "gene_list=['LSM2']\n",
      "match=None\n",
      "['LSM2']:     ENSG00000204392\n",
      "gene_string='LSM3'\n",
      "gene_list=['LSM3']\n",
      "match=None\n",
      "['LSM3']:     ENSG00000170860\n",
      "gene_string='LSM4'\n",
      "gene_list=['LSM4']\n",
      "match=None\n",
      "['LSM4']:     ENSG00000130520\n",
      "gene_string='LSM5'\n",
      "gene_list=['LSM5']\n",
      "match=None\n",
      "['LSM5']:     ENSG00000106355\n",
      "gene_string='LSM6'\n",
      "gene_list=['LSM6']\n",
      "match=None\n",
      "['LSM6']:     ENSG00000164167\n",
      "gene_string='LSM7'\n",
      "gene_list=['LSM7']\n",
      "match=None\n",
      "['LSM7']:     ENSG00000130332\n",
      "gene_string='LSM10'\n",
      "gene_list=['LSM10']\n",
      "match=None\n",
      "['LSM10']:     ENSG00000181817\n",
      "gene_string='LSM11'\n",
      "gene_list=['LSM11']\n",
      "match=None\n",
      "['LSM11']:     ENSG00000155858\n",
      "gene_string='LSM12'\n",
      "gene_list=['LSM12']\n",
      "match=None\n",
      "['LSM12']:     ENSG00000161654\n",
      "gene_string='LST1'\n",
      "gene_list=['LST1']\n",
      "match=None\n",
      "['LST1']:     ENSG00000204482\n",
      "gene_string='LTB4R2'\n",
      "gene_list=['LTB4R2']\n",
      "match=None\n",
      "['LTB4R2']:     ENSG00000213906\n",
      "gene_string='LTBP3'\n",
      "gene_list=['LTBP3']\n",
      "match=None\n",
      "['LTBP3']:     ENSG00000168056\n",
      "gene_string='LTV1'\n",
      "gene_list=['LTV1']\n",
      "match=None\n",
      "['LTV1']:     ENSG00000135521\n",
      "gene_string='LUC7L3'\n",
      "gene_list=['LUC7L3']\n",
      "match=None\n",
      "['LUC7L3']:     ENSG00000108848\n",
      "gene_string='LURAP1'\n",
      "gene_list=['LURAP1']\n",
      "match=None\n",
      "['LURAP1']:     ENSG00000171357\n",
      "gene_string='LYRM4'\n",
      "gene_list=['LYRM4']\n",
      "match=None\n",
      "['LYRM4']:     ENSG00000214113\n",
      "gene_string='MAD2L1'\n",
      "gene_list=['MAD2L1']\n",
      "match=None\n",
      "['MAD2L1']:     ENSG00000164109\n",
      "gene_string='MAD2L2'\n",
      "gene_list=['MAD2L2']\n",
      "match=None\n",
      "['MAD2L2']:     ENSG00000116670\n",
      "gene_string='MAGOH'\n",
      "gene_list=['MAGOH']\n",
      "match=None\n",
      "['MAGOH']:     ENSG00000162385\n",
      "gene_string='MAK16'\n",
      "gene_list=['MAK16']\n",
      "match=None\n",
      "['MAK16']:     ENSG00000198042\n",
      "gene_string='MANF'\n",
      "gene_list=['MANF']\n",
      "match=None\n",
      "['MANF']:     ENSG00000145050\n",
      "gene_string='MARS'\n",
      "gene_list=['MARS']\n",
      "match=None\n",
      "['MARS']:     ENSG00000166986\n",
      "gene_string='MARS2'\n",
      "gene_list=['MARS2']\n",
      "match=None\n",
      "['MARS2']:     ENSG00000247626\n",
      "gene_string='MASTL'\n",
      "gene_list=['MASTL']\n",
      "match=None\n",
      "['MASTL']:     ENSG00000120539\n",
      "gene_string='MAT2A'\n",
      "gene_list=['MAT2A']\n",
      "match=None\n",
      "['MAT2A']:     ENSG00000168906\n",
      "gene_string='MAU2'\n",
      "gene_list=['MAU2']\n",
      "match=None\n",
      "['MAU2']:     ENSG00000129933\n",
      "gene_string='MAX'\n",
      "gene_list=['MAX']\n",
      "match=None\n",
      "['MAX']:     ENSG00000125952\n",
      "gene_string='MBTPS1'\n",
      "gene_list=['MBTPS1']\n",
      "match=None\n",
      "['MBTPS1']:     ENSG00000140943\n",
      "gene_string='MBTPS2'\n",
      "gene_list=['MBTPS2']\n",
      "match=None\n",
      "['MBTPS2']:     ENSG00000012174\n",
      "gene_string='MCL1'\n",
      "gene_list=['MCL1']\n",
      "match=None\n",
      "['MCL1']:     ENSG00000143384\n",
      "gene_string='MCM2'\n",
      "gene_list=['MCM2']\n",
      "match=None\n",
      "['MCM2']:     ENSG00000073111\n",
      "gene_string='MCM3'\n",
      "gene_list=['MCM3']\n",
      "match=None\n",
      "['MCM3']:     ENSG00000112118\n",
      "gene_string='MCM3AP'\n",
      "gene_list=['MCM3AP']\n",
      "match=None\n",
      "['MCM3AP']:     ENSG00000160294\n",
      "gene_string='MCM4'\n",
      "gene_list=['MCM4']\n",
      "match=None\n",
      "['MCM4']:     ENSG00000104738\n",
      "gene_string='MCM5'\n",
      "gene_list=['MCM5']\n",
      "match=None\n",
      "['MCM5']:     ENSG00000100297\n",
      "gene_string='MCM6'\n",
      "gene_list=['MCM6']\n",
      "match=None\n",
      "['MCM6']:     ENSG00000076003\n",
      "gene_string='MCM7'\n",
      "gene_list=['MCM7']\n",
      "match=None\n",
      "['MCM7']:     ENSG00000166508\n",
      "gene_string='MCM10'\n",
      "gene_list=['MCM10']\n",
      "match=None\n",
      "['MCM10']:     ENSG00000065328\n",
      "gene_string='MCMBP'\n",
      "gene_list=['MCMBP']\n",
      "match=None\n",
      "['MCMBP']:     ENSG00000197771\n",
      "gene_string='MCRS1'\n",
      "gene_list=['MCRS1']\n",
      "match=None\n",
      "['MCRS1']:     ENSG00000187778\n",
      "gene_string='MDN1'\n",
      "gene_list=['MDN1']\n",
      "match=None\n",
      "['MDN1']:     ENSG00000112159\n",
      "gene_string='MED1'\n",
      "gene_list=['MED1']\n",
      "match=None\n",
      "['MED1']:     ENSG00000125686\n",
      "gene_string='MED4'\n",
      "gene_list=['MED4']\n",
      "match=None\n",
      "['MED4']:     ENSG00000136146\n",
      "gene_string='MED6'\n",
      "gene_list=['MED6']\n",
      "match=None\n",
      "['MED6']:     ENSG00000133997\n",
      "gene_string='MED7'\n",
      "gene_list=['MED7']\n",
      "match=None\n",
      "['MED7']:     ENSG00000155868\n",
      "gene_string='MED8'\n",
      "gene_list=['MED8']\n",
      "match=None\n",
      "['MED8']:     ENSG00000159479\n",
      "gene_string='MED9'\n",
      "gene_list=['MED9']\n",
      "match=None\n",
      "['MED9']:     ENSG00000141026\n",
      "gene_string='MED10'\n",
      "gene_list=['MED10']\n",
      "match=None\n",
      "['MED10']:     ENSG00000133398\n",
      "gene_string='MED11'\n",
      "gene_list=['MED11']\n",
      "match=None\n",
      "['MED11']:     ENSG00000161920\n",
      "gene_string='MED12'\n",
      "gene_list=['MED12']\n",
      "match=None\n",
      "['MED12']:     ENSG00000184634\n",
      "gene_string='MED14'\n",
      "gene_list=['MED14']\n",
      "match=None\n",
      "['MED14']:     ENSG00000180182\n",
      "gene_string='MED17'\n",
      "gene_list=['MED17']\n",
      "match=None\n",
      "['MED17']:     ENSG00000042429\n",
      "gene_string='MED18'\n",
      "gene_list=['MED18']\n",
      "match=None\n",
      "['MED18']:     ENSG00000130772\n",
      "gene_string='MED19'\n",
      "gene_list=['MED19']\n",
      "match=None\n",
      "['MED19']:     ENSG00000156603\n",
      "gene_string='MED20'\n",
      "gene_list=['MED20']\n",
      "match=None\n",
      "['MED20']:     ENSG00000124641\n",
      "gene_string='MED21'\n",
      "gene_list=['MED21']\n",
      "match=None\n",
      "['MED21']:     ENSG00000152944\n",
      "gene_string='MED22'\n",
      "gene_list=['MED22']\n",
      "match=None\n",
      "['MED22']:     ENSG00000148297\n",
      "gene_string='MED26'\n",
      "gene_list=['MED26']\n",
      "match=None\n",
      "['MED26']:     ENSG00000105085\n",
      "gene_string='MED27'\n",
      "gene_list=['MED27']\n",
      "match=None\n",
      "['MED27']:     ENSG00000160563\n",
      "gene_string='MED28'\n",
      "gene_list=['MED28']\n",
      "match=None\n",
      "['MED28']:     ENSG00000118579\n",
      "gene_string='MED29'\n",
      "gene_list=['MED29']\n",
      "match=None\n",
      "['MED29']:     ENSG00000063322\n",
      "gene_string='MED30'\n",
      "gene_list=['MED30']\n",
      "match=None\n",
      "['MED30']:     ENSG00000164758\n",
      "gene_string='MED31'\n",
      "gene_list=['MED31']\n",
      "match=None\n",
      "['MED31']:     ENSG00000108590\n",
      "gene_string='MEPCE'\n",
      "gene_list=['MEPCE']\n",
      "match=None\n",
      "['MEPCE']:     ENSG00000146834\n",
      "gene_string='METAP1'\n",
      "gene_list=['METAP1']\n",
      "match=None\n",
      "['METAP1']:     ENSG00000164024\n",
      "gene_string='METAP2'\n",
      "gene_list=['METAP2']\n",
      "match=None\n",
      "['METAP2']:     ENSG00000111142\n",
      "gene_string='METTL1'\n",
      "gene_list=['METTL1']\n",
      "match=None\n",
      "['METTL1']:     ENSG00000037897\n",
      "gene_string='METTL3'\n",
      "gene_list=['METTL3']\n",
      "match=None\n",
      "['METTL3']:     ENSG00000165819\n",
      "gene_string='METTL14'\n",
      "gene_list=['METTL14']\n",
      "match=None\n",
      "['METTL14']:     ENSG00000145388\n",
      "gene_string='METTL16'\n",
      "gene_list=['METTL16']\n",
      "match=None\n",
      "['METTL16']:     ENSG00000127804\n",
      "gene_string='METTL17'\n",
      "gene_list=['METTL17']\n",
      "match=None\n",
      "['METTL17']:     ENSG00000165792\n",
      "gene_string='METTL23'\n",
      "gene_list=['METTL23']\n",
      "match=None\n",
      "['METTL23']:     ENSG00000181038\n",
      "gene_string='MEX3A'\n",
      "gene_list=['MEX3A']\n",
      "match=None\n",
      "['MEX3A']:     ENSG00000254726\n",
      "gene_string='MFAP1'\n",
      "gene_list=['MFAP1']\n",
      "match=None\n",
      "['MFAP1']:     ENSG00000140259\n",
      "gene_string='MFN2'\n",
      "gene_list=['MFN2']\n",
      "match=None\n",
      "['MFN2']:     ENSG00000116688\n",
      "gene_string='MICOS10'\n",
      "gene_list=['MICOS10']\n",
      "match=None\n",
      "['MICOS10']:     ENSG00000173436\n",
      "gene_string='MIOS'\n",
      "gene_list=['MIOS']\n",
      "match=None\n",
      "['MIOS']:     ENSG00000164654\n",
      "gene_string='MIPEP'\n",
      "gene_list=['MIPEP']\n",
      "match=None\n",
      "['MIPEP']:     ENSG00000027001\n",
      "gene_string='MIS12'\n",
      "gene_list=['MIS12']\n",
      "match=None\n",
      "['MIS12']:     ENSG00000167842\n",
      "gene_string='MIS18A'\n",
      "gene_list=['MIS18A']\n",
      "match=None\n",
      "['MIS18A']:     ENSG00000159055\n",
      "gene_string='MIS18BP1'\n",
      "gene_list=['MIS18BP1']\n",
      "match=None\n",
      "['MIS18BP1']:     ENSG00000129534\n",
      "gene_string='MKRN1'\n",
      "gene_list=['MKRN1']\n",
      "match=None\n",
      "['MKRN1']:     ENSG00000133606\n",
      "gene_string='MLST8'\n",
      "gene_list=['MLST8']\n",
      "match=None\n",
      "['MLST8']:     ENSG00000167965\n",
      "gene_string='MMGT1'\n",
      "gene_list=['MMGT1']\n",
      "match=None\n",
      "['MMGT1']:     ENSG00000169446\n",
      "gene_string='MMS19'\n",
      "gene_list=['MMS19']\n",
      "match=None\n",
      "['MMS19']:     ENSG00000155229\n",
      "gene_string='MMS22L'\n",
      "gene_list=['MMS22L']\n",
      "match=None\n",
      "['MMS22L']:     ENSG00000146263\n",
      "gene_string='MNAT1'\n",
      "gene_list=['MNAT1']\n",
      "match=None\n",
      "['MNAT1']:     ENSG00000020426\n",
      "gene_string='MOB4'\n",
      "gene_list=['MOB4']\n",
      "match=None\n",
      "['MOB4']:     ENSG00000115540\n",
      "gene_string='MOCS3'\n",
      "gene_list=['MOCS3']\n",
      "match=None\n",
      "['MOCS3']:     ENSG00000124217\n",
      "gene_string='MOG'\n",
      "gene_list=['MOG']\n",
      "match=None\n",
      "['MOG']:     ENSG00000204655\n",
      "gene_string='MOK'\n",
      "gene_list=['MOK']\n",
      "match=None\n",
      "['MOK']:     ENSG00000080823\n",
      "gene_string='MPHOSPH10'\n",
      "gene_list=['MPHOSPH10']\n",
      "match=None\n",
      "['MPHOSPH10']:     ENSG00000124383\n",
      "gene_string='MRGBP'\n",
      "gene_list=['MRGBP']\n",
      "match=None\n",
      "['MRGBP']:     ENSG00000101189\n",
      "gene_string='MRPL1'\n",
      "gene_list=['MRPL1']\n",
      "match=None\n",
      "['MRPL1']:     ENSG00000169288\n",
      "gene_string='MRPL2'\n",
      "gene_list=['MRPL2']\n",
      "match=None\n",
      "['MRPL2']:     ENSG00000112651\n",
      "gene_string='MRPL3'\n",
      "gene_list=['MRPL3']\n",
      "match=None\n",
      "['MRPL3']:     ENSG00000114686\n",
      "gene_string='MRPL4'\n",
      "gene_list=['MRPL4']\n",
      "match=None\n",
      "['MRPL4']:     ENSG00000105364\n",
      "gene_string='MRPL9'\n",
      "gene_list=['MRPL9']\n",
      "match=None\n",
      "['MRPL9']:     ENSG00000143436\n",
      "gene_string='MRPL10'\n",
      "gene_list=['MRPL10']\n",
      "match=None\n",
      "['MRPL10']:     ENSG00000159111\n",
      "gene_string='MRPL13'\n",
      "gene_list=['MRPL13']\n",
      "match=None\n",
      "['MRPL13']:     ENSG00000172172\n",
      "gene_string='MRPL14'\n",
      "gene_list=['MRPL14']\n",
      "match=None\n",
      "['MRPL14']:     ENSG00000180992\n",
      "gene_string='MRPL15'\n",
      "gene_list=['MRPL15']\n",
      "match=None\n",
      "['MRPL15']:     ENSG00000137547\n",
      "gene_string='MRPL16'\n",
      "gene_list=['MRPL16']\n",
      "match=None\n",
      "['MRPL16']:     ENSG00000166902\n",
      "gene_string='MRPL17'\n",
      "gene_list=['MRPL17']\n",
      "match=None\n",
      "['MRPL17']:     ENSG00000158042\n",
      "gene_string='MRPL18'\n",
      "gene_list=['MRPL18']\n",
      "match=None\n",
      "['MRPL18']:     ENSG00000112110\n",
      "gene_string='MRPL19'\n",
      "gene_list=['MRPL19']\n",
      "match=None\n",
      "['MRPL19']:     ENSG00000115364\n",
      "gene_string='MRPL20'\n",
      "gene_list=['MRPL20']\n",
      "match=None\n",
      "['MRPL20']:     ENSG00000242485\n",
      "gene_string='MRPL21'\n",
      "gene_list=['MRPL21']\n",
      "match=None\n",
      "['MRPL21']:     ENSG00000197345\n",
      "gene_string='MRPL22'\n",
      "gene_list=['MRPL22']\n",
      "match=None\n",
      "['MRPL22']:     ENSG00000082515\n",
      "gene_string='MRPL24'\n",
      "gene_list=['MRPL24']\n",
      "match=None\n",
      "['MRPL24']:     ENSG00000143314\n",
      "gene_string='MRPL27'\n",
      "gene_list=['MRPL27']\n",
      "match=None\n",
      "['MRPL27']:     ENSG00000182154\n",
      "gene_string='MRPL28'\n",
      "gene_list=['MRPL28']\n",
      "match=None\n",
      "['MRPL28']:     ENSG00000086504\n",
      "gene_string='MRPL32'\n",
      "gene_list=['MRPL32']\n",
      "match=None\n",
      "['MRPL32']:     ENSG00000106591\n",
      "gene_string='MRPL33'\n",
      "gene_list=['MRPL33']\n",
      "match=None\n",
      "['MRPL33']:     ENSG00000243147\n",
      "gene_string='MRPL34'\n",
      "gene_list=['MRPL34']\n",
      "match=None\n",
      "['MRPL34']:     ENSG00000130312\n",
      "gene_string='MRPL35'\n",
      "gene_list=['MRPL35']\n",
      "match=None\n",
      "['MRPL35']:     ENSG00000132313\n",
      "gene_string='MRPL36'\n",
      "gene_list=['MRPL36']\n",
      "match=None\n",
      "['MRPL36']:     ENSG00000171421\n",
      "gene_string='MRPL37'\n",
      "gene_list=['MRPL37']\n",
      "match=None\n",
      "['MRPL37']:     ENSG00000116221\n",
      "gene_string='MRPL38'\n",
      "gene_list=['MRPL38']\n",
      "match=None\n",
      "['MRPL38']:     ENSG00000204316\n",
      "gene_string='MRPL39'\n",
      "gene_list=['MRPL39']\n",
      "match=None\n",
      "['MRPL39']:     ENSG00000154719\n",
      "gene_string='MRPL41'\n",
      "gene_list=['MRPL41']\n",
      "match=None\n",
      "['MRPL41']:     ENSG00000182154\n",
      "gene_string='MRPL43'\n",
      "gene_list=['MRPL43']\n",
      "match=None\n",
      "['MRPL43']:     ENSG00000055950\n",
      "gene_string='MRPL44'\n",
      "gene_list=['MRPL44']\n",
      "match=None\n",
      "['MRPL44']:     ENSG00000135900\n",
      "gene_string='MRPL45'\n",
      "gene_list=['MRPL45']\n",
      "match=None\n",
      "['MRPL45']:     ENSG00000278845\n",
      "gene_string='MRPL49'\n",
      "gene_list=['MRPL49']\n",
      "match=None\n",
      "['MRPL49']:     ENSG00000149792\n",
      "gene_string='MRPL50'\n",
      "gene_list=['MRPL50']\n",
      "match=None\n",
      "['MRPL50']:     ENSG00000136897\n",
      "gene_string='MRPL51'\n",
      "gene_list=['MRPL51']\n",
      "match=None\n",
      "['MRPL51']:     ENSG00000111639\n",
      "gene_string='MRPL53'\n",
      "gene_list=['MRPL53']\n",
      "match=None\n",
      "['MRPL53']:     ENSG00000204822\n",
      "gene_string='MRPL54'\n",
      "gene_list=['MRPL54']\n",
      "match=None\n",
      "['MRPL54']:     ENSG00000183617\n",
      "gene_string='MRPL55'\n",
      "gene_list=['MRPL55']\n",
      "match=None\n",
      "['MRPL55']:     ENSG00000162910\n",
      "gene_string='MRPS2'\n",
      "gene_list=['MRPS2']\n",
      "match=None\n",
      "['MRPS2']:     ENSG00000122140\n",
      "gene_string='MRPS5'\n",
      "gene_list=['MRPS5']\n",
      "match=None\n",
      "['MRPS5']:     ENSG00000144029\n",
      "gene_string='MRPS6'\n",
      "gene_list=['MRPS6']\n",
      "match=None\n",
      "['MRPS6']:     ENSG00000243927\n",
      "gene_string='MRPS7'\n",
      "gene_list=['MRPS7']\n",
      "match=None\n",
      "['MRPS7']:     ENSG00000125445\n",
      "gene_string='MRPS9'\n",
      "gene_list=['MRPS9']\n",
      "match=None\n",
      "['MRPS9']:     ENSG00000135972\n",
      "gene_string='MRPS10'\n",
      "gene_list=['MRPS10']\n",
      "match=None\n",
      "['MRPS10']:     ENSG00000048544\n",
      "gene_string='MRPS11'\n",
      "gene_list=['MRPS11']\n",
      "match=None\n",
      "['MRPS11']:     ENSG00000181991\n",
      "gene_string='MRPS12'\n",
      "gene_list=['MRPS12']\n",
      "match=None\n",
      "['MRPS12']:     ENSG00000128626\n",
      "gene_string='MRPS14'\n",
      "gene_list=['MRPS14']\n",
      "match=None\n",
      "['MRPS14']:     ENSG00000120333\n",
      "gene_string='MRPS16'\n",
      "gene_list=['MRPS16']\n",
      "match=None\n",
      "['MRPS16']:     ENSG00000182180\n",
      "gene_string='MRPS18A'\n",
      "gene_list=['MRPS18A']\n",
      "match=None\n",
      "['MRPS18A']:     ENSG00000096080\n",
      "gene_string='MRPS18B'\n",
      "gene_list=['MRPS18B']\n",
      "match=None\n",
      "['MRPS18B']:     ENSG00000204568\n",
      "gene_string='MRPS21'\n",
      "gene_list=['MRPS21']\n",
      "match=None\n",
      "['MRPS21']:     ENSG00000266472\n",
      "gene_string='MRPS22'\n",
      "gene_list=['MRPS22']\n",
      "match=None\n",
      "['MRPS22']:     ENSG00000175110\n",
      "gene_string='MRPS23'\n",
      "gene_list=['MRPS23']\n",
      "match=None\n",
      "['MRPS23']:     ENSG00000181610\n",
      "gene_string='MRPS24'\n",
      "gene_list=['MRPS24']\n",
      "match=None\n",
      "['MRPS24']:     ENSG00000062582\n",
      "gene_string='MRPS25'\n",
      "gene_list=['MRPS25']\n",
      "match=None\n",
      "['MRPS25']:     ENSG00000131368\n",
      "gene_string='MRPS26'\n",
      "gene_list=['MRPS26']\n",
      "match=None\n",
      "['MRPS26']:     ENSG00000125901\n",
      "gene_string='MRPS27'\n",
      "gene_list=['MRPS27']\n",
      "match=None\n",
      "['MRPS27']:     ENSG00000113048\n",
      "gene_string='MRPS28'\n",
      "gene_list=['MRPS28']\n",
      "match=None\n",
      "['MRPS28']:     ENSG00000061794\n",
      "gene_string='MRPS31'\n",
      "gene_list=['MRPS31']\n",
      "match=None\n",
      "['MRPS31']:     ENSG00000102738\n",
      "gene_string='MRPS34'\n",
      "gene_list=['MRPS34']\n",
      "match=None\n",
      "['MRPS34']:     ENSG00000074071\n",
      "gene_string='MRPS35'\n",
      "gene_list=['MRPS35']\n",
      "match=None\n",
      "['MRPS35']:     ENSG00000061794\n",
      "gene_string='MRTO4'\n",
      "gene_list=['MRTO4']\n",
      "match=None\n",
      "['MRTO4']:     ENSG00000053372\n",
      "gene_string='MSRB1'\n",
      "gene_list=['MSRB1']\n",
      "match=None\n",
      "['MSRB1']:     ENSG00000198736\n",
      "gene_string='MST1'\n",
      "gene_list=['MST1']\n",
      "match=None\n",
      "['MST1']:     ENSG00000101109\n",
      "gene_string='MSTO1'\n",
      "gene_list=['MSTO1']\n",
      "match=None\n",
      "['MSTO1']:     ENSG00000125459\n",
      "gene_string='MT2A'\n",
      "gene_list=['MT2A']\n",
      "match=None\n",
      "['MT2A']:     ENSG00000125148\n",
      "gene_string='MTBP'\n",
      "gene_list=['MTBP']\n",
      "match=None\n",
      "['MTBP']:     ENSG00000172167\n",
      "gene_string='MTG2'\n",
      "gene_list=['MTG2']\n",
      "match=None\n",
      "['MTG2']:     ENSG00000101181\n",
      "gene_string='MTOR'\n",
      "gene_list=['MTOR']\n",
      "match=None\n",
      "['MTOR']:     ENSG00000198793\n",
      "gene_string='MTPAP'\n",
      "gene_list=['MTPAP']\n",
      "match=None\n",
      "['MTPAP']:     ENSG00000107951\n",
      "gene_string='MTRNR2L1'\n",
      "gene_list=['MTRNR2L1']\n",
      "match=None\n",
      "['MTRNR2L1']:     None\n",
      "gene_string='MVD'\n",
      "gene_list=['MVD']\n",
      "match=None\n",
      "['MVD']:     ENSG00000167508\n",
      "gene_string='MVK'\n",
      "gene_list=['MVK']\n",
      "match=None\n",
      "['MVK']:     ENSG00000110921\n",
      "gene_string='MYBBP1A'\n",
      "gene_list=['MYBBP1A']\n",
      "match=None\n",
      "['MYBBP1A']:     ENSG00000132382\n",
      "gene_string='MYBL2'\n",
      "gene_list=['MYBL2']\n",
      "match=None\n",
      "['MYBL2']:     ENSG00000101057\n",
      "gene_string='MYC'\n",
      "gene_list=['MYC']\n",
      "match=None\n",
      "['MYC']:     ENSG00000136997\n",
      "gene_string='MYCBP'\n",
      "gene_list=['MYCBP']\n",
      "match=None\n",
      "['MYCBP']:     ENSG00000214114\n",
      "gene_string='MYO1H'\n",
      "gene_list=['MYO1H']\n",
      "match=None\n",
      "['MYO1H']:     ENSG00000174527\n",
      "gene_string='MZF1'\n",
      "gene_list=['MZF1']\n",
      "match=None\n",
      "['MZF1']:     ENSG00000099326\n",
      "gene_string='MZT1'\n",
      "gene_list=['MZT1']\n",
      "match=None\n",
      "['MZT1']:     ENSG00000204899\n",
      "gene_string='N6AMT1'\n",
      "gene_list=['N6AMT1']\n",
      "match=None\n",
      "['N6AMT1']:     ENSG00000156239\n",
      "gene_string='NAA10'\n",
      "gene_list=['NAA10']\n",
      "match=None\n",
      "['NAA10']:     ENSG00000102030\n",
      "gene_string='NAA15'\n",
      "gene_list=['NAA15']\n",
      "match=None\n",
      "['NAA15']:     ENSG00000164134\n",
      "gene_string='NAA20'\n",
      "gene_list=['NAA20']\n",
      "match=None\n",
      "['NAA20']:     ENSG00000173418\n",
      "gene_string='NAA25'\n",
      "gene_list=['NAA25']\n",
      "match=None\n",
      "['NAA25']:     ENSG00000111300\n",
      "gene_string='NAA35'\n",
      "gene_list=['NAA35']\n",
      "match=None\n",
      "['NAA35']:     ENSG00000135040\n",
      "gene_string='NAA38'\n",
      "gene_list=['NAA38']\n",
      "match=None\n",
      "['NAA38']:     ENSG00000183011\n",
      "gene_string='NAA50'\n",
      "gene_list=['NAA50']\n",
      "match=None\n",
      "['NAA50']:     ENSG00000121579\n",
      "gene_string='NACA'\n",
      "gene_list=['NACA']\n",
      "match=None\n",
      "['NACA']:     ENSG00000196531\n",
      "gene_string='NAE1'\n",
      "gene_list=['NAE1']\n",
      "match=None\n",
      "['NAE1']:     ENSG00000159593\n",
      "gene_string='NAF1'\n",
      "gene_list=['NAF1']\n",
      "match=None\n",
      "['NAF1']:     ENSG00000145901\n",
      "gene_string='NAGLU'\n",
      "gene_list=['NAGLU']\n",
      "match=None\n",
      "['NAGLU']:     ENSG00000108784\n",
      "gene_string='NANOG'\n",
      "gene_list=['NANOG']\n",
      "match=None\n",
      "['NANOG']:     ENSG00000111704\n",
      "gene_string='NAPA'\n",
      "gene_list=['NAPA']\n",
      "match=None\n",
      "['NAPA']:     ENSG00000105402\n",
      "gene_string='NAPG'\n",
      "gene_list=['NAPG']\n",
      "match=None\n",
      "['NAPG']:     ENSG00000134265\n",
      "gene_string='NARS'\n",
      "gene_list=['NARS']\n",
      "match=None\n",
      "['NARS']:     ENSG00000134440\n",
      "gene_string='NASP'\n",
      "gene_list=['NASP']\n",
      "match=None\n",
      "['NASP']:     ENSG00000132780\n",
      "gene_string='NAT10'\n",
      "gene_list=['NAT10']\n",
      "match=None\n",
      "['NAT10']:     ENSG00000135372\n",
      "gene_string='NBAS'\n",
      "gene_list=['NBAS']\n",
      "match=None\n",
      "['NBAS']:     ENSG00000151779\n",
      "gene_string='NBEAL1'\n",
      "gene_list=['NBEAL1']\n",
      "match=None\n",
      "['NBEAL1']:     ENSG00000144426\n",
      "gene_string='NBPF3'\n",
      "gene_list=['NBPF3']\n",
      "match=None\n",
      "['NBPF3']:     ENSG00000142794\n",
      "gene_string='NBPF12'\n",
      "gene_list=['NBPF12']\n",
      "match=None\n",
      "['NBPF12']:     ENSG00000268043\n",
      "gene_string='NBPF15'\n",
      "gene_list=['NBPF15']\n",
      "match=None\n",
      "['NBPF15']:     ENSG00000266338\n",
      "gene_string='NCAPD2'\n",
      "gene_list=['NCAPD2']\n",
      "match=None\n",
      "['NCAPD2']:     ENSG00000010292\n",
      "gene_string='NCAPD3'\n",
      "gene_list=['NCAPD3']\n",
      "match=None\n",
      "['NCAPD3']:     ENSG00000151503\n",
      "gene_string='NCAPG'\n",
      "gene_list=['NCAPG']\n",
      "match=None\n",
      "['NCAPG']:     ENSG00000109805\n",
      "gene_string='NCAPG2'\n",
      "gene_list=['NCAPG2']\n",
      "match=None\n",
      "['NCAPG2']:     ENSG00000146918\n",
      "gene_string='NCAPH'\n",
      "gene_list=['NCAPH']\n",
      "match=None\n",
      "['NCAPH']:     ENSG00000121152\n",
      "gene_string='NCAPH2'\n",
      "gene_list=['NCAPH2']\n",
      "match=None\n",
      "['NCAPH2']:     ENSG00000025770\n",
      "gene_string='NCBP1'\n",
      "gene_list=['NCBP1']\n",
      "match=None\n",
      "['NCBP1']:     ENSG00000136937\n",
      "gene_string='NCBP2'\n",
      "gene_list=['NCBP2']\n",
      "match=None\n",
      "['NCBP2']:     ENSG00000114503\n",
      "gene_string='NCKAP1'\n",
      "gene_list=['NCKAP1']\n",
      "match=None\n",
      "['NCKAP1']:     ENSG00000061676\n",
      "gene_string='NCL'\n",
      "gene_list=['NCL']\n",
      "match=None\n",
      "['NCL']:     ENSG00000115053\n",
      "gene_string='NCOA4'\n",
      "gene_list=['NCOA4']\n",
      "match=None\n",
      "['NCOA4']:     ENSG00000266412\n",
      "gene_string='NDC1'\n",
      "gene_list=['NDC1']\n",
      "match=None\n",
      "['NDC1']:     ENSG00000058804\n",
      "gene_string='NDC80'\n",
      "gene_list=['NDC80']\n",
      "match=None\n",
      "['NDC80']:     ENSG00000080986\n",
      "gene_string='NDOR1'\n",
      "gene_list=['NDOR1']\n",
      "match=None\n",
      "['NDOR1']:     ENSG00000188566\n",
      "gene_string='NDUFA2'\n",
      "gene_list=['NDUFA2']\n",
      "match=None\n",
      "['NDUFA2']:     ENSG00000131495\n",
      "gene_string='NDUFA4'\n",
      "gene_list=['NDUFA4']\n",
      "match=None\n",
      "['NDUFA4']:     ENSG00000189043\n",
      "gene_string='NDUFA6'\n",
      "gene_list=['NDUFA6']\n",
      "match=None\n",
      "['NDUFA6']:     ENSG00000184983\n",
      "gene_string='NDUFA8'\n",
      "gene_list=['NDUFA8']\n",
      "match=None\n",
      "['NDUFA8']:     ENSG00000119421\n",
      "gene_string='NDUFA11'\n",
      "gene_list=['NDUFA11']\n",
      "match=None\n",
      "['NDUFA11']:     ENSG00000174886\n",
      "gene_string='NDUFAB1'\n",
      "gene_list=['NDUFAB1']\n",
      "match=None\n",
      "['NDUFAB1']:     ENSG00000004779\n",
      "gene_string='NDUFAF3'\n",
      "gene_list=['NDUFAF3']\n",
      "match=None\n",
      "['NDUFAF3']:     ENSG00000178057\n",
      "gene_string='NDUFAF7'\n",
      "gene_list=['NDUFAF7']\n",
      "match=None\n",
      "['NDUFAF7']:     ENSG00000003509\n",
      "gene_string='NDUFB3'\n",
      "gene_list=['NDUFB3']\n",
      "match=None\n",
      "['NDUFB3']:     ENSG00000119013\n",
      "gene_string='NDUFB4'\n",
      "gene_list=['NDUFB4']\n",
      "match=None\n",
      "['NDUFB4']:     ENSG00000065518\n",
      "gene_string='NDUFB6'\n",
      "gene_list=['NDUFB6']\n",
      "match=None\n",
      "['NDUFB6']:     ENSG00000165264\n",
      "gene_string='NDUFB7'\n",
      "gene_list=['NDUFB7']\n",
      "match=None\n",
      "['NDUFB7']:     ENSG00000099795\n",
      "gene_string='NDUFB8'\n",
      "gene_list=['NDUFB8']\n",
      "match=None\n",
      "['NDUFB8']:     ENSG00000166136\n",
      "gene_string='NDUFB10'\n",
      "gene_list=['NDUFB10']\n",
      "match=None\n",
      "['NDUFB10']:     ENSG00000140990\n",
      "gene_string='NDUFS3'\n",
      "gene_list=['NDUFS3']\n",
      "match=None\n",
      "['NDUFS3']:     ENSG00000213619\n",
      "gene_string='NDUFS5'\n",
      "gene_list=['NDUFS5']\n",
      "match=None\n",
      "['NDUFS5']:     ENSG00000168653\n",
      "gene_string='NEDD1'\n",
      "gene_list=['NEDD1']\n",
      "match=None\n",
      "['NEDD1']:     ENSG00000139350\n",
      "gene_string='NEDD8'\n",
      "gene_list=['NEDD8']\n",
      "match=None\n",
      "['NEDD8']:     ENSG00000129559\n",
      "gene_string='NELFA'\n",
      "gene_list=['NELFA']\n",
      "match=None\n",
      "['NELFA']:     ENSG00000185049\n",
      "gene_string='NELFB'\n",
      "gene_list=['NELFB']\n",
      "match=None\n",
      "['NELFB']:     ENSG00000188986\n",
      "gene_string='NELFCD'\n",
      "gene_list=['NELFCD']\n",
      "match=None\n",
      "['NELFCD']:     ENSG00000101158\n",
      "gene_string='NELFE'\n",
      "gene_list=['NELFE']\n",
      "match=None\n",
      "['NELFE']:     ENSG00000204356\n",
      "gene_string='NEMF'\n",
      "gene_list=['NEMF']\n",
      "match=None\n",
      "['NEMF']:     ENSG00000165525\n",
      "gene_string='NEPRO'\n",
      "gene_list=['NEPRO']\n",
      "match=None\n",
      "['NEPRO']:     ENSG00000163608\n",
      "gene_string='NFATC2IP'\n",
      "gene_list=['NFATC2IP']\n",
      "match=None\n",
      "['NFATC2IP']:     ENSG00000176953\n",
      "gene_string='NFKBIE'\n",
      "gene_list=['NFKBIE']\n",
      "match=None\n",
      "['NFKBIE']:     ENSG00000146232\n",
      "gene_string='NFRKB'\n",
      "gene_list=['NFRKB']\n",
      "match=None\n",
      "['NFRKB']:     ENSG00000170322\n",
      "gene_string='NFS1'\n",
      "gene_list=['NFS1']\n",
      "match=None\n",
      "['NFS1']:     ENSG00000244005\n",
      "gene_string='NFYB'\n",
      "gene_list=['NFYB']\n",
      "match=None\n",
      "['NFYB']:     ENSG00000120837\n",
      "gene_string='NFYC'\n",
      "gene_list=['NFYC']\n",
      "match=None\n",
      "['NFYC']:     ENSG00000066136\n",
      "gene_string='NHLRC2'\n",
      "gene_list=['NHLRC2']\n",
      "match=None\n",
      "['NHLRC2']:     ENSG00000196865\n",
      "gene_string='NHP2'\n",
      "gene_list=['NHP2']\n",
      "match=None\n",
      "['NHP2']:     ENSG00000145912\n",
      "gene_string='NIFK'\n",
      "gene_list=['NIFK']\n",
      "match=None\n",
      "['NIFK']:     ENSG00000155438\n",
      "gene_string='NIP7'\n",
      "gene_list=['NIP7']\n",
      "match=None\n",
      "['NIP7']:     ENSG00000132603\n",
      "gene_string='NISCH'\n",
      "gene_list=['NISCH']\n",
      "match=None\n",
      "['NISCH']:     ENSG00000010322\n",
      "gene_string='NKAP'\n",
      "gene_list=['NKAP']\n",
      "match=None\n",
      "['NKAP']:     ENSG00000101882\n",
      "gene_string='NLE1'\n",
      "gene_list=['NLE1']\n",
      "match=None\n",
      "['NLE1']:     ENSG00000073536\n",
      "gene_string='NMD3'\n",
      "gene_list=['NMD3']\n",
      "match=None\n",
      "['NMD3']:     ENSG00000169251\n",
      "gene_string='NMT1'\n",
      "gene_list=['NMT1']\n",
      "match=None\n",
      "['NMT1']:     ENSG00000136448\n",
      "gene_string='NOB1'\n",
      "gene_list=['NOB1']\n",
      "match=None\n",
      "['NOB1']:     ENSG00000141101\n",
      "gene_string='NOC2L'\n",
      "gene_list=['NOC2L']\n",
      "match=None\n",
      "['NOC2L']:     ENSG00000188976\n",
      "gene_string='NOC3L'\n",
      "gene_list=['NOC3L']\n",
      "match=None\n",
      "['NOC3L']:     ENSG00000173145\n",
      "gene_string='NOC4L'\n",
      "gene_list=['NOC4L']\n",
      "match=None\n",
      "['NOC4L']:     ENSG00000184967\n",
      "gene_string='NOL6'\n",
      "gene_list=['NOL6']\n",
      "match=None\n",
      "['NOL6']:     ENSG00000165271\n",
      "gene_string='NOL7'\n",
      "gene_list=['NOL7']\n",
      "match=None\n",
      "['NOL7']:     ENSG00000225921\n",
      "gene_string='NOL8'\n",
      "gene_list=['NOL8']\n",
      "match=None\n",
      "['NOL8']:     ENSG00000198000\n",
      "gene_string='NOL9'\n",
      "gene_list=['NOL9']\n",
      "match=None\n",
      "['NOL9']:     ENSG00000162408\n",
      "gene_string='NOL10'\n",
      "gene_list=['NOL10']\n",
      "match=None\n",
      "['NOL10']:     ENSG00000115761\n",
      "gene_string='NOL11'\n",
      "gene_list=['NOL11']\n",
      "match=None\n",
      "['NOL11']:     ENSG00000130935\n",
      "gene_string='NOL12'\n",
      "gene_list=['NOL12']\n",
      "match=None\n",
      "['NOL12']:     ENSG00000273899\n",
      "gene_string='NOLC1'\n",
      "gene_list=['NOLC1']\n",
      "match=None\n",
      "['NOLC1']:     ENSG00000166197\n",
      "gene_string='NOM1'\n",
      "gene_list=['NOM1']\n",
      "match=None\n",
      "['NOM1']:     ENSG00000146909\n",
      "gene_string='NOMO3'\n",
      "gene_list=['NOMO3']\n",
      "match=None\n",
      "['NOMO3']:     ENSG00000103226\n",
      "gene_string='NOP2'\n",
      "gene_list=['NOP2']\n",
      "match=None\n",
      "['NOP2']:     ENSG00000111641\n",
      "gene_string='NOP9'\n",
      "gene_list=['NOP9']\n",
      "match=None\n",
      "['NOP9']:     ENSG00000196943\n",
      "gene_string='NOP10'\n",
      "gene_list=['NOP10']\n",
      "match=None\n",
      "['NOP10']:     ENSG00000182117\n",
      "gene_string='NOP14'\n",
      "gene_list=['NOP14']\n",
      "match=None\n",
      "['NOP14']:     ENSG00000087269\n",
      "gene_string='NOP16'\n",
      "gene_list=['NOP16']\n",
      "match=None\n",
      "['NOP16']:     ENSG00000048162\n",
      "gene_string='NOP56'\n",
      "gene_list=['NOP56']\n",
      "match=None\n",
      "['NOP56']:     ENSG00000101361\n",
      "gene_string='NOP58'\n",
      "gene_list=['NOP58']\n",
      "match=None\n",
      "['NOP58']:     ENSG00000055044\n",
      "gene_string='NOS1AP'\n",
      "gene_list=['NOS1AP']\n",
      "match=None\n",
      "['NOS1AP']:     ENSG00000198929\n",
      "gene_string='NPAT'\n",
      "gene_list=['NPAT']\n",
      "match=None\n",
      "['NPAT']:     ENSG00000149308\n",
      "gene_string='NPB'\n",
      "gene_list=['NPB']\n",
      "match=None\n",
      "['NPB']:     ENSG00000183979\n",
      "gene_string='NPEPPS'\n",
      "gene_list=['NPEPPS']\n",
      "match=None\n",
      "['NPEPPS']:     ENSG00000141279\n",
      "gene_string='NPLOC4'\n",
      "gene_list=['NPLOC4']\n",
      "match=None\n",
      "['NPLOC4']:     ENSG00000182446\n",
      "gene_string='NPM1'\n",
      "gene_list=['NPM1']\n",
      "match=None\n",
      "['NPM1']:     ENSG00000181163\n",
      "gene_string='NPM3'\n",
      "gene_list=['NPM3']\n",
      "match=None\n",
      "['NPM3']:     ENSG00000107833\n",
      "gene_string='NR2C2AP'\n",
      "gene_list=['NR2C2AP']\n",
      "match=None\n",
      "['NR2C2AP']:     ENSG00000184162\n",
      "gene_string='NRBP1'\n",
      "gene_list=['NRBP1']\n",
      "match=None\n",
      "['NRBP1']:     ENSG00000115216\n",
      "gene_string='NRDE2'\n",
      "gene_list=['NRDE2']\n",
      "match=None\n",
      "['NRDE2']:     ENSG00000119720\n",
      "gene_string='NRF1'\n",
      "gene_list=['NRF1']\n",
      "match=None\n",
      "['NRF1']:     ENSG00000082641\n",
      "gene_string='NSA2'\n",
      "gene_list=['NSA2']\n",
      "match=None\n",
      "['NSA2']:     ENSG00000164346\n",
      "gene_string='NSF'\n",
      "gene_list=['NSF']\n",
      "match=None\n",
      "['NSF']:     ENSG00000073969\n",
      "gene_string='NSL1'\n",
      "gene_list=['NSL1']\n",
      "match=None\n",
      "['NSL1']:     ENSG00000120071\n",
      "gene_string='NSMCE1'\n",
      "gene_list=['NSMCE1']\n",
      "match=None\n",
      "['NSMCE1']:     ENSG00000169189\n",
      "gene_string='NSMCE2'\n",
      "gene_list=['NSMCE2']\n",
      "match=None\n",
      "['NSMCE2']:     ENSG00000156831\n",
      "gene_string='NSRP1'\n",
      "gene_list=['NSRP1']\n",
      "match=None\n",
      "['NSRP1']:     ENSG00000126653\n",
      "gene_string='NSUN4'\n",
      "gene_list=['NSUN4']\n",
      "match=None\n",
      "['NSUN4']:     ENSG00000117481\n",
      "gene_string='NUBP1'\n",
      "gene_list=['NUBP1']\n",
      "match=None\n",
      "['NUBP1']:     ENSG00000103274\n",
      "gene_string='NUBP2'\n",
      "gene_list=['NUBP2']\n",
      "match=None\n",
      "['NUBP2']:     ENSG00000095906\n",
      "gene_string='NUDC'\n",
      "gene_list=['NUDC']\n",
      "match=None\n",
      "['NUDC']:     ENSG00000090273\n",
      "gene_string='NUDCD3'\n",
      "gene_list=['NUDCD3']\n",
      "match=None\n",
      "['NUDCD3']:     ENSG00000015676\n",
      "gene_string='NUDT21'\n",
      "gene_list=['NUDT21']\n",
      "match=None\n",
      "['NUDT21']:     ENSG00000167005\n",
      "gene_string='NUF2'\n",
      "gene_list=['NUF2']\n",
      "match=None\n",
      "['NUF2']:     ENSG00000143228\n",
      "gene_string='NUFIP1'\n",
      "gene_list=['NUFIP1']\n",
      "match=None\n",
      "['NUFIP1']:     ENSG00000083635\n",
      "gene_string='NUMA1'\n",
      "gene_list=['NUMA1']\n",
      "match=None\n",
      "['NUMA1']:     ENSG00000137497\n",
      "gene_string='NUP35'\n",
      "gene_list=['NUP35']\n",
      "match=None\n",
      "['NUP35']:     ENSG00000163002\n",
      "gene_string='NUP43'\n",
      "gene_list=['NUP43']\n",
      "match=None\n",
      "['NUP43']:     ENSG00000120253\n",
      "gene_string='NUP54'\n",
      "gene_list=['NUP54']\n",
      "match=None\n",
      "['NUP54']:     ENSG00000138750\n",
      "gene_string='NUP62'\n",
      "gene_list=['NUP62']\n",
      "match=None\n",
      "['NUP62']:     ENSG00000213024\n",
      "gene_string='NUP85'\n",
      "gene_list=['NUP85']\n",
      "match=None\n",
      "['NUP85']:     ENSG00000125450\n",
      "gene_string='NUP88'\n",
      "gene_list=['NUP88']\n",
      "match=None\n",
      "['NUP88']:     ENSG00000108559\n",
      "gene_string='NUP93'\n",
      "gene_list=['NUP93']\n",
      "match=None\n",
      "['NUP93']:     ENSG00000102900\n",
      "gene_string='NUP98'\n",
      "gene_list=['NUP98']\n",
      "match=None\n",
      "['NUP98']:     ENSG00000110713\n",
      "gene_string='NUP107'\n",
      "gene_list=['NUP107']\n",
      "match=None\n",
      "['NUP107']:     ENSG00000111581\n",
      "gene_string='NUP133'\n",
      "gene_list=['NUP133']\n",
      "match=None\n",
      "['NUP133']:     ENSG00000069248\n",
      "gene_string='NUP153'\n",
      "gene_list=['NUP153']\n",
      "match=None\n",
      "['NUP153']:     ENSG00000124789\n",
      "gene_string='NUP155'\n",
      "gene_list=['NUP155']\n",
      "match=None\n",
      "['NUP155']:     ENSG00000113569\n",
      "gene_string='NUP160'\n",
      "gene_list=['NUP160']\n",
      "match=None\n",
      "['NUP160']:     ENSG00000030066\n",
      "gene_string='NUP205'\n",
      "gene_list=['NUP205']\n",
      "match=None\n",
      "['NUP205']:     ENSG00000155561\n",
      "gene_string='NUP214'\n",
      "gene_list=['NUP214']\n",
      "match=None\n",
      "['NUP214']:     ENSG00000126883\n",
      "gene_string='NUS1'\n",
      "gene_list=['NUS1']\n",
      "match=None\n",
      "['NUS1']:     ENSG00000153989\n",
      "gene_string='NUTF2'\n",
      "gene_list=['NUTF2']\n",
      "match=None\n",
      "['NUTF2']:     ENSG00000102898\n",
      "gene_string='NVL'\n",
      "gene_list=['NVL']\n",
      "match=None\n",
      "['NVL']:     ENSG00000143748\n",
      "gene_string='NXF1'\n",
      "gene_list=['NXF1']\n",
      "match=None\n",
      "['NXF1']:     ENSG00000162231\n",
      "gene_string='OGFOD1'\n",
      "gene_list=['OGFOD1']\n",
      "match=None\n",
      "['OGFOD1']:     ENSG00000087263\n",
      "gene_string='OGT'\n",
      "gene_list=['OGT']\n",
      "match=None\n",
      "['OGT']:     ENSG00000147162\n",
      "gene_string='OIP5'\n",
      "gene_list=['OIP5']\n",
      "match=None\n",
      "['OIP5']:     ENSG00000104147\n",
      "gene_string='OLFML3'\n",
      "gene_list=['OLFML3']\n",
      "match=None\n",
      "['OLFML3']:     ENSG00000116774\n",
      "gene_string='OPA1'\n",
      "gene_list=['OPA1']\n",
      "match=None\n",
      "['OPA1']:     ENSG00000198836\n",
      "gene_string='OR1D2'\n",
      "gene_list=['OR1D2']\n",
      "match=None\n",
      "['OR1D2']:     ENSG00000184166\n",
      "gene_string='OR1E2'\n",
      "gene_list=['OR1E2']\n",
      "match=None\n",
      "['OR1E2']:     ENSG00000127780\n",
      "gene_string='OR2A1'\n",
      "gene_list=['OR2A1']\n",
      "match=None\n",
      "['OR2A1']:     ENSG00000212807\n",
      "gene_string='OR2T5'\n",
      "gene_list=['OR2T5']\n",
      "match=None\n",
      "['OR2T5']:     ENSG00000182783\n",
      "gene_string='OR2T29'\n",
      "gene_list=['OR2T29']\n",
      "match=None\n",
      "['OR2T29']:     ENSG00000203661\n",
      "gene_string='OR4F4'\n",
      "gene_list=['OR4F4']\n",
      "match=None\n",
      "['OR4F4']:     ENSG00000177693\n",
      "gene_string='OR4K1'\n",
      "gene_list=['OR4K1']\n",
      "match=None\n",
      "['OR4K1']:     ENSG00000155249\n",
      "gene_string='OR4N2'\n",
      "gene_list=['OR4N2']\n",
      "match=None\n",
      "['OR4N2']:     ENSG00000176294\n",
      "gene_string='OR7A10'\n",
      "gene_list=['OR7A10']\n",
      "match=None\n",
      "['OR7A10']:     ENSG00000127515\n",
      "gene_string='OR7A17'\n",
      "gene_list=['OR7A17']\n",
      "match=None\n",
      "['OR7A17']:     ENSG00000185385\n",
      "gene_string='ORC1'\n",
      "gene_list=['ORC1']\n",
      "match=None\n",
      "['ORC1']:     ENSG00000102743\n",
      "gene_string='ORC3'\n",
      "gene_list=['ORC3']\n",
      "match=None\n",
      "['ORC3']:     ENSG00000135336\n",
      "gene_string='ORC4'\n",
      "gene_list=['ORC4']\n",
      "match=None\n",
      "['ORC4']:     ENSG00000115947\n",
      "gene_string='ORC5'\n",
      "gene_list=['ORC5']\n",
      "match=None\n",
      "['ORC5']:     ENSG00000164815\n",
      "gene_string='ORC6'\n",
      "gene_list=['ORC6']\n",
      "match=None\n",
      "['ORC6']:     ENSG00000091651\n",
      "gene_string='OSBP'\n",
      "gene_list=['OSBP']\n",
      "match=None\n",
      "['OSBP']:     ENSG00000110048\n",
      "gene_string='OSGEP'\n",
      "gene_list=['OSGEP']\n",
      "match=None\n",
      "['OSGEP']:     ENSG00000092094\n",
      "gene_string='OSTC'\n",
      "gene_list=['OSTC']\n",
      "match=None\n",
      "['OSTC']:     ENSG00000198856\n",
      "gene_string='OTOP1'\n",
      "gene_list=['OTOP1']\n",
      "match=None\n",
      "['OTOP1']:     ENSG00000163982\n",
      "gene_string='OTX1'\n",
      "gene_list=['OTX1']\n",
      "match=None\n",
      "['OTX1']:     ENSG00000115507\n",
      "gene_string='OVCA2'\n",
      "gene_list=['OVCA2']\n",
      "match=None\n",
      "['OVCA2']:     ENSG00000262664\n",
      "gene_string='OXA1L'\n",
      "gene_list=['OXA1L']\n",
      "match=None\n",
      "['OXA1L']:     ENSG00000155463\n",
      "gene_string='PABPC1'\n",
      "gene_list=['PABPC1']\n",
      "match=None\n",
      "['PABPC1']:     ENSG00000070756\n",
      "gene_string='PABPN1'\n",
      "gene_list=['PABPN1']\n",
      "match=None\n",
      "['PABPN1']:     ENSG00000100836\n",
      "gene_string='PAF1'\n",
      "gene_list=['PAF1']\n",
      "match=None\n",
      "['PAF1']:     ENSG00000006712\n",
      "gene_string='PAFAH1B1'\n",
      "gene_list=['PAFAH1B1']\n",
      "match=None\n",
      "['PAFAH1B1']:     ENSG00000007168\n",
      "gene_string='PAK1IP1'\n",
      "gene_list=['PAK1IP1']\n",
      "match=None\n",
      "['PAK1IP1']:     ENSG00000111845\n",
      "gene_string='PALB2'\n",
      "gene_list=['PALB2']\n",
      "match=None\n",
      "['PALB2']:     ENSG00000083093\n",
      "gene_string='PAM16'\n",
      "gene_list=['PAM16']\n",
      "match=None\n",
      "['PAM16']:     ENSG00000103426\n",
      "gene_string='PARN'\n",
      "gene_list=['PARN']\n",
      "match=None\n",
      "['PARN']:     ENSG00000140694\n",
      "gene_string='PARS2'\n",
      "gene_list=['PARS2']\n",
      "match=None\n",
      "['PARS2']:     ENSG00000162396\n",
      "gene_string='PAXBP1'\n",
      "gene_list=['PAXBP1']\n",
      "match=None\n",
      "['PAXBP1']:     ENSG00000159086\n",
      "gene_string='PC'\n",
      "gene_list=['PC']\n",
      "match=None\n",
      "['PC']:     ENSG00000173599\n",
      "gene_string='PCBP1'\n",
      "gene_list=['PCBP1']\n",
      "match=None\n",
      "['PCBP1']:     ENSG00000169564\n",
      "gene_string='PCBP2'\n",
      "gene_list=['PCBP2']\n",
      "match=None\n",
      "['PCBP2']:     ENSG00000197111\n",
      "gene_string='PCF11'\n",
      "gene_list=['PCF11']\n",
      "match=None\n",
      "['PCF11']:     ENSG00000165494\n",
      "gene_string='PCID2'\n",
      "gene_list=['PCID2']\n",
      "match=None\n",
      "['PCID2']:     ENSG00000126226\n",
      "gene_string='PCNA'\n",
      "gene_list=['PCNA']\n",
      "match=None\n",
      "['PCNA']:     ENSG00000132646\n",
      "gene_string='PDAP1'\n",
      "gene_list=['PDAP1']\n",
      "match=None\n",
      "['PDAP1']:     ENSG00000106244\n",
      "gene_string='PDCD2'\n",
      "gene_list=['PDCD2']\n",
      "match=None\n",
      "['PDCD2']:     ENSG00000071994\n",
      "gene_string='PDCD5'\n",
      "gene_list=['PDCD5']\n",
      "match=None\n",
      "['PDCD5']:     ENSG00000105185\n",
      "gene_string='PDCD6'\n",
      "gene_list=['PDCD6']\n",
      "match=None\n",
      "['PDCD6']:     ENSG00000249915\n",
      "gene_string='PDCD6IP'\n",
      "gene_list=['PDCD6IP']\n",
      "match=None\n",
      "['PDCD6IP']:     ENSG00000170248\n",
      "gene_string='PDCD7'\n",
      "gene_list=['PDCD7']\n",
      "match=None\n",
      "['PDCD7']:     ENSG00000090470\n",
      "gene_string='PDCD11'\n",
      "gene_list=['PDCD11']\n",
      "match=None\n",
      "['PDCD11']:     ENSG00000148843\n",
      "gene_string='PDPK1'\n",
      "gene_list=['PDPK1']\n",
      "match=None\n",
      "['PDPK1']:     ENSG00000140992\n",
      "gene_string='PDRG1'\n",
      "gene_list=['PDRG1']\n",
      "match=None\n",
      "['PDRG1']:     ENSG00000088356\n",
      "gene_string='PEF1'\n",
      "gene_list=['PEF1']\n",
      "match=None\n",
      "['PEF1']:     ENSG00000162517\n",
      "gene_string='PELO'\n",
      "gene_list=['PELO']\n",
      "match=None\n",
      "['PELO']:     ENSG00000152684\n",
      "gene_string='PELP1'\n",
      "gene_list=['PELP1']\n",
      "match=None\n",
      "['PELP1']:     ENSG00000141456\n",
      "gene_string='PES1'\n",
      "gene_list=['PES1']\n",
      "match=None\n",
      "['PES1']:     ENSG00000100029\n",
      "gene_string='PET117'\n",
      "gene_list=['PET117']\n",
      "match=None\n",
      "['PET117']:     ENSG00000232838\n",
      "gene_string='PEX1'\n",
      "gene_list=['PEX1']\n",
      "match=None\n",
      "['PEX1']:     ENSG00000127980\n",
      "gene_string='PFDN1'\n",
      "gene_list=['PFDN1']\n",
      "match=None\n",
      "['PFDN1']:     ENSG00000113068\n",
      "gene_string='PFDN2'\n",
      "gene_list=['PFDN2']\n",
      "match=None\n",
      "['PFDN2']:     ENSG00000143256\n",
      "gene_string='PFDN4'\n",
      "gene_list=['PFDN4']\n",
      "match=None\n",
      "['PFDN4']:     ENSG00000101132\n",
      "gene_string='PFDN5'\n",
      "gene_list=['PFDN5']\n",
      "match=None\n",
      "['PFDN5']:     ENSG00000123349\n",
      "gene_string='PFDN6'\n",
      "gene_list=['PFDN6']\n",
      "match=None\n",
      "['PFDN6']:     ENSG00000204220\n",
      "gene_string='PFN1'\n",
      "gene_list=['PFN1']\n",
      "match=None\n",
      "['PFN1']:     ENSG00000108518\n",
      "gene_string='PGAM1'\n",
      "gene_list=['PGAM1']\n",
      "match=None\n",
      "['PGAM1']:     ENSG00000226784\n",
      "gene_string='PGD'\n",
      "gene_list=['PGD']\n",
      "match=None\n",
      "['PGD']:     ENSG00000142657\n",
      "gene_string='PGK1'\n",
      "gene_list=['PGK1']\n",
      "match=None\n",
      "['PGK1']:     ENSG00000102144\n",
      "gene_string='PGPEP1'\n",
      "gene_list=['PGPEP1']\n",
      "match=None\n",
      "['PGPEP1']:     ENSG00000130517\n",
      "gene_string='PGS1'\n",
      "gene_list=['PGS1']\n",
      "match=None\n",
      "['PGS1']:     ENSG00000087157\n",
      "gene_string='PHAX'\n",
      "gene_list=['PHAX']\n",
      "match=None\n",
      "['PHAX']:     ENSG00000164902\n",
      "gene_string='PHB'\n",
      "gene_list=['PHB']\n",
      "match=None\n",
      "['PHB']:     ENSG00000167085\n",
      "gene_string='PHB2'\n",
      "gene_list=['PHB2']\n",
      "match=None\n",
      "['PHB2']:     ENSG00000215021\n",
      "gene_string='PHF5A'\n",
      "gene_list=['PHF5A']\n",
      "match=None\n",
      "['PHF5A']:     ENSG00000100410\n",
      "gene_string='PHF12'\n",
      "gene_list=['PHF12']\n",
      "match=None\n",
      "['PHF12']:     ENSG00000109118\n",
      "gene_string='PI4KA'\n",
      "gene_list=['PI4KA']\n",
      "match=None\n",
      "['PI4KA']:     ENSG00000241973\n",
      "gene_string='PIAS4'\n",
      "gene_list=['PIAS4']\n",
      "match=None\n",
      "['PIAS4']:     ENSG00000105229\n",
      "gene_string='PIGH'\n",
      "gene_list=['PIGH']\n",
      "match=None\n",
      "['PIGH']:     ENSG00000100564\n",
      "gene_string='PIK3C3'\n",
      "gene_list=['PIK3C3']\n",
      "match=None\n",
      "['PIK3C3']:     ENSG00000078142\n",
      "gene_string='PIK3R4'\n",
      "gene_list=['PIK3R4']\n",
      "match=None\n",
      "['PIK3R4']:     ENSG00000196455\n",
      "gene_string='PKM'\n",
      "gene_list=['PKM']\n",
      "match=None\n",
      "['PKM']:     ENSG00000067225\n",
      "gene_string='PKMYT1'\n",
      "gene_list=['PKMYT1']\n",
      "match=None\n",
      "['PKMYT1']:     ENSG00000127564\n",
      "gene_string='PLA2G10'\n",
      "gene_list=['PLA2G10']\n",
      "match=None\n",
      "['PLA2G10']:     ENSG00000069764\n",
      "gene_string='PLEKHN1'\n",
      "gene_list=['PLEKHN1']\n",
      "match=None\n",
      "['PLEKHN1']:     ENSG00000187583\n",
      "gene_string='PLK1'\n",
      "gene_list=['PLK1']\n",
      "match=None\n",
      "['PLK1']:     ENSG00000166851\n",
      "gene_string='PLK4'\n",
      "gene_list=['PLK4']\n",
      "match=None\n",
      "['PLK4']:     ENSG00000142731\n",
      "gene_string='PLRG1'\n",
      "gene_list=['PLRG1']\n",
      "match=None\n",
      "['PLRG1']:     ENSG00000171566\n",
      "gene_string='PMF1'\n",
      "gene_list=['PMF1']\n",
      "match=None\n",
      "['PMF1']:     ENSG00000160783\n",
      "gene_string='PMPCA'\n",
      "gene_list=['PMPCA']\n",
      "match=None\n",
      "['PMPCA']:     ENSG00000165688\n",
      "gene_string='PMPCB'\n",
      "gene_list=['PMPCB']\n",
      "match=None\n",
      "['PMPCB']:     ENSG00000105819\n",
      "gene_string='PNISR'\n",
      "gene_list=['PNISR']\n",
      "match=None\n",
      "['PNISR']:     ENSG00000132424\n",
      "gene_string='PNKP'\n",
      "gene_list=['PNKP']\n",
      "match=None\n",
      "['PNKP']:     ENSG00000039650\n",
      "gene_string='PNN'\n",
      "gene_list=['PNN']\n",
      "match=None\n",
      "['PNN']:     ENSG00000100941\n",
      "gene_string='PNO1'\n",
      "gene_list=['PNO1']\n",
      "match=None\n",
      "['PNO1']:     ENSG00000115946\n",
      "gene_string='PNPT1'\n",
      "gene_list=['PNPT1']\n",
      "match=None\n",
      "['PNPT1']:     ENSG00000138035\n",
      "gene_string='POGZ'\n",
      "gene_list=['POGZ']\n",
      "match=None\n",
      "['POGZ']:     ENSG00000143442\n",
      "gene_string='POLA1'\n",
      "gene_list=['POLA1']\n",
      "match=None\n",
      "['POLA1']:     ENSG00000101868\n",
      "gene_string='POLA2'\n",
      "gene_list=['POLA2']\n",
      "match=None\n",
      "['POLA2']:     ENSG00000014138\n",
      "gene_string='POLD1'\n",
      "gene_list=['POLD1']\n",
      "match=None\n",
      "['POLD1']:     ENSG00000062822\n",
      "gene_string='POLD2'\n",
      "gene_list=['POLD2']\n",
      "match=None\n",
      "['POLD2']:     ENSG00000106628\n",
      "gene_string='POLD3'\n",
      "gene_list=['POLD3']\n",
      "match=None\n",
      "['POLD3']:     ENSG00000077514\n",
      "gene_string='POLE'\n",
      "gene_list=['POLE']\n",
      "match=None\n",
      "['POLE']:     ENSG00000177084\n",
      "gene_string='POLE2'\n",
      "gene_list=['POLE2']\n",
      "match=None\n",
      "['POLE2']:     ENSG00000100479\n",
      "gene_string='POLG2'\n",
      "gene_list=['POLG2']\n",
      "match=None\n",
      "['POLG2']:     ENSG00000256525\n",
      "gene_string='POLR1A'\n",
      "gene_list=['POLR1A']\n",
      "match=None\n",
      "['POLR1A']:     ENSG00000068654\n",
      "gene_string='POLR1B'\n",
      "gene_list=['POLR1B']\n",
      "match=None\n",
      "['POLR1B']:     ENSG00000125630\n",
      "gene_string='POLR1C'\n",
      "gene_list=['POLR1C']\n",
      "match=None\n",
      "['POLR1C']:     ENSG00000171453\n",
      "gene_string='POLR1E'\n",
      "gene_list=['POLR1E']\n",
      "match=None\n",
      "['POLR1E']:     ENSG00000137054\n",
      "gene_string='POLR2A'\n",
      "gene_list=['POLR2A']\n",
      "match=None\n",
      "['POLR2A']:     ENSG00000181222\n",
      "gene_string='POLR2B'\n",
      "gene_list=['POLR2B']\n",
      "match=None\n",
      "['POLR2B']:     ENSG00000047315\n",
      "gene_string='POLR2C'\n",
      "gene_list=['POLR2C']\n",
      "match=None\n",
      "['POLR2C']:     ENSG00000102978\n",
      "gene_string='POLR2D'\n",
      "gene_list=['POLR2D']\n",
      "match=None\n",
      "['POLR2D']:     ENSG00000144231\n",
      "gene_string='POLR2E'\n",
      "gene_list=['POLR2E']\n",
      "match=None\n",
      "['POLR2E']:     ENSG00000099817\n",
      "gene_string='POLR2F'\n",
      "gene_list=['POLR2F']\n",
      "match=None\n",
      "['POLR2F']:     ENSG00000100142\n",
      "gene_string='POLR2G'\n",
      "gene_list=['POLR2G']\n",
      "match=None\n",
      "['POLR2G']:     ENSG00000168002\n",
      "gene_string='POLR2H'\n",
      "gene_list=['POLR2H']\n",
      "match=None\n",
      "['POLR2H']:     ENSG00000163882\n",
      "gene_string='POLR2I'\n",
      "gene_list=['POLR2I']\n",
      "match=None\n",
      "['POLR2I']:     ENSG00000105258\n",
      "gene_string='POLR2K'\n",
      "gene_list=['POLR2K']\n",
      "match=None\n",
      "['POLR2K']:     ENSG00000147669\n",
      "gene_string='POLR2L'\n",
      "gene_list=['POLR2L']\n",
      "match=None\n",
      "['POLR2L']:     ENSG00000177700\n",
      "gene_string='POLR3A'\n",
      "gene_list=['POLR3A']\n",
      "match=None\n",
      "['POLR3A']:     ENSG00000148606\n",
      "gene_string='POLR3B'\n",
      "gene_list=['POLR3B']\n",
      "match=None\n",
      "['POLR3B']:     ENSG00000013503\n",
      "gene_string='POLR3C'\n",
      "gene_list=['POLR3C']\n",
      "match=None\n",
      "['POLR3C']:     ENSG00000186141\n",
      "gene_string='POLR3D'\n",
      "gene_list=['POLR3D']\n",
      "match=None\n",
      "['POLR3D']:     ENSG00000168495\n",
      "gene_string='POLR3E'\n",
      "gene_list=['POLR3E']\n",
      "match=None\n",
      "['POLR3E']:     ENSG00000058600\n",
      "gene_string='POLR3F'\n",
      "gene_list=['POLR3F']\n",
      "match=None\n",
      "['POLR3F']:     ENSG00000132664\n",
      "gene_string='POLR3H'\n",
      "gene_list=['POLR3H']\n",
      "match=None\n",
      "['POLR3H']:     ENSG00000100413\n",
      "gene_string='POLR3K'\n",
      "gene_list=['POLR3K']\n",
      "match=None\n",
      "['POLR3K']:     ENSG00000161980\n",
      "gene_string='POLRMT'\n",
      "gene_list=['POLRMT']\n",
      "match=None\n",
      "['POLRMT']:     ENSG00000099821\n",
      "gene_string='POP1'\n",
      "gene_list=['POP1']\n",
      "match=None\n",
      "['POP1']:     ENSG00000112276\n",
      "gene_string='POP4'\n",
      "gene_list=['POP4']\n",
      "match=None\n",
      "['POP4']:     ENSG00000105171\n",
      "gene_string='POP5'\n",
      "gene_list=['POP5']\n",
      "match=None\n",
      "['POP5']:     ENSG00000167272\n",
      "gene_string='POP7'\n",
      "gene_list=['POP7']\n",
      "match=None\n",
      "['POP7']:     ENSG00000172336\n",
      "gene_string='POT1'\n",
      "gene_list=['POT1']\n",
      "match=None\n",
      "['POT1']:     ENSG00000128513\n",
      "gene_string='POTEI'\n",
      "gene_list=['POTEI']\n",
      "match=None\n",
      "['POTEI']:     ENSG00000196834\n",
      "gene_string='POU5F1B'\n",
      "gene_list=['POU5F1B']\n",
      "match=None\n",
      "['POU5F1B']:     ENSG00000212993\n",
      "gene_string='PPA1'\n",
      "gene_list=['PPA1']\n",
      "match=None\n",
      "['PPA1']:     ENSG00000180817\n",
      "gene_string='PPA2'\n",
      "gene_list=['PPA2']\n",
      "match=None\n",
      "['PPA2']:     ENSG00000138777\n",
      "gene_string='PPAN'\n",
      "gene_list=['PPAN']\n",
      "match=None\n",
      "['PPAN']:     ENSG00000130810\n",
      "gene_string='PPARGC1B'\n",
      "gene_list=['PPARGC1B']\n",
      "match=None\n",
      "['PPARGC1B']:     ENSG00000155846\n",
      "gene_string='PPIA'\n",
      "gene_list=['PPIA']\n",
      "match=None\n",
      "['PPIA']:     ENSG00000196262\n",
      "gene_string='PPIE'\n",
      "gene_list=['PPIE']\n",
      "match=None\n",
      "['PPIE']:     ENSG00000084072\n",
      "gene_string='PPIH'\n",
      "gene_list=['PPIH']\n",
      "match=None\n",
      "['PPIH']:     ENSG00000171960\n",
      "gene_string='PPIL1'\n",
      "gene_list=['PPIL1']\n",
      "match=None\n",
      "['PPIL1']:     ENSG00000137168\n",
      "gene_string='PPIL2'\n",
      "gene_list=['PPIL2']\n",
      "match=None\n",
      "['PPIL2']:     ENSG00000100023\n",
      "gene_string='PPIL4'\n",
      "gene_list=['PPIL4']\n",
      "match=None\n",
      "['PPIL4']:     ENSG00000131013\n",
      "gene_string='PPME1'\n",
      "gene_list=['PPME1']\n",
      "match=None\n",
      "['PPME1']:     ENSG00000214517\n",
      "gene_string='PPP1CA'\n",
      "gene_list=['PPP1CA']\n",
      "match=None\n",
      "['PPP1CA']:     ENSG00000172531\n",
      "gene_string='PPP1CB'\n",
      "gene_list=['PPP1CB']\n",
      "match=None\n",
      "['PPP1CB']:     ENSG00000213639\n",
      "gene_string='PPP1R2'\n",
      "gene_list=['PPP1R2']\n",
      "match=None\n",
      "['PPP1R2']:     ENSG00000184203\n",
      "gene_string='PPP1R7'\n",
      "gene_list=['PPP1R7']\n",
      "match=None\n",
      "['PPP1R7']:     ENSG00000115685\n",
      "gene_string='PPP1R8'\n",
      "gene_list=['PPP1R8']\n",
      "match=None\n",
      "['PPP1R8']:     ENSG00000117751\n",
      "gene_string='PPP1R10'\n",
      "gene_list=['PPP1R10']\n",
      "match=None\n",
      "['PPP1R10']:     ENSG00000204569\n",
      "gene_string='PPP1R11'\n",
      "gene_list=['PPP1R11']\n",
      "match=None\n",
      "['PPP1R11']:     ENSG00000204619\n",
      "gene_string='PPP1R12A'\n",
      "gene_list=['PPP1R12A']\n",
      "match=None\n",
      "['PPP1R12A']:     ENSG00000058272\n",
      "gene_string='PPP1R15B'\n",
      "gene_list=['PPP1R15B']\n",
      "match=None\n",
      "['PPP1R15B']:     ENSG00000158615\n",
      "gene_string='PPP2CA'\n",
      "gene_list=['PPP2CA']\n",
      "match=None\n",
      "['PPP2CA']:     ENSG00000113575\n",
      "gene_string='PPP2R1A'\n",
      "gene_list=['PPP2R1A']\n",
      "match=None\n",
      "['PPP2R1A']:     ENSG00000105568\n",
      "gene_string='PPP2R3C'\n",
      "gene_list=['PPP2R3C']\n",
      "match=None\n",
      "['PPP2R3C']:     ENSG00000092020\n",
      "gene_string='PPP4C'\n",
      "gene_list=['PPP4C']\n",
      "match=None\n",
      "['PPP4C']:     ENSG00000149923\n",
      "gene_string='PPP4R2'\n",
      "gene_list=['PPP4R2']\n",
      "match=None\n",
      "['PPP4R2']:     ENSG00000163605\n",
      "gene_string='PPP6C'\n",
      "gene_list=['PPP6C']\n",
      "match=None\n",
      "['PPP6C']:     ENSG00000119414\n",
      "gene_string='PPRC1'\n",
      "gene_list=['PPRC1']\n",
      "match=None\n",
      "['PPRC1']:     ENSG00000148840\n",
      "gene_string='PPWD1'\n",
      "gene_list=['PPWD1']\n",
      "match=None\n",
      "['PPWD1']:     ENSG00000113593\n",
      "gene_string='PPY'\n",
      "gene_list=['PPY']\n",
      "match=None\n",
      "['PPY']:     ENSG00000108849\n",
      "gene_string='PRC1'\n",
      "gene_list=['PRC1']\n",
      "match=None\n",
      "['PRC1']:     ENSG00000198901\n",
      "gene_string='PRDM10'\n",
      "gene_list=['PRDM10']\n",
      "match=None\n",
      "['PRDM10']:     ENSG00000170325\n",
      "gene_string='PREB'\n",
      "gene_list=['PREB']\n",
      "match=None\n",
      "['PREB']:     ENSG00000138073\n",
      "gene_string='PRELID1'\n",
      "gene_list=['PRELID1']\n",
      "match=None\n",
      "['PRELID1']:     ENSG00000169230\n",
      "gene_string='PRELID3B'\n",
      "gene_list=['PRELID3B']\n",
      "match=None\n",
      "['PRELID3B']:     ENSG00000101166\n",
      "gene_string='PRIM1'\n",
      "gene_list=['PRIM1']\n",
      "match=None\n",
      "['PRIM1']:     ENSG00000198056\n",
      "gene_string='PRIM2'\n",
      "gene_list=['PRIM2']\n",
      "match=None\n",
      "['PRIM2']:     ENSG00000146143\n",
      "gene_string='PRKRA'\n",
      "gene_list=['PRKRA']\n",
      "match=None\n",
      "['PRKRA']:     ENSG00000180228\n",
      "gene_string='PRKRIP1'\n",
      "gene_list=['PRKRIP1']\n",
      "match=None\n",
      "['PRKRIP1']:     ENSG00000128563\n",
      "gene_string='PRMT1'\n",
      "gene_list=['PRMT1']\n",
      "match=None\n",
      "['PRMT1']:     ENSG00000126457\n",
      "gene_string='PRMT5'\n",
      "gene_list=['PRMT5']\n",
      "match=None\n",
      "['PRMT5']:     ENSG00000100462\n",
      "gene_string='PRODH'\n",
      "gene_list=['PRODH']\n",
      "match=None\n",
      "['PRODH']:     ENSG00000100033\n",
      "gene_string='PRORP'\n",
      "gene_list=['PRORP']\n",
      "match=None\n",
      "['PRORP']:     ENSG00000100890\n",
      "gene_string='PRPF3'\n",
      "gene_list=['PRPF3']\n",
      "match=None\n",
      "['PRPF3']:     ENSG00000117360\n",
      "gene_string='PRPF4'\n",
      "gene_list=['PRPF4']\n",
      "match=None\n",
      "['PRPF4']:     ENSG00000136875\n",
      "gene_string='PRPF4B'\n",
      "gene_list=['PRPF4B']\n",
      "match=None\n",
      "['PRPF4B']:     ENSG00000112739\n",
      "gene_string='PRPF6'\n",
      "gene_list=['PRPF6']\n",
      "match=None\n",
      "['PRPF6']:     ENSG00000101161\n",
      "gene_string='PRPF8'\n",
      "gene_list=['PRPF8']\n",
      "match=None\n",
      "['PRPF8']:     ENSG00000174231\n",
      "gene_string='PRPF18'\n",
      "gene_list=['PRPF18']\n",
      "match=None\n",
      "['PRPF18']:     ENSG00000165630\n",
      "gene_string='PRPF19'\n",
      "gene_list=['PRPF19']\n",
      "match=None\n",
      "['PRPF19']:     ENSG00000110107\n",
      "gene_string='PRPF31'\n",
      "gene_list=['PRPF31']\n",
      "match=None\n",
      "['PRPF31']:     ENSG00000105618\n",
      "gene_string='PRPF38A'\n",
      "gene_list=['PRPF38A']\n",
      "match=None\n",
      "['PRPF38A']:     ENSG00000134748\n",
      "gene_string='PRPF38B'\n",
      "gene_list=['PRPF38B']\n",
      "match=None\n",
      "['PRPF38B']:     ENSG00000134186\n",
      "gene_string='PRPF40A'\n",
      "gene_list=['PRPF40A']\n",
      "match=None\n",
      "['PRPF40A']:     ENSG00000196504\n",
      "gene_string='PRRC2A'\n",
      "gene_list=['PRRC2A']\n",
      "match=None\n",
      "['PRRC2A']:     ENSG00000204469\n",
      "gene_string='PSAT1'\n",
      "gene_list=['PSAT1']\n",
      "match=None\n",
      "['PSAT1']:     ENSG00000135069\n",
      "gene_string='PSMA1'\n",
      "gene_list=['PSMA1']\n",
      "match=None\n",
      "['PSMA1']:     ENSG00000129084\n",
      "gene_string='PSMA2'\n",
      "gene_list=['PSMA2']\n",
      "match=None\n",
      "['PSMA2']:     ENSG00000256646\n",
      "gene_string='PSMA3'\n",
      "gene_list=['PSMA3']\n",
      "match=None\n",
      "['PSMA3']:     ENSG00000100567\n",
      "gene_string='PSMA4'\n",
      "gene_list=['PSMA4']\n",
      "match=None\n",
      "['PSMA4']:     ENSG00000041357\n",
      "gene_string='PSMA5'\n",
      "gene_list=['PSMA5']\n",
      "match=None\n",
      "['PSMA5']:     ENSG00000143106\n",
      "gene_string='PSMA6'\n",
      "gene_list=['PSMA6']\n",
      "match=None\n",
      "['PSMA6']:     ENSG00000100902\n",
      "gene_string='PSMA7'\n",
      "gene_list=['PSMA7']\n",
      "match=None\n",
      "['PSMA7']:     ENSG00000101182\n",
      "gene_string='PSMB1'\n",
      "gene_list=['PSMB1']\n",
      "match=None\n",
      "['PSMB1']:     ENSG00000008018\n",
      "gene_string='PSMB2'\n",
      "gene_list=['PSMB2']\n",
      "match=None\n",
      "['PSMB2']:     ENSG00000126067\n",
      "gene_string='PSMB3'\n",
      "gene_list=['PSMB3']\n",
      "match=None\n",
      "['PSMB3']:     ENSG00000277791\n",
      "gene_string='PSMB4'\n",
      "gene_list=['PSMB4']\n",
      "match=None\n",
      "['PSMB4']:     ENSG00000159377\n",
      "gene_string='PSMB5'\n",
      "gene_list=['PSMB5']\n",
      "match=None\n",
      "['PSMB5']:     ENSG00000100804\n",
      "gene_string='PSMB6'\n",
      "gene_list=['PSMB6']\n",
      "match=None\n",
      "['PSMB6']:     ENSG00000142507\n",
      "gene_string='PSMB7'\n",
      "gene_list=['PSMB7']\n",
      "match=None\n",
      "['PSMB7']:     ENSG00000136930\n",
      "gene_string='PSMC1'\n",
      "gene_list=['PSMC1']\n",
      "match=None\n",
      "['PSMC1']:     ENSG00000100764\n",
      "gene_string='PSMC2'\n",
      "gene_list=['PSMC2']\n",
      "match=None\n",
      "['PSMC2']:     ENSG00000161057\n",
      "gene_string='PSMC3'\n",
      "gene_list=['PSMC3']\n",
      "match=None\n",
      "['PSMC3']:     ENSG00000165916\n",
      "gene_string='PSMC4'\n",
      "gene_list=['PSMC4']\n",
      "match=None\n",
      "['PSMC4']:     ENSG00000013275\n",
      "gene_string='PSMC5'\n",
      "gene_list=['PSMC5']\n",
      "match=None\n",
      "['PSMC5']:     ENSG00000087191\n",
      "gene_string='PSMC6'\n",
      "gene_list=['PSMC6']\n",
      "match=None\n",
      "['PSMC6']:     ENSG00000100519\n",
      "gene_string='PSMD1'\n",
      "gene_list=['PSMD1']\n",
      "match=None\n",
      "['PSMD1']:     ENSG00000173692\n",
      "gene_string='PSMD2'\n",
      "gene_list=['PSMD2']\n",
      "match=None\n",
      "['PSMD2']:     ENSG00000175166\n",
      "gene_string='PSMD3'\n",
      "gene_list=['PSMD3']\n",
      "match=None\n",
      "['PSMD3']:     ENSG00000108344\n",
      "gene_string='PSMD4'\n",
      "gene_list=['PSMD4']\n",
      "match=None\n",
      "['PSMD4']:     ENSG00000159352\n",
      "gene_string='PSMD6'\n",
      "gene_list=['PSMD6']\n",
      "match=None\n",
      "['PSMD6']:     ENSG00000163636\n",
      "gene_string='PSMD7'\n",
      "gene_list=['PSMD7']\n",
      "match=None\n",
      "['PSMD7']:     ENSG00000103035\n",
      "gene_string='PSMD8'\n",
      "gene_list=['PSMD8']\n",
      "match=None\n",
      "['PSMD8']:     ENSG00000099341\n",
      "gene_string='PSMD9'\n",
      "gene_list=['PSMD9']\n",
      "match=None\n",
      "['PSMD9']:     ENSG00000110801\n",
      "gene_string='PSMD11'\n",
      "gene_list=['PSMD11']\n",
      "match=None\n",
      "['PSMD11']:     ENSG00000108671\n",
      "gene_string='PSMD12'\n",
      "gene_list=['PSMD12']\n",
      "match=None\n",
      "['PSMD12']:     ENSG00000197170\n",
      "gene_string='PSMD13'\n",
      "gene_list=['PSMD13']\n",
      "match=None\n",
      "['PSMD13']:     ENSG00000185627\n",
      "gene_string='PSMD14'\n",
      "gene_list=['PSMD14']\n",
      "match=None\n",
      "['PSMD14']:     ENSG00000115233\n",
      "gene_string='PSME1'\n",
      "gene_list=['PSME1']\n",
      "match=None\n",
      "['PSME1']:     ENSG00000092010\n",
      "gene_string='PSME2'\n",
      "gene_list=['PSME2']\n",
      "match=None\n",
      "['PSME2']:     ENSG00000100911\n",
      "gene_string='PSMG1'\n",
      "gene_list=['PSMG1']\n",
      "match=None\n",
      "['PSMG1']:     ENSG00000183527\n",
      "gene_string='PSMG2'\n",
      "gene_list=['PSMG2']\n",
      "match=None\n",
      "['PSMG2']:     ENSG00000128789\n",
      "gene_string='PSMG3'\n",
      "gene_list=['PSMG3']\n",
      "match=None\n",
      "['PSMG3']:     ENSG00000157778\n",
      "gene_string='PSMG4'\n",
      "gene_list=['PSMG4']\n",
      "match=None\n",
      "['PSMG4']:     ENSG00000180822\n",
      "gene_string='PSTK'\n",
      "gene_list=['PSTK']\n",
      "match=None\n",
      "['PSTK']:     ENSG00000179988\n",
      "gene_string='PTBP1'\n",
      "gene_list=['PTBP1']\n",
      "match=None\n",
      "['PTBP1']:     ENSG00000011304\n",
      "gene_string='PTCD1'\n",
      "gene_list=['PTCD1']\n",
      "match=None\n",
      "['PTCD1']:     ENSG00000106246\n",
      "gene_string='PTCD3'\n",
      "gene_list=['PTCD3']\n",
      "match=None\n",
      "['PTCD3']:     ENSG00000132300\n",
      "gene_string='PTK2'\n",
      "gene_list=['PTK2']\n",
      "match=None\n",
      "['PTK2']:     ENSG00000169398\n",
      "gene_string='PTMA'\n",
      "gene_list=['PTMA']\n",
      "match=None\n",
      "['PTMA']:     ENSG00000187514\n",
      "gene_string='PTPMT1'\n",
      "gene_list=['PTPMT1']\n",
      "match=None\n",
      "['PTPMT1']:     ENSG00000110536\n",
      "gene_string='PTPN23'\n",
      "gene_list=['PTPN23']\n",
      "match=None\n",
      "['PTPN23']:     ENSG00000076201\n",
      "gene_string='PUF60'\n",
      "gene_list=['PUF60']\n",
      "match=None\n",
      "['PUF60']:     ENSG00000179950\n",
      "gene_string='PWP1'\n",
      "gene_list=['PWP1']\n",
      "match=None\n",
      "['PWP1']:     ENSG00000136045\n",
      "gene_string='PWP2'\n",
      "gene_list=['PWP2']\n",
      "match=None\n",
      "['PWP2']:     ENSG00000241945\n",
      "gene_string='PXN'\n",
      "gene_list=['PXN']\n",
      "match=None\n",
      "['PXN']:     ENSG00000089159\n",
      "gene_string='PYROXD1'\n",
      "gene_list=['PYROXD1']\n",
      "match=None\n",
      "['PYROXD1']:     ENSG00000121350\n",
      "gene_string='QARS'\n",
      "gene_list=['QARS']\n",
      "match=None\n",
      "['QARS']:     ENSG00000172053\n",
      "gene_string='QRSL1'\n",
      "gene_list=['QRSL1']\n",
      "match=None\n",
      "['QRSL1']:     ENSG00000130348\n",
      "gene_string='RAB4A'\n",
      "gene_list=['RAB4A']\n",
      "match=None\n",
      "['RAB4A']:     ENSG00000168118\n",
      "gene_string='RAB6A'\n",
      "gene_list=['RAB6A']\n",
      "match=None\n",
      "['RAB6A']:     ENSG00000175582\n",
      "gene_string='RAB18'\n",
      "gene_list=['RAB18']\n",
      "match=None\n",
      "['RAB18']:     ENSG00000099246\n",
      "gene_string='RABGGTA'\n",
      "gene_list=['RABGGTA']\n",
      "match=None\n",
      "['RABGGTA']:     ENSG00000100949\n",
      "gene_string='RABGGTB'\n",
      "gene_list=['RABGGTB']\n",
      "match=None\n",
      "['RABGGTB']:     ENSG00000137955\n",
      "gene_string='RAC1'\n",
      "gene_list=['RAC1']\n",
      "match=None\n",
      "['RAC1']:     ENSG00000136238\n",
      "gene_string='RAC3'\n",
      "gene_list=['RAC3']\n",
      "match=None\n",
      "['RAC3']:     ENSG00000169750\n",
      "gene_string='RACGAP1'\n",
      "gene_list=['RACGAP1']\n",
      "match=None\n",
      "['RACGAP1']:     ENSG00000161800\n",
      "gene_string='RAD9A'\n",
      "gene_list=['RAD9A']\n",
      "match=None\n",
      "['RAD9A']:     ENSG00000172613\n",
      "gene_string='RAD17'\n",
      "gene_list=['RAD17']\n",
      "match=None\n",
      "['RAD17']:     ENSG00000152942\n",
      "gene_string='RAD21'\n",
      "gene_list=['RAD21']\n",
      "match=None\n",
      "['RAD21']:     ENSG00000164754\n",
      "gene_string='RAD51'\n",
      "gene_list=['RAD51']\n",
      "match=None\n",
      "['RAD51']:     ENSG00000051180\n",
      "gene_string='RAD51C'\n",
      "gene_list=['RAD51C']\n",
      "match=None\n",
      "['RAD51C']:     ENSG00000108384\n",
      "gene_string='RAD51D'\n",
      "gene_list=['RAD51D']\n",
      "match=None\n",
      "['RAD51D']:     ENSG00000185379\n",
      "gene_string='RAE1'\n",
      "gene_list=['RAE1']\n",
      "match=None\n",
      "['RAE1']:     ENSG00000165799\n",
      "gene_string='RAN'\n",
      "gene_list=['RAN']\n",
      "match=None\n",
      "['RAN']:     ENSG00000132341\n",
      "gene_string='RANBP2'\n",
      "gene_list=['RANBP2']\n",
      "match=None\n",
      "['RANBP2']:     ENSG00000153201\n",
      "gene_string='RANGAP1'\n",
      "gene_list=['RANGAP1']\n",
      "match=None\n",
      "['RANGAP1']:     ENSG00000100401\n",
      "gene_string='RARS'\n",
      "gene_list=['RARS']\n",
      "match=None\n",
      "['RARS']:     ENSG00000113643\n",
      "gene_string='RARS2'\n",
      "gene_list=['RARS2']\n",
      "match=None\n",
      "['RARS2']:     ENSG00000146282\n",
      "gene_string='RBBP4'\n",
      "gene_list=['RBBP4']\n",
      "match=None\n",
      "['RBBP4']:     ENSG00000162521\n",
      "gene_string='RBBP5'\n",
      "gene_list=['RBBP5']\n",
      "match=None\n",
      "['RBBP5']:     ENSG00000117222\n",
      "gene_string='RBBP6'\n",
      "gene_list=['RBBP6']\n",
      "match=None\n",
      "['RBBP6']:     ENSG00000122257\n",
      "gene_string='RBBP8'\n",
      "gene_list=['RBBP8']\n",
      "match=None\n",
      "['RBBP8']:     ENSG00000101773\n",
      "gene_string='RBM4'\n",
      "gene_list=['RBM4']\n",
      "match=None\n",
      "['RBM4']:     ENSG00000173933\n",
      "gene_string='RBM8A'\n",
      "gene_list=['RBM8A']\n",
      "match=None\n",
      "['RBM8A']:     ENSG00000265241\n",
      "gene_string='RBM10'\n",
      "gene_list=['RBM10']\n",
      "match=None\n",
      "['RBM10']:     ENSG00000182872\n",
      "gene_string='RBM14'\n",
      "gene_list=['RBM14']\n",
      "match=None\n",
      "['RBM14']:     ENSG00000239306\n",
      "gene_string='RBM17'\n",
      "gene_list=['RBM17']\n",
      "match=None\n",
      "['RBM17']:     ENSG00000134453\n",
      "gene_string='RBM19'\n",
      "gene_list=['RBM19']\n",
      "match=None\n",
      "['RBM19']:     ENSG00000122965\n",
      "gene_string='RBM22'\n",
      "gene_list=['RBM22']\n",
      "match=None\n",
      "['RBM22']:     ENSG00000086589\n",
      "gene_string='RBM25'\n",
      "gene_list=['RBM25']\n",
      "match=None\n",
      "['RBM25']:     ENSG00000119707\n",
      "gene_string='RBM28'\n",
      "gene_list=['RBM28']\n",
      "match=None\n",
      "['RBM28']:     ENSG00000106344\n",
      "gene_string='RBM39'\n",
      "gene_list=['RBM39']\n",
      "match=None\n",
      "['RBM39']:     ENSG00000131051\n",
      "gene_string='RBM42'\n",
      "gene_list=['RBM42']\n",
      "match=None\n",
      "['RBM42']:     ENSG00000126254\n",
      "gene_string='RBM48'\n",
      "gene_list=['RBM48']\n",
      "match=None\n",
      "['RBM48']:     ENSG00000127993\n",
      "gene_string='RBMX'\n",
      "gene_list=['RBMX']\n",
      "match=None\n",
      "['RBMX']:     ENSG00000147274\n",
      "gene_string='RBMX2'\n",
      "gene_list=['RBMX2']\n",
      "match=None\n",
      "['RBMX2']:     ENSG00000134597\n",
      "gene_string='RBMXL1'\n",
      "gene_list=['RBMXL1']\n",
      "match=None\n",
      "['RBMXL1']:     ENSG00000213516\n",
      "gene_string='RBX1'\n",
      "gene_list=['RBX1']\n",
      "match=None\n",
      "['RBX1']:     ENSG00000100387\n",
      "gene_string='RCC1'\n",
      "gene_list=['RCC1']\n",
      "match=None\n",
      "['RCC1']:     ENSG00000180198\n",
      "gene_string='RCL1'\n",
      "gene_list=['RCL1']\n",
      "match=None\n",
      "['RCL1']:     ENSG00000120158\n",
      "gene_string='RCOR1'\n",
      "gene_list=['RCOR1']\n",
      "match=None\n",
      "['RCOR1']:     ENSG00000089902\n",
      "gene_string='REV3L'\n",
      "gene_list=['REV3L']\n",
      "match=None\n",
      "['REV3L']:     ENSG00000009413\n",
      "gene_string='REXO2'\n",
      "gene_list=['REXO2']\n",
      "match=None\n",
      "['REXO2']:     ENSG00000076043\n",
      "gene_string='RFC1'\n",
      "gene_list=['RFC1']\n",
      "match=None\n",
      "['RFC1']:     ENSG00000035928\n",
      "gene_string='RFC2'\n",
      "gene_list=['RFC2']\n",
      "match=None\n",
      "['RFC2']:     ENSG00000049541\n",
      "gene_string='RFC3'\n",
      "gene_list=['RFC3']\n",
      "match=None\n",
      "['RFC3']:     ENSG00000133119\n",
      "gene_string='RFC4'\n",
      "gene_list=['RFC4']\n",
      "match=None\n",
      "['RFC4']:     ENSG00000163918\n",
      "gene_string='RFC5'\n",
      "gene_list=['RFC5']\n",
      "match=None\n",
      "['RFC5']:     ENSG00000111445\n",
      "gene_string='RFFL'\n",
      "gene_list=['RFFL']\n",
      "match=None\n",
      "['RFFL']:     ENSG00000092871\n",
      "gene_string='RFT1'\n",
      "gene_list=['RFT1']\n",
      "match=None\n",
      "['RFT1']:     ENSG00000163933\n",
      "gene_string='RGP1'\n",
      "gene_list=['RGP1']\n",
      "match=None\n",
      "['RGP1']:     ENSG00000187627\n",
      "gene_string='RGPD6'\n",
      "gene_list=['RGPD6']\n",
      "match=None\n",
      "['RGPD6']:     ENSG00000183054\n",
      "gene_string='RHOQ'\n",
      "gene_list=['RHOQ']\n",
      "match=None\n",
      "['RHOQ']:     ENSG00000119729\n",
      "gene_string='RHPN1'\n",
      "gene_list=['RHPN1']\n",
      "match=None\n",
      "['RHPN1']:     ENSG00000158106\n",
      "gene_string='RINT1'\n",
      "gene_list=['RINT1']\n",
      "match=None\n",
      "['RINT1']:     ENSG00000135249\n",
      "gene_string='RIOK1'\n",
      "gene_list=['RIOK1']\n",
      "match=None\n",
      "['RIOK1']:     ENSG00000124784\n",
      "gene_string='RIOK2'\n",
      "gene_list=['RIOK2']\n",
      "match=None\n",
      "['RIOK2']:     ENSG00000058729\n",
      "gene_string='RMI1'\n",
      "gene_list=['RMI1']\n",
      "match=None\n",
      "['RMI1']:     ENSG00000178966\n",
      "gene_string='RNASEH2C'\n",
      "gene_list=['RNASEH2C']\n",
      "match=None\n",
      "['RNASEH2C']:     ENSG00000172922\n",
      "gene_string='RNF4'\n",
      "gene_list=['RNF4']\n",
      "match=None\n",
      "['RNF4']:     ENSG00000063978\n",
      "gene_string='RNF8'\n",
      "gene_list=['RNF8']\n",
      "match=None\n",
      "['RNF8']:     ENSG00000112130\n",
      "gene_string='RNF14'\n",
      "gene_list=['RNF14']\n",
      "match=None\n",
      "['RNF14']:     ENSG00000013561\n",
      "gene_string='RNF20'\n",
      "gene_list=['RNF20']\n",
      "match=None\n",
      "['RNF20']:     ENSG00000155827\n",
      "gene_string='RNF40'\n",
      "gene_list=['RNF40']\n",
      "match=None\n",
      "['RNF40']:     ENSG00000103549\n",
      "gene_string='RNF103'\n",
      "gene_list=['RNF103']\n",
      "match=None\n",
      "['RNF103']:     ENSG00000239305\n",
      "gene_string='RNF113A'\n",
      "gene_list=['RNF113A']\n",
      "match=None\n",
      "['RNF113A']:     ENSG00000125352\n",
      "gene_string='RNF123'\n",
      "gene_list=['RNF123']\n",
      "match=None\n",
      "['RNF123']:     ENSG00000164068\n",
      "gene_string='RNF168'\n",
      "gene_list=['RNF168']\n",
      "match=None\n",
      "['RNF168']:     ENSG00000163961\n",
      "gene_string='RNGTT'\n",
      "gene_list=['RNGTT']\n",
      "match=None\n",
      "['RNGTT']:     ENSG00000111880\n",
      "gene_string='RNMT'\n",
      "gene_list=['RNMT']\n",
      "match=None\n",
      "['RNMT']:     ENSG00000101654\n",
      "gene_string='RNPC3'\n",
      "gene_list=['RNPC3']\n",
      "match=None\n",
      "['RNPC3']:     ENSG00000185946\n",
      "gene_string='RNPS1'\n",
      "gene_list=['RNPS1']\n",
      "match=None\n",
      "['RNPS1']:     ENSG00000205937\n",
      "gene_string='ROMO1'\n",
      "gene_list=['ROMO1']\n",
      "match=None\n",
      "['ROMO1']:     ENSG00000125995\n",
      "gene_string='RPA1'\n",
      "gene_list=['RPA1']\n",
      "match=None\n",
      "['RPA1']:     ENSG00000132383\n",
      "gene_string='RPA2'\n",
      "gene_list=['RPA2']\n",
      "match=None\n",
      "['RPA2']:     ENSG00000125630\n",
      "gene_string='RPA3'\n",
      "gene_list=['RPA3']\n",
      "match=None\n",
      "['RPA3']:     ENSG00000106399\n",
      "gene_string='RPAIN'\n",
      "gene_list=['RPAIN']\n",
      "match=None\n",
      "['RPAIN']:     ENSG00000129197\n",
      "gene_string='RPAP1'\n",
      "gene_list=['RPAP1']\n",
      "match=None\n",
      "['RPAP1']:     ENSG00000103932\n",
      "gene_string='RPAP2'\n",
      "gene_list=['RPAP2']\n",
      "match=None\n",
      "['RPAP2']:     ENSG00000122484\n",
      "gene_string='RPAP3'\n",
      "gene_list=['RPAP3']\n",
      "match=None\n",
      "['RPAP3']:     ENSG00000005175\n",
      "gene_string='RPE'\n",
      "gene_list=['RPE']\n",
      "match=None\n",
      "['RPE']:     ENSG00000197713\n",
      "gene_string='RPF1'\n",
      "gene_list=['RPF1']\n",
      "match=None\n",
      "['RPF1']:     ENSG00000069869\n",
      "gene_string='RPF2'\n",
      "gene_list=['RPF2']\n",
      "match=None\n",
      "['RPF2']:     ENSG00000197498\n",
      "gene_string='RPL3'\n",
      "gene_list=['RPL3']\n",
      "match=None\n",
      "['RPL3']:     ENSG00000100316\n",
      "gene_string='RPL4'\n",
      "gene_list=['RPL4']\n",
      "match=None\n",
      "['RPL4']:     ENSG00000174444\n",
      "gene_string='RPL5'\n",
      "gene_list=['RPL5']\n",
      "match=None\n",
      "['RPL5']:     ENSG00000122406\n",
      "gene_string='RPL6'\n",
      "gene_list=['RPL6']\n",
      "match=None\n",
      "['RPL6']:     ENSG00000089009\n",
      "gene_string='RPL7'\n",
      "gene_list=['RPL7']\n",
      "match=None\n",
      "['RPL7']:     ENSG00000147604\n",
      "gene_string='RPL7A'\n",
      "gene_list=['RPL7A']\n",
      "match=None\n",
      "['RPL7A']:     ENSG00000148303\n",
      "gene_string='RPL7L1'\n",
      "gene_list=['RPL7L1']\n",
      "match=None\n",
      "['RPL7L1']:     ENSG00000146223\n",
      "gene_string='RPL8'\n",
      "gene_list=['RPL8']\n",
      "match=None\n",
      "['RPL8']:     ENSG00000161016\n",
      "gene_string='RPL9'\n",
      "gene_list=['RPL9']\n",
      "match=None\n",
      "['RPL9']:     ENSG00000163682\n",
      "gene_string='RPL10'\n",
      "gene_list=['RPL10']\n",
      "match=None\n",
      "['RPL10']:     ENSG00000174748\n",
      "gene_string='RPL10A'\n",
      "gene_list=['RPL10A']\n",
      "match=None\n",
      "['RPL10A']:     ENSG00000198755\n",
      "gene_string='RPL11'\n",
      "gene_list=['RPL11']\n",
      "match=None\n",
      "['RPL11']:     ENSG00000142676\n",
      "gene_string='RPL12'\n",
      "gene_list=['RPL12']\n",
      "match=None\n",
      "['RPL12']:     ENSG00000197958\n",
      "gene_string='RPL13'\n",
      "gene_list=['RPL13']\n",
      "match=None\n",
      "['RPL13']:     ENSG00000172172\n",
      "gene_string='RPL13A'\n",
      "gene_list=['RPL13A']\n",
      "match=None\n",
      "['RPL13A']:     ENSG00000142541\n",
      "gene_string='RPL14'\n",
      "gene_list=['RPL14']\n",
      "match=None\n",
      "['RPL14']:     ENSG00000188846\n",
      "gene_string='RPL15'\n",
      "gene_list=['RPL15']\n",
      "match=None\n",
      "['RPL15']:     ENSG00000174748\n",
      "gene_string='RPL17'\n",
      "gene_list=['RPL17']\n",
      "match=None\n",
      "['RPL17']:     ENSG00000265681\n",
      "gene_string='RPL18'\n",
      "gene_list=['RPL18']\n",
      "match=None\n",
      "['RPL18']:     ENSG00000063177\n",
      "gene_string='RPL18A'\n",
      "gene_list=['RPL18A']\n",
      "match=None\n",
      "['RPL18A']:     ENSG00000105640\n",
      "gene_string='RPL19'\n",
      "gene_list=['RPL19']\n",
      "match=None\n",
      "['RPL19']:     ENSG00000108298\n",
      "gene_string='RPL21'\n",
      "gene_list=['RPL21']\n",
      "match=None\n",
      "['RPL21']:     ENSG00000122026\n",
      "gene_string='RPL23'\n",
      "gene_list=['RPL23']\n",
      "match=None\n",
      "['RPL23']:     ENSG00000125691\n",
      "gene_string='RPL23A'\n",
      "gene_list=['RPL23A']\n",
      "match=None\n",
      "['RPL23A']:     ENSG00000198242\n",
      "gene_string='RPL24'\n",
      "gene_list=['RPL24']\n",
      "match=None\n",
      "['RPL24']:     ENSG00000137876\n",
      "gene_string='RPL26'\n",
      "gene_list=['RPL26']\n",
      "match=None\n",
      "['RPL26']:     ENSG00000161970\n",
      "gene_string='RPL26L1'\n",
      "gene_list=['RPL26L1']\n",
      "match=None\n",
      "['RPL26L1']:     ENSG00000037241\n",
      "gene_string='RPL27'\n",
      "gene_list=['RPL27']\n",
      "match=None\n",
      "['RPL27']:     ENSG00000131469\n",
      "gene_string='RPL27A'\n",
      "gene_list=['RPL27A']\n",
      "match=None\n",
      "['RPL27A']:     ENSG00000166441\n",
      "gene_string='RPL28'\n",
      "gene_list=['RPL28']\n",
      "match=None\n",
      "['RPL28']:     ENSG00000108107\n",
      "gene_string='RPL29'\n",
      "gene_list=['RPL29']\n",
      "match=None\n",
      "['RPL29']:     ENSG00000162244\n",
      "gene_string='RPL30'\n",
      "gene_list=['RPL30']\n",
      "match=None\n",
      "['RPL30']:     ENSG00000156482\n",
      "gene_string='RPL31'\n",
      "gene_list=['RPL31']\n",
      "match=None\n",
      "['RPL31']:     ENSG00000071082\n",
      "gene_string='RPL32'\n",
      "gene_list=['RPL32']\n",
      "match=None\n",
      "['RPL32']:     ENSG00000144713\n",
      "gene_string='RPL34'\n",
      "gene_list=['RPL34']\n",
      "match=None\n",
      "['RPL34']:     ENSG00000109475\n",
      "gene_string='RPL35'\n",
      "gene_list=['RPL35']\n",
      "match=None\n",
      "['RPL35']:     ENSG00000136942\n",
      "gene_string='RPL35A'\n",
      "gene_list=['RPL35A']\n",
      "match=None\n",
      "['RPL35A']:     ENSG00000182899\n",
      "gene_string='RPL36'\n",
      "gene_list=['RPL36']\n",
      "match=None\n",
      "['RPL36']:     ENSG00000130255\n",
      "gene_string='RPL36A'\n",
      "gene_list=['RPL36A']\n",
      "match=None\n",
      "['RPL36A']:     ENSG00000241343\n",
      "gene_string='RPL36AL'\n",
      "gene_list=['RPL36AL']\n",
      "match=None\n",
      "['RPL36AL']:     ENSG00000165502\n",
      "gene_string='RPL37'\n",
      "gene_list=['RPL37']\n",
      "match=None\n",
      "['RPL37']:     ENSG00000145592\n",
      "gene_string='RPL37A'\n",
      "gene_list=['RPL37A']\n",
      "match=None\n",
      "['RPL37A']:     ENSG00000197756\n",
      "gene_string='RPL38'\n",
      "gene_list=['RPL38']\n",
      "match=None\n",
      "['RPL38']:     ENSG00000172809\n",
      "gene_string='RPL39'\n",
      "gene_list=['RPL39']\n",
      "match=None\n",
      "['RPL39']:     ENSG00000198918\n",
      "gene_string='RPL41'\n",
      "gene_list=['RPL41']\n",
      "match=None\n",
      "['RPL41']:     ENSG00000229117\n",
      "gene_string='RPLP0'\n",
      "gene_list=['RPLP0']\n",
      "match=None\n",
      "['RPLP0']:     ENSG00000089157\n",
      "gene_string='RPLP1'\n",
      "gene_list=['RPLP1']\n",
      "match=None\n",
      "['RPLP1']:     ENSG00000137818\n",
      "gene_string='RPLP2'\n",
      "gene_list=['RPLP2']\n",
      "match=None\n",
      "['RPLP2']:     ENSG00000177600\n",
      "gene_string='RPN1'\n",
      "gene_list=['RPN1']\n",
      "match=None\n",
      "['RPN1']:     ENSG00000175166\n",
      "gene_string='RPN2'\n",
      "gene_list=['RPN2']\n",
      "match=None\n",
      "['RPN2']:     ENSG00000173692\n",
      "gene_string='RPP14'\n",
      "gene_list=['RPP14']\n",
      "match=None\n",
      "['RPP14']:     ENSG00000163684\n",
      "gene_string='RPP21'\n",
      "gene_list=['RPP21']\n",
      "match=None\n",
      "['RPP21']:     ENSG00000241370\n",
      "gene_string='RPP30'\n",
      "gene_list=['RPP30']\n",
      "match=None\n",
      "['RPP30']:     ENSG00000148688\n",
      "gene_string='RPP38'\n",
      "gene_list=['RPP38']\n",
      "match=None\n",
      "['RPP38']:     ENSG00000152464\n",
      "gene_string='RPP40'\n",
      "gene_list=['RPP40']\n",
      "match=None\n",
      "['RPP40']:     ENSG00000124787\n",
      "gene_string='RPS2'\n",
      "gene_list=['RPS2']\n",
      "match=None\n",
      "['RPS2']:     ENSG00000140988\n",
      "gene_string='RPS3'\n",
      "gene_list=['RPS3']\n",
      "match=None\n",
      "['RPS3']:     ENSG00000149273\n",
      "gene_string='RPS3A'\n",
      "gene_list=['RPS3A']\n",
      "match=None\n",
      "['RPS3A']:     ENSG00000145425\n",
      "gene_string='RPS4X'\n",
      "gene_list=['RPS4X']\n",
      "match=None\n",
      "['RPS4X']:     ENSG00000198034\n",
      "gene_string='RPS5'\n",
      "gene_list=['RPS5']\n",
      "match=None\n",
      "['RPS5']:     ENSG00000083845\n",
      "gene_string='RPS6'\n",
      "gene_list=['RPS6']\n",
      "match=None\n",
      "['RPS6']:     ENSG00000137154\n",
      "gene_string='RPS7'\n",
      "gene_list=['RPS7']\n",
      "match=None\n",
      "['RPS7']:     ENSG00000171863\n",
      "gene_string='RPS8'\n",
      "gene_list=['RPS8']\n",
      "match=None\n",
      "['RPS8']:     ENSG00000142937\n",
      "gene_string='RPS9'\n",
      "gene_list=['RPS9']\n",
      "match=None\n",
      "['RPS9']:     ENSG00000170889\n",
      "gene_string='RPS10'\n",
      "gene_list=['RPS10']\n",
      "match=None\n",
      "['RPS10']:     ENSG00000124614\n",
      "gene_string='RPS10-NUDT3'\n",
      "gene_list=['RPS10-NUDT3']\n",
      "match=None\n",
      "['RPS10-NUDT3']:     ENSG00000270800\n",
      "gene_string='RPS11'\n",
      "gene_list=['RPS11']\n",
      "match=None\n",
      "['RPS11']:     ENSG00000142534\n",
      "gene_string='RPS12'\n",
      "gene_list=['RPS12']\n",
      "match=None\n",
      "['RPS12']:     ENSG00000128626\n",
      "gene_string='RPS13'\n",
      "gene_list=['RPS13']\n",
      "match=None\n",
      "['RPS13']:     ENSG00000110700\n",
      "gene_string='RPS14'\n",
      "gene_list=['RPS14']\n",
      "match=None\n",
      "['RPS14']:     ENSG00000164587\n",
      "gene_string='RPS15'\n",
      "gene_list=['RPS15']\n",
      "match=None\n",
      "['RPS15']:     ENSG00000115268\n",
      "gene_string='RPS15A'\n",
      "gene_list=['RPS15A']\n",
      "match=None\n",
      "['RPS15A']:     ENSG00000134419\n",
      "gene_string='RPS16'\n",
      "gene_list=['RPS16']\n",
      "match=None\n",
      "['RPS16']:     ENSG00000105193\n",
      "gene_string='RPS17'\n",
      "gene_list=['RPS17']\n",
      "match=None\n",
      "['RPS17']:     ENSG00000182774\n",
      "gene_string='RPS18'\n",
      "gene_list=['RPS18']\n",
      "match=None\n",
      "['RPS18']:     ENSG00000231500\n",
      "gene_string='RPS19'\n",
      "gene_list=['RPS19']\n",
      "match=None\n",
      "['RPS19']:     ENSG00000105372\n",
      "gene_string='RPS19BP1'\n",
      "gene_list=['RPS19BP1']\n",
      "match=None\n",
      "['RPS19BP1']:     ENSG00000187051\n",
      "gene_string='RPS20'\n",
      "gene_list=['RPS20']\n",
      "match=None\n",
      "['RPS20']:     ENSG00000008988\n",
      "gene_string='RPS21'\n",
      "gene_list=['RPS21']\n",
      "match=None\n",
      "['RPS21']:     ENSG00000171858\n",
      "gene_string='RPS23'\n",
      "gene_list=['RPS23']\n",
      "match=None\n",
      "['RPS23']:     ENSG00000186468\n",
      "gene_string='RPS24'\n",
      "gene_list=['RPS24']\n",
      "match=None\n",
      "['RPS24']:     ENSG00000138326\n",
      "gene_string='RPS25'\n",
      "gene_list=['RPS25']\n",
      "match=None\n",
      "['RPS25']:     ENSG00000118181\n",
      "gene_string='RPS26'\n",
      "gene_list=['RPS26']\n",
      "match=None\n",
      "['RPS26']:     ENSG00000197728\n",
      "gene_string='RPS27'\n",
      "gene_list=['RPS27']\n",
      "match=None\n",
      "['RPS27']:     ENSG00000177954\n",
      "gene_string='RPS27A'\n",
      "gene_list=['RPS27A']\n",
      "match=None\n",
      "['RPS27A']:     ENSG00000143947\n",
      "gene_string='RPS28'\n",
      "gene_list=['RPS28']\n",
      "match=None\n",
      "['RPS28']:     ENSG00000233927\n",
      "gene_string='RPS29'\n",
      "gene_list=['RPS29']\n",
      "match=None\n",
      "['RPS29']:     ENSG00000213741\n",
      "gene_string='RPSA'\n",
      "gene_list=['RPSA']\n",
      "match=None\n",
      "['RPSA']:     ENSG00000168028\n",
      "gene_string='RPTOR'\n",
      "gene_list=['RPTOR']\n",
      "match=None\n",
      "['RPTOR']:     ENSG00000141564\n",
      "gene_string='RPUSD3'\n",
      "gene_list=['RPUSD3']\n",
      "match=None\n",
      "['RPUSD3']:     ENSG00000156990\n",
      "gene_string='RPUSD4'\n",
      "gene_list=['RPUSD4']\n",
      "match=None\n",
      "['RPUSD4']:     ENSG00000165526\n",
      "gene_string='RRM1'\n",
      "gene_list=['RRM1']\n",
      "match=None\n",
      "['RRM1']:     ENSG00000167325\n",
      "gene_string='RRM2'\n",
      "gene_list=['RRM2']\n",
      "match=None\n",
      "['RRM2']:     ENSG00000171848\n",
      "gene_string='RRN3'\n",
      "gene_list=['RRN3']\n",
      "match=None\n",
      "['RRN3']:     ENSG00000085721\n",
      "gene_string='RRP1'\n",
      "gene_list=['RRP1']\n",
      "match=None\n",
      "['RRP1']:     ENSG00000160214\n",
      "gene_string='RRP7A'\n",
      "gene_list=['RRP7A']\n",
      "match=None\n",
      "['RRP7A']:     ENSG00000189306\n",
      "gene_string='RRP9'\n",
      "gene_list=['RRP9']\n",
      "match=None\n",
      "['RRP9']:     ENSG00000114767\n",
      "gene_string='RRP12'\n",
      "gene_list=['RRP12']\n",
      "match=None\n",
      "['RRP12']:     ENSG00000052749\n",
      "gene_string='RRP15'\n",
      "gene_list=['RRP15']\n",
      "match=None\n",
      "['RRP15']:     ENSG00000067533\n",
      "gene_string='RRS1'\n",
      "gene_list=['RRS1']\n",
      "match=None\n",
      "['RRS1']:     ENSG00000179041\n",
      "gene_string='RSL1D1'\n",
      "gene_list=['RSL1D1']\n",
      "match=None\n",
      "['RSL1D1']:     ENSG00000171490\n",
      "gene_string='RSL24D1'\n",
      "gene_list=['RSL24D1']\n",
      "match=None\n",
      "['RSL24D1']:     ENSG00000137876\n",
      "gene_string='RSRC2'\n",
      "gene_list=['RSRC2']\n",
      "match=None\n",
      "['RSRC2']:     ENSG00000111011\n",
      "gene_string='RTCB'\n",
      "gene_list=['RTCB']\n",
      "match=None\n",
      "['RTCB']:     ENSG00000100220\n",
      "gene_string='RTEL1'\n",
      "gene_list=['RTEL1']\n",
      "match=None\n",
      "['RTEL1']:     ENSG00000258366\n",
      "gene_string='RTF1'\n",
      "gene_list=['RTF1']\n",
      "match=None\n",
      "['RTF1']:     ENSG00000137815\n",
      "gene_string='RTTN'\n",
      "gene_list=['RTTN']\n",
      "match=None\n",
      "['RTTN']:     ENSG00000176225\n",
      "gene_string='RUVBL1'\n",
      "gene_list=['RUVBL1']\n",
      "match=None\n",
      "['RUVBL1']:     ENSG00000175792\n",
      "gene_string='RUVBL2'\n",
      "gene_list=['RUVBL2']\n",
      "match=None\n",
      "['RUVBL2']:     ENSG00000183207\n",
      "gene_string='SACM1L'\n",
      "gene_list=['SACM1L']\n",
      "match=None\n",
      "['SACM1L']:     ENSG00000211456\n",
      "gene_string='SAE1'\n",
      "gene_list=['SAE1']\n",
      "match=None\n",
      "['SAE1']:     ENSG00000142230\n",
      "gene_string='SAMM50'\n",
      "gene_list=['SAMM50']\n",
      "match=None\n",
      "['SAMM50']:     ENSG00000100347\n",
      "gene_string='SAP18'\n",
      "gene_list=['SAP18']\n",
      "match=None\n",
      "['SAP18']:     ENSG00000150459\n",
      "gene_string='SAP30BP'\n",
      "gene_list=['SAP30BP']\n",
      "match=None\n",
      "['SAP30BP']:     ENSG00000161526\n",
      "gene_string='SAP130'\n",
      "gene_list=['SAP130']\n",
      "match=None\n",
      "['SAP130']:     ENSG00000189091\n",
      "gene_string='SARNP'\n",
      "gene_list=['SARNP']\n",
      "match=None\n",
      "['SARNP']:     ENSG00000205323\n",
      "gene_string='SARS'\n",
      "gene_list=['SARS']\n",
      "match=None\n",
      "['SARS']:     ENSG00000104835\n",
      "gene_string='SARS2'\n",
      "gene_list=['SARS2']\n",
      "match=None\n",
      "['SARS2']:     ENSG00000104835\n",
      "gene_string='SART1'\n",
      "gene_list=['SART1']\n",
      "match=None\n",
      "['SART1']:     ENSG00000175467\n",
      "gene_string='SART3'\n",
      "gene_list=['SART3']\n",
      "match=None\n",
      "['SART3']:     ENSG00000075856\n",
      "gene_string='SASS6'\n",
      "gene_list=['SASS6']\n",
      "match=None\n",
      "['SASS6']:     ENSG00000156876\n",
      "gene_string='SBDS'\n",
      "gene_list=['SBDS']\n",
      "match=None\n",
      "['SBDS']:     ENSG00000126524\n",
      "gene_string='SBNO1'\n",
      "gene_list=['SBNO1']\n",
      "match=None\n",
      "['SBNO1']:     ENSG00000139697\n",
      "gene_string='SCAF1'\n",
      "gene_list=['SCAF1']\n",
      "match=None\n",
      "['SCAF1']:     ENSG00000115944\n",
      "gene_string='SCAP'\n",
      "gene_list=['SCAP']\n",
      "match=None\n",
      "['SCAP']:     ENSG00000114650\n",
      "gene_string='SCD'\n",
      "gene_list=['SCD']\n",
      "match=None\n",
      "['SCD']:     ENSG00000099194\n",
      "gene_string='SCFD1'\n",
      "gene_list=['SCFD1']\n",
      "match=None\n",
      "['SCFD1']:     ENSG00000092108\n",
      "gene_string='SCNM1'\n",
      "gene_list=['SCNM1']\n",
      "match=None\n",
      "['SCNM1']:     ENSG00000163156\n",
      "gene_string='SDAD1'\n",
      "gene_list=['SDAD1']\n",
      "match=None\n",
      "['SDAD1']:     ENSG00000198301\n",
      "gene_string='SDC1'\n",
      "gene_list=['SDC1']\n",
      "match=None\n",
      "['SDC1']:     ENSG00000115884\n",
      "gene_string='SDE2'\n",
      "gene_list=['SDE2']\n",
      "match=None\n",
      "['SDE2']:     ENSG00000143751\n",
      "gene_string='SDHA'\n",
      "gene_list=['SDHA']\n",
      "match=None\n",
      "['SDHA']:     ENSG00000073578\n",
      "gene_string='SDHAF2'\n",
      "gene_list=['SDHAF2']\n",
      "match=None\n",
      "['SDHAF2']:     ENSG00000167985\n",
      "gene_string='SDHC'\n",
      "gene_list=['SDHC']\n",
      "match=None\n",
      "['SDHC']:     ENSG00000143252\n",
      "gene_string='SDHD'\n",
      "gene_list=['SDHD']\n",
      "match=None\n",
      "['SDHD']:     ENSG00000204370\n",
      "gene_string='SEC13'\n",
      "gene_list=['SEC13']\n",
      "match=None\n",
      "['SEC13']:     ENSG00000157020\n",
      "gene_string='SEC16A'\n",
      "gene_list=['SEC16A']\n",
      "match=None\n",
      "['SEC16A']:     ENSG00000148396\n",
      "gene_string='SEC61A1'\n",
      "gene_list=['SEC61A1']\n",
      "match=None\n",
      "['SEC61A1']:     ENSG00000058262\n",
      "gene_string='SEC61B'\n",
      "gene_list=['SEC61B']\n",
      "match=None\n",
      "['SEC61B']:     ENSG00000106803\n",
      "gene_string='SEC61G'\n",
      "gene_list=['SEC61G']\n",
      "match=None\n",
      "['SEC61G']:     ENSG00000132432\n",
      "gene_string='SEC62'\n",
      "gene_list=['SEC62']\n",
      "match=None\n",
      "['SEC62']:     ENSG00000008952\n",
      "gene_string='SEC63'\n",
      "gene_list=['SEC63']\n",
      "match=None\n",
      "['SEC63']:     ENSG00000025796\n",
      "gene_string='SEH1L'\n",
      "gene_list=['SEH1L']\n",
      "match=None\n",
      "['SEH1L']:     ENSG00000085415\n",
      "gene_string='SEM1'\n",
      "gene_list=['SEM1']\n",
      "match=None\n",
      "['SEM1']:     ENSG00000127922\n",
      "gene_string='SENP6'\n",
      "gene_list=['SENP6']\n",
      "match=None\n",
      "['SENP6']:     ENSG00000112701\n",
      "gene_string='SEPHS2'\n",
      "gene_list=['SEPHS2']\n",
      "match=None\n",
      "['SEPHS2']:     ENSG00000179918\n",
      "gene_string='SEPSECS'\n",
      "gene_list=['SEPSECS']\n",
      "match=None\n",
      "['SEPSECS']:     ENSG00000109618\n",
      "gene_string='SERBP1'\n",
      "gene_list=['SERBP1']\n",
      "match=None\n",
      "['SERBP1']:     ENSG00000142864\n",
      "gene_string='SET'\n",
      "gene_list=['SET']\n",
      "match=None\n",
      "['SET']:     ENSG00000119335\n",
      "gene_string='SETD1A'\n",
      "gene_list=['SETD1A']\n",
      "match=None\n",
      "['SETD1A']:     ENSG00000099381\n",
      "gene_string='SETD2'\n",
      "gene_list=['SETD2']\n",
      "match=None\n",
      "['SETD2']:     ENSG00000181555\n",
      "gene_string='SETDB1'\n",
      "gene_list=['SETDB1']\n",
      "match=None\n",
      "['SETDB1']:     ENSG00000143379\n",
      "gene_string='SF1'\n",
      "gene_list=['SF1']\n",
      "match=None\n",
      "['SF1']:     ENSG00000168066\n",
      "gene_string='SF3A1'\n",
      "gene_list=['SF3A1']\n",
      "match=None\n",
      "['SF3A1']:     ENSG00000099995\n",
      "gene_string='SF3A2'\n",
      "gene_list=['SF3A2']\n",
      "match=None\n",
      "['SF3A2']:     ENSG00000104897\n",
      "gene_string='SF3A3'\n",
      "gene_list=['SF3A3']\n",
      "match=None\n",
      "['SF3A3']:     ENSG00000183431\n",
      "gene_string='SF3B1'\n",
      "gene_list=['SF3B1']\n",
      "match=None\n",
      "['SF3B1']:     ENSG00000087365\n",
      "gene_string='SF3B2'\n",
      "gene_list=['SF3B2']\n",
      "match=None\n",
      "['SF3B2']:     ENSG00000087365\n",
      "gene_string='SF3B3'\n",
      "gene_list=['SF3B3']\n",
      "match=None\n",
      "['SF3B3']:     ENSG00000189091\n",
      "gene_string='SF3B4'\n",
      "gene_list=['SF3B4']\n",
      "match=None\n",
      "['SF3B4']:     ENSG00000143368\n",
      "gene_string='SF3B5'\n",
      "gene_list=['SF3B5']\n",
      "match=None\n",
      "['SF3B5']:     ENSG00000169976\n",
      "gene_string='SFPQ'\n",
      "gene_list=['SFPQ']\n",
      "match=None\n",
      "['SFPQ']:     ENSG00000116560\n",
      "gene_string='SFSWAP'\n",
      "gene_list=['SFSWAP']\n",
      "match=None\n",
      "['SFSWAP']:     ENSG00000061936\n",
      "gene_string='SGPP1'\n",
      "gene_list=['SGPP1']\n",
      "match=None\n",
      "['SGPP1']:     ENSG00000126821\n",
      "gene_string='SHQ1'\n",
      "gene_list=['SHQ1']\n",
      "match=None\n",
      "['SHQ1']:     ENSG00000144736\n",
      "gene_string='SIGLEC14'\n",
      "gene_list=['SIGLEC14']\n",
      "match=None\n",
      "['SIGLEC14']:     ENSG00000254415\n",
      "gene_string='SIN3A'\n",
      "gene_list=['SIN3A']\n",
      "match=None\n",
      "['SIN3A']:     ENSG00000169375\n",
      "gene_string='SKA1'\n",
      "gene_list=['SKA1']\n",
      "match=None\n",
      "['SKA1']:     ENSG00000154839\n",
      "gene_string='SKA3'\n",
      "gene_list=['SKA3']\n",
      "match=None\n",
      "['SKA3']:     ENSG00000165480\n",
      "gene_string='SKP1'\n",
      "gene_list=['SKP1']\n",
      "match=None\n",
      "['SKP1']:     ENSG00000113558\n",
      "gene_string='SKP2'\n",
      "gene_list=['SKP2']\n",
      "match=None\n",
      "['SKP2']:     ENSG00000126756\n",
      "gene_string='SLBP'\n",
      "gene_list=['SLBP']\n",
      "match=None\n",
      "['SLBP']:     ENSG00000163950\n",
      "gene_string='SLC1A5'\n",
      "gene_list=['SLC1A5']\n",
      "match=None\n",
      "['SLC1A5']:     ENSG00000105281\n",
      "gene_string='SLC4A5'\n",
      "gene_list=['SLC4A5']\n",
      "match=None\n",
      "['SLC4A5']:     ENSG00000080493\n",
      "gene_string='SLC7A5'\n",
      "gene_list=['SLC7A5']\n",
      "match=None\n",
      "['SLC7A5']:     ENSG00000103257\n",
      "gene_string='SLC7A6OS'\n",
      "gene_list=['SLC7A6OS']\n",
      "match=None\n",
      "['SLC7A6OS']:     ENSG00000103061\n",
      "gene_string='SLC9B1'\n",
      "gene_list=['SLC9B1']\n",
      "match=None\n",
      "['SLC9B1']:     ENSG00000164037\n",
      "gene_string='SLC16A5'\n",
      "gene_list=['SLC16A5']\n",
      "match=None\n",
      "['SLC16A5']:     ENSG00000170190\n",
      "gene_string='SLC25A3'\n",
      "gene_list=['SLC25A3']\n",
      "match=None\n",
      "['SLC25A3']:     ENSG00000075415\n",
      "gene_string='SLC25A10'\n",
      "gene_list=['SLC25A10']\n",
      "match=None\n",
      "['SLC25A10']:     ENSG00000183048\n",
      "gene_string='SLC25A26'\n",
      "gene_list=['SLC25A26']\n",
      "match=None\n",
      "['SLC25A26']:     ENSG00000144741\n",
      "gene_string='SLC35B1'\n",
      "gene_list=['SLC35B1']\n",
      "match=None\n",
      "['SLC35B1']:     ENSG00000121073\n",
      "gene_string='SLC35G6'\n",
      "gene_list=['SLC35G6']\n",
      "match=None\n",
      "['SLC35G6']:     ENSG00000259224\n",
      "gene_string='SLC39A7'\n",
      "gene_list=['SLC39A7']\n",
      "match=None\n",
      "['SLC39A7']:     ENSG00000112473\n",
      "gene_string='SLC39A9'\n",
      "gene_list=['SLC39A9']\n",
      "match=None\n",
      "['SLC39A9']:     ENSG00000029364\n",
      "gene_string='SLC39A10'\n",
      "gene_list=['SLC39A10']\n",
      "match=None\n",
      "['SLC39A10']:     ENSG00000196950\n",
      "gene_string='SLU7'\n",
      "gene_list=['SLU7']\n",
      "match=None\n",
      "['SLU7']:     ENSG00000164609\n",
      "gene_string='SMAGP'\n",
      "gene_list=['SMAGP']\n",
      "match=None\n",
      "['SMAGP']:     ENSG00000170545\n",
      "gene_string='SMARCB1'\n",
      "gene_list=['SMARCB1']\n",
      "match=None\n",
      "['SMARCB1']:     ENSG00000099956\n",
      "gene_string='SMARCE1'\n",
      "gene_list=['SMARCE1']\n",
      "match=None\n",
      "['SMARCE1']:     ENSG00000073584\n",
      "gene_string='SMC1A'\n",
      "gene_list=['SMC1A']\n",
      "match=None\n",
      "['SMC1A']:     ENSG00000072501\n",
      "gene_string='SMC2'\n",
      "gene_list=['SMC2']\n",
      "match=None\n",
      "['SMC2']:     ENSG00000136824\n",
      "gene_string='SMC3'\n",
      "gene_list=['SMC3']\n",
      "match=None\n",
      "['SMC3']:     ENSG00000108055\n",
      "gene_string='SMC4'\n",
      "gene_list=['SMC4']\n",
      "match=None\n",
      "['SMC4']:     ENSG00000113810\n",
      "gene_string='SMC5'\n",
      "gene_list=['SMC5']\n",
      "match=None\n",
      "['SMC5']:     ENSG00000198887\n",
      "gene_string='SMC6'\n",
      "gene_list=['SMC6']\n",
      "match=None\n",
      "['SMC6']:     ENSG00000163029\n",
      "gene_string='SMG1'\n",
      "gene_list=['SMG1']\n",
      "match=None\n",
      "['SMG1']:     ENSG00000157106\n",
      "gene_string='SMG5'\n",
      "gene_list=['SMG5']\n",
      "match=None\n",
      "['SMG5']:     ENSG00000198952\n",
      "gene_string='SMG6'\n",
      "gene_list=['SMG6']\n",
      "match=None\n",
      "['SMG6']:     ENSG00000070366\n",
      "gene_string='SMN2'\n",
      "gene_list=['SMN2']\n",
      "match=None\n",
      "['SMN2']:     ENSG00000172062\n",
      "gene_string='SMNDC1'\n",
      "gene_list=['SMNDC1']\n",
      "match=None\n",
      "['SMNDC1']:     ENSG00000119953\n",
      "gene_string='SMU1'\n",
      "gene_list=['SMU1']\n",
      "match=None\n",
      "['SMU1']:     ENSG00000122692\n",
      "gene_string='SMUG1'\n",
      "gene_list=['SMUG1']\n",
      "match=None\n",
      "['SMUG1']:     ENSG00000123415\n",
      "gene_string='SNAPC1'\n",
      "gene_list=['SNAPC1']\n",
      "match=None\n",
      "['SNAPC1']:     ENSG00000023608\n",
      "gene_string='SNAPC2'\n",
      "gene_list=['SNAPC2']\n",
      "match=None\n",
      "['SNAPC2']:     ENSG00000104976\n",
      "gene_string='SNAPC3'\n",
      "gene_list=['SNAPC3']\n",
      "match=None\n",
      "['SNAPC3']:     ENSG00000164975\n",
      "gene_string='SNAPC4'\n",
      "gene_list=['SNAPC4']\n",
      "match=None\n",
      "['SNAPC4']:     ENSG00000165684\n",
      "gene_string='SNAPC5'\n",
      "gene_list=['SNAPC5']\n",
      "match=None\n",
      "['SNAPC5']:     ENSG00000174446\n",
      "gene_string='SNF8'\n",
      "gene_list=['SNF8']\n",
      "match=None\n",
      "['SNF8']:     ENSG00000159210\n",
      "gene_string='SNIP1'\n",
      "gene_list=['SNIP1']\n",
      "match=None\n",
      "['SNIP1']:     ENSG00000163877\n",
      "gene_string='SNRNP25'\n",
      "gene_list=['SNRNP25']\n",
      "match=None\n",
      "['SNRNP25']:     ENSG00000161981\n",
      "gene_string='SNRNP27'\n",
      "gene_list=['SNRNP27']\n",
      "match=None\n",
      "['SNRNP27']:     ENSG00000124380\n",
      "gene_string='SNRNP35'\n",
      "gene_list=['SNRNP35']\n",
      "match=None\n",
      "['SNRNP35']:     ENSG00000184209\n",
      "gene_string='SNRNP40'\n",
      "gene_list=['SNRNP40']\n",
      "match=None\n",
      "['SNRNP40']:     ENSG00000060688\n",
      "gene_string='SNRNP48'\n",
      "gene_list=['SNRNP48']\n",
      "match=None\n",
      "['SNRNP48']:     ENSG00000168566\n",
      "gene_string='SNRNP70'\n",
      "gene_list=['SNRNP70']\n",
      "match=None\n",
      "['SNRNP70']:     ENSG00000104852\n",
      "gene_string='SNRNP200'\n",
      "gene_list=['SNRNP200']\n",
      "match=None\n",
      "['SNRNP200']:     ENSG00000144028\n",
      "gene_string='SNRPA1'\n",
      "gene_list=['SNRPA1']\n",
      "match=None\n",
      "['SNRPA1']:     ENSG00000131876\n",
      "gene_string='SNRPB'\n",
      "gene_list=['SNRPB']\n",
      "match=None\n",
      "['SNRPB']:     ENSG00000125835\n",
      "gene_string='SNRPC'\n",
      "gene_list=['SNRPC']\n",
      "match=None\n",
      "['SNRPC']:     ENSG00000124562\n",
      "gene_string='SNRPD1'\n",
      "gene_list=['SNRPD1']\n",
      "match=None\n",
      "['SNRPD1']:     ENSG00000167088\n",
      "gene_string='SNRPD2'\n",
      "gene_list=['SNRPD2']\n",
      "match=None\n",
      "['SNRPD2']:     ENSG00000125743\n",
      "gene_string='SNRPD3'\n",
      "gene_list=['SNRPD3']\n",
      "match=None\n",
      "['SNRPD3']:     ENSG00000286070\n",
      "gene_string='SNRPE'\n",
      "gene_list=['SNRPE']\n",
      "match=None\n",
      "['SNRPE']:     ENSG00000182004\n",
      "gene_string='SNRPF'\n",
      "gene_list=['SNRPF']\n",
      "match=None\n",
      "['SNRPF']:     ENSG00000139343\n",
      "gene_string='SNRPG'\n",
      "gene_list=['SNRPG']\n",
      "match=None\n",
      "['SNRPG']:     ENSG00000143977\n",
      "gene_string='SNUPN'\n",
      "gene_list=['SNUPN']\n",
      "match=None\n",
      "['SNUPN']:     ENSG00000169371\n",
      "gene_string='SNW1'\n",
      "gene_list=['SNW1']\n",
      "match=None\n",
      "['SNW1']:     ENSG00000100603\n",
      "gene_string='SNX15'\n",
      "gene_list=['SNX15']\n",
      "match=None\n",
      "['SNX15']:     ENSG00000110025\n",
      "gene_string='SOD1'\n",
      "gene_list=['SOD1']\n",
      "match=None\n",
      "['SOD1']:     ENSG00000142168\n",
      "gene_string='SOD2'\n",
      "gene_list=['SOD2']\n",
      "match=None\n",
      "['SOD2']:     ENSG00000291237\n",
      "gene_string='SON'\n",
      "gene_list=['SON']\n",
      "match=None\n",
      "['SON']:     ENSG00000159140\n",
      "gene_string='SP2'\n",
      "gene_list=['SP2']\n",
      "match=None\n",
      "['SP2']:     ENSG00000167182\n",
      "gene_string='SPATA5'\n",
      "gene_list=['SPATA5']\n",
      "match=None\n",
      "['SPATA5']:     ENSG00000145375\n",
      "gene_string='SPATA5L1'\n",
      "gene_list=['SPATA5L1']\n",
      "match=None\n",
      "['SPATA5L1']:     ENSG00000171763\n",
      "gene_string='SPC24'\n",
      "gene_list=['SPC24']\n",
      "match=None\n",
      "['SPC24']:     ENSG00000161888\n",
      "gene_string='SPC25'\n",
      "gene_list=['SPC25']\n",
      "match=None\n",
      "['SPC25']:     ENSG00000152253\n",
      "gene_string='SPCS2'\n",
      "gene_list=['SPCS2']\n",
      "match=None\n",
      "['SPCS2']:     ENSG00000118363\n",
      "gene_string='SPCS3'\n",
      "gene_list=['SPCS3']\n",
      "match=None\n",
      "['SPCS3']:     ENSG00000129128\n",
      "gene_string='SPDL1'\n",
      "gene_list=['SPDL1']\n",
      "match=None\n",
      "['SPDL1']:     ENSG00000040275\n",
      "gene_string='SPDYE2'\n",
      "gene_list=['SPDYE2']\n",
      "match=None\n",
      "['SPDYE2']:     ENSG00000205238\n",
      "gene_string='SPRTN'\n",
      "gene_list=['SPRTN']\n",
      "match=None\n",
      "['SPRTN']:     ENSG00000010072\n",
      "gene_string='SPTLC1'\n",
      "gene_list=['SPTLC1']\n",
      "match=None\n",
      "['SPTLC1']:     ENSG00000090054\n",
      "gene_string='SPTSSA'\n",
      "gene_list=['SPTSSA']\n",
      "match=None\n",
      "['SPTSSA']:     ENSG00000165389\n",
      "gene_string='SRBD1'\n",
      "gene_list=['SRBD1']\n",
      "match=None\n",
      "['SRBD1']:     ENSG00000068784\n",
      "gene_string='SRCAP'\n",
      "gene_list=['SRCAP']\n",
      "match=None\n",
      "['SRCAP']:     ENSG00000080603\n",
      "gene_string='SRF'\n",
      "gene_list=['SRF']\n",
      "match=None\n",
      "['SRF']:     ENSG00000112658\n",
      "gene_string='SRFBP1'\n",
      "gene_list=['SRFBP1']\n",
      "match=None\n",
      "['SRFBP1']:     ENSG00000151304\n",
      "gene_string='SRP9'\n",
      "gene_list=['SRP9']\n",
      "match=None\n",
      "['SRP9']:     ENSG00000143742\n",
      "gene_string='SRP14'\n",
      "gene_list=['SRP14']\n",
      "match=None\n",
      "['SRP14']:     ENSG00000140319\n",
      "gene_string='SRP19'\n",
      "gene_list=['SRP19']\n",
      "match=None\n",
      "['SRP19']:     ENSG00000153037\n",
      "gene_string='SRP54'\n",
      "gene_list=['SRP54']\n",
      "match=None\n",
      "['SRP54']:     ENSG00000100883\n",
      "gene_string='SRP68'\n",
      "gene_list=['SRP68']\n",
      "match=None\n",
      "['SRP68']:     ENSG00000167881\n",
      "gene_string='SRP72'\n",
      "gene_list=['SRP72']\n",
      "match=None\n",
      "['SRP72']:     ENSG00000174780\n",
      "gene_string='SRPRB'\n",
      "gene_list=['SRPRB']\n",
      "match=None\n",
      "['SRPRB']:     ENSG00000144867\n",
      "gene_string='SRRM1'\n",
      "gene_list=['SRRM1']\n",
      "match=None\n",
      "['SRRM1']:     ENSG00000133226\n",
      "gene_string='SRRM2'\n",
      "gene_list=['SRRM2']\n",
      "match=None\n",
      "['SRRM2']:     ENSG00000167978\n",
      "gene_string='SRRT'\n",
      "gene_list=['SRRT']\n",
      "match=None\n",
      "['SRRT']:     ENSG00000087087\n",
      "gene_string='SRSF1'\n",
      "gene_list=['SRSF1']\n",
      "match=None\n",
      "['SRSF1']:     ENSG00000136450\n",
      "gene_string='SRSF2'\n",
      "gene_list=['SRSF2']\n",
      "match=None\n",
      "['SRSF2']:     ENSG00000161547\n",
      "gene_string='SRSF3'\n",
      "gene_list=['SRSF3']\n",
      "match=None\n",
      "['SRSF3']:     ENSG00000112081\n",
      "gene_string='SRSF6'\n",
      "gene_list=['SRSF6']\n",
      "match=None\n",
      "['SRSF6']:     ENSG00000124193\n",
      "gene_string='SRSF7'\n",
      "gene_list=['SRSF7']\n",
      "match=None\n",
      "['SRSF7']:     ENSG00000115875\n",
      "gene_string='SRSF10'\n",
      "gene_list=['SRSF10']\n",
      "match=None\n",
      "['SRSF10']:     ENSG00000188529\n",
      "gene_string='SRSF11'\n",
      "gene_list=['SRSF11']\n",
      "match=None\n",
      "['SRSF11']:     ENSG00000116754\n",
      "gene_string='SS18L2'\n",
      "gene_list=['SS18L2']\n",
      "match=None\n",
      "['SS18L2']:     ENSG00000008324\n",
      "gene_string='SSBP1'\n",
      "gene_list=['SSBP1']\n",
      "match=None\n",
      "['SSBP1']:     ENSG00000106028\n",
      "gene_string='SSBP3'\n",
      "gene_list=['SSBP3']\n",
      "match=None\n",
      "['SSBP3']:     ENSG00000157216\n",
      "gene_string='SSRP1'\n",
      "gene_list=['SSRP1']\n",
      "match=None\n",
      "['SSRP1']:     ENSG00000149136\n",
      "gene_string='SSU72'\n",
      "gene_list=['SSU72']\n",
      "match=None\n",
      "['SSU72']:     ENSG00000160075\n",
      "gene_string='SSX4'\n",
      "gene_list=['SSX4']\n",
      "match=None\n",
      "['SSX4']:     ENSG00000268009\n",
      "gene_string='STARD7'\n",
      "gene_list=['STARD7']\n",
      "match=None\n",
      "['STARD7']:     ENSG00000084090\n",
      "gene_string='STAT5A'\n",
      "gene_list=['STAT5A']\n",
      "match=None\n",
      "['STAT5A']:     ENSG00000126561\n",
      "gene_string='STIL'\n",
      "gene_list=['STIL']\n",
      "match=None\n",
      "['STIL']:     ENSG00000123473\n",
      "gene_string='STRAP'\n",
      "gene_list=['STRAP']\n",
      "match=None\n",
      "['STRAP']:     ENSG00000151304\n",
      "gene_string='STRIP1'\n",
      "gene_list=['STRIP1']\n",
      "match=None\n",
      "['STRIP1']:     ENSG00000143093\n",
      "gene_string='STX5'\n",
      "gene_list=['STX5']\n",
      "match=None\n",
      "['STX5']:     ENSG00000162236\n",
      "gene_string='STX18'\n",
      "gene_list=['STX18']\n",
      "match=None\n",
      "['STX18']:     ENSG00000168818\n",
      "gene_string='SUGP1'\n",
      "gene_list=['SUGP1']\n",
      "match=None\n",
      "['SUGP1']:     ENSG00000105705\n",
      "gene_string='SUGT1'\n",
      "gene_list=['SUGT1']\n",
      "match=None\n",
      "['SUGT1']:     ENSG00000165416\n",
      "gene_string='SUMO2'\n",
      "gene_list=['SUMO2']\n",
      "match=None\n",
      "['SUMO2']:     ENSG00000188612\n",
      "gene_string='SUPT4H1'\n",
      "gene_list=['SUPT4H1']\n",
      "match=None\n",
      "['SUPT4H1']:     ENSG00000213246\n",
      "gene_string='SUPT5H'\n",
      "gene_list=['SUPT5H']\n",
      "match=None\n",
      "['SUPT5H']:     ENSG00000196235\n",
      "gene_string='SUPT6H'\n",
      "gene_list=['SUPT6H']\n",
      "match=None\n",
      "['SUPT6H']:     ENSG00000109111\n",
      "gene_string='SUPT16H'\n",
      "gene_list=['SUPT16H']\n",
      "match=None\n",
      "['SUPT16H']:     ENSG00000092201\n",
      "gene_string='SUPV3L1'\n",
      "gene_list=['SUPV3L1']\n",
      "match=None\n",
      "['SUPV3L1']:     ENSG00000156502\n",
      "gene_string='SURF6'\n",
      "gene_list=['SURF6']\n",
      "match=None\n",
      "['SURF6']:     ENSG00000148296\n",
      "gene_string='SYF2'\n",
      "gene_list=['SYF2']\n",
      "match=None\n",
      "['SYF2']:     ENSG00000117614\n",
      "gene_string='SYMPK'\n",
      "gene_list=['SYMPK']\n",
      "match=None\n",
      "['SYMPK']:     ENSG00000125755\n",
      "gene_string='SYS1'\n",
      "gene_list=['SYS1']\n",
      "match=None\n",
      "['SYS1']:     ENSG00000204070\n",
      "gene_string='TACC3'\n",
      "gene_list=['TACC3']\n",
      "match=None\n",
      "['TACC3']:     ENSG00000013810\n",
      "gene_string='TADA2A'\n",
      "gene_list=['TADA2A']\n",
      "match=None\n",
      "['TADA2A']:     ENSG00000276234\n",
      "gene_string='TADA3'\n",
      "gene_list=['TADA3']\n",
      "match=None\n",
      "['TADA3']:     ENSG00000171148\n",
      "gene_string='TAF1'\n",
      "gene_list=['TAF1']\n",
      "match=None\n",
      "['TAF1']:     ENSG00000147133\n",
      "gene_string='TAF1B'\n",
      "gene_list=['TAF1B']\n",
      "match=None\n",
      "['TAF1B']:     ENSG00000115750\n",
      "gene_string='TAF1C'\n",
      "gene_list=['TAF1C']\n",
      "match=None\n",
      "['TAF1C']:     ENSG00000103168\n",
      "gene_string='TAF1D'\n",
      "gene_list=['TAF1D']\n",
      "match=None\n",
      "['TAF1D']:     ENSG00000166012\n",
      "gene_string='TAF2'\n",
      "gene_list=['TAF2']\n",
      "match=None\n",
      "['TAF2']:     ENSG00000064313\n",
      "gene_string='TAF5'\n",
      "gene_list=['TAF5']\n",
      "match=None\n",
      "['TAF5']:     ENSG00000148835\n",
      "gene_string='TAF6'\n",
      "gene_list=['TAF6']\n",
      "match=None\n",
      "['TAF6']:     ENSG00000106290\n",
      "gene_string='TAF7'\n",
      "gene_list=['TAF7']\n",
      "match=None\n",
      "['TAF7']:     ENSG00000178913\n",
      "gene_string='TAF8'\n",
      "gene_list=['TAF8']\n",
      "match=None\n",
      "['TAF8']:     ENSG00000137413\n",
      "gene_string='TAF10'\n",
      "gene_list=['TAF10']\n",
      "match=None\n",
      "['TAF10']:     ENSG00000166337\n",
      "gene_string='TAF12'\n",
      "gene_list=['TAF12']\n",
      "match=None\n",
      "['TAF12']:     ENSG00000120656\n",
      "gene_string='TAMM41'\n",
      "gene_list=['TAMM41']\n",
      "match=None\n",
      "['TAMM41']:     ENSG00000144559\n",
      "gene_string='TANGO6'\n",
      "gene_list=['TANGO6']\n",
      "match=None\n",
      "['TANGO6']:     ENSG00000103047\n",
      "gene_string='TARDBP'\n",
      "gene_list=['TARDBP']\n",
      "match=None\n",
      "['TARDBP']:     ENSG00000120948\n",
      "gene_string='TARS'\n",
      "gene_list=['TARS']\n",
      "match=None\n",
      "['TARS']:     ENSG00000113407\n",
      "gene_string='TARS2'\n",
      "gene_list=['TARS2']\n",
      "match=None\n",
      "['TARS2']:     ENSG00000143374\n",
      "gene_string='TBC1D3'\n",
      "gene_list=['TBC1D3']\n",
      "match=None\n",
      "['TBC1D3']:     ENSG00000274512\n",
      "gene_string='TBC1D28'\n",
      "gene_list=['TBC1D28']\n",
      "match=None\n",
      "['TBC1D28']:     ENSG00000189375\n",
      "gene_string='TBCA'\n",
      "gene_list=['TBCA']\n",
      "match=None\n",
      "['TBCA']:     ENSG00000171530\n",
      "gene_string='TBCB'\n",
      "gene_list=['TBCB']\n",
      "match=None\n",
      "['TBCB']:     ENSG00000105254\n",
      "gene_string='TBCC'\n",
      "gene_list=['TBCC']\n",
      "match=None\n",
      "['TBCC']:     ENSG00000124659\n",
      "gene_string='TBCD'\n",
      "gene_list=['TBCD']\n",
      "match=None\n",
      "['TBCD']:     ENSG00000141556\n",
      "gene_string='TBCE'\n",
      "gene_list=['TBCE']\n",
      "match=None\n",
      "['TBCE']:     ENSG00000284770\n",
      "gene_string='TBL3'\n",
      "gene_list=['TBL3']\n",
      "match=None\n",
      "['TBL3']:     ENSG00000183751\n",
      "gene_string='TBP'\n",
      "gene_list=['TBP']\n",
      "match=None\n",
      "['TBP']:     ENSG00000112592\n",
      "gene_string='TCERG1'\n",
      "gene_list=['TCERG1']\n",
      "match=None\n",
      "['TCERG1']:     ENSG00000113649\n",
      "gene_string='TCOF1'\n",
      "gene_list=['TCOF1']\n",
      "match=None\n",
      "['TCOF1']:     ENSG00000070814\n",
      "gene_string='TCP1'\n",
      "gene_list=['TCP1']\n",
      "match=None\n",
      "['TCP1']:     ENSG00000120438\n",
      "gene_string='TDGF1'\n",
      "gene_list=['TDGF1']\n",
      "match=None\n",
      "['TDGF1']:     ENSG00000241186\n",
      "gene_string='TEAD3'\n",
      "gene_list=['TEAD3']\n",
      "match=None\n",
      "['TEAD3']:     ENSG00000007866\n",
      "gene_string='TEFM'\n",
      "gene_list=['TEFM']\n",
      "match=None\n",
      "['TEFM']:     ENSG00000172171\n",
      "gene_string='TELO2'\n",
      "gene_list=['TELO2']\n",
      "match=None\n",
      "['TELO2']:     ENSG00000100726\n",
      "gene_string='TEN1'\n",
      "gene_list=['TEN1']\n",
      "match=None\n",
      "['TEN1']:     ENSG00000009694\n",
      "gene_string='TERF1'\n",
      "gene_list=['TERF1']\n",
      "match=None\n",
      "['TERF1']:     ENSG00000147601\n",
      "gene_string='TERF2'\n",
      "gene_list=['TERF2']\n",
      "match=None\n",
      "['TERF2']:     ENSG00000132604\n",
      "gene_string='TEX10'\n",
      "gene_list=['TEX10']\n",
      "match=None\n",
      "['TEX10']:     ENSG00000136891\n",
      "gene_string='TFAM'\n",
      "gene_list=['TFAM']\n",
      "match=None\n",
      "['TFAM']:     ENSG00000108064\n",
      "gene_string='TFB2M'\n",
      "gene_list=['TFB2M']\n",
      "match=None\n",
      "['TFB2M']:     ENSG00000162851\n",
      "gene_string='TFDP1'\n",
      "gene_list=['TFDP1']\n",
      "match=None\n",
      "['TFDP1']:     ENSG00000198176\n",
      "gene_string='TFIP11'\n",
      "gene_list=['TFIP11']\n",
      "match=None\n",
      "['TFIP11']:     ENSG00000100109\n",
      "gene_string='TFRC'\n",
      "gene_list=['TFRC']\n",
      "match=None\n",
      "['TFRC']:     ENSG00000072274\n",
      "gene_string='TGS1'\n",
      "gene_list=['TGS1']\n",
      "match=None\n",
      "['TGS1']:     ENSG00000137574\n",
      "gene_string='THAP1'\n",
      "gene_list=['THAP1']\n",
      "match=None\n",
      "['THAP1']:     ENSG00000131931\n",
      "gene_string='THAP11'\n",
      "gene_list=['THAP11']\n",
      "match=None\n",
      "['THAP11']:     ENSG00000168286\n",
      "gene_string='THG1L'\n",
      "gene_list=['THG1L']\n",
      "match=None\n",
      "['THG1L']:     ENSG00000113272\n",
      "gene_string='THOC1'\n",
      "gene_list=['THOC1']\n",
      "match=None\n",
      "['THOC1']:     ENSG00000079134\n",
      "gene_string='THOC2'\n",
      "gene_list=['THOC2']\n",
      "match=None\n",
      "['THOC2']:     ENSG00000125676\n",
      "gene_string='THOC3'\n",
      "gene_list=['THOC3']\n",
      "match=None\n",
      "['THOC3']:     ENSG00000051596\n",
      "gene_string='THOC5'\n",
      "gene_list=['THOC5']\n",
      "match=None\n",
      "['THOC5']:     ENSG00000100296\n",
      "gene_string='THOC6'\n",
      "gene_list=['THOC6']\n",
      "match=None\n",
      "['THOC6']:     ENSG00000131652\n",
      "gene_string='THOC7'\n",
      "gene_list=['THOC7']\n",
      "match=None\n",
      "['THOC7']:     ENSG00000163634\n",
      "gene_string='TICRR'\n",
      "gene_list=['TICRR']\n",
      "match=None\n",
      "['TICRR']:     ENSG00000140534\n",
      "gene_string='TIGD1'\n",
      "gene_list=['TIGD1']\n",
      "match=None\n",
      "['TIGD1']:     ENSG00000221944\n",
      "gene_string='TIMELESS'\n",
      "gene_list=['TIMELESS']\n",
      "match=None\n",
      "['TIMELESS']:     ENSG00000111602\n",
      "gene_string='TIMM8A'\n",
      "gene_list=['TIMM8A']\n",
      "match=None\n",
      "['TIMM8A']:     ENSG00000126953\n",
      "gene_string='TIMM9'\n",
      "gene_list=['TIMM9']\n",
      "match=None\n",
      "['TIMM9']:     ENSG00000100575\n",
      "gene_string='TIMM10'\n",
      "gene_list=['TIMM10']\n",
      "match=None\n",
      "['TIMM10']:     ENSG00000134809\n",
      "gene_string='TIMM13'\n",
      "gene_list=['TIMM13']\n",
      "match=None\n",
      "['TIMM13']:     ENSG00000099800\n",
      "gene_string='TIMM22'\n",
      "gene_list=['TIMM22']\n",
      "match=None\n",
      "['TIMM22']:     ENSG00000177370\n",
      "gene_string='TIMM23B'\n",
      "gene_list=['TIMM23B']\n",
      "match=None\n",
      "['TIMM23B']:     ENSG00000204152\n",
      "gene_string='TIMM44'\n",
      "gene_list=['TIMM44']\n",
      "match=None\n",
      "['TIMM44']:     ENSG00000104980\n",
      "gene_string='TINF2'\n",
      "gene_list=['TINF2']\n",
      "match=None\n",
      "['TINF2']:     ENSG00000092330\n",
      "gene_string='TIPIN'\n",
      "gene_list=['TIPIN']\n",
      "match=None\n",
      "['TIPIN']:     ENSG00000075131\n",
      "gene_string='TKT'\n",
      "gene_list=['TKT']\n",
      "match=None\n",
      "['TKT']:     ENSG00000163931\n",
      "gene_string='TLCD1'\n",
      "gene_list=['TLCD1']\n",
      "match=None\n",
      "['TLCD1']:     ENSG00000160606\n",
      "gene_string='TLN1'\n",
      "gene_list=['TLN1']\n",
      "match=None\n",
      "['TLN1']:     ENSG00000137076\n",
      "gene_string='TM7SF2'\n",
      "gene_list=['TM7SF2']\n",
      "match=None\n",
      "['TM7SF2']:     ENSG00000149809\n",
      "gene_string='TMA16'\n",
      "gene_list=['TMA16']\n",
      "match=None\n",
      "['TMA16']:     ENSG00000198498\n",
      "gene_string='TMED2'\n",
      "gene_list=['TMED2']\n",
      "match=None\n",
      "['TMED2']:     ENSG00000086598\n",
      "gene_string='TMEM127'\n",
      "gene_list=['TMEM127']\n",
      "match=None\n",
      "['TMEM127']:     ENSG00000135956\n",
      "gene_string='TMEM199'\n",
      "gene_list=['TMEM199']\n",
      "match=None\n",
      "['TMEM199']:     ENSG00000244045\n",
      "gene_string='TMEM240'\n",
      "gene_list=['TMEM240']\n",
      "match=None\n",
      "['TMEM240']:     ENSG00000205090\n",
      "gene_string='TMEM258'\n",
      "gene_list=['TMEM258']\n",
      "match=None\n",
      "['TMEM258']:     ENSG00000134825\n",
      "gene_string='TMSB10'\n",
      "gene_list=['TMSB10']\n",
      "match=None\n",
      "['TMSB10']:     ENSG00000034510\n",
      "gene_string='TMX2'\n",
      "gene_list=['TMX2']\n",
      "match=None\n",
      "['TMX2']:     ENSG00000213593\n",
      "gene_string='TNNT2'\n",
      "gene_list=['TNNT2']\n",
      "match=None\n",
      "['TNNT2']:     ENSG00000118194\n",
      "gene_string='TNPO3'\n",
      "gene_list=['TNPO3']\n",
      "match=None\n",
      "['TNPO3']:     ENSG00000064419\n",
      "gene_string='TOE1'\n",
      "gene_list=['TOE1']\n",
      "match=None\n",
      "['TOE1']:     ENSG00000132773\n",
      "gene_string='TOMM20'\n",
      "gene_list=['TOMM20']\n",
      "match=None\n",
      "['TOMM20']:     ENSG00000173726\n",
      "gene_string='TOMM22'\n",
      "gene_list=['TOMM22']\n",
      "match=None\n",
      "['TOMM22']:     ENSG00000100216\n",
      "gene_string='TOMM40'\n",
      "gene_list=['TOMM40']\n",
      "match=None\n",
      "['TOMM40']:     ENSG00000130204\n",
      "gene_string='TONSL'\n",
      "gene_list=['TONSL']\n",
      "match=None\n",
      "['TONSL']:     ENSG00000160949\n",
      "gene_string='TOP1'\n",
      "gene_list=['TOP1']\n",
      "match=None\n",
      "['TOP1']:     ENSG00000198900\n",
      "gene_string='TOP2A'\n",
      "gene_list=['TOP2A']\n",
      "match=None\n",
      "['TOP2A']:     ENSG00000131747\n",
      "gene_string='TOP3A'\n",
      "gene_list=['TOP3A']\n",
      "match=None\n",
      "['TOP3A']:     ENSG00000177302\n",
      "gene_string='TOPBP1'\n",
      "gene_list=['TOPBP1']\n",
      "match=None\n",
      "['TOPBP1']:     ENSG00000163781\n",
      "gene_string='TOR2A'\n",
      "gene_list=['TOR2A']\n",
      "match=None\n",
      "['TOR2A']:     ENSG00000160404\n",
      "gene_string='TP53RK'\n",
      "gene_list=['TP53RK']\n",
      "match=None\n",
      "['TP53RK']:     ENSG00000172315\n",
      "gene_string='TPI1'\n",
      "gene_list=['TPI1']\n",
      "match=None\n",
      "['TPI1']:     ENSG00000111669\n",
      "gene_string='TPR'\n",
      "gene_list=['TPR']\n",
      "match=None\n",
      "['TPR']:     ENSG00000047410\n",
      "gene_string='TPRKB'\n",
      "gene_list=['TPRKB']\n",
      "match=None\n",
      "['TPRKB']:     ENSG00000172315\n",
      "gene_string='TPT1'\n",
      "gene_list=['TPT1']\n",
      "match=None\n",
      "['TPT1']:     ENSG00000133112\n",
      "gene_string='TPX2'\n",
      "gene_list=['TPX2']\n",
      "match=None\n",
      "['TPX2']:     ENSG00000088325\n",
      "gene_string='TRA2B'\n",
      "gene_list=['TRA2B']\n",
      "match=None\n",
      "['TRA2B']:     ENSG00000136527\n",
      "gene_string='TRAIP'\n",
      "gene_list=['TRAIP']\n",
      "match=None\n",
      "['TRAIP']:     ENSG00000183763\n",
      "gene_string='TRAPPC1'\n",
      "gene_list=['TRAPPC1']\n",
      "match=None\n",
      "['TRAPPC1']:     ENSG00000170043\n",
      "gene_string='TRAPPC3'\n",
      "gene_list=['TRAPPC3']\n",
      "match=None\n",
      "['TRAPPC3']:     ENSG00000054116\n",
      "gene_string='TRAPPC4'\n",
      "gene_list=['TRAPPC4']\n",
      "match=None\n",
      "['TRAPPC4']:     ENSG00000196655\n",
      "gene_string='TRAPPC5'\n",
      "gene_list=['TRAPPC5']\n",
      "match=None\n",
      "['TRAPPC5']:     ENSG00000181029\n",
      "gene_string='TRAPPC8'\n",
      "gene_list=['TRAPPC8']\n",
      "match=None\n",
      "['TRAPPC8']:     ENSG00000153339\n",
      "gene_string='TRAPPC11'\n",
      "gene_list=['TRAPPC11']\n",
      "match=None\n",
      "['TRAPPC11']:     ENSG00000168538\n",
      "gene_string='TREML2'\n",
      "gene_list=['TREML2']\n",
      "match=None\n",
      "['TREML2']:     ENSG00000112195\n",
      "gene_string='TREX2'\n",
      "gene_list=['TREX2']\n",
      "match=None\n",
      "['TREX2']:     ENSG00000183479\n",
      "gene_string='TRIAP1'\n",
      "gene_list=['TRIAP1']\n",
      "match=None\n",
      "['TRIAP1']:     ENSG00000170855\n",
      "gene_string='TRIM49C'\n",
      "gene_list=['TRIM49C']\n",
      "match=None\n",
      "['TRIM49C']:     ENSG00000204449\n",
      "gene_string='TRMT5'\n",
      "gene_list=['TRMT5']\n",
      "match=None\n",
      "['TRMT5']:     ENSG00000126814\n",
      "gene_string='TRMT6'\n",
      "gene_list=['TRMT6']\n",
      "match=None\n",
      "['TRMT6']:     ENSG00000089195\n",
      "gene_string='TRMT10C'\n",
      "gene_list=['TRMT10C']\n",
      "match=None\n",
      "['TRMT10C']:     ENSG00000174173\n",
      "gene_string='TRMT112'\n",
      "gene_list=['TRMT112']\n",
      "match=None\n",
      "['TRMT112']:     ENSG00000173113\n",
      "gene_string='TRNP1'\n",
      "gene_list=['TRNP1']\n",
      "match=None\n",
      "['TRNP1']:     ENSG00000253368\n",
      "gene_string='TRNT1'\n",
      "gene_list=['TRNT1']\n",
      "match=None\n",
      "['TRNT1']:     ENSG00000072756\n",
      "gene_string='TRPM7'\n",
      "gene_list=['TRPM7']\n",
      "match=None\n",
      "['TRPM7']:     ENSG00000092439\n",
      "gene_string='TRRAP'\n",
      "gene_list=['TRRAP']\n",
      "match=None\n",
      "['TRRAP']:     ENSG00000196367\n",
      "gene_string='TSEN2'\n",
      "gene_list=['TSEN2']\n",
      "match=None\n",
      "['TSEN2']:     ENSG00000154743\n",
      "gene_string='TSEN34'\n",
      "gene_list=['TSEN34']\n",
      "match=None\n",
      "['TSEN34']:     ENSG00000170892\n",
      "gene_string='TSEN54'\n",
      "gene_list=['TSEN54']\n",
      "match=None\n",
      "['TSEN54']:     ENSG00000182173\n",
      "gene_string='TSFM'\n",
      "gene_list=['TSFM']\n",
      "match=None\n",
      "['TSFM']:     ENSG00000123297\n",
      "gene_string='TSG101'\n",
      "gene_list=['TSG101']\n",
      "match=None\n",
      "['TSG101']:     ENSG00000074319\n",
      "gene_string='TSPYL5'\n",
      "gene_list=['TSPYL5']\n",
      "match=None\n",
      "['TSPYL5']:     ENSG00000180543\n",
      "gene_string='TSR1'\n",
      "gene_list=['TSR1']\n",
      "match=None\n",
      "['TSR1']:     ENSG00000167721\n",
      "gene_string='TSR2'\n",
      "gene_list=['TSR2']\n",
      "match=None\n",
      "['TSR2']:     ENSG00000158526\n",
      "gene_string='TSSK3'\n",
      "gene_list=['TSSK3']\n",
      "match=None\n",
      "['TSSK3']:     ENSG00000162526\n",
      "gene_string='TTC1'\n",
      "gene_list=['TTC1']\n",
      "match=None\n",
      "['TTC1']:     ENSG00000113312\n",
      "gene_string='TTC4'\n",
      "gene_list=['TTC4']\n",
      "match=None\n",
      "['TTC4']:     ENSG00000243725\n",
      "gene_string='TTC27'\n",
      "gene_list=['TTC27']\n",
      "match=None\n",
      "['TTC27']:     ENSG00000018699\n",
      "gene_string='TTF2'\n",
      "gene_list=['TTF2']\n",
      "match=None\n",
      "['TTF2']:     ENSG00000116830\n",
      "gene_string='TTI1'\n",
      "gene_list=['TTI1']\n",
      "match=None\n",
      "['TTI1']:     ENSG00000101407\n",
      "gene_string='TTI2'\n",
      "gene_list=['TTI2']\n",
      "match=None\n",
      "['TTI2']:     ENSG00000129696\n",
      "gene_string='TTK'\n",
      "gene_list=['TTK']\n",
      "match=None\n",
      "['TTK']:     ENSG00000112742\n",
      "gene_string='TUBA1B'\n",
      "gene_list=['TUBA1B']\n",
      "match=None\n",
      "['TUBA1B']:     ENSG00000123416\n",
      "gene_string='TUBA1C'\n",
      "gene_list=['TUBA1C']\n",
      "match=None\n",
      "['TUBA1C']:     ENSG00000167553\n",
      "gene_string='TUBA3D'\n",
      "gene_list=['TUBA3D']\n",
      "match=None\n",
      "['TUBA3D']:     ENSG00000075886\n",
      "gene_string='TUBB'\n",
      "gene_list=['TUBB']\n",
      "match=None\n",
      "['TUBB']:     ENSG00000137267\n",
      "gene_string='TUBD1'\n",
      "gene_list=['TUBD1']\n",
      "match=None\n",
      "['TUBD1']:     ENSG00000108423\n",
      "gene_string='TUBE1'\n",
      "gene_list=['TUBE1']\n",
      "match=None\n",
      "['TUBE1']:     ENSG00000074935\n",
      "gene_string='TUBG1'\n",
      "gene_list=['TUBG1']\n",
      "match=None\n",
      "['TUBG1']:     ENSG00000131462\n",
      "gene_string='TUBGCP2'\n",
      "gene_list=['TUBGCP2']\n",
      "match=None\n",
      "['TUBGCP2']:     ENSG00000130640\n",
      "gene_string='TUBGCP3'\n",
      "gene_list=['TUBGCP3']\n",
      "match=None\n",
      "['TUBGCP3']:     ENSG00000126216\n",
      "gene_string='TUBGCP4'\n",
      "gene_list=['TUBGCP4']\n",
      "match=None\n",
      "['TUBGCP4']:     ENSG00000137822\n",
      "gene_string='TUBGCP5'\n",
      "gene_list=['TUBGCP5']\n",
      "match=None\n",
      "['TUBGCP5']:     ENSG00000275835\n",
      "gene_string='TUBGCP6'\n",
      "gene_list=['TUBGCP6']\n",
      "match=None\n",
      "['TUBGCP6']:     ENSG00000128159\n",
      "gene_string='TULP1'\n",
      "gene_list=['TULP1']\n",
      "match=None\n",
      "['TULP1']:     ENSG00000112041\n",
      "gene_string='TUT1'\n",
      "gene_list=['TUT1']\n",
      "match=None\n",
      "['TUT1']:     ENSG00000149016\n",
      "gene_string='TVP23C'\n",
      "gene_list=['TVP23C']\n",
      "match=None\n",
      "['TVP23C']:     ENSG00000259024\n",
      "gene_string='TWF1'\n",
      "gene_list=['TWF1']\n",
      "match=None\n",
      "['TWF1']:     ENSG00000151239\n",
      "gene_string='TWISTNB'\n",
      "gene_list=['TWISTNB']\n",
      "match=None\n",
      "['TWISTNB']:     ENSG00000105849\n",
      "gene_string='TXN'\n",
      "gene_list=['TXN']\n",
      "match=None\n",
      "['TXN']:     ENSG00000136810\n",
      "gene_string='TXNL4A'\n",
      "gene_list=['TXNL4A']\n",
      "match=None\n",
      "['TXNL4A']:     ENSG00000141759\n",
      "gene_string='TXNL4B'\n",
      "gene_list=['TXNL4B']\n",
      "match=None\n",
      "['TXNL4B']:     ENSG00000140830\n",
      "gene_string='U2AF1'\n",
      "gene_list=['U2AF1']\n",
      "match=None\n",
      "['U2AF1']:     ENSG00000160201\n",
      "gene_string='U2AF2'\n",
      "gene_list=['U2AF2']\n",
      "match=None\n",
      "['U2AF2']:     ENSG00000063244\n",
      "gene_string='U2SURP'\n",
      "gene_list=['U2SURP']\n",
      "match=None\n",
      "['U2SURP']:     ENSG00000163714\n",
      "gene_string='UBA1'\n",
      "gene_list=['UBA1']\n",
      "match=None\n",
      "['UBA1']:     ENSG00000130985\n",
      "gene_string='UBA2'\n",
      "gene_list=['UBA2']\n",
      "match=None\n",
      "['UBA2']:     ENSG00000126261\n",
      "gene_string='UBA3'\n",
      "gene_list=['UBA3']\n",
      "match=None\n",
      "['UBA3']:     ENSG00000144744\n",
      "gene_string='UBA5'\n",
      "gene_list=['UBA5']\n",
      "match=None\n",
      "['UBA5']:     ENSG00000081307\n",
      "gene_string='UBA52'\n",
      "gene_list=['UBA52']\n",
      "match=None\n",
      "['UBA52']:     ENSG00000221983\n",
      "gene_string='UBAP1'\n",
      "gene_list=['UBAP1']\n",
      "match=None\n",
      "['UBAP1']:     ENSG00000165006\n",
      "gene_string='UBC'\n",
      "gene_list=['UBC']\n",
      "match=None\n",
      "['UBC']:     ENSG00000150991\n",
      "gene_string='UBE2D3'\n",
      "gene_list=['UBE2D3']\n",
      "match=None\n",
      "['UBE2D3']:     ENSG00000109332\n",
      "gene_string='UBE2H'\n",
      "gene_list=['UBE2H']\n",
      "match=None\n",
      "['UBE2H']:     ENSG00000186591\n",
      "gene_string='UBE2I'\n",
      "gene_list=['UBE2I']\n",
      "match=None\n",
      "['UBE2I']:     ENSG00000103275\n",
      "gene_string='UBE2L3'\n",
      "gene_list=['UBE2L3']\n",
      "match=None\n",
      "['UBE2L3']:     ENSG00000185651\n",
      "gene_string='UBE2M'\n",
      "gene_list=['UBE2M']\n",
      "match=None\n",
      "['UBE2M']:     ENSG00000130725\n",
      "gene_string='UBE2N'\n",
      "gene_list=['UBE2N']\n",
      "match=None\n",
      "['UBE2N']:     ENSG00000177889\n",
      "gene_string='UBE2Z'\n",
      "gene_list=['UBE2Z']\n",
      "match=None\n",
      "['UBE2Z']:     ENSG00000159202\n",
      "gene_string='UBL5'\n",
      "gene_list=['UBL5']\n",
      "match=None\n",
      "['UBL5']:     ENSG00000198258\n",
      "gene_string='UBQLN4'\n",
      "gene_list=['UBQLN4']\n",
      "match=None\n",
      "['UBQLN4']:     ENSG00000160803\n",
      "gene_string='UBR4'\n",
      "gene_list=['UBR4']\n",
      "match=None\n",
      "['UBR4']:     ENSG00000127481\n",
      "gene_string='UBR5'\n",
      "gene_list=['UBR5']\n",
      "match=None\n",
      "['UBR5']:     ENSG00000104517\n",
      "gene_string='UBTF'\n",
      "gene_list=['UBTF']\n",
      "match=None\n",
      "['UBTF']:     ENSG00000108312\n",
      "gene_string='UFL1'\n",
      "gene_list=['UFL1']\n",
      "match=None\n",
      "['UFL1']:     ENSG00000014123\n",
      "gene_string='UFM1'\n",
      "gene_list=['UFM1']\n",
      "match=None\n",
      "['UFM1']:     ENSG00000120686\n",
      "gene_string='UNC45A'\n",
      "gene_list=['UNC45A']\n",
      "match=None\n",
      "['UNC45A']:     ENSG00000140553\n",
      "gene_string='UNC50'\n",
      "gene_list=['UNC50']\n",
      "match=None\n",
      "['UNC50']:     ENSG00000115446\n",
      "gene_string='UPF1'\n",
      "gene_list=['UPF1']\n",
      "match=None\n",
      "['UPF1']:     ENSG00000005007\n",
      "gene_string='UPF2'\n",
      "gene_list=['UPF2']\n",
      "match=None\n",
      "['UPF2']:     ENSG00000151461\n",
      "gene_string='UPF3A'\n",
      "gene_list=['UPF3A']\n",
      "match=None\n",
      "['UPF3A']:     ENSG00000169062\n",
      "gene_string='UQCC2'\n",
      "gene_list=['UQCC2']\n",
      "match=None\n",
      "['UQCC2']:     ENSG00000137288\n",
      "gene_string='UQCRB'\n",
      "gene_list=['UQCRB']\n",
      "match=None\n",
      "['UQCRB']:     ENSG00000156467\n",
      "gene_string='UQCRC1'\n",
      "gene_list=['UQCRC1']\n",
      "match=None\n",
      "['UQCRC1']:     ENSG00000010256\n",
      "gene_string='UQCRC2'\n",
      "gene_list=['UQCRC2']\n",
      "match=None\n",
      "['UQCRC2']:     ENSG00000140740\n",
      "gene_string='UQCRFS1'\n",
      "gene_list=['UQCRFS1']\n",
      "match=None\n",
      "['UQCRFS1']:     ENSG00000169021\n",
      "gene_string='UQCRH'\n",
      "gene_list=['UQCRH']\n",
      "match=None\n",
      "['UQCRH']:     ENSG00000173660\n",
      "gene_string='UQCRQ'\n",
      "gene_list=['UQCRQ']\n",
      "match=None\n",
      "['UQCRQ']:     ENSG00000164405\n",
      "gene_string='URB1'\n",
      "gene_list=['URB1']\n",
      "match=None\n",
      "['URB1']:     ENSG00000142207\n",
      "gene_string='URB2'\n",
      "gene_list=['URB2']\n",
      "match=None\n",
      "['URB2']:     ENSG00000135763\n",
      "gene_string='URI1'\n",
      "gene_list=['URI1']\n",
      "match=None\n",
      "['URI1']:     ENSG00000105176\n",
      "gene_string='URM1'\n",
      "gene_list=['URM1']\n",
      "match=None\n",
      "['URM1']:     ENSG00000167118\n",
      "gene_string='UROD'\n",
      "gene_list=['UROD']\n",
      "match=None\n",
      "['UROD']:     ENSG00000126088\n",
      "gene_string='USF2'\n",
      "gene_list=['USF2']\n",
      "match=None\n",
      "['USF2']:     ENSG00000105698\n",
      "gene_string='USP5'\n",
      "gene_list=['USP5']\n",
      "match=None\n",
      "['USP5']:     ENSG00000111667\n",
      "gene_string='USP7'\n",
      "gene_list=['USP7']\n",
      "match=None\n",
      "['USP7']:     ENSG00000187555\n",
      "gene_string='USP8'\n",
      "gene_list=['USP8']\n",
      "match=None\n",
      "['USP8']:     ENSG00000138592\n",
      "gene_string='USP9X'\n",
      "gene_list=['USP9X']\n",
      "match=None\n",
      "['USP9X']:     ENSG00000124486\n",
      "gene_string='USP10'\n",
      "gene_list=['USP10']\n",
      "match=None\n",
      "['USP10']:     ENSG00000170832\n",
      "gene_string='USP19'\n",
      "gene_list=['USP19']\n",
      "match=None\n",
      "['USP19']:     ENSG00000172046\n",
      "gene_string='USP36'\n",
      "gene_list=['USP36']\n",
      "match=None\n",
      "['USP36']:     ENSG00000055483\n",
      "gene_string='USP37'\n",
      "gene_list=['USP37']\n",
      "match=None\n",
      "['USP37']:     ENSG00000135913\n",
      "gene_string='USP39'\n",
      "gene_list=['USP39']\n",
      "match=None\n",
      "['USP39']:     ENSG00000168883\n",
      "gene_string='USPL1'\n",
      "gene_list=['USPL1']\n",
      "match=None\n",
      "['USPL1']:     ENSG00000132952\n",
      "gene_string='UTP3'\n",
      "gene_list=['UTP3']\n",
      "match=None\n",
      "['UTP3']:     ENSG00000132467\n",
      "gene_string='UTP6'\n",
      "gene_list=['UTP6']\n",
      "match=None\n",
      "['UTP6']:     ENSG00000108651\n",
      "gene_string='UTP15'\n",
      "gene_list=['UTP15']\n",
      "match=None\n",
      "['UTP15']:     ENSG00000164338\n",
      "gene_string='UTP18'\n",
      "gene_list=['UTP18']\n",
      "match=None\n",
      "['UTP18']:     ENSG00000011260\n",
      "gene_string='UTP20'\n",
      "gene_list=['UTP20']\n",
      "match=None\n",
      "['UTP20']:     ENSG00000120800\n",
      "gene_string='UTP23'\n",
      "gene_list=['UTP23']\n",
      "match=None\n",
      "['UTP23']:     ENSG00000147679\n",
      "gene_string='UXS1'\n",
      "gene_list=['UXS1']\n",
      "match=None\n",
      "['UXS1']:     ENSG00000115652\n",
      "gene_string='UXT'\n",
      "gene_list=['UXT']\n",
      "match=None\n",
      "['UXT']:     ENSG00000126756\n",
      "gene_string='VARS'\n",
      "gene_list=['VARS']\n",
      "match=None\n",
      "['VARS']:     ENSG00000204394\n",
      "gene_string='VARS2'\n",
      "gene_list=['VARS2']\n",
      "match=None\n",
      "['VARS2']:     ENSG00000204394\n",
      "gene_string='VBP1'\n",
      "gene_list=['VBP1']\n",
      "match=None\n",
      "['VBP1']:     ENSG00000155959\n",
      "gene_string='VCP'\n",
      "gene_list=['VCP']\n",
      "match=None\n",
      "['VCP']:     ENSG00000165280\n",
      "gene_string='VEZT'\n",
      "gene_list=['VEZT']\n",
      "match=None\n",
      "['VEZT']:     ENSG00000028203\n",
      "gene_string='VHL'\n",
      "gene_list=['VHL']\n",
      "match=None\n",
      "['VHL']:     ENSG00000134086\n",
      "gene_string='VMP1'\n",
      "gene_list=['VMP1']\n",
      "match=None\n",
      "['VMP1']:     ENSG00000062716\n",
      "gene_string='VPS13D'\n",
      "gene_list=['VPS13D']\n",
      "match=None\n",
      "['VPS13D']:     ENSG00000048707\n",
      "gene_string='VPS18'\n",
      "gene_list=['VPS18']\n",
      "match=None\n",
      "['VPS18']:     ENSG00000104142\n",
      "gene_string='VPS25'\n",
      "gene_list=['VPS25']\n",
      "match=None\n",
      "['VPS25']:     ENSG00000131475\n",
      "gene_string='VPS28'\n",
      "gene_list=['VPS28']\n",
      "match=None\n",
      "['VPS28']:     ENSG00000160948\n",
      "gene_string='VPS29'\n",
      "gene_list=['VPS29']\n",
      "match=None\n",
      "['VPS29']:     ENSG00000111237\n",
      "gene_string='VPS33A'\n",
      "gene_list=['VPS33A']\n",
      "match=None\n",
      "['VPS33A']:     ENSG00000139719\n",
      "gene_string='VPS35'\n",
      "gene_list=['VPS35']\n",
      "match=None\n",
      "['VPS35']:     ENSG00000069329\n",
      "gene_string='VPS37A'\n",
      "gene_list=['VPS37A']\n",
      "match=None\n",
      "['VPS37A']:     ENSG00000155975\n",
      "gene_string='VPS37C'\n",
      "gene_list=['VPS37C']\n",
      "match=None\n",
      "['VPS37C']:     ENSG00000167987\n",
      "gene_string='VPS41'\n",
      "gene_list=['VPS41']\n",
      "match=None\n",
      "['VPS41']:     ENSG00000006715\n",
      "gene_string='VPS51'\n",
      "gene_list=['VPS51']\n",
      "match=None\n",
      "['VPS51']:     ENSG00000149823\n",
      "gene_string='VPS54'\n",
      "gene_list=['VPS54']\n",
      "match=None\n",
      "['VPS54']:     ENSG00000143952\n",
      "gene_string='VPS72'\n",
      "gene_list=['VPS72']\n",
      "match=None\n",
      "['VPS72']:     ENSG00000163159\n",
      "gene_string='WAC'\n",
      "gene_list=['WAC']\n",
      "match=None\n",
      "['WAC']:     ENSG00000095787\n",
      "gene_string='WARS'\n",
      "gene_list=['WARS']\n",
      "match=None\n",
      "['WARS']:     ENSG00000140105\n",
      "gene_string='WBP1'\n",
      "gene_list=['WBP1']\n",
      "match=None\n",
      "['WBP1']:     ENSG00000239779\n",
      "gene_string='WBP1L'\n",
      "gene_list=['WBP1L']\n",
      "match=None\n",
      "['WBP1L']:     ENSG00000166272\n",
      "gene_string='WBP11'\n",
      "gene_list=['WBP11']\n",
      "match=None\n",
      "['WBP11']:     ENSG00000084463\n",
      "gene_string='WDHD1'\n",
      "gene_list=['WDHD1']\n",
      "match=None\n",
      "['WDHD1']:     ENSG00000198554\n",
      "gene_string='WDR1'\n",
      "gene_list=['WDR1']\n",
      "match=None\n",
      "['WDR1']:     ENSG00000071127\n",
      "gene_string='WDR3'\n",
      "gene_list=['WDR3']\n",
      "match=None\n",
      "['WDR3']:     ENSG00000065183\n",
      "gene_string='WDR4'\n",
      "gene_list=['WDR4']\n",
      "match=None\n",
      "['WDR4']:     ENSG00000160193\n",
      "gene_string='WDR5'\n",
      "gene_list=['WDR5']\n",
      "match=None\n",
      "['WDR5']:     ENSG00000196363\n",
      "gene_string='WDR7'\n",
      "gene_list=['WDR7']\n",
      "match=None\n",
      "['WDR7']:     ENSG00000091157\n",
      "gene_string='WDR12'\n",
      "gene_list=['WDR12']\n",
      "match=None\n",
      "['WDR12']:     ENSG00000138442\n",
      "gene_string='WDR18'\n",
      "gene_list=['WDR18']\n",
      "match=None\n",
      "['WDR18']:     ENSG00000065268\n",
      "gene_string='WDR24'\n",
      "gene_list=['WDR24']\n",
      "match=None\n",
      "['WDR24']:     ENSG00000127580\n",
      "gene_string='WDR25'\n",
      "gene_list=['WDR25']\n",
      "match=None\n",
      "['WDR25']:     ENSG00000176473\n",
      "gene_string='WDR26'\n",
      "gene_list=['WDR26']\n",
      "match=None\n",
      "['WDR26']:     ENSG00000162923\n",
      "gene_string='WDR33'\n",
      "gene_list=['WDR33']\n",
      "match=None\n",
      "['WDR33']:     ENSG00000136709\n",
      "gene_string='WDR36'\n",
      "gene_list=['WDR36']\n",
      "match=None\n",
      "['WDR36']:     ENSG00000134987\n",
      "gene_string='WDR43'\n",
      "gene_list=['WDR43']\n",
      "match=None\n",
      "['WDR43']:     ENSG00000163811\n",
      "gene_string='WDR46'\n",
      "gene_list=['WDR46']\n",
      "match=None\n",
      "['WDR46']:     ENSG00000227057\n",
      "gene_string='WDR54'\n",
      "gene_list=['WDR54']\n",
      "match=None\n",
      "['WDR54']:     ENSG00000005448\n",
      "gene_string='WDR55'\n",
      "gene_list=['WDR55']\n",
      "match=None\n",
      "['WDR55']:     ENSG00000120314\n",
      "gene_string='WDR61'\n",
      "gene_list=['WDR61']\n",
      "match=None\n",
      "['WDR61']:     ENSG00000140395\n",
      "gene_string='WDR70'\n",
      "gene_list=['WDR70']\n",
      "match=None\n",
      "['WDR70']:     ENSG00000082068\n",
      "gene_string='WDR74'\n",
      "gene_list=['WDR74']\n",
      "match=None\n",
      "['WDR74']:     ENSG00000133316\n",
      "gene_string='WDR75'\n",
      "gene_list=['WDR75']\n",
      "match=None\n",
      "['WDR75']:     ENSG00000115368\n",
      "gene_string='WDR77'\n",
      "gene_list=['WDR77']\n",
      "match=None\n",
      "['WDR77']:     ENSG00000116455\n",
      "gene_string='WDR82'\n",
      "gene_list=['WDR82']\n",
      "match=None\n",
      "['WDR82']:     ENSG00000164091\n",
      "gene_string='WDR83OS'\n",
      "gene_list=['WDR83OS']\n",
      "match=None\n",
      "['WDR83OS']:     ENSG00000105583\n",
      "gene_string='WDR92'\n",
      "gene_list=['WDR92']\n",
      "match=None\n",
      "['WDR92']:     ENSG00000243667\n",
      "gene_string='WDTC1'\n",
      "gene_list=['WDTC1']\n",
      "match=None\n",
      "['WDTC1']:     ENSG00000142784\n",
      "gene_string='WEE1'\n",
      "gene_list=['WEE1']\n",
      "match=None\n",
      "['WEE1']:     ENSG00000166483\n",
      "gene_string='WNK1'\n",
      "gene_list=['WNK1']\n",
      "match=None\n",
      "['WNK1']:     ENSG00000060237\n",
      "gene_string='WTAP'\n",
      "gene_list=['WTAP']\n",
      "match=None\n",
      "['WTAP']:     ENSG00000146457\n",
      "gene_string='XAB2'\n",
      "gene_list=['XAB2']\n",
      "match=None\n",
      "['XAB2']:     ENSG00000076924\n",
      "gene_string='XPO1'\n",
      "gene_list=['XPO1']\n",
      "match=None\n",
      "['XPO1']:     ENSG00000082898\n",
      "gene_string='XPO5'\n",
      "gene_list=['XPO5']\n",
      "match=None\n",
      "['XPO5']:     ENSG00000124571\n",
      "gene_string='XRCC1'\n",
      "gene_list=['XRCC1']\n",
      "match=None\n",
      "['XRCC1']:     ENSG00000073050\n",
      "gene_string='XRCC2'\n",
      "gene_list=['XRCC2']\n",
      "match=None\n",
      "['XRCC2']:     ENSG00000196584\n",
      "gene_string='XRCC3'\n",
      "gene_list=['XRCC3']\n",
      "match=None\n",
      "['XRCC3']:     ENSG00000126215\n",
      "gene_string='XRCC5'\n",
      "gene_list=['XRCC5']\n",
      "match=None\n",
      "['XRCC5']:     ENSG00000079246\n",
      "gene_string='XRCC6'\n",
      "gene_list=['XRCC6']\n",
      "match=None\n",
      "['XRCC6']:     ENSG00000196419\n",
      "gene_string='XRN1'\n",
      "gene_list=['XRN1']\n",
      "match=None\n",
      "['XRN1']:     ENSG00000114127\n",
      "gene_string='XRN2'\n",
      "gene_list=['XRN2']\n",
      "match=None\n",
      "['XRN2']:     ENSG00000088930\n",
      "gene_string='YARS'\n",
      "gene_list=['YARS']\n",
      "match=None\n",
      "['YARS']:     ENSG00000134684\n",
      "gene_string='YARS2'\n",
      "gene_list=['YARS2']\n",
      "match=None\n",
      "['YARS2']:     ENSG00000139131\n",
      "gene_string='YBX1'\n",
      "gene_list=['YBX1']\n",
      "match=None\n",
      "['YBX1']:     ENSG00000065978\n",
      "gene_string='YEATS2'\n",
      "gene_list=['YEATS2']\n",
      "match=None\n",
      "['YEATS2']:     ENSG00000163872\n",
      "gene_string='YEATS4'\n",
      "gene_list=['YEATS4']\n",
      "match=None\n",
      "['YEATS4']:     ENSG00000127337\n",
      "gene_string='YJEFN3'\n",
      "gene_list=['YJEFN3']\n",
      "match=None\n",
      "['YJEFN3']:     ENSG00000250067\n",
      "gene_string='YKT6'\n",
      "gene_list=['YKT6']\n",
      "match=None\n",
      "['YKT6']:     ENSG00000106636\n",
      "gene_string='YPEL1'\n",
      "gene_list=['YPEL1']\n",
      "match=None\n",
      "['YPEL1']:     ENSG00000100027\n",
      "gene_string='YPEL5'\n",
      "gene_list=['YPEL5']\n",
      "match=None\n",
      "['YPEL5']:     ENSG00000119801\n",
      "gene_string='YRDC'\n",
      "gene_list=['YRDC']\n",
      "match=None\n",
      "['YRDC']:     ENSG00000196449\n",
      "gene_string='YTHDC1'\n",
      "gene_list=['YTHDC1']\n",
      "match=None\n",
      "['YTHDC1']:     ENSG00000083896\n",
      "gene_string='YY1'\n",
      "gene_list=['YY1']\n",
      "match=None\n",
      "['YY1']:     ENSG00000100811\n",
      "gene_string='ZBTB11'\n",
      "gene_list=['ZBTB11']\n",
      "match=None\n",
      "['ZBTB11']:     ENSG00000066422\n",
      "gene_string='ZBTB17'\n",
      "gene_list=['ZBTB17']\n",
      "match=None\n",
      "['ZBTB17']:     ENSG00000116809\n",
      "gene_string='ZC3H8'\n",
      "gene_list=['ZC3H8']\n",
      "match=None\n",
      "['ZC3H8']:     ENSG00000144161\n",
      "gene_string='ZC3H18'\n",
      "gene_list=['ZC3H18']\n",
      "match=None\n",
      "['ZC3H18']:     ENSG00000158545\n",
      "gene_string='ZCCHC9'\n",
      "gene_list=['ZCCHC9']\n",
      "match=None\n",
      "['ZCCHC9']:     ENSG00000131732\n",
      "gene_string='ZCRB1'\n",
      "gene_list=['ZCRB1']\n",
      "match=None\n",
      "['ZCRB1']:     ENSG00000139168\n",
      "gene_string='ZFC3H1'\n",
      "gene_list=['ZFC3H1']\n",
      "match=None\n",
      "['ZFC3H1']:     ENSG00000133858\n",
      "gene_string='ZFP69B'\n",
      "gene_list=['ZFP69B']\n",
      "match=None\n",
      "['ZFP69B']:     ENSG00000187801\n",
      "gene_string='ZMAT2'\n",
      "gene_list=['ZMAT2']\n",
      "match=None\n",
      "['ZMAT2']:     ENSG00000146007\n",
      "gene_string='ZMAT5'\n",
      "gene_list=['ZMAT5']\n",
      "match=None\n",
      "['ZMAT5']:     ENSG00000100319\n",
      "gene_string='ZNF24'\n",
      "gene_list=['ZNF24']\n",
      "match=None\n",
      "['ZNF24']:     ENSG00000172466\n",
      "gene_string='ZNF100'\n",
      "gene_list=['ZNF100']\n",
      "match=None\n",
      "['ZNF100']:     ENSG00000197020\n",
      "gene_string='ZNF131'\n",
      "gene_list=['ZNF131']\n",
      "match=None\n",
      "['ZNF131']:     ENSG00000172262\n",
      "gene_string='ZNF133'\n",
      "gene_list=['ZNF133']\n",
      "match=None\n",
      "['ZNF133']:     ENSG00000125846\n",
      "gene_string='ZNF207'\n",
      "gene_list=['ZNF207']\n",
      "match=None\n",
      "['ZNF207']:     ENSG00000010244\n",
      "gene_string='ZNF236'\n",
      "gene_list=['ZNF236']\n",
      "match=None\n",
      "['ZNF236']:     ENSG00000130856\n",
      "gene_string='ZNF253'\n",
      "gene_list=['ZNF253']\n",
      "match=None\n",
      "['ZNF253']:     ENSG00000256771\n",
      "gene_string='ZNF284'\n",
      "gene_list=['ZNF284']\n",
      "match=None\n",
      "['ZNF284']:     ENSG00000186026\n",
      "gene_string='ZNF317'\n",
      "gene_list=['ZNF317']\n",
      "match=None\n",
      "['ZNF317']:     ENSG00000130803\n",
      "gene_string='ZNF335'\n",
      "gene_list=['ZNF335']\n",
      "match=None\n",
      "['ZNF335']:     ENSG00000198026\n",
      "gene_string='ZNF407'\n",
      "gene_list=['ZNF407']\n",
      "match=None\n",
      "['ZNF407']:     ENSG00000215421\n",
      "gene_string='ZNF468'\n",
      "gene_list=['ZNF468']\n",
      "match=None\n",
      "['ZNF468']:     ENSG00000204604\n",
      "gene_string='ZNF492'\n",
      "gene_list=['ZNF492']\n",
      "match=None\n",
      "['ZNF492']:     ENSG00000229676\n",
      "gene_string='ZNF506'\n",
      "gene_list=['ZNF506']\n",
      "match=None\n",
      "['ZNF506']:     ENSG00000081665\n",
      "gene_string='ZNF559'\n",
      "gene_list=['ZNF559']\n",
      "match=None\n",
      "['ZNF559']:     ENSG00000188321\n",
      "gene_string='ZNF574'\n",
      "gene_list=['ZNF574']\n",
      "match=None\n",
      "['ZNF574']:     ENSG00000105732\n",
      "gene_string='ZNF658'\n",
      "gene_list=['ZNF658']\n",
      "match=None\n",
      "['ZNF658']:     ENSG00000274349\n",
      "gene_string='ZNF674'\n",
      "gene_list=['ZNF674']\n",
      "match=None\n",
      "['ZNF674']:     ENSG00000251192\n",
      "gene_string='ZNF676'\n",
      "gene_list=['ZNF676']\n",
      "match=None\n",
      "['ZNF676']:     ENSG00000196109\n",
      "gene_string='ZNF687'\n",
      "gene_list=['ZNF687']\n",
      "match=None\n",
      "['ZNF687']:     ENSG00000143373\n",
      "gene_string='ZNF718'\n",
      "gene_list=['ZNF718']\n",
      "match=None\n",
      "['ZNF718']:     ENSG00000250312\n",
      "gene_string='ZNF720'\n",
      "gene_list=['ZNF720']\n",
      "match=None\n",
      "['ZNF720']:     ENSG00000197302\n",
      "gene_string='ZNF763'\n",
      "gene_list=['ZNF763']\n",
      "match=None\n",
      "['ZNF763']:     ENSG00000197054\n",
      "gene_string='ZNF830'\n",
      "gene_list=['ZNF830']\n",
      "match=None\n",
      "['ZNF830']:     ENSG00000198783\n",
      "gene_string='ZNHIT2'\n",
      "gene_list=['ZNHIT2']\n",
      "match=None\n",
      "['ZNHIT2']:     ENSG00000174276\n",
      "gene_string='ZNHIT3'\n",
      "gene_list=['ZNHIT3']\n",
      "match=None\n",
      "['ZNHIT3']:     ENSG00000273611\n",
      "gene_string='ZNHIT6'\n",
      "gene_list=['ZNHIT6']\n",
      "match=None\n",
      "['ZNHIT6']:     ENSG00000117174\n",
      "gene_string='ZNRD1'\n",
      "gene_list=['ZNRD1']\n",
      "match=None\n",
      "['ZNRD1']:     ENSG00000066379\n",
      "gene_string='ZRANB2'\n",
      "gene_list=['ZRANB2']\n",
      "match=None\n",
      "['ZRANB2']:     ENSG00000132485\n",
      "gene_string='ZRSR2'\n",
      "gene_list=['ZRSR2']\n",
      "match=None\n",
      "['ZRSR2']:     ENSG00000169249\n",
      "gene_string='ZW10'\n",
      "gene_list=['ZW10']\n",
      "match=None\n",
      "['ZW10']:     ENSG00000086827\n",
      "gene_string='ZWINT'\n",
      "gene_list=['ZWINT']\n",
      "match=None\n",
      "['ZWINT']:     ENSG00000122952\n",
      "gene_string='control'\n",
      "gene_list=['control']\n",
      "match=None\n",
      "['control']:     control\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>batch</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript</th>\n",
       "      <th>gene_transcript</th>\n",
       "      <th>guide_id</th>\n",
       "      <th>percent_mito</th>\n",
       "      <th>UMI_count</th>\n",
       "      <th>z_gemgroup_UMI</th>\n",
       "      <th>core_scale_factor</th>\n",
       "      <th>...</th>\n",
       "      <th>perturbation</th>\n",
       "      <th>organism</th>\n",
       "      <th>perturbation_type</th>\n",
       "      <th>tissue_type</th>\n",
       "      <th>ncounts</th>\n",
       "      <th>ngenes</th>\n",
       "      <th>nperts</th>\n",
       "      <th>percent_ribo</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>gene_list</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGAAATCCA-27</th>\n",
       "      <td>27</td>\n",
       "      <td>NAF1</td>\n",
       "      <td>ENSG00000145414</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>5449_NAF1_P1P2_ENSG00000145414</td>\n",
       "      <td>NAF1_+_164087918.23-P1P2|NAF1_-_164087674.23-P1P2</td>\n",
       "      <td>0.112083</td>\n",
       "      <td>11438.0</td>\n",
       "      <td>0.013047</td>\n",
       "      <td>0.813253</td>\n",
       "      <td>...</td>\n",
       "      <td>NAF1</td>\n",
       "      <td>human</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>11324.0</td>\n",
       "      <td>3332</td>\n",
       "      <td>1</td>\n",
       "      <td>0.225362</td>\n",
       "      <td>3332</td>\n",
       "      <td>ENSG00000145901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGAACTTCC-31</th>\n",
       "      <td>31</td>\n",
       "      <td>BUB1</td>\n",
       "      <td>ENSG00000169679</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>935_BUB1_P1P2_ENSG00000169679</td>\n",
       "      <td>BUB1_-_111435363.23-P1P2|BUB1_-_111435372.23-P1P2</td>\n",
       "      <td>0.179895</td>\n",
       "      <td>5342.0</td>\n",
       "      <td>-1.522247</td>\n",
       "      <td>0.844107</td>\n",
       "      <td>...</td>\n",
       "      <td>BUB1</td>\n",
       "      <td>human</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>5257.0</td>\n",
       "      <td>2192</td>\n",
       "      <td>1</td>\n",
       "      <td>0.129732</td>\n",
       "      <td>2192</td>\n",
       "      <td>ENSG00000169679</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGAAGCCAC-34</th>\n",
       "      <td>34</td>\n",
       "      <td>UBL5</td>\n",
       "      <td>ENSG00000198258</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>9534_UBL5_P1P2_ENSG00000198258</td>\n",
       "      <td>UBL5_-_9938639.23-P1P2|UBL5_+_9938801.23-P1P2</td>\n",
       "      <td>0.105287</td>\n",
       "      <td>17305.0</td>\n",
       "      <td>0.384157</td>\n",
       "      <td>1.091537</td>\n",
       "      <td>...</td>\n",
       "      <td>UBL5</td>\n",
       "      <td>human</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>17135.0</td>\n",
       "      <td>4002</td>\n",
       "      <td>1</td>\n",
       "      <td>0.236825</td>\n",
       "      <td>4002</td>\n",
       "      <td>ENSG00000198258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGAATAGTC-43</th>\n",
       "      <td>43</td>\n",
       "      <td>C9orf16</td>\n",
       "      <td>ENSG00000171159</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>1131_C9orf16_P1P2_ENSG00000171159</td>\n",
       "      <td>C9orf16_+_130922603.23-P1P2|C9orf16_+_13092264...</td>\n",
       "      <td>0.099359</td>\n",
       "      <td>30244.0</td>\n",
       "      <td>3.721912</td>\n",
       "      <td>0.948277</td>\n",
       "      <td>...</td>\n",
       "      <td>C9orf16</td>\n",
       "      <td>human</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>29717.0</td>\n",
       "      <td>5358</td>\n",
       "      <td>1</td>\n",
       "      <td>0.246828</td>\n",
       "      <td>5358</td>\n",
       "      <td>ENSG00000171159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGACAGCGT-28</th>\n",
       "      <td>28</td>\n",
       "      <td>TIMM9</td>\n",
       "      <td>ENSG00000100575</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>8927_TIMM9_P1P2_ENSG00000100575</td>\n",
       "      <td>TIMM9_-_58893843.23-P1P2|TIMM9_-_58893848.23-P1P2</td>\n",
       "      <td>0.137623</td>\n",
       "      <td>8407.0</td>\n",
       "      <td>-0.975371</td>\n",
       "      <td>0.868942</td>\n",
       "      <td>...</td>\n",
       "      <td>TIMM9</td>\n",
       "      <td>human</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>8261.0</td>\n",
       "      <td>2944</td>\n",
       "      <td>1</td>\n",
       "      <td>0.183392</td>\n",
       "      <td>2944</td>\n",
       "      <td>ENSG00000100575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGTCTGTCGTC-45</th>\n",
       "      <td>45</td>\n",
       "      <td>ATP6V1D</td>\n",
       "      <td>ENSG00000100554</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>682_ATP6V1D_P1P2_ENSG00000100554</td>\n",
       "      <td>ATP6V1D_+_67826485.23-P1P2|ATP6V1D_+_67826497....</td>\n",
       "      <td>0.100272</td>\n",
       "      <td>18350.0</td>\n",
       "      <td>0.428227</td>\n",
       "      <td>1.115052</td>\n",
       "      <td>...</td>\n",
       "      <td>ATP6V1D</td>\n",
       "      <td>human</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>18044.0</td>\n",
       "      <td>4300</td>\n",
       "      <td>1</td>\n",
       "      <td>0.253547</td>\n",
       "      <td>4300</td>\n",
       "      <td>ENSG00000100554</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGTCTGTCTCG-27</th>\n",
       "      <td>27</td>\n",
       "      <td>CNOT3</td>\n",
       "      <td>ENSG00000088038</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>1718_CNOT3_P1P2_ENSG00000088038</td>\n",
       "      <td>CNOT3_+_54641532.23-P1P2|CNOT3_-_54641691.23-P1P2</td>\n",
       "      <td>0.093876</td>\n",
       "      <td>8671.0</td>\n",
       "      <td>-0.633593</td>\n",
       "      <td>0.813253</td>\n",
       "      <td>...</td>\n",
       "      <td>CNOT3</td>\n",
       "      <td>human</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>8510.0</td>\n",
       "      <td>3158</td>\n",
       "      <td>1</td>\n",
       "      <td>0.183196</td>\n",
       "      <td>3158</td>\n",
       "      <td>ENSG00000088038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGTCTGTGCGG-44</th>\n",
       "      <td>44</td>\n",
       "      <td>METTL3</td>\n",
       "      <td>ENSG00000165819</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>5004_METTL3_P1P2_ENSG00000165819</td>\n",
       "      <td>METTL3_+_21979431.23-P1P2|METTL3_-_21979084.23...</td>\n",
       "      <td>0.107983</td>\n",
       "      <td>20568.0</td>\n",
       "      <td>1.054624</td>\n",
       "      <td>0.973352</td>\n",
       "      <td>...</td>\n",
       "      <td>METTL3</td>\n",
       "      <td>human</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>20355.0</td>\n",
       "      <td>4247</td>\n",
       "      <td>1</td>\n",
       "      <td>0.256595</td>\n",
       "      <td>4247</td>\n",
       "      <td>ENSG00000165819</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGTCTTGCAGA-14</th>\n",
       "      <td>14</td>\n",
       "      <td>RPL5</td>\n",
       "      <td>ENSG00000122406</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>7475_RPL5_P1P2_ENSG00000122406</td>\n",
       "      <td>RPL5_+_93297664.23-P1P2|RPL5_-_93297968.23-P1P2</td>\n",
       "      <td>0.128225</td>\n",
       "      <td>23568.0</td>\n",
       "      <td>1.676254</td>\n",
       "      <td>1.050055</td>\n",
       "      <td>...</td>\n",
       "      <td>RPL5</td>\n",
       "      <td>human</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>23169.0</td>\n",
       "      <td>4690</td>\n",
       "      <td>1</td>\n",
       "      <td>0.210713</td>\n",
       "      <td>4690</td>\n",
       "      <td>ENSG00000122406</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGTCTTTACAC-25</th>\n",
       "      <td>25</td>\n",
       "      <td>SEC61B</td>\n",
       "      <td>ENSG00000106803</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>7752_SEC61B_P1P2_ENSG00000106803</td>\n",
       "      <td>SEC61B_+_101984577.23-P1P2|SEC61B_-_101984591....</td>\n",
       "      <td>0.095009</td>\n",
       "      <td>13504.0</td>\n",
       "      <td>0.341625</td>\n",
       "      <td>0.855594</td>\n",
       "      <td>...</td>\n",
       "      <td>SEC61B</td>\n",
       "      <td>human</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>13314.0</td>\n",
       "      <td>3794</td>\n",
       "      <td>1</td>\n",
       "      <td>0.211432</td>\n",
       "      <td>3794</td>\n",
       "      <td>ENSG00000106803</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>310385 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     batch     gene          gene_id transcript  \\\n",
       "cell_barcode                                                      \n",
       "AAACCCAAGAAATCCA-27     27     NAF1  ENSG00000145414       P1P2   \n",
       "AAACCCAAGAACTTCC-31     31     BUB1  ENSG00000169679       P1P2   \n",
       "AAACCCAAGAAGCCAC-34     34     UBL5  ENSG00000198258       P1P2   \n",
       "AAACCCAAGAATAGTC-43     43  C9orf16  ENSG00000171159       P1P2   \n",
       "AAACCCAAGACAGCGT-28     28    TIMM9  ENSG00000100575       P1P2   \n",
       "...                    ...      ...              ...        ...   \n",
       "TTTGTTGTCTGTCGTC-45     45  ATP6V1D  ENSG00000100554       P1P2   \n",
       "TTTGTTGTCTGTCTCG-27     27    CNOT3  ENSG00000088038       P1P2   \n",
       "TTTGTTGTCTGTGCGG-44     44   METTL3  ENSG00000165819       P1P2   \n",
       "TTTGTTGTCTTGCAGA-14     14     RPL5  ENSG00000122406       P1P2   \n",
       "TTTGTTGTCTTTACAC-25     25   SEC61B  ENSG00000106803       P1P2   \n",
       "\n",
       "                                       gene_transcript  \\\n",
       "cell_barcode                                             \n",
       "AAACCCAAGAAATCCA-27     5449_NAF1_P1P2_ENSG00000145414   \n",
       "AAACCCAAGAACTTCC-31      935_BUB1_P1P2_ENSG00000169679   \n",
       "AAACCCAAGAAGCCAC-34     9534_UBL5_P1P2_ENSG00000198258   \n",
       "AAACCCAAGAATAGTC-43  1131_C9orf16_P1P2_ENSG00000171159   \n",
       "AAACCCAAGACAGCGT-28    8927_TIMM9_P1P2_ENSG00000100575   \n",
       "...                                                ...   \n",
       "TTTGTTGTCTGTCGTC-45   682_ATP6V1D_P1P2_ENSG00000100554   \n",
       "TTTGTTGTCTGTCTCG-27    1718_CNOT3_P1P2_ENSG00000088038   \n",
       "TTTGTTGTCTGTGCGG-44   5004_METTL3_P1P2_ENSG00000165819   \n",
       "TTTGTTGTCTTGCAGA-14     7475_RPL5_P1P2_ENSG00000122406   \n",
       "TTTGTTGTCTTTACAC-25   7752_SEC61B_P1P2_ENSG00000106803   \n",
       "\n",
       "                                                              guide_id  \\\n",
       "cell_barcode                                                             \n",
       "AAACCCAAGAAATCCA-27  NAF1_+_164087918.23-P1P2|NAF1_-_164087674.23-P1P2   \n",
       "AAACCCAAGAACTTCC-31  BUB1_-_111435363.23-P1P2|BUB1_-_111435372.23-P1P2   \n",
       "AAACCCAAGAAGCCAC-34      UBL5_-_9938639.23-P1P2|UBL5_+_9938801.23-P1P2   \n",
       "AAACCCAAGAATAGTC-43  C9orf16_+_130922603.23-P1P2|C9orf16_+_13092264...   \n",
       "AAACCCAAGACAGCGT-28  TIMM9_-_58893843.23-P1P2|TIMM9_-_58893848.23-P1P2   \n",
       "...                                                                ...   \n",
       "TTTGTTGTCTGTCGTC-45  ATP6V1D_+_67826485.23-P1P2|ATP6V1D_+_67826497....   \n",
       "TTTGTTGTCTGTCTCG-27  CNOT3_+_54641532.23-P1P2|CNOT3_-_54641691.23-P1P2   \n",
       "TTTGTTGTCTGTGCGG-44  METTL3_+_21979431.23-P1P2|METTL3_-_21979084.23...   \n",
       "TTTGTTGTCTTGCAGA-14    RPL5_+_93297664.23-P1P2|RPL5_-_93297968.23-P1P2   \n",
       "TTTGTTGTCTTTACAC-25  SEC61B_+_101984577.23-P1P2|SEC61B_-_101984591....   \n",
       "\n",
       "                     percent_mito  UMI_count  z_gemgroup_UMI  \\\n",
       "cell_barcode                                                   \n",
       "AAACCCAAGAAATCCA-27      0.112083    11438.0        0.013047   \n",
       "AAACCCAAGAACTTCC-31      0.179895     5342.0       -1.522247   \n",
       "AAACCCAAGAAGCCAC-34      0.105287    17305.0        0.384157   \n",
       "AAACCCAAGAATAGTC-43      0.099359    30244.0        3.721912   \n",
       "AAACCCAAGACAGCGT-28      0.137623     8407.0       -0.975371   \n",
       "...                           ...        ...             ...   \n",
       "TTTGTTGTCTGTCGTC-45      0.100272    18350.0        0.428227   \n",
       "TTTGTTGTCTGTCTCG-27      0.093876     8671.0       -0.633593   \n",
       "TTTGTTGTCTGTGCGG-44      0.107983    20568.0        1.054624   \n",
       "TTTGTTGTCTTGCAGA-14      0.128225    23568.0        1.676254   \n",
       "TTTGTTGTCTTTACAC-25      0.095009    13504.0        0.341625   \n",
       "\n",
       "                     core_scale_factor  ...  perturbation organism  \\\n",
       "cell_barcode                            ...                          \n",
       "AAACCCAAGAAATCCA-27           0.813253  ...          NAF1    human   \n",
       "AAACCCAAGAACTTCC-31           0.844107  ...          BUB1    human   \n",
       "AAACCCAAGAAGCCAC-34           1.091537  ...          UBL5    human   \n",
       "AAACCCAAGAATAGTC-43           0.948277  ...       C9orf16    human   \n",
       "AAACCCAAGACAGCGT-28           0.868942  ...         TIMM9    human   \n",
       "...                                ...  ...           ...      ...   \n",
       "TTTGTTGTCTGTCGTC-45           1.115052  ...       ATP6V1D    human   \n",
       "TTTGTTGTCTGTCTCG-27           0.813253  ...         CNOT3    human   \n",
       "TTTGTTGTCTGTGCGG-44           0.973352  ...        METTL3    human   \n",
       "TTTGTTGTCTTGCAGA-14           1.050055  ...          RPL5    human   \n",
       "TTTGTTGTCTTTACAC-25           0.855594  ...        SEC61B    human   \n",
       "\n",
       "                     perturbation_type tissue_type  ncounts  ngenes nperts  \\\n",
       "cell_barcode                                                                 \n",
       "AAACCCAAGAAATCCA-27             CRISPR   cell_line  11324.0    3332      1   \n",
       "AAACCCAAGAACTTCC-31             CRISPR   cell_line   5257.0    2192      1   \n",
       "AAACCCAAGAAGCCAC-34             CRISPR   cell_line  17135.0    4002      1   \n",
       "AAACCCAAGAATAGTC-43             CRISPR   cell_line  29717.0    5358      1   \n",
       "AAACCCAAGACAGCGT-28             CRISPR   cell_line   8261.0    2944      1   \n",
       "...                                ...         ...      ...     ...    ...   \n",
       "TTTGTTGTCTGTCGTC-45             CRISPR   cell_line  18044.0    4300      1   \n",
       "TTTGTTGTCTGTCTCG-27             CRISPR   cell_line   8510.0    3158      1   \n",
       "TTTGTTGTCTGTGCGG-44             CRISPR   cell_line  20355.0    4247      1   \n",
       "TTTGTTGTCTTGCAGA-14             CRISPR   cell_line  23169.0    4690      1   \n",
       "TTTGTTGTCTTTACAC-25             CRISPR   cell_line  13314.0    3794      1   \n",
       "\n",
       "                    percent_ribo n_genes        gene_list  \n",
       "cell_barcode                                               \n",
       "AAACCCAAGAAATCCA-27     0.225362    3332  ENSG00000145901  \n",
       "AAACCCAAGAACTTCC-31     0.129732    2192  ENSG00000169679  \n",
       "AAACCCAAGAAGCCAC-34     0.236825    4002  ENSG00000198258  \n",
       "AAACCCAAGAATAGTC-43     0.246828    5358  ENSG00000171159  \n",
       "AAACCCAAGACAGCGT-28     0.183392    2944  ENSG00000100575  \n",
       "...                          ...     ...              ...  \n",
       "TTTGTTGTCTGTCGTC-45     0.253547    4300  ENSG00000100554  \n",
       "TTTGTTGTCTGTCTCG-27     0.183196    3158  ENSG00000088038  \n",
       "TTTGTTGTCTGTGCGG-44     0.256595    4247  ENSG00000165819  \n",
       "TTTGTTGTCTTGCAGA-14     0.210713    4690  ENSG00000122406  \n",
       "TTTGTTGTCTTTACAC-25     0.211432    3794  ENSG00000106803  \n",
       "\n",
       "[310385 rows x 26 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs['gene_list'] = adata.obs['perturbation'].apply(utils.convert_genomic_location_to_ensembl_ids)\n",
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 310342 × 8563\n",
       "    obs: 'batch', 'gene', 'gene_id', 'transcript', 'gene_transcript', 'guide_id', 'percent_mito', 'UMI_count', 'z_gemgroup_UMI', 'core_scale_factor', 'core_adjusted_UMI_count', 'disease', 'cancer', 'cell_line', 'sex', 'age', 'perturbation', 'organism', 'perturbation_type', 'tissue_type', 'ncounts', 'ngenes', 'nperts', 'percent_ribo', 'n_genes', 'gene_list'\n",
       "    var: 'chr', 'start', 'end', 'class', 'strand', 'length', 'in_matrix', 'mean', 'std', 'cv', 'fano', 'ensembl_id', 'ncounts', 'ncells', 'n_cells'\n",
       "    uns: 'log1p'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "valid_indices = adata.obs[(~adata.obs['gene_list'].str.contains('_')) & (adata.obs['gene_list'] != 'None')].index\n",
    "adata = adata[valid_indices, :]\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/anndata/_core/anndata.py:1138: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n"
     ]
    }
   ],
   "source": [
    "adata.write(f\"../data/{perturbation_screen_name}_saved_2.h5ad\", compression='gzip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 118 × 2742\n",
       "    obs: 'sample', 'total_umis', 'Size_Factor', 'gene', 'all_gene', 'barcode', 'read_count', 'umi_count', 'proportion', 'guide_count', 'sample_directory', 'ko_barcode_file', 'id', 'prep_batch', 'within_batch_chip', 'within_chip_lane', 'percent.mito', 'tissue_type', 'cell_line', 'cancer', 'disease', 'perturbation_type', 'celltype', 'organism', 'perturbation', 'nperts', 'ngenes', 'ncounts', 'percent_mito', 'percent_ribo', 'n_genes', 'gene_list'\n",
       "    var: 'ensembl_id', 'ncounts', 'ncells', 'n_cells', 'highly_variable', 'means', 'dispersions', 'dispersions_norm'\n",
       "    uns: 'hvg', 'log1p'"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata = sc.read_h5ad(f\"../data/{perturbation_screen_name}_saved.h5ad\")\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 310342 × 8563\n",
       "    obs: 'batch', 'gene', 'gene_id', 'transcript', 'gene_transcript', 'guide_id', 'percent_mito', 'UMI_count', 'z_gemgroup_UMI', 'core_scale_factor', 'core_adjusted_UMI_count', 'disease', 'cancer', 'cell_line', 'sex', 'age', 'perturbation', 'organism', 'perturbation_type', 'tissue_type', 'ncounts', 'ngenes', 'nperts', 'percent_ribo', 'n_genes', 'gene_list'\n",
       "    var: 'chr', 'start', 'end', 'class', 'strand', 'length', 'in_matrix', 'mean', 'std', 'cv', 'fano', 'ensembl_id', 'ncounts', 'ncells', 'n_cells'\n",
       "    uns: 'log1p'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "perturbation_counts = adata.obs['gene_list'].value_counts()\n",
    "\n",
    "min_cells = 3\n",
    "valid_gene_lists = perturbation_counts[perturbation_counts >= min_cells].index\n",
    "\n",
    "adata = adata[adata.obs['gene_list'].isin(valid_gene_lists)]\n",
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Option 1: E-Distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>perturbation</th>\n",
       "      <th>e_distance</th>\n",
       "      <th>p_value</th>\n",
       "      <th>significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000005339</td>\n",
       "      <td>0.193873</td>\n",
       "      <td>0.008</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000154447</td>\n",
       "      <td>0.199291</td>\n",
       "      <td>0.612</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000147133</td>\n",
       "      <td>0.100834</td>\n",
       "      <td>0.194</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000168246</td>\n",
       "      <td>0.204343</td>\n",
       "      <td>0.040</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000151474</td>\n",
       "      <td>4.298759</td>\n",
       "      <td>0.000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179</th>\n",
       "      <td>ENSG00000135999</td>\n",
       "      <td>0.161179</td>\n",
       "      <td>0.210</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>ENSG00000127884</td>\n",
       "      <td>0.153476</td>\n",
       "      <td>0.194</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>ENSG00000073050</td>\n",
       "      <td>0.110636</td>\n",
       "      <td>0.486</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>ENSG00000044574</td>\n",
       "      <td>0.357272</td>\n",
       "      <td>0.080</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183</th>\n",
       "      <td>ENSG00000054523</td>\n",
       "      <td>0.413045</td>\n",
       "      <td>0.050</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>184 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        perturbation  e_distance  p_value  significant\n",
       "0    ENSG00000005339    0.193873    0.008         True\n",
       "1    ENSG00000154447    0.199291    0.612        False\n",
       "2    ENSG00000147133    0.100834    0.194        False\n",
       "3    ENSG00000168246    0.204343    0.040         True\n",
       "4    ENSG00000151474    4.298759    0.000         True\n",
       "..               ...         ...      ...          ...\n",
       "179  ENSG00000135999    0.161179    0.210        False\n",
       "180  ENSG00000127884    0.153476    0.194        False\n",
       "181  ENSG00000073050    0.110636    0.486        False\n",
       "182  ENSG00000044574    0.357272    0.080        False\n",
       "183  ENSG00000054523    0.413045    0.050        False\n",
       "\n",
       "[184 rows x 4 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edist_df = pd.read_csv('../data/perturbation_screens/TianKampmann2021_CRISPRi_e_distances.csv', index_col=0) #TianKampmann2021_CRISPRi_e_distances # '../data/perturbation_screens/GasperiniShendure2019_lowMOI_e_distances.csv'\n",
    "edist_df['significant'] = edist_df['p_value'] < 0.05\n",
    "edist_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.lines.Line2D at 0x7ff2b54a1ea0>"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjMAAAGxCAYAAACXwjeMAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjEsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvc2/+5QAAAAlwSFlzAAAPYQAAD2EBqD+naQAAI6hJREFUeJzt3XtwU3X+//FXoDUtLC33EqRCUUCpy0UuCqxyEeqKqF3H24qKru7CQrnYWZVakMsIHdyxsnJzYb3NLrdBhWVmEUEULIJcChUsKAioCMEKAglQijTn94c/+rW2pGnI5XzK8zGTGXPySXjnQzVPT5rWYVmWJQAAAEPVivYAAAAAl4KYAQAARiNmAACA0YgZAABgNGIGAAAYjZgBAABGI2YAAIDRiBkAAGC0mGgPEG4+n0+HDx9WvXr15HA4oj0OAAAIgGVZ8nq9at68uWrV8n/upcbHzOHDh5WcnBztMQAAQBAOHjyoFi1a+F1T42OmXr16kn7ejISEhChPE4DTp6XmzX/+58OHpbp1ozsPAABR4PF4lJycXPY67k+Nj5kLby0lJCSYETO1a//fPyckEDMAgMtaIN8iwjcAAwAAoxEzAADAaMQMAAAwGjEDAACMRswAAACjETMAAMBoxAwAADAaMQMAAIxGzAAAAKMRMwAAwGjEDAAAMBoxAwAAjEbMAAAAoxEzAADAaMQMAAAwWky0BzBdaodOcrvdfte4XC4V7iiIzEAAAFxmiJlL5Ha7lTZlmd81q7LTIzILAACXI95mAgAARiNmAACA0YgZAABgNGIGAAAYjZgBAABGI2YAAIDRiBkAAGA0YgYAABiNmAEAAEYjZgAAgNGIGQAAYDRiBgAAGI2YAQAARiNmAACA0YgZAABgNGIGAAAYjZgBAABGI2YAAIDRiBkAAGA0YgYAABiNmAEAAEYjZgAAgNGIGQAAYDRiBgAAGI2YAQAARotqzHz88ce688471bx5czkcDi1btqzc7ZZlaeLEiWrevLni4+PVp08fFRYWRmdYAABgS1GNmdOnT6tjx46aOXNmpbe/+OKLys3N1cyZM7VlyxY1a9ZMAwYMkNfrjfCkAADArmKi+Yfffvvtuv322yu9zbIsTZ8+XdnZ2brnnnskSW+99ZaSkpK0YMECDR06NJKjAgAAm7Lt98wcOHBAR44cUVpaWtkxp9Op3r17a8OGDVGcDAAA2ElUz8z4c+TIEUlSUlJSueNJSUn65ptvLnq/kpISlZSUlF33eDzhGRAAANiCbWPmAofDUe66ZVkVjv1STk6OJk2aFO6xbCu1Qye53W6/a1wulwp3FERmIAAAwsy2MdOsWTNJP5+hcblcZceLiooqnK35paysLGVmZpZd93g8Sk5ODt+gNuN2u5U2ZZnfNauy0yMyCwAAkWDb75lJSUlRs2bNtHr16rJj586d07p169SzZ8+L3s/pdCohIaHcBQAA1FxRPTNz6tQpffXVV2XXDxw4oIKCAjVs2FBXXXWVxowZo6lTp6pNmzZq06aNpk6dqjp16uihhx6K4tQAAMBOohozW7duVd++fcuuX3h7aMiQIXrzzTf1zDPPqLi4WMOHD9fx48d14403atWqVapXr160RgYAADYT1Zjp06ePLMu66O0Oh0MTJ07UxIkTIzcUAAAwim2/ZwYAACAQxAwAADAaMQMAAIxGzAAAAKMRMwAAwGjEDAAAMBoxAwAAjEbMAAAAoxEzAADAaMQMAAAwGjEDAACMRswAAACjETMAAMBoxAwAADAaMQMAAIxGzAAAAKMRMwAAwGjEDAAAMBoxAwAAjEbMAAAAoxEzAADAaMQMAAAwGjEDAACMRswAAACjETMAAMBoxAwAADBaTLQHAAKR2qGT3G633zUul0uFOwoiMxAAwDaIGRjB7XYrbcoyv2tWZadHZBYAgL3wNhMAADAaMQMAAIxGzAAAAKMRMwAAwGjEDAAAMBoxAwAAjEbMAAAAoxEzAADAaMQMAAAwGjEDAACMRswAAACjETMAAMBoxAwAADAaMQMAAIxGzAAAAKMRMwAAwGjEDAAAMBoxAwAAjEbMAAAAoxEzAADAaMQMAAAwGjEDAACMRswAAACjETMAAMBoxAwAADCarWPm/PnzGjdunFJSUhQfH6/WrVtr8uTJ8vl80R4NAADYREy0B/Bn2rRpevXVV/XWW28pNTVVW7du1eOPP67ExESNHj062uMBAAAbsHXMbNy4UXfffbfuuOMOSVKrVq20cOFCbd26NcqTAQAAu7D120y/+93vtGbNGu3Zs0eS9Nlnn2n9+vUaOHBglCcDAAB2YeszM88++6xOnjypa6+9VrVr11ZpaammTJmiP/7xjxe9T0lJiUpKSsquezyeSIwKAACixNYxs3jxYv3nP//RggULlJqaqoKCAo0ZM0bNmzfXkCFDKr1PTk6OJk2aFOFJ/fN4T6lhkyS/a1wulwp3FERmIAAAahBbx8zTTz+tsWPH6sEHH5Qk/fa3v9U333yjnJyci8ZMVlaWMjMzy657PB4lJydHZN6LsXw+pU1Z5nfNquz0iMwCAEBNY+uYOXPmjGrVKv9tPbVr1/b70Wyn0ymn0xnu0QAAgE3YOmbuvPNOTZkyRVdddZVSU1O1fft25ebm6k9/+lO0RwMAADZh65iZMWOGxo8fr+HDh6uoqEjNmzfX0KFD9fzzz0d7NAAAYBO2jpl69epp+vTpmj59erRHAQAANmXrnzMDAABQFWIGAAAYjZgBAABGI2YAAIDRiBkAAGA0YgYAABiNmAEAAEYjZgAAgNGIGQAAYDRiBgAAGI2YAQAARiNmAACA0YgZAABgNGIGAAAYjZgBAABGI2YAAIDRiBkAAGA0YgYAABiNmAEAAEYjZgAAgNGIGQAAYDRiBgAAGI2YAQAARiNmAACA0YgZAABgNGIGAAAYLSbaA8BcqR06ye12+13jcrlUuKMgMgMBAC5LxAyC5na7lTZlmd81q7LTIzILAODyxdtMAADAaMQMAAAwGjEDAACMRswAAACjETMAAMBoxAwAADAaMQMAAIxGzAAAAKMRMwAAwGjEDAAAMBoxAwAAjEbMAAAAoxEzAADAaMQMAAAwGjEDAACMRswAAACjETMAAMBoxAwAADBaUDHTunVrHTt2rMLxEydOqHXr1pc8FAAAQKCCipmvv/5apaWlFY6XlJTo0KFDlzwUAABAoGKqs3j58uVl//z+++8rMTGx7HppaanWrFmjVq1ahWw4AACAqlQrZtLT0yVJDodDQ4YMKXdbbGysWrVqpZdeeilkwwEAAFSlWjHj8/kkSSkpKdqyZYsaN24clqEAAAACVa2YueDAgQOhngMAACAoQcWMJK1Zs0Zr1qxRUVFR2RmbC15//fVLHgwAACAQQX2aadKkSUpLS9OaNWt09OhRHT9+vNwllA4dOqSHH35YjRo1Up06ddSpUyfl5+eH9M8AAADmCurMzKuvvqo333xTjzzySKjnKef48ePq1auX+vbtq/fee09NmzbVvn37VL9+/bD+uQAAwBxBxcy5c+fUs2fPUM9SwbRp05ScnKw33nij7Bgf/QYAAL8U1NtMTz75pBYsWBDqWSpYvny5unbtqvvuu09NmzZV586dNW/evLD/uQAAwBxBnZk5e/as5s6dqw8++EAdOnRQbGxsudtzc3NDMtz+/fs1Z84cZWZm6rnnntPmzZs1atQoOZ1OPfroo5Xep6SkRCUlJWXXPR5PSGYBAAD2FFTM7NixQ506dZIkff755+VuczgclzzUBT6fT127dtXUqVMlSZ07d1ZhYaHmzJlz0ZjJycnRpEmTQjZDpHi8p9SwSZLqWJa++//HWrRM0Zlf7OeZ4rOqEx9XxeN4wzhleKR26CS32+13TSDP68Ie+uNyuVS4o+CS5wnkcQAAkRFUzHz00UehnqNSLpdL7du3L3fsuuuu0zvvvHPR+2RlZSkzM7PsusfjUXJycthmDBXL51PalGVylhRLo26VJN06YZFKnPFla5Zk9FNa7kq/j7Mko19Y5wwHt9uttCnL/K4J5Hld2EN/VmWnh2SeQB4HABAZQf+cmUjo1auXvvzyy3LH9uzZo5YtW170Pk6nU06nM9yjAQAAmwgqZvr27ev37aQPP/ww6IF+6amnnlLPnj01depU3X///dq8ebPmzp2ruXPnhuTxAQCA+YKKmQvfL3PBTz/9pIKCAn3++ecVfgHlpejWrZuWLl2qrKwsTZ48WSkpKZo+fboGDx4csj8DAACYLaiYefnllys9PnHiRJ06deqSBvq1QYMGadCgQSF9TAAAUHME9XNmLubhhx/m9zIBAICICmnMbNy4UXFx/j86DAAAEEpBvc10zz33lLtuWZbcbre2bt2q8ePHh2QwAACAQAQVM4mJieWu16pVS+3atdPkyZOVlpYWksEAAAACEVTM/PIXPwIAAETTJf3QvPz8fO3evVsOh0Pt27dX586dQzUXAABAQIKKmaKiIj344INau3at6tevL8uydPLkSfXt21eLFi1SkyZNQj0nAABApYL6NNPIkSPl8XhUWFioH3/8UcePH9fnn38uj8ejUaNGhXpGAACAiwrqzMzKlSv1wQcf6Lrrris71r59e82aNYtvAAYAABEV1JkZn8+n2NjYCsdjY2Pl8/kueSgAAIBABRUz/fr10+jRo3X48OGyY4cOHdJTTz2lW2+9NWTDAQAAVCWomJk5c6a8Xq9atWqlq6++Wtdcc41SUlLk9Xo1Y8aMUM8IAABwUUF9z0xycrK2bdum1atX64svvpBlWWrfvr369+8f6vkAAAD8qtaZmQ8//FDt27eXx+ORJA0YMEAjR47UqFGj1K1bN6WmpiovLy8sgwIAAFSmWjEzffp0/fnPf1ZCQkKF2xITEzV06FDl5uaGbDgAAICqVCtmPvvsM/3+97+/6O1paWnKz8+/5KEAAAACVa2Y+f777yv9SPYFMTEx+uGHHy55KAAAgEBVK2auvPJK7dy586K379ixQy6X65KHAgAACFS1YmbgwIF6/vnndfbs2Qq3FRcXa8KECRo0aFDIhgMAAKhKtT6aPW7cOL377rtq27atMjIy1K5dOzkcDu3evVuzZs1SaWmpsrOzwzUrAABABdWKmaSkJG3YsEF//etflZWVJcuyJEkOh0O33XabZs+eraSkpLAMCgAAUJlq/9C8li1basWKFTp+/Li++uorWZalNm3aqEGDBuGYDwAAwK+gfgKwJDVo0EDdunUL5SwAAADVFtTvZgIAALALYgYAABiNmAEAAEYL+ntmgEB4vKfUsIn/T7h5vN4ITWO/eSIptUMnud1uv2tcLpcKdxREZqAICuS5nyk+qzrxcX7XBLI/kdxnu/2d2m0eXD6IGYSV5fMpbcoyv2uWZPSLzDCy3zyR5Ha7q3zuq7LTIzJLpAXy3Jdk9FNa7kq/awLZn0jus93+Tu02Dy4fvM0EAACMRswAAACjETMAAMBoxAwAADAaMQMAAIxGzAAAAKMRMwAAwGjEDAAAMBoxAwAAjEbMAAAAoxEzAADAaMQMAAAwGjEDAACMRswAAACjETMAAMBoxAwAADAaMQMAAIxGzAAAAKMRMwAAwGjEDAAAMBoxAwAAjEbMAAAAoxEzAADAaMQMAAAwGjEDAACMZlTM5OTkyOFwaMyYMdEeBQAA2IQxMbNlyxbNnTtXHTp0iPYoAADARoyImVOnTmnw4MGaN2+eGjRoEO1xAACAjRgRMyNGjNAdd9yh/v37R3sUAABgMzHRHqAqixYt0rZt27Rly5aA1peUlKikpKTsusfjCddoAADABmwdMwcPHtTo0aO1atUqxcXFBXSfnJwcTZo0KcyTmc3jPaWGTZL8rjlTfFZ14v3vucfrDeVYuMykdugkt9vtd00gX4cul0uFOwpCOBkQHoF8zfP1HBxbx0x+fr6KiorUpUuXsmOlpaX6+OOPNXPmTJWUlKh27drl7pOVlaXMzMyy6x6PR8nJyRGb2QSWz6e0Kcv8rlmS0U9puSurXAMEy+12h+TrcFV2euiGAsIokK95vp6DY+uYufXWW7Vz585yxx5//HFde+21evbZZyuEjCQ5nU45nc5IjQgAAKLM1jFTr149XX/99eWO1a1bV40aNapwHAAAXJ6M+DQTAADAxdj6zExl1q5dG+0RAACAjXBmBgAAGI2YAQAARiNmAACA0YgZAABgNGIGAAAYjZgBAABGI2YAAIDRiBkAAGA0YgYAABiNmAEAAEYjZgAAgNGIGQAAYDRiBgAAGI2YAQAARiNmAACA0YgZAABgNGIGAAAYjZgBAABGI2YAAIDRiBkAAGA0YgYAABiNmAEAAEYjZgAAgNGIGQAAYDRiBgAAGI2YAQAARouJ9gDA5Sy1Qye53W6/a1wulwp3FERkHo/3lBo2SbLNPHYTyP54vN6QPM7lvM/wz27/3bADYgaIIrfbrbQpy/yuWZWdHpFZJMny+Ww1j90Esj9LMvqF5HEu532Gf3b774Yd8DYTAAAwGjEDAACMRswAAACjETMAAMBoxAwAADAaMQMAAIxGzAAAAKMRMwAAwGjEDAAAMBoxAwAAjEbMAAAAoxEzAADAaMQMAAAwGjEDAACMRswAAACjETMAAMBoxAwAADAaMQMAAIxGzAAAAKMRMwAAwGjEDAAAMBoxAwAAjEbMAAAAoxEzAADAaMQMAAAwmq1jJicnR926dVO9evXUtGlTpaen68svv4z2WAAAwEZsHTPr1q3TiBEj9Omnn2r16tU6f/680tLSdPr06WiPBgAAbCIm2gP4s3LlynLX33jjDTVt2lT5+fm65ZZbojQVAACwE1ufmfm1kydPSpIaNmwY5UkAAIBd2PrMzC9ZlqXMzEz97ne/0/XXX3/RdSUlJSopKSm77vF4IjEeAACIEmNiJiMjQzt27ND69ev9rsvJydGkSZMiNBUuVx7vKTVskuR3zZnis6oTH1fF43hDOVZEXM7PPZIC2edQ7WFqh05yu91+14Tq7zSQ5+VyuVS4o8DvmlDNHMiaQOYJlUj+vdckRsTMyJEjtXz5cn388cdq0aKF37VZWVnKzMwsu+7xeJScnBzuEXGZsXw+pU1Z5nfNkox+SstdWeUa01zOzz2SAt3nUHC73RH7Ow3kea3KTq/ycUI5c1VrApknVCL5916T2DpmLMvSyJEjtXTpUq1du1YpKSlV3sfpdMrpdEZgOgAAYAe2jpkRI0ZowYIF+u9//6t69erpyJEjkqTExETFx8dHeToAAGAHtv4005w5c3Ty5En16dNHLper7LJ48eJojwYAAGzC1mdmLMuK9ggAAMDmbH1mBgAAoCrEDAAAMBoxAwAAjEbMAAAAoxEzAADAaMQMAAAwGjEDAACMRswAAACjETMAAMBoxAwAADAaMQMAAIxGzAAAAKMRMwAAwGjEDAAAMBoxAwAAjEbMAAAAoxEzAADAaMQMAAAwGjEDAACMRswAAACjETMAAMBoxAwAADAaMQMAAIxGzAAAAKMRMwAAwGjEDAAAMFpMtAcA4J/He0oNmyT5XXOm+KzqxMdV8TjeUI5lG4HsT0197oEwcX/sNrOJ/w6GauZA1rhcLhXuKKjuiCFFzAA2Z/l8SpuyzO+aJRn9lJa7sso1NVGg+3O5MnF/7Dazif8OhnLmqtasyk6v5nShx9tMAADAaMQMAAAwGjEDAACMRswAAACjETMAAMBoxAwAADAaMQMAAIxGzAAAAKMRMwAAwGjEDAAAMBoxAwAAjEbMAAAAoxEzAADAaMQMAAAwGjEDAACMRswAAACjETMAAMBoxAwAADAaMQMAAIxGzAAAAKMRMwAAwGjEDAAAMBoxAwAAjEbMAAAAoxEzAADAaEbEzOzZs5WSkqK4uDh16dJFeXl50R4JAADYhO1jZvHixRozZoyys7O1fft23Xzzzbr99tv17bffRns0AABgA7aPmdzcXD3xxBN68skndd1112n69OlKTk7WnDlzoj0aAACwAVvHzLlz55Sfn6+0tLRyx9PS0rRhw4YoTQUAAOwkJtoD+HP06FGVlpYqKSmp3PGkpCQdOXKk0vuUlJSopKSk7PrJkyclSR6PJywzWj6ffio+7X+NZQW8xlFSrAuTnis+rZ98vqAehzWsYQ1rWMOaiKzx+cLyGnvhMS3LqnqxZWOHDh2yJFkbNmwod/yFF16w2rVrV+l9JkyYYEniwoULFy5cuNSAy8GDB6vsBVufmWncuLFq165d4SxMUVFRhbM1F2RlZSkzM7Psus/n048//qhGjRrJ4XCEdD6Px6Pk5GQdPHhQCQkJIX1s/B/2OTLY58hgnyODfY6McO6zZVnyer1q3rx5lWttHTNXXHGFunTpotWrV+sPf/hD2fHVq1fr7rvvrvQ+TqdTTqez3LH69euHc0wlJCTwL0sEsM+RwT5HBvscGexzZIRrnxMTEwNaZ+uYkaTMzEw98sgj6tq1q3r06KG5c+fq22+/1bBhw6I9GgAAsAHbx8wDDzygY8eOafLkyXK73br++uu1YsUKtWzZMtqjAQAAG7B9zEjS8OHDNXz48GiPUYHT6dSECRMqvK2F0GKfI4N9jgz2OTLY58iwyz47LCuQzzwBAADYk61/aB4AAEBViBkAAGA0YgYAABiNmPFj9uzZSklJUVxcnLp06aK8vDy/69etW6cuXbooLi5OrVu31quvvhqhSc1Xnb1+9913NWDAADVp0kQJCQnq0aOH3n///QhOa67qfk1f8MknnygmJkadOnUK74A1RHX3uaSkRNnZ2WrZsqWcTqeuvvpqvf766xGa1lzV3ef58+erY8eOqlOnjlwulx5//HEdO3YsQtOa6eOPP9add96p5s2by+FwaNmyZVXeJyqvhZf+SwdqpkWLFlmxsbHWvHnzrF27dlmjR4+26tata33zzTeVrt+/f79Vp04da/To0dauXbusefPmWbGxsdbbb78d4cnNU929Hj16tDVt2jRr8+bN1p49e6ysrCwrNjbW2rZtW4QnN0t19/mCEydOWK1bt7bS0tKsjh07RmZYgwWzz3fddZd14403WqtXr7YOHDhgbdq0yfrkk08iOLV5qrvPeXl5Vq1atax//OMf1v79+628vDwrNTXVSk9Pj/DkZlmxYoWVnZ1tvfPOO5Yka+nSpX7XR+u1kJi5iO7du1vDhg0rd+zaa6+1xo4dW+n6Z555xrr22mvLHRs6dKh10003hW3GmqK6e12Z9u3bW5MmTQr1aDVKsPv8wAMPWOPGjbMmTJhAzASguvv83nvvWYmJidaxY8ciMV6NUd19/vvf/261bt263LFXXnnFatGiRdhmrGkCiZlovRbyNlMlzp07p/z8fKWlpZU7npaWpg0bNlR6n40bN1ZYf9ttt2nr1q366aefwjar6YLZ61/z+Xzyer1q2LBhOEasEYLd5zfeeEP79u3ThAkTwj1ijRDMPi9fvlxdu3bViy++qCuvvFJt27bV3/72NxUXF0diZCMFs889e/bUd999pxUrVsiyLH3//fd6++23dccdd0Ri5MtGtF4LjfiheZF29OhRlZaWVvhllklJSRV+6eUFR44cqXT9+fPndfToUblcrrDNa7Jg9vrXXnrpJZ0+fVr3339/OEasEYLZ571792rs2LHKy8tTTAz/qQhEMPu8f/9+rV+/XnFxcVq6dKmOHj2q4cOH68cff+T7Zi4imH3u2bOn5s+frwceeEBnz57V+fPnddddd2nGjBmRGPmyEa3XQs7M+PHr37JtWZbf37xd2frKjqOi6u71BQsXLtTEiRO1ePFiNW3aNFzj1RiB7nNpaakeeughTZo0SW3bto3UeDVGdb6efT6fHA6H5s+fr+7du2vgwIHKzc3Vm2++ydmZKlRnn3ft2qVRo0bp+eefV35+vlauXKkDBw7we/7CIBqvhfzvViUaN26s2rVrVyj8oqKiCsV5QbNmzSpdHxMTo0aNGoVtVtMFs9cXLF68WE888YSWLFmi/v37h3NM41V3n71er7Zu3art27crIyND0s8vupZlKSYmRqtWrVK/fv0iMrtJgvl6drlcuvLKK8v9duDrrrtOlmXpu+++U5s2bcI6s4mC2eecnBz16tVLTz/9tCSpQ4cOqlu3rm6++Wa98MILnD0PkWi9FnJmphJXXHGFunTpotWrV5c7vnr1avXs2bPS+/To0aPC+lWrVqlr166KjY0N26ymC2avpZ/PyDz22GNasGAB73kHoLr7nJCQoJ07d6qgoKDsMmzYMLVr104FBQW68cYbIzW6UYL5eu7Vq5cOHz6sU6dOlR3bs2ePatWqpRYtWoR1XlMFs89nzpxRrVrlX/Jq164t6f/OHODSRe21MKzfXmywCx/7e+2116xdu3ZZY8aMserWrWt9/fXXlmVZ1tixY61HHnmkbP2Fj6M99dRT1q5du6zXXnuNj2YHqLp7vWDBAismJsaaNWuW5Xa7yy4nTpyI1lMwQnX3+df4NFNgqrvPXq/XatGihXXvvfdahYWF1rp166w2bdpYTz75ZLSeghGqu89vvPGGFRMTY82ePdvat2+ftX79eqtr165W9+7do/UUjOD1eq3t27db27dvtyRZubm51vbt28s+Am+X10Jixo9Zs2ZZLVu2tK644grrhhtusNatW1d225AhQ6zevXuXW7927Vqrc+fO1hVXXGG1atXKmjNnToQnNld19rp3796WpAqXIUOGRH5ww1T3a/qXiJnAVXefd+/ebfXv39+Kj4+3WrRoYWVmZlpnzpyJ8NTmqe4+v/LKK1b79u2t+Ph4y+VyWYMHD7a+++67CE9tlo8++sjvf2/t8lrIb80GAABG43tmAACA0YgZAABgNGIGAAAYjZgBAABGI2YAAIDRiBkAAGA0YgYAABiNmAEAAEYjZgDUSA6HQ8uWLYv2GAAigJgBAABGI2YAAIDRiBkAYdGnTx9lZGQoIyND9evXV6NGjTRu3DhV9evgsrKydNNNN1U43qFDB02YMEGStGXLFg0YMECNGzdWYmKievfurW3btl30MdeuXSuHw6ETJ06UHSsoKJDD4dDXX39ddmzDhg265ZZbFB8fr+TkZI0aNUqnT5+u3hMHEHHEDICweeuttxQTE6NNmzbplVde0csvv6x//etffu8zePBgbdq0Sfv27Ss7VlhYqJ07d2rw4MGSJK/XqyFDhigvL0+ffvqp2rRpo4EDB8rr9QY9686dO3Xbbbfpnnvu0Y4dO7R48WKtX79eGRkZQT8mgMjgt2YDCIs+ffqoqKhIhYWFcjgckqSxY8dq+fLl2rVrl9/7duzYUffee6/Gjx8vSXruuef0wQcfaPPmzZWuLy0tVYMGDbRgwQINGjRI0s/fALx06VKlp6dr7dq16tu3r44fP6769etL+vnMTOfOnXXgwAG1atVKjz76qOLj4/XPf/6z7HHXr1+v3r176/Tp04qLi7vULQEQJpyZARA2N910U1nISFKPHj20d+9elZaW+r3f4MGDNX/+fEmSZVlauHBh2VkZSSoqKtKwYcPUtm1bJSYmKjExUadOndK3334b9Kz5+fl688039Zvf/Kbsctttt8nn8+nAgQNBPy6A8IuJ9gAA8GsPPfSQxo4dq23btqm4uFgHDx7Ugw8+WHb7Y489ph9++EHTp09Xy5Yt5XQ61aNHD507d67Sx6tV6+f/b/vlieiffvqp3Bqfz6ehQ4dq1KhRFe5/1VVXheJpAQgTYgZA2Hz66acVrrdp00a1a9f2e78WLVrolltu0fz581VcXKz+/fsrKSmp7Pa8vDzNnj1bAwcOlCQdPHhQR48evejjNWnSRJLkdrvVoEEDST+/zfRLN9xwgwoLC3XNNdcE/PwA2ANvMwEIm4MHDyozM1NffvmlFi5cqBkzZmj06NEB3Xfw4MFatGiRlixZoocffrjcbddcc43+/e9/a/fu3dq0aZMGDx6s+Pj4iz7WNddco+TkZE2cOFF79uzR//73P7300kvl1jz77LPauHGjRowYoYKCAu3du1fLly/XyJEjq//EAUQUMQMgbB599FEVFxere/fuGjFihEaOHKm//OUvAd33vvvu07Fjx3TmzBmlp6eXu+3111/X8ePH1blzZz3yyCMaNWqUmjZtetHHio2N1cKFC/XFF1+oY8eOmjZtml544YVyazp06KB169Zp7969uvnmm9W5c2eNHz9eLper2s8bQGTxaSYAYdGnTx916tRJ06dPj/YoAGo4zswAAACjETMAIiovL6/cx59/fQGA6uJtJgARVVxcrEOHDl30dj5NBKC6iBkAAGA03mYCAABGI2YAAIDRiBkAAGA0YgYAABiNmAEAAEYjZgAAgNGIGQAAYDRiBgAAGO3/AZdxl5E++9jlAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.histplot(edist_df, x='p_value', bins=50)\n",
    "plt.axvline(x=0.05, color='red')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Axes: xlabel='e_distance', ylabel='Count'>"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjMAAAGxCAYAAACXwjeMAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjEsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvc2/+5QAAAAlwSFlzAAAPYQAAD2EBqD+naQAANfhJREFUeJzt3X9UVHXCx/HPiDiCAuYvBgqSDExF06R1pR/qmrRqnpTWp9Ray+poaC3yuBbR5mQtFK1Em+lmp0V2fVh7nt1qfbZyJUtqI3eBRF0015KEpyC0EFAQEu7zh4dZJxABgTvX3q9z7jnM937nzoehk5/zvXfu2AzDMAQAAGBRvcwOAAAAcCEoMwAAwNIoMwAAwNIoMwAAwNIoMwAAwNIoMwAAwNIoMwAAwNIoMwAAwNJ6mx2guzU1NenLL7+Un5+fbDab2XEAAEA7GIahmpoaBQcHq1evttdeLvoy8+WXXyokJMTsGAAAoBNKS0t12WWXtTnnoi8zfn5+ks68Gf7+/ianAQAA7VFdXa2QkBDXv+NtuejLTPOpJX9/f8oMAAAW055LRLgAGAAAWBplBgAAWBplBgAAWNpFf80MAABdrbGxUd9++63ZMSzN29tbXl5eXXIsygwAAO1kGIbKy8t1/Phxs6NcFAYMGCCHw3HB94GjzAAA0E7NRWbo0KHy9fXlZqydZBiGamtrVVFRIUkKCgq6oONRZgAAaIfGxkZXkRk0aJDZcSzPx8dHklRRUaGhQ4de0CknLgAGAKAdmq+R8fX1NTnJxaP5vbzQ648oMwAAdACnlrpOV72XlBkAALrZ3XffrTlz5nTpMT///HPZbDYVFha6xj788EONGTNG3t7emjNnjnbu3CmbzXbRX7DMNTMAAHSz559/XoZhdOkxQ0JCVFZWpsGDB7vGEhISNG7cOL399tvq37+/fH19VVZWpoCAgC597bZs2rRJ8fHxPVqgKDMAAHSz7igTXl5ecjgcbmOfffaZli5d6vYt09+dczHiNBMAAF3kj3/8o8aMGSMfHx8NGjRIN910k06ePNniNFNNTY0WLlyofv36KSgoSM8995ymTJmi+Ph415xhw4YpOTlZixcvlp+fn0JDQ7Vx40bX/rNPMzX//PXXX2vx4sWy2WzatGlTq6eZPvzwQ02ePFm+vr665JJLdPPNN6uyslKStG3bNl1//fUaMGCABg0apFtuuUWfffZZi9d87bXXNHXqVPn6+urqq6/WRx99JEnauXOn7rnnHlVVVclms8lms8npdHbLe302ygwAAF2grKxM8+fP1+LFi3XgwAHt3LlTsbGxrZ5eSkhI0IcffqitW7cqOztbH3zwgT7++OMW89auXauoqCjt3r1bcXFxeuCBB/TJJ5+0mNd8ysnf31/p6ekqKyvT7bff3mJeYWGhpk2bptGjR+ujjz7S3/72N82ePVuNjY2SpJMnTyohIUF5eXnasWOHevXqpblz56qpqcntOElJSVq5cqUKCwsVERGh+fPn6/Tp04qOjlZ6err8/f1VVlamsrIyrVy5srNvabtxmgkAgC5QVlam06dPKzY2VpdffrkkacyYMS3m1dTUKDMzU1lZWZo2bZokKSMjQ8HBwS3mzpw5U3FxcZKkhx9+WM8995x27typq666ym1e8yknm82mgICAc55aSk1NVVRUlNavX+8aGz16tOvn2267zW3+K6+8oqFDh2r//v2KjIx0ja9cuVKzZs2SJD3xxBMaPXq0Pv30U1111VUKCAiQzWbr0dNbrMxcoDHjxmtIoKPNbcy48WbHBAB0s6uvvlrTpk3TmDFjNG/ePL388suu0zdnO3z4sL799lv94Ac/cI0FBARoxIgRLeaOHTvW9XNzQWi+a25nNK/MnMtnn32mBQsW6IorrpC/v7/CwsIkSSUlJefM1Xz33gvJdaFYmblA5WVlenTz+23OSb7zxh5KAwAwi5eXl7Kzs5Wbm6vt27frhRdeUFJSkv7+97+7zWs+7fTde6y0djrK29vb7bHNZmtxyqcjmu+6ey6zZ89WSEiIXn75ZQUHB6upqUmRkZFqaGg4Z67m3+NCcl0oVmYAAOgiNptN1113nZ544gnt3r1bffr00euvv+42Z/jw4fL29tY//vEP11h1dbUOHTrU7fnGjh2rHTt2tLrv66+/1oEDB/TYY49p2rRpGjlyZKsrS+fTp08f1zU4PYWVGQAAusDf//537dixQzExMRo6dKj+/ve/6+jRoxo5cqT27t3rmufn56dFixbp5z//uQYOHKihQ4dq9erV6tWrV7ffXTgxMVFjxoxRXFycli5dqj59+ui9997TvHnzNHDgQA0aNEgbN25UUFCQSkpK9Mgjj3T4NYYNG6YTJ05ox44duvrqq+Xr69vtXwHBygwAAF3A399f77//vmbOnKmIiAg99thjWrt2rWbMmNFiblpamiZNmqRbbrlFN910k6677jqNHDlSffv27daMERER2r59u/bs2aMf/OAHmjRpkv785z+rd+/e6tWrl7Zs2aKCggJFRkZqxYoVevbZZzv8GtHR0Vq6dKluv/12DRkyRKmpqd3wm7izGV19S0IPU11drYCAAFVVVcnf37/Ljz8k0NGua2aOflXe5a8NAOg5p06dUnFxscLCwrq8dJw8eVKXXnqp1q5dq3vvvbdLj+3J2npPO/LvN6eZAADoYbt379Ynn3yiH/zgB6qqqtKaNWskSbfeeqvJyayJMgMAgAl+9atf6eDBg+rTp48mTJigDz74wO17ltB+pl4zc/r0aT322GMKCwuTj4+PrrjiCq1Zs8bt412GYcjpdCo4OFg+Pj6aMmWKioqKTEwNAMCFGT9+vAoKCnTixAl98803ys7ObvUGe2gfU8vMM888o9/85jdat26dDhw4oNTUVD377LN64YUXXHNSU1OVlpamdevWKS8vTw6HQ9OnT1dNTY2JyQEAgKcwtcx89NFHuvXWWzVr1iwNGzZMP/nJTxQTE6P8/HxJZ1Zl0tPTlZSUpNjYWEVGRiozM1O1tbXKysoyMzoAAPAQppaZ66+/Xjt27NC//vUvSdKePXv0t7/9TTNnzpQkFRcXq7y8XDExMa7n2O12TZ48Wbm5ua0es76+XtXV1W4bAAC4eJl6AfDDDz+sqqoqXXXVVfLy8lJjY6N++ctfav78+ZKk8vIzH2cODAx0e15gYKCOHDnS6jFTUlL0xBNPdG9wAADgMUxdmXn11Ve1efNmZWVl6eOPP1ZmZqZ+9atfKTMz021ea99fca67JCYmJqqqqsq1lZaWdlt+AABgPlNXZn7+85/rkUce0R133CHpzFelHzlyRCkpKVq0aJHr68PLy8td38opnflmzu+u1jSz2+2y2+3dHx4AAHgEU1dmamtr1auXewQvLy/XR7PDwsLkcDiUnZ3t2t/Q0KCcnBxFR0f3aFYAAL5vNm3apAEDBpgd47xMXZmZPXu2fvnLXyo0NFSjR4/W7t27lZaWpsWLF0s6c3opPj5eycnJCg8PV3h4uJKTk+Xr66sFCxaYGR0AAJeSkhIdO3asx15v8ODBCg0Nbff8u+++u8UlHJJ06NAhXXnllV0ZzRSmlpkXXnhBv/jFLxQXF6eKigoFBwdryZIlevzxx11zVq1apbq6OsXFxamyslITJ07U9u3b5efnZ2JyAADOKCkp0VUjR6qutrbHXtPH11efHDjQoULz4x//WBkZGW5jQ4YM6epopjC1zPj5+Sk9PV3p6ennnGOz2eR0OuV0OnssFwAA7XXs2DHV1dZq4cPPKjB0eLe/3lcln+m/nvm5jh071qEyY7fbXdeiNktLS1NGRoYOHz6sgQMHavbs2UpNTVX//v1bPcaePXsUHx+v/Px82Ww2hYeH66WXXlJUVJQkKTc3V4888ojy8vI0ePBgzZ07VykpKerXr1/nf+F24LuZAADoAoGhw3VZ+GizY3RIr1699Otf/1rDhg1TcXGx4uLitGrVKq1fv77V+QsXLtT48eO1YcMGeXl5qbCwUN7e3pKkffv26eabb9aTTz6pV155RUePHtXy5cu1fPnyFitCXY0yAwDA98Bf/vIXtxWXGTNm6H/+539cj8PCwvTkk0/qgQceOGeZKSkp0c9//nNdddVVkqTw8HDXvmeffVYLFixQfHy8a9+vf/1rTZ48WRs2bFDfvn274bc6gzIDAMD3wNSpU7VhwwbX4379+um9995TcnKy9u/fr+rqap0+fVqnTp3SyZMnWz01lJCQoPvuu0+///3vddNNN2nevHkaPvzMqbWCggJ9+umn+q//+i/XfMMw1NTUpOLiYo0cObLbfjdTP5oNAAB6Rr9+/XTllVe6toaGBs2cOVORkZH605/+pIKCAr344ouSpG+//bbVYzidThUVFWnWrFl69913NWrUKL3++uuSpKamJi1ZskSFhYWubc+ePTp06JCr8HQXVmYAAPgeys/P1+nTp7V27VrXPd/++7//+7zPi4iIUEREhFasWKH58+crIyNDc+fO1TXXXKOioiJTPurNygwAAN9Dw4cP1+nTp/XCCy/o8OHD+v3vf6/f/OY355xfV1en5cuXa+fOnTpy5Ig+/PBD5eXluU4fPfzww/roo4+0bNkyFRYW6tChQ9q6dasefPDBbv9dWJkBAKALfFXymaVeZ9y4cUpLS9MzzzyjxMRE3XjjjUpJSdFPf/rTVud7eXnp66+/1k9/+lN99dVXGjx4sGJjY11f7jx27Fjl5OQoKSlJN9xwgwzD0PDhw3X77bd3Sd622AzDMLr9VUxUXV2tgIAAVVVVyd/fv8uPPyTQoUc3v9/mnOQ7b9TRr8q7/LUBAD3n1KlTKi4uVlhYmNsnc6xy0zxPdK73VOrYv9+szAAAcAFCQ0P1yYEDHv11Bhc7ygwAABcoNDSUcmEiLgAGAACWRpkBAACWRpkBAACWRpkBAACWRpkBAACWRpkBAACWRpkBAACWRpkBAACWxk3zAAC4QCUlJR57B2Cbzdbm/kWLFmnTpk1dkMo8lBkAAC5ASUmJRo68SrW1dT32mr6+Pjpw4JN2FZqysjLXz6+++qoef/xxHTx40DXm4+PjNv/bb7+Vt7d314XtAZQZAAAuwLFjx1RbW6fNj/6HRoYO6fbXO1ByVHcm/7eOHTvWrjLjcDhcPwcEBMhms7nGPv/8cwUFBenVV1/V+vXrtWvXLm3YsEFHjhzRG2+8ocLCQtdz09PTlZ6ers8//9w1lpGRodTUVBUXF2vYsGF66KGHFBcX12W/a3tRZgAA6AIjQ4fomohLzY7RKQ8//LDWrl2rjIwM2e12bdy48bzPefnll7V69WqtW7dO48eP1+7du3X//ferX79+WrRoUQ+k/jfKDAAA33Px8fGKjY3t0HOefPJJrV271vW8sLAw7d+/Xy+99BJlBgAA9KyoqKgOzT969KhKS0t177336v7773eNnz59WgEBAV0d77woMwAAfM/169fP7XGvXr1kGIbb2Lfffuv6uampSdKZU00TJ050m+fl5dVNKc+NMgMAANwMGTJE5eXlMgzD9dHusy8GDgwM1KWXXqrDhw9r4cKFJqX8N8oMAABwM2XKFB09elSpqan6yU9+om3btuntt9+Wv7+/a47T6dRDDz0kf39/zZgxQ/X19crPz1dlZaUSEhJ6NC9lpgccP16lIYGONuc4goK0r3B3DyUCAHS1AyVHL5rXGTlypNavX6/k5GQ9+eSTuu2227Ry5Uq3Tzndd9998vX11bPPPqtVq1apX79+GjNmjOLj47s933fZjO+eFLvIVFdXKyAgQFVVVW6NsqsMCXTo0c3vtzln5cwx+tVb+9qck3znjTr6VXlXRgMAdKFTp06puLhYYWFh6tu3r2vc02+a58nO9Z5KHfv3m5UZAAAuQGhoqA4c+MRjv87g+4AyAwDABQoNDaVcmMjUb80eNmyYbDZbi23ZsmWSJMMw5HQ6FRwcLB8fH02ZMkVFRUVmRgYAAB7G1DKTl5ensrIy15adnS1JmjdvniQpNTVVaWlpWrdunfLy8uRwODR9+nTV1NSYGRsAAHgQU8vMkCFD5HA4XNtf/vIXDR8+XJMnT5ZhGEpPT1dSUpJiY2MVGRmpzMxM1dbWKisry8zYAADAg5haZs7W0NCgzZs3a/HixbLZbCouLlZ5ebliYmJcc+x2uyZPnqzc3FwTkwIAvs8u8g8B96iuei895gLgN954Q8ePH9fdd98tSSovP/Mx5cDAQLd5gYGBOnLkyDmPU19fr/r6etfj6urqrg8LAPje8fb2liTV1tbKx8fH5DQXh9raWkn/fm87y2PKzCuvvKIZM2YoODjYbbz5NsrNzr61cmtSUlL0xBNPdEtGAMD3l5eXlwYMGKCKigpJkq+vb5v/HuHcDMNQbW2tKioqNGDAgAv+PiePKDNHjhzRO++8o9dee8015nCcuWNueXm5goKCXOMVFRUtVmvOlpiY6HYb5erqaoWEhHRDagDA903zv03NhQYXZsCAAa739EJ4RJnJyMjQ0KFDNWvWLNdYWFiYHA6HsrOzNX78eElnrqvJycnRM888c85j2e122e32bs8MAPj+sdlsCgoK0tChQ92+RRod5+3t3WXfsG16mWlqalJGRoYWLVqk3r3/Hcdmsyk+Pl7JyckKDw9XeHi4kpOT5evrqwULFpiYGADwfefl5dVl/xDjwpleZt555x2VlJRo8eLFLfatWrVKdXV1iouLU2VlpSZOnKjt27fLz8/PhKQAAMATmV5mYmJizvnRLJvNJqfTKafT2bOhAACAZXjMfWYAAAA6gzIDAAAsjTIDAAAsjTIDAAAsjTIDAAAsjTIDAAAsjTIDAAAsjTIDAAAsjTIDAAAsjTIDAAAsjTIDAAAsjTIDAAAsjTIDAAAsjTIDAAAsjTIDAAAsjTIDAAAsjTIDAAAsjTIDAAAsjTIDAAAsjTIDAAAsjTIDAAAsjTIDAAAsjTIDAAAsjTIDAAAsjTIDAAAsjTIDAAAsjTIDAAAsjTIDAAAsjTIDAAAsjTIDAAAsjTIDAAAsjTIDAAAszfQy88UXX+jOO+/UoEGD5Ovrq3HjxqmgoMC13zAMOZ1OBQcHy8fHR1OmTFFRUZGJiQEAgCcxtcxUVlbquuuuk7e3t95++23t379fa9eu1YABA1xzUlNTlZaWpnXr1ikvL08Oh0PTp09XTU2NecEBAIDH6G3miz/zzDMKCQlRRkaGa2zYsGGunw3DUHp6upKSkhQbGytJyszMVGBgoLKysrRkyZKejgwAADyMqSszW7duVVRUlObNm6ehQ4dq/Pjxevnll137i4uLVV5erpiYGNeY3W7X5MmTlZub2+ox6+vrVV1d7bYBAICLl6ll5vDhw9qwYYPCw8P117/+VUuXLtVDDz2k3/3ud5Kk8vJySVJgYKDb8wIDA137vislJUUBAQGuLSQkpHt/CQAAYCpTy0xTU5OuueYaJScna/z48VqyZInuv/9+bdiwwW2ezWZze2wYRouxZomJiaqqqnJtpaWl3ZYfAACYz9QyExQUpFGjRrmNjRw5UiUlJZIkh8MhSS1WYSoqKlqs1jSz2+3y9/d32wAAwMXL1DJz3XXX6eDBg25j//rXv3T55ZdLksLCwuRwOJSdne3a39DQoJycHEVHR/doVgAA4JlM/TTTihUrFB0dreTkZP3Hf/yH/vGPf2jjxo3auHGjpDOnl+Lj45WcnKzw8HCFh4crOTlZvr6+WrBggZnRAQCAhzC1zFx77bV6/fXXlZiYqDVr1igsLEzp6elauHCha86qVatUV1enuLg4VVZWauLEidq+fbv8/PxMTA4AADyFqWVGkm655Rbdcsst59xvs9nkdDrldDp7LhQAALAM07/OAAAA4EJQZgAAgKVRZgAAgKVRZgAAgKVRZgAAgKVRZgAAgKVRZgAAgKVRZgAAgKVRZgAAgKVRZgAAgKVRZgAAgKVRZgAAgKVRZgAAgKVRZgAAgKVRZgAAgKVRZgAAgKVRZgAAgKVRZgAAgKVRZgAAgKVRZgAAgKVRZgAAgKVRZgAAgKVRZgAAgKVRZgAAgKVRZgAAgKVRZgAAgKVRZgAAgKVRZgAAgKVRZgAAgKVRZgAAgKVRZgAAgKWZWmacTqdsNpvb5nA4XPsNw5DT6VRwcLB8fHw0ZcoUFRUVmZgYAAB4GtNXZkaPHq2ysjLXtm/fPte+1NRUpaWlad26dcrLy5PD4dD06dNVU1NjYmIAAOBJTC8zvXv3lsPhcG1DhgyRdGZVJj09XUlJSYqNjVVkZKQyMzNVW1urrKwsk1MDAABPYXqZOXTokIKDgxUWFqY77rhDhw8fliQVFxervLxcMTExrrl2u12TJ09Wbm6uWXEBAICH6W3mi0+cOFG/+93vFBERoa+++kpPPfWUoqOjVVRUpPLycklSYGCg23MCAwN15MiRcx6zvr5e9fX1rsfV1dXdEx4AAHgEU8vMjBkzXD+PGTNGkyZN0vDhw5WZmakf/vCHkiSbzeb2HMMwWoydLSUlRU888UT3BAYAAB7H9NNMZ+vXr5/GjBmjQ4cOuT7V1LxC06yioqLFas3ZEhMTVVVV5dpKS0u7NTMAADCXR5WZ+vp6HThwQEFBQQoLC5PD4VB2drZrf0NDg3JychQdHX3OY9jtdvn7+7ttAADg4mXqaaaVK1dq9uzZCg0NVUVFhZ566ilVV1dr0aJFstlsio+PV3JyssLDwxUeHq7k5GT5+vpqwYIFZsYGAAAexNQy83//93+aP3++jh07piFDhuiHP/yhdu3apcsvv1yStGrVKtXV1SkuLk6VlZWaOHGitm/fLj8/PzNjAwAAD2JqmdmyZUub+202m5xOp5xOZ88EAgAAluNR18wAAAB0FGUGAABYWqfKzBVXXKGvv/66xfjx48d1xRVXXHAoAACA9upUmfn888/V2NjYYry+vl5ffPHFBYcCAABorw5dALx161bXz3/9618VEBDgetzY2KgdO3Zo2LBhXRYOAADgfDpUZubMmSPpzKeMFi1a5LbP29tbw4YN09q1a7ssHAAAwPl0qMw0NTVJksLCwpSXl6fBgwd3SygAAID26tR9ZoqLi7s6BwAAQKd0+qZ5O3bs0I4dO1RRUeFasWn229/+9oKDAQAAtEenyswTTzyhNWvWKCoqSkFBQbLZbF2dCwAAoF06VWZ+85vfaNOmTbrrrru6Og8AAECHdOo+Mw0NDYqOju7qLAAAAB3WqTJz3333KSsrq6uzAAAAdFinTjOdOnVKGzdu1DvvvKOxY8fK29vbbX9aWlqXhAMAADifTpWZvXv3aty4cZKkf/7zn277uBgYAAD0pE6Vmffee6+rcwAAAHRKp66ZAQAA8BSdWpmZOnVqm6eT3n333U4HAgAA6IhOlZnm62WaffvttyosLNQ///nPFl9ACQAA0J06VWaee+65VsedTqdOnDhxQYEAAAA6okuvmbnzzjv5XiYAANCjurTMfPTRR+rbt29XHhIAAKBNnTrNFBsb6/bYMAyVlZUpPz9fv/jFL7okGAAAQHt0qswEBAS4Pe7Vq5dGjBihNWvWKCYmpkuCAQAAtEenykxGRkZX5wAAAOiUTpWZZgUFBTpw4IBsNptGjRql8ePHd1UuAACAdulUmamoqNAdd9yhnTt3asCAATIMQ1VVVZo6daq2bNmiIUOGdHVOAACAVnXq00wPPvigqqurVVRUpG+++UaVlZX65z//qerqaj300ENdnREAAOCcOrUys23bNr3zzjsaOXKka2zUqFF68cUXuQAYAAD0qE6tzDQ1Ncnb27vFuLe3t5qami44FAAAQHt1qsz86Ec/0s9+9jN9+eWXrrEvvvhCK1as0LRp07osHAAAwPl0qsysW7dONTU1GjZsmIYPH64rr7xSYWFhqqmp0QsvvNCpICkpKbLZbIqPj3eNGYYhp9Op4OBg+fj4aMqUKSoqKurU8QEAwMWpU9fMhISE6OOPP1Z2drY++eQTGYahUaNG6aabbupUiLy8PG3cuFFjx451G09NTVVaWpo2bdqkiIgIPfXUU5o+fboOHjwoPz+/Tr0WAAC4uHRoZebdd9/VqFGjVF1dLUmaPn26HnzwQT300EO69tprNXr0aH3wwQcdCnDixAktXLhQL7/8si655BLXuGEYSk9PV1JSkmJjYxUZGanMzEzV1tYqKyurQ68BAAAuXh0qM+np6br//vvl7+/fYl9AQICWLFmitLS0DgVYtmyZZs2a1WJVp7i4WOXl5W6fjrLb7Zo8ebJyc3M79BoAAODi1aEys2fPHv34xz8+5/6YmBgVFBS0+3hbtmzRxx9/rJSUlBb7ysvLJUmBgYFu44GBga59ramvr1d1dbXbBgAALl4dKjNfffVVqx/Jbta7d28dPXq0XccqLS3Vz372M23evFl9+/Y95zybzeb22DCMFmNnS0lJUUBAgGsLCQlpVx4AAGBNHSozl156qfbt23fO/Xv37lVQUFC7jlVQUKCKigpNmDBBvXv3Vu/evZWTk6Nf//rX6t27t2tF5rurMBUVFS1Wa86WmJioqqoq11ZaWtquPAAAwJo6VGZmzpypxx9/XKdOnWqxr66uTqtXr9Ytt9zSrmNNmzZN+/btU2FhoWuLiorSwoULVVhYqCuuuEIOh0PZ2dmu5zQ0NCgnJ0fR0dHnPK7dbpe/v7/bBgAALl4d+mj2Y489ptdee00RERFavny5RowYIZvNpgMHDujFF19UY2OjkpKS2nUsPz8/RUZGuo3169dPgwYNco3Hx8crOTlZ4eHhCg8PV3Jysnx9fbVgwYKOxAYAABexDpWZwMBA5ebm6oEHHlBiYqIMw5B05rqWm2++WevXr2/zFFBHrVq1SnV1dYqLi1NlZaUmTpyo7du3c48ZAADg0uGb5l1++eV66623VFlZqU8//VSGYSg8PNztHjGdtXPnTrfHNptNTqdTTqfzgo8NAAAuTp26A7AkXXLJJbr22mu7MgsAAECHdeq7mQAAADwFZQYAAFgaZQYAAFgaZQYAAFgaZQYAAFgaZQYAAFgaZQYAAFgaZQYAAFgaZQYAAFgaZQYAAFgaZQYAAFgaZQYAAFgaZQYAAFgaZQYAAFgaZQYAAFgaZQYAAFgaZQYAAFgaZQYAAFgaZQYAAFgaZQYAAFgaZQYAAFgaZQYAAFgaZQYAAFgaZQYAAFgaZQYAAFgaZQYAAFgaZQYAAFgaZQYAAFgaZQYAAFgaZQYAAFgaZQYAAFiaqWVmw4YNGjt2rPz9/eXv769Jkybp7bffdu03DENOp1PBwcHy8fHRlClTVFRUZGJiAADgaUwtM5dddpmefvpp5efnKz8/Xz/60Y906623ugpLamqq0tLStG7dOuXl5cnhcGj69OmqqakxMzYAAPAgppaZ2bNna+bMmYqIiFBERIR++ctfqn///tq1a5cMw1B6erqSkpIUGxuryMhIZWZmqra2VllZWWbGBgAAHsRjrplpbGzUli1bdPLkSU2aNEnFxcUqLy9XTEyMa47dbtfkyZOVm5t7zuPU19erurrabQMAABcv08vMvn371L9/f9ntdi1dulSvv/66Ro0apfLycklSYGCg2/zAwEDXvtakpKQoICDAtYWEhHRrfgAAYC7Ty8yIESNUWFioXbt26YEHHtCiRYu0f/9+136bzeY23zCMFmNnS0xMVFVVlWsrLS3ttuwAAMB8vc0O0KdPH1155ZWSpKioKOXl5en555/Xww8/LEkqLy9XUFCQa35FRUWL1Zqz2e122e327g0NAAA8hukrM99lGIbq6+sVFhYmh8Oh7Oxs176Ghgbl5OQoOjraxIQAAMCTmLoy8+ijj2rGjBkKCQlRTU2NtmzZop07d2rbtm2y2WyKj49XcnKywsPDFR4eruTkZPn6+mrBggVmxgYAAB7E1DLz1Vdf6a677lJZWZkCAgI0duxYbdu2TdOnT5ckrVq1SnV1dYqLi1NlZaUmTpyo7du3y8/Pz8zYAADAg5haZl555ZU299tsNjmdTjmdzp4JBAAALMfjrpkBAADoCMoMAACwNMoMAACwNMoMAACwNMoMAACwNMoMAACwNMoMAACwNMoMAACwNMoMAACwNMoMAACwNMoMAACwNMoMAACwNMoMAACwNMoMAACwNMoMAACwNMoMAACwNMoMAACwNMoMAACwNMoMAACwNMoMAACwNMoMAACwNMoMAACwNMoMAACwNMoMAACwNMoMAACwNMoMAACwNMoMAACwNMoMAACwNMoMAACwNMoMAACwNMoMAACwNFPLTEpKiq699lr5+flp6NChmjNnjg4ePOg2xzAMOZ1OBQcHy8fHR1OmTFFRUZFJiQEAgKcxtczk5ORo2bJl2rVrl7Kzs3X69GnFxMTo5MmTrjmpqalKS0vTunXrlJeXJ4fDoenTp6umpsbE5AAAwFP0NvPFt23b5vY4IyNDQ4cOVUFBgW688UYZhqH09HQlJSUpNjZWkpSZmanAwEBlZWVpyZIlZsQGAAAexKOumamqqpIkDRw4UJJUXFys8vJyxcTEuObY7XZNnjxZubm5pmQEAACexdSVmbMZhqGEhARdf/31ioyMlCSVl5dLkgIDA93mBgYG6siRI60ep76+XvX19a7H1dXV3ZQYAAB4Ao9ZmVm+fLn27t2rP/zhDy322Ww2t8eGYbQYa5aSkqKAgADXFhIS0i15AQCAZ/CIMvPggw9q69ateu+993TZZZe5xh0Oh6R/r9A0q6ioaLFa0ywxMVFVVVWurbS0tPuCAwAA05laZgzD0PLly/Xaa6/p3XffVVhYmNv+sLAwORwOZWdnu8YaGhqUk5Oj6OjoVo9pt9vl7+/vtgEAgIuXqdfMLFu2TFlZWfrzn/8sPz8/1wpMQECAfHx8ZLPZFB8fr+TkZIWHhys8PFzJycny9fXVggULzIwOAAA8hKllZsOGDZKkKVOmuI1nZGTo7rvvliStWrVKdXV1iouLU2VlpSZOnKjt27fLz8+vh9MCAABPZGqZMQzjvHNsNpucTqecTmf3BwIAAJbjERcAAwAAdBZlBgAAWBplBgAAWBplBgAAWBplBgAAWBplBgAAWBplBgAAWBplBgAAWBplBgAAWBplBgAAWBplBgAAWBplBgAAWBplBgAAWBplBgAAWBplBgAAWBplBgAAWBplBgAAWBplBgAAWBplBgAAWBplBgAAWBplBgAAWBplBgAAWBplBgAAWBplBgAAWBplBgAAWBplBgAAWBplBgAAWBplBgAAWBplBgAAWBplBgAAWBplBgAAWBplBgAAWJqpZeb999/X7NmzFRwcLJvNpjfeeMNtv2EYcjqdCg4Olo+Pj6ZMmaKioiJzwnaz48ePKzhwcJvbhHFjzI4JAIDH6W3mi588eVJXX3217rnnHt12220t9qempiotLU2bNm1SRESEnnrqKU2fPl0HDx6Un5+fCYm7T1NTk77c8p9tzgm+Y20PpQEAwDpMLTMzZszQjBkzWt1nGIbS09OVlJSk2NhYSVJmZqYCAwOVlZWlJUuW9GRUAADgoTz2mpni4mKVl5crJibGNWa32zV58mTl5uae83n19fWqrq522wAAwMXLY8tMeXm5JCkwMNBtPDAw0LWvNSkpKQoICHBtISEh3ZoTAACYy2PLTDObzeb22DCMFmNnS0xMVFVVlWsrLS3t7ogAAMBEpl4z0xaHwyHpzApNUFCQa7yioqLFas3Z7Ha77HZ7t+cDAACewWNXZsLCwuRwOJSdne0aa2hoUE5OjqKjo01MBgAAPImpKzMnTpzQp59+6npcXFyswsJCDRw4UKGhoYqPj1dycrLCw8MVHh6u5ORk+fr6asGCBSamBgAAnsTUMpOfn6+pU6e6HickJEiSFi1apE2bNmnVqlWqq6tTXFycKisrNXHiRG3fvv2iu8cMAADoPFPLzJQpU2QYxjn322w2OZ1OOZ3OngsFAAAsxWOvmQEAAGgPygwAALA0j/1o9sXEMAyl/Sr1vHMAAEDHUWZ6gCFDCT9p++Pk/7kjs4fSAABwceE0EwAAsDTKDAAAsDTKDAAAsDTKDAAAsDTKDAAAsDTKDAAAsDTKDAAAsDTKDAAAsDTKDAAAsDTKDAAAsDS+zsBDGLJpyG1Ptznn+Mn6HkoDAIB1UGY8hqFHn3muzRkr45b2UBYAAKyD00wAAMDSKDMAAMDSKDMAAMDSKDMAAMDSKDMAAMDSKDMAAMDSKDMAAMDSKDMAAMDSKDMAAMDSKDMAAMDSKDMAAMDSKDMAAMDSKDMAAMDSKDMAAMDSKDMAAMDSepsdoD3Wr1+vZ599VmVlZRo9erTS09N1ww03mB2rxzUZkpd3n/PMsmngwEvanOEICtK+wt1dF6wNE8aNUVlZWZtzgoKCVFC4r0fyAADazyr/D/f4MvPqq68qPj5e69ev13XXXaeXXnpJM2bM0P79+xUaGmp2vB5m6FfrX2pzRsIDS/To5vfbnJN8541dGapNZWVl+nLLf7Y5J/iOtT2UBgDQEVb5f7jHn2ZKS0vTvffeq/vuu08jR45Uenq6QkJCtGHDBrOjAQAAD+DRZaahoUEFBQWKiYlxG4+JiVFubq5JqQAAgCfx6NNMx44dU2NjowIDA93GAwMDVV5e3upz6uvrVV9f73pcVVUlSaquru6WjE1NTTp18kTbkwxDp+rqem7OefI0NTV12/vR6mudPOUxeQAA7Wfm/8Obj2kYxvknGx7siy++MCQZubm5buNPPfWUMWLEiFafs3r1akMSGxsbGxsb20WwlZaWnrcvePTKzODBg+Xl5dViFaaioqLFak2zxMREJSQkuB43NTXpm2++0aBBg2Sz2Tqcobq6WiEhISotLZW/v3+Hn4+ewd/JOvhbWQN/J+u4WP9WhmGopqZGwcHB553r0WWmT58+mjBhgrKzszV37lzXeHZ2tm699dZWn2O322W3293GBgwYcMFZ/P39L6r/SC5W/J2sg7+VNfB3so6L8W8VEBDQrnkeXWYkKSEhQXfddZeioqI0adIkbdy4USUlJVq6dKnZ0QAAgAfw+DJz++236+uvv9aaNWtUVlamyMhIvfXWW7r88svNjgYAADyAx5cZSYqLi1NcXJwpr22327V69eoWp67gWfg7WQd/K2vg72Qd/K0km2G05zNPAAAAnsmjb5oHAABwPpQZAABgaZQZAABgaZSZNqxfv15hYWHq27evJkyYoA8++MDsSPiO999/X7Nnz1ZwcLBsNpveeOMNsyOhFSkpKbr22mvl5+enoUOHas6cOTp48KDZsdCKDRs2aOzYsa57lkyaNElvv/222bFwHikpKbLZbIqPjzc7iikoM+fw6quvKj4+XklJSdq9e7duuOEGzZgxQyUlJWZHw1lOnjypq6++WuvWrTM7CtqQk5OjZcuWadeuXcrOztbp06cVExOjkydPmh0N33HZZZfp6aefVn5+vvLz8/WjH/1It956q4qKisyOhnPIy8vTxo0bNXbsWLOjmIZPM53DxIkTdc0112jDhg2usZEjR2rOnDlKSUkxMRnOxWaz6fXXX9ecOXPMjoLzOHr0qIYOHaqcnBzdeOONZsfBeQwcOFDPPvus7r33XrOj4DtOnDiha665RuvXr9dTTz2lcePGKT093exYPY6VmVY0NDSooKBAMTExbuMxMTHKzc01KRVw8Wj+NvuBAweanARtaWxs1JYtW3Ty5ElNmjTJ7DhoxbJlyzRr1izddNNNZkcxlSVumtfTjh07psbGxhZfZhkYGNjiSy8BdIxhGEpISND111+vyMhIs+OgFfv27dOkSZN06tQp9e/fX6+//rpGjRpldix8x5YtW/Txxx8rLy/P7Cimo8y04bvfsm0YRqe+eRvAvy1fvlx79+7V3/72N7Oj4BxGjBihwsJCHT9+XH/605+0aNEi5eTkUGg8SGlpqX72s59p+/bt6tu3r9lxTEeZacXgwYPl5eXVYhWmoqKixWoNgPZ78MEHtXXrVr3//vu67LLLzI6Dc+jTp4+uvPJKSVJUVJTy8vL0/PPP66WXXjI5GZoVFBSooqJCEyZMcI01Njbq/fff17p161RfXy8vLy8TE/YsrplpRZ8+fTRhwgRlZ2e7jWdnZys6OtqkVIB1GYah5cuX67XXXtO7776rsLAwsyOhAwzDUH19vdkxcJZp06Zp3759KiwsdG1RUVFauHChCgsLv1dFRmJl5pwSEhJ01113KSoqSpMmTdLGjRtVUlKipUuXmh0NZzlx4oQ+/fRT1+Pi4mIVFhZq4MCBCg0NNTEZzrZs2TJlZWXpz3/+s/z8/FyrngEBAfLx8TE5Hc726KOPasaMGQoJCVFNTY22bNminTt3atu2bWZHw1n8/PxaXHPWr18/DRo06Ht5LRpl5hxuv/12ff3111qzZo3KysoUGRmpt956S5dffrnZ0XCW/Px8TZ061fU4ISFBkrRo0SJt2rTJpFT4ruZbHEyZMsVtPCMjQ3fffXfPB8I5ffXVV7rrrrtUVlamgIAAjR07Vtu2bdP06dPNjgacE/eZAQAAlsY1MwAAwNIoMwAAwNIoMwAAwNIoMwAAwNIoMwAAwNIoMwAAwNIoMwAAwNIoMwAAwNIoMwBMtXPnTtlsNh0/flyStGnTJg0YMMDUTACshTIDwKPcfvvt+te//tWuuRQfABLfzQTAw/j4+PDlkwA6hJUZAB1mGIZSU1N1xRVXyMfHR1dffbX++Mc/tuu5b731liIiIuTj46OpU6fq888/d9v/3dWWPXv2aOrUqfLz85O/v78mTJig/Px87dy5U/fcc4+qqqpks9lks9nkdDolSZs3b1ZUVJT8/PzkcDi0YMECVVRUuI7ZfGprx44dioqKkq+vr6Kjo3Xw4EG3LFu3blVUVJT69u2rwYMHKzY21rWvoaFBq1at0qWXXqp+/fpp4sSJ2rlzZ4feRwBdgzIDoMMee+wxZWRkaMOGDSoqKtKKFSt05513Kicnp83nlZaWKjY2VjNnzlRhYaHuu+8+PfLII20+Z+HChbrsssuUl5engoICPfLII/L29lZ0dLTS09Pl7++vsrIylZWVaeXKlZLOFI0nn3xSe/bs0RtvvKHi4uJWv507KSlJa9euVX5+vnr37q3Fixe79r355puKjY3VrFmztHv3blfxaXbPPffoww8/1JYtW7R3717NmzdPP/7xj3Xo0KEOvJMAuoQBAB1w4sQJo2/fvkZubq7b+L333mvMnz+/zecmJiYaI0eONJqamlxjDz/8sCHJqKysNAzDMDIyMoyAgADXfj8/P2PTpk2tHu+7c8/lH//4hyHJqKmpMQzDMN577z1DkvHOO++45rz55puGJKOurs4wDMOYNGmSsXDhwlaP9+mnnxo2m8344osv3ManTZtmJCYmnjcPgK7FNTMAOmT//v06deqUpk+f7jbe0NCg8ePHt/ncAwcO6Ic//KFsNptrbNKkSW0+JyEhQffdd59+//vf66abbtK8efM0fPjwNp+ze/duOZ1OFRYW6ptvvlFTU5MkqaSkRKNGjXLNGzt2rOvnoKAgSVJFRYVCQ0NVWFio+++/v9Xjf/zxxzIMQxEREW7j9fX1GjRoUJvZAHQ9ygyADmkuBm+++aYuvfRSt312u73N5xqG0eHXczqdWrBggd588029/fbbWr16tbZs2aK5c+e2Ov/kyZOKiYlRTEyMNm/erCFDhqikpEQ333yzGhoa3OZ6e3u7fm4uWM2/X1sXITc1NcnLy0sFBQXy8vJy29e/f/8O/44ALgxlBkCHjBo1Sna7XSUlJZo8eXKHn/vGG2+4je3ateu8z4uIiFBERIRWrFih+fPnKyMjQ3PnzlWfPn3U2NjoNveTTz7RsWPH9PTTTyskJESSlJ+f36Gc0plVmx07duiee+5psW/8+PFqbGxURUWFbrjhhg4fG0DX4gJgAB3i5+enlStXasWKFcrMzNRnn32m3bt368UXX1RmZmabz126dKk+++wzJSQk6ODBg8rKytKmTZvOOb+urk7Lly/Xzp07deTIEX344YfKy8vTyJEjJUnDhg3TiRMntGPHDh07dky1tbUKDQ1Vnz599MILL+jw4cPaunWrnnzyyQ7/nqtXr9Yf/vAHrV69WgcOHNC+ffuUmpoq6Uy5WrhwoX7605/qtddeU3FxsfLy8vTMM8/orbfe6vBrAbhAZl+0A8B6mpqajOeff94YMWKE4e3tbQwZMsS4+eabjZycnPM+93//93+NK6+80rDb7cYNN9xg/Pa3vz3nBcD19fXGHXfcYYSEhBh9+vQxgoODjeXLl7su0jUMw1i6dKkxaNAgQ5KxevVqwzAMIysryxg2bJhht9uNSZMmGVu3bjUkGbt37zYM498XADe/pmEYxu7duw1JRnFxsWvsT3/6kzFu3DijT58+xuDBg43Y2FjXvoaGBuPxxx83hg0bZnh7exsOh8OYO3eusXfv3k69pwA6z2YYnTiJDQAA4CE4zQQAACyNMgOgyyxdulT9+/dvdVu6dKnZ8QBcpDjNBKDLVFRUqLq6utV9/v7+Gjp0aA8nAvB9QJkBAACWxmkmAABgaZQZAABgaZQZAABgaZQZAABgaZQZAABgaZQZAABgaZQZAABgaZQZAABgaf8Pz7ZW/TjOZfwAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.histplot(edist_df, x='e_distance', hue='significant', bins=50)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Alternative: other distance metrics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:669: ImplicitModificationWarning: Trying to modify attribute `._uns` of view, initializing view as actual.\n",
      "  adata.uns[key_added] = {}\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:460: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:463: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:473: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/users/thiemea/micromamba/envs/nen_env/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:484: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n"
     ]
    }
   ],
   "source": [
    "sc.tl.rank_genes_groups(adata, groupby='gene_list', reference='control', method='t-test')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSG00000137845</th>\n",
       "      <th>ENSG00000185736</th>\n",
       "      <th>ENSG00000141385</th>\n",
       "      <th>ENSG00000156709</th>\n",
       "      <th>ENSG00000127914</th>\n",
       "      <th>ENSG00000159461</th>\n",
       "      <th>ENSG00000160746</th>\n",
       "      <th>ENSG00000183020</th>\n",
       "      <th>ENSG00000100823</th>\n",
       "      <th>ENSG00000049618</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000143952</th>\n",
       "      <th>ENSG00000196998</th>\n",
       "      <th>ENSG00000158023</th>\n",
       "      <th>ENSG00000186153</th>\n",
       "      <th>ENSG00000101966</th>\n",
       "      <th>ENSG00000073050</th>\n",
       "      <th>ENSG00000134684</th>\n",
       "      <th>ENSG00000188994</th>\n",
       "      <th>ENSG00000166716</th>\n",
       "      <th>ENSG00000153896</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NR2F1</td>\n",
       "      <td>UNC80</td>\n",
       "      <td>NHLRC3</td>\n",
       "      <td>APBB2</td>\n",
       "      <td>FAM184A</td>\n",
       "      <td>NHLRC3</td>\n",
       "      <td>GTF2H2</td>\n",
       "      <td>LNPK</td>\n",
       "      <td>PPFIA2</td>\n",
       "      <td>SGIP1</td>\n",
       "      <td>...</td>\n",
       "      <td>ID2</td>\n",
       "      <td>KMT2B</td>\n",
       "      <td>RPL10</td>\n",
       "      <td>RPF2</td>\n",
       "      <td>SYT4</td>\n",
       "      <td>RRAGB</td>\n",
       "      <td>LCORL</td>\n",
       "      <td>SYT1</td>\n",
       "      <td>MTMR4</td>\n",
       "      <td>IRX3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>IGFBP2</td>\n",
       "      <td>WDR70</td>\n",
       "      <td>GTF2H2</td>\n",
       "      <td>CEP290</td>\n",
       "      <td>BRINP3</td>\n",
       "      <td>CASKIN1</td>\n",
       "      <td>KLHL34</td>\n",
       "      <td>GNB3</td>\n",
       "      <td>ZNF528</td>\n",
       "      <td>ERBB4</td>\n",
       "      <td>...</td>\n",
       "      <td>NHLRC3</td>\n",
       "      <td>GSE1</td>\n",
       "      <td>RPS27</td>\n",
       "      <td>SPAG16</td>\n",
       "      <td>KITLG</td>\n",
       "      <td>KLHL29</td>\n",
       "      <td>ESF1</td>\n",
       "      <td>CRABP2</td>\n",
       "      <td>TNRC6C</td>\n",
       "      <td>DMWD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>N4BP2L2</td>\n",
       "      <td>CLVS2</td>\n",
       "      <td>SAMD5</td>\n",
       "      <td>KCNQ1OT1</td>\n",
       "      <td>LUZP2</td>\n",
       "      <td>PHKG1</td>\n",
       "      <td>SPIN2A</td>\n",
       "      <td>BRSK2</td>\n",
       "      <td>UBE2QL1</td>\n",
       "      <td>ANAPC7</td>\n",
       "      <td>...</td>\n",
       "      <td>AC008522.1</td>\n",
       "      <td>NNAT</td>\n",
       "      <td>ZNF397</td>\n",
       "      <td>UBE2QL1</td>\n",
       "      <td>PPFIA2</td>\n",
       "      <td>UBE2QL1</td>\n",
       "      <td>TANC2</td>\n",
       "      <td>UNC13A</td>\n",
       "      <td>ECM1</td>\n",
       "      <td>CARHSP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PPP4R4</td>\n",
       "      <td>STARD4-AS1</td>\n",
       "      <td>KCNJ6</td>\n",
       "      <td>GABPB1-AS1</td>\n",
       "      <td>ITGBL1</td>\n",
       "      <td>GTF2H2</td>\n",
       "      <td>CAMK1D</td>\n",
       "      <td>ZNF7</td>\n",
       "      <td>DCBLD2</td>\n",
       "      <td>LPAR2</td>\n",
       "      <td>...</td>\n",
       "      <td>CIC</td>\n",
       "      <td>KCNJ6</td>\n",
       "      <td>VIPR2</td>\n",
       "      <td>RRAGB</td>\n",
       "      <td>YIPF4</td>\n",
       "      <td>MTMR4</td>\n",
       "      <td>PTPRN</td>\n",
       "      <td>APP</td>\n",
       "      <td>DNAJC22</td>\n",
       "      <td>SRRM3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PTTG1</td>\n",
       "      <td>CACNB4</td>\n",
       "      <td>N4BP2L2</td>\n",
       "      <td>DLG2</td>\n",
       "      <td>SGK494</td>\n",
       "      <td>BIRC6</td>\n",
       "      <td>RIMS2</td>\n",
       "      <td>NPIPB5</td>\n",
       "      <td>EHMT1</td>\n",
       "      <td>GTF2H2</td>\n",
       "      <td>...</td>\n",
       "      <td>KCNJ6</td>\n",
       "      <td>PRKCB</td>\n",
       "      <td>PIK3R3</td>\n",
       "      <td>MEF2C</td>\n",
       "      <td>DCBLD2</td>\n",
       "      <td>WLS</td>\n",
       "      <td>CCDC186</td>\n",
       "      <td>PAX2</td>\n",
       "      <td>WDR70</td>\n",
       "      <td>RAPGEF6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HMGB2</td>\n",
       "      <td>CHD6</td>\n",
       "      <td>DIP2C</td>\n",
       "      <td>CCNL1</td>\n",
       "      <td>MDK</td>\n",
       "      <td>YIPF4</td>\n",
       "      <td>CLVS2</td>\n",
       "      <td>CCNT2</td>\n",
       "      <td>ULK4</td>\n",
       "      <td>NPHP3</td>\n",
       "      <td>...</td>\n",
       "      <td>PPP6R2</td>\n",
       "      <td>MTHFD2</td>\n",
       "      <td>AFF3</td>\n",
       "      <td>NGFR</td>\n",
       "      <td>EBF3</td>\n",
       "      <td>AC021504.1</td>\n",
       "      <td>TH</td>\n",
       "      <td>DNAJC22</td>\n",
       "      <td>MTRNR2L8</td>\n",
       "      <td>XRRA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>TTYH1</td>\n",
       "      <td>SPATA6</td>\n",
       "      <td>HGSNAT</td>\n",
       "      <td>AC092683.1</td>\n",
       "      <td>TENT5A</td>\n",
       "      <td>FSTL5</td>\n",
       "      <td>RPF2</td>\n",
       "      <td>NFATC2IP</td>\n",
       "      <td>SNCA</td>\n",
       "      <td>MTHFD2</td>\n",
       "      <td>...</td>\n",
       "      <td>ZIC2</td>\n",
       "      <td>ZNF432</td>\n",
       "      <td>PTPRG</td>\n",
       "      <td>KCNC1</td>\n",
       "      <td>KSR1</td>\n",
       "      <td>ZNF385D</td>\n",
       "      <td>LNPK</td>\n",
       "      <td>CEP78</td>\n",
       "      <td>CRNDE</td>\n",
       "      <td>ANKRD45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>FAM184A</td>\n",
       "      <td>ADAR</td>\n",
       "      <td>FABP5</td>\n",
       "      <td>AL590705.1</td>\n",
       "      <td>PRKCB</td>\n",
       "      <td>CTNNB1</td>\n",
       "      <td>AL023284.4</td>\n",
       "      <td>NDST4</td>\n",
       "      <td>CALY</td>\n",
       "      <td>SCN2A</td>\n",
       "      <td>...</td>\n",
       "      <td>TTC7A</td>\n",
       "      <td>CASKIN1</td>\n",
       "      <td>UBE2QL1</td>\n",
       "      <td>DPY19L2</td>\n",
       "      <td>CCNL1</td>\n",
       "      <td>PCLAF</td>\n",
       "      <td>GTF2H2</td>\n",
       "      <td>LPAR2</td>\n",
       "      <td>PILRA</td>\n",
       "      <td>TTYH3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>PTN</td>\n",
       "      <td>ELF2</td>\n",
       "      <td>P4HA3</td>\n",
       "      <td>CCDC136</td>\n",
       "      <td>DCBLD2</td>\n",
       "      <td>WSB1</td>\n",
       "      <td>HGSNAT</td>\n",
       "      <td>STX18-AS1</td>\n",
       "      <td>SMC2</td>\n",
       "      <td>TTLL7</td>\n",
       "      <td>...</td>\n",
       "      <td>PILRA</td>\n",
       "      <td>MCF2L</td>\n",
       "      <td>STK4</td>\n",
       "      <td>ZGLP1</td>\n",
       "      <td>WLS</td>\n",
       "      <td>THSD7A</td>\n",
       "      <td>NEAT1</td>\n",
       "      <td>DHDDS</td>\n",
       "      <td>AL080317.2</td>\n",
       "      <td>CIC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CCNA2</td>\n",
       "      <td>CALD1</td>\n",
       "      <td>AC006480.2</td>\n",
       "      <td>MFSD14C</td>\n",
       "      <td>GLRA2</td>\n",
       "      <td>FAM184A</td>\n",
       "      <td>RFX1</td>\n",
       "      <td>AC142472.1</td>\n",
       "      <td>TAOK3</td>\n",
       "      <td>KMT2B</td>\n",
       "      <td>...</td>\n",
       "      <td>AL023284.4</td>\n",
       "      <td>AL450384.2</td>\n",
       "      <td>CCDC186</td>\n",
       "      <td>WDFY3</td>\n",
       "      <td>POU2F2</td>\n",
       "      <td>LRRN3</td>\n",
       "      <td>FAM184A</td>\n",
       "      <td>CEBPG</td>\n",
       "      <td>CYB5R4</td>\n",
       "      <td>CCND2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NFATC2IP</td>\n",
       "      <td>DIP2A</td>\n",
       "      <td>ADAR</td>\n",
       "      <td>PAXBP1</td>\n",
       "      <td>DNAJC22</td>\n",
       "      <td>CARHSP1</td>\n",
       "      <td>AGO1</td>\n",
       "      <td>PPIAL4G</td>\n",
       "      <td>CAMK2N1</td>\n",
       "      <td>DDX27</td>\n",
       "      <td>...</td>\n",
       "      <td>AFF3</td>\n",
       "      <td>CCDC186</td>\n",
       "      <td>DNAJC22</td>\n",
       "      <td>AC137932.3</td>\n",
       "      <td>PUM2</td>\n",
       "      <td>PCSK1N</td>\n",
       "      <td>VPS13A</td>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>LCORL</td>\n",
       "      <td>PEAK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CD99</td>\n",
       "      <td>DLG2</td>\n",
       "      <td>PDHX</td>\n",
       "      <td>TF</td>\n",
       "      <td>GTF2H2</td>\n",
       "      <td>AC093525.7</td>\n",
       "      <td>LRRC7</td>\n",
       "      <td>RPLP1</td>\n",
       "      <td>STAG1</td>\n",
       "      <td>ESF1</td>\n",
       "      <td>...</td>\n",
       "      <td>WDR60</td>\n",
       "      <td>RRAGB</td>\n",
       "      <td>JUND</td>\n",
       "      <td>PNISR</td>\n",
       "      <td>GPHN</td>\n",
       "      <td>CARHSP1</td>\n",
       "      <td>TMEM191C</td>\n",
       "      <td>ITM2B</td>\n",
       "      <td>IRX3</td>\n",
       "      <td>RBCK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>BIRC5</td>\n",
       "      <td>AL139424.2</td>\n",
       "      <td>PCNX2</td>\n",
       "      <td>GNB3</td>\n",
       "      <td>CNTN5</td>\n",
       "      <td>RIMS2</td>\n",
       "      <td>CTNNAL1</td>\n",
       "      <td>ZNF708</td>\n",
       "      <td>POU2F2</td>\n",
       "      <td>IRX3</td>\n",
       "      <td>...</td>\n",
       "      <td>GTF2H2</td>\n",
       "      <td>TAOK3</td>\n",
       "      <td>LSAMP</td>\n",
       "      <td>HCN1</td>\n",
       "      <td>LRRC1</td>\n",
       "      <td>MDK</td>\n",
       "      <td>ZNF397</td>\n",
       "      <td>SAT1</td>\n",
       "      <td>LAG3</td>\n",
       "      <td>STARD4-AS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>GNG5</td>\n",
       "      <td>TTC19</td>\n",
       "      <td>AC009133.1</td>\n",
       "      <td>DOCK5</td>\n",
       "      <td>FSTL5</td>\n",
       "      <td>MAGI3</td>\n",
       "      <td>AL450384.2</td>\n",
       "      <td>CEP152</td>\n",
       "      <td>ABCA5</td>\n",
       "      <td>SPRY3</td>\n",
       "      <td>...</td>\n",
       "      <td>GNB3</td>\n",
       "      <td>WDR66</td>\n",
       "      <td>HRK</td>\n",
       "      <td>POLE</td>\n",
       "      <td>UBE2QL1</td>\n",
       "      <td>OLFM3</td>\n",
       "      <td>AC087190.1</td>\n",
       "      <td>RPS6</td>\n",
       "      <td>PAX2</td>\n",
       "      <td>ZNF816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>HGSNAT</td>\n",
       "      <td>ZRANB2</td>\n",
       "      <td>IGFBP2</td>\n",
       "      <td>AL450384.2</td>\n",
       "      <td>SMC5</td>\n",
       "      <td>AATK</td>\n",
       "      <td>CCDC78</td>\n",
       "      <td>TSC2</td>\n",
       "      <td>LHX9</td>\n",
       "      <td>KSR1</td>\n",
       "      <td>...</td>\n",
       "      <td>ITGBL1</td>\n",
       "      <td>ULK4</td>\n",
       "      <td>GBA2</td>\n",
       "      <td>AFF1</td>\n",
       "      <td>TMEM259</td>\n",
       "      <td>MADD</td>\n",
       "      <td>FOXP1</td>\n",
       "      <td>LHX5-AS1</td>\n",
       "      <td>NHLRC3</td>\n",
       "      <td>MAST1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>CRNDE</td>\n",
       "      <td>CCZ1</td>\n",
       "      <td>GNB3</td>\n",
       "      <td>TNRC6C</td>\n",
       "      <td>UPP1</td>\n",
       "      <td>AC027031.2</td>\n",
       "      <td>BRINP3</td>\n",
       "      <td>AC024610.2</td>\n",
       "      <td>YIPF4</td>\n",
       "      <td>LCORL</td>\n",
       "      <td>...</td>\n",
       "      <td>CHTF18</td>\n",
       "      <td>HDAC7</td>\n",
       "      <td>ATP6V0A1</td>\n",
       "      <td>LYPD1</td>\n",
       "      <td>CALY</td>\n",
       "      <td>ZRANB2</td>\n",
       "      <td>ANKRD45</td>\n",
       "      <td>CCDC186</td>\n",
       "      <td>LMBR1L</td>\n",
       "      <td>SEMA3D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NAIP</td>\n",
       "      <td>OSBPL10</td>\n",
       "      <td>IRX3</td>\n",
       "      <td>ZDHHC14</td>\n",
       "      <td>AL023284.4</td>\n",
       "      <td>ZIC2</td>\n",
       "      <td>SYNJ2</td>\n",
       "      <td>WLS</td>\n",
       "      <td>AAK1</td>\n",
       "      <td>SLITRK5</td>\n",
       "      <td>...</td>\n",
       "      <td>DCC</td>\n",
       "      <td>TNRC6C</td>\n",
       "      <td>ZIC2</td>\n",
       "      <td>SUN1</td>\n",
       "      <td>TMEM39A</td>\n",
       "      <td>WDR70</td>\n",
       "      <td>S100A10</td>\n",
       "      <td>PXDN</td>\n",
       "      <td>SNX22</td>\n",
       "      <td>LPAR2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ZDHHC11</td>\n",
       "      <td>ABCA5</td>\n",
       "      <td>WDR70</td>\n",
       "      <td>INTS6</td>\n",
       "      <td>MYO19</td>\n",
       "      <td>TSC2</td>\n",
       "      <td>ERAP2</td>\n",
       "      <td>HGSNAT</td>\n",
       "      <td>DAB1</td>\n",
       "      <td>NEBL</td>\n",
       "      <td>...</td>\n",
       "      <td>PTPRK</td>\n",
       "      <td>HRK</td>\n",
       "      <td>ELAVL4</td>\n",
       "      <td>DBNL</td>\n",
       "      <td>RANGAP1</td>\n",
       "      <td>LYPD1</td>\n",
       "      <td>PPFIBP1</td>\n",
       "      <td>XRRA1</td>\n",
       "      <td>FEZF1-AS1</td>\n",
       "      <td>ULK4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ZNF816</td>\n",
       "      <td>RPF2</td>\n",
       "      <td>UCN</td>\n",
       "      <td>ITGBL1</td>\n",
       "      <td>TLE1</td>\n",
       "      <td>ENDOV</td>\n",
       "      <td>TH</td>\n",
       "      <td>AC009133.1</td>\n",
       "      <td>THSD7A</td>\n",
       "      <td>EGFL8</td>\n",
       "      <td>...</td>\n",
       "      <td>SAMD5</td>\n",
       "      <td>SGPP2</td>\n",
       "      <td>CARHSP1</td>\n",
       "      <td>PACS2</td>\n",
       "      <td>AL590705.1</td>\n",
       "      <td>AL133492.1</td>\n",
       "      <td>AL023284.4</td>\n",
       "      <td>PUM2</td>\n",
       "      <td>ERBB4</td>\n",
       "      <td>ZMYM2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>LAMB1</td>\n",
       "      <td>PDE1B</td>\n",
       "      <td>TTC14</td>\n",
       "      <td>RFX3</td>\n",
       "      <td>GALR1</td>\n",
       "      <td>AC005224.3</td>\n",
       "      <td>VARS2</td>\n",
       "      <td>CCDC78</td>\n",
       "      <td>AL138899.1</td>\n",
       "      <td>ATP8A1</td>\n",
       "      <td>...</td>\n",
       "      <td>TTYH3</td>\n",
       "      <td>TMEM39A</td>\n",
       "      <td>ZNF638</td>\n",
       "      <td>TRIM65</td>\n",
       "      <td>TIAM2</td>\n",
       "      <td>UBAP2</td>\n",
       "      <td>SNCG</td>\n",
       "      <td>PEAK1</td>\n",
       "      <td>CCDC78</td>\n",
       "      <td>SORBS2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20 rows × 184 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   ENSG00000137845 ENSG00000185736 ENSG00000141385 ENSG00000156709  \\\n",
       "0            NR2F1           UNC80          NHLRC3           APBB2   \n",
       "1           IGFBP2           WDR70          GTF2H2          CEP290   \n",
       "2          N4BP2L2           CLVS2           SAMD5        KCNQ1OT1   \n",
       "3           PPP4R4      STARD4-AS1           KCNJ6      GABPB1-AS1   \n",
       "4            PTTG1          CACNB4         N4BP2L2            DLG2   \n",
       "5            HMGB2            CHD6           DIP2C           CCNL1   \n",
       "6            TTYH1          SPATA6          HGSNAT      AC092683.1   \n",
       "7          FAM184A            ADAR           FABP5      AL590705.1   \n",
       "8              PTN            ELF2           P4HA3         CCDC136   \n",
       "9            CCNA2           CALD1      AC006480.2         MFSD14C   \n",
       "10        NFATC2IP           DIP2A            ADAR          PAXBP1   \n",
       "11            CD99            DLG2            PDHX              TF   \n",
       "12           BIRC5      AL139424.2           PCNX2            GNB3   \n",
       "13            GNG5           TTC19      AC009133.1           DOCK5   \n",
       "14          HGSNAT          ZRANB2          IGFBP2      AL450384.2   \n",
       "15           CRNDE            CCZ1            GNB3          TNRC6C   \n",
       "16            NAIP         OSBPL10            IRX3         ZDHHC14   \n",
       "17         ZDHHC11           ABCA5           WDR70           INTS6   \n",
       "18          ZNF816            RPF2             UCN          ITGBL1   \n",
       "19           LAMB1           PDE1B           TTC14            RFX3   \n",
       "\n",
       "   ENSG00000127914 ENSG00000159461 ENSG00000160746 ENSG00000183020  \\\n",
       "0          FAM184A          NHLRC3          GTF2H2            LNPK   \n",
       "1           BRINP3         CASKIN1          KLHL34            GNB3   \n",
       "2            LUZP2           PHKG1          SPIN2A           BRSK2   \n",
       "3           ITGBL1          GTF2H2          CAMK1D            ZNF7   \n",
       "4           SGK494           BIRC6           RIMS2          NPIPB5   \n",
       "5              MDK           YIPF4           CLVS2           CCNT2   \n",
       "6           TENT5A           FSTL5            RPF2        NFATC2IP   \n",
       "7            PRKCB          CTNNB1      AL023284.4           NDST4   \n",
       "8           DCBLD2            WSB1          HGSNAT       STX18-AS1   \n",
       "9            GLRA2         FAM184A            RFX1      AC142472.1   \n",
       "10         DNAJC22         CARHSP1            AGO1         PPIAL4G   \n",
       "11          GTF2H2      AC093525.7           LRRC7           RPLP1   \n",
       "12           CNTN5           RIMS2         CTNNAL1          ZNF708   \n",
       "13           FSTL5           MAGI3      AL450384.2          CEP152   \n",
       "14            SMC5            AATK          CCDC78            TSC2   \n",
       "15            UPP1      AC027031.2          BRINP3      AC024610.2   \n",
       "16      AL023284.4            ZIC2           SYNJ2             WLS   \n",
       "17           MYO19            TSC2           ERAP2          HGSNAT   \n",
       "18            TLE1           ENDOV              TH      AC009133.1   \n",
       "19           GALR1      AC005224.3           VARS2          CCDC78   \n",
       "\n",
       "   ENSG00000100823 ENSG00000049618  ... ENSG00000143952 ENSG00000196998  \\\n",
       "0           PPFIA2           SGIP1  ...             ID2           KMT2B   \n",
       "1           ZNF528           ERBB4  ...          NHLRC3            GSE1   \n",
       "2          UBE2QL1          ANAPC7  ...      AC008522.1            NNAT   \n",
       "3           DCBLD2           LPAR2  ...             CIC           KCNJ6   \n",
       "4            EHMT1          GTF2H2  ...           KCNJ6           PRKCB   \n",
       "5             ULK4           NPHP3  ...          PPP6R2          MTHFD2   \n",
       "6             SNCA          MTHFD2  ...            ZIC2          ZNF432   \n",
       "7             CALY           SCN2A  ...           TTC7A         CASKIN1   \n",
       "8             SMC2           TTLL7  ...           PILRA           MCF2L   \n",
       "9            TAOK3           KMT2B  ...      AL023284.4      AL450384.2   \n",
       "10         CAMK2N1           DDX27  ...            AFF3         CCDC186   \n",
       "11           STAG1            ESF1  ...           WDR60           RRAGB   \n",
       "12          POU2F2            IRX3  ...          GTF2H2           TAOK3   \n",
       "13           ABCA5           SPRY3  ...            GNB3           WDR66   \n",
       "14            LHX9            KSR1  ...          ITGBL1            ULK4   \n",
       "15           YIPF4           LCORL  ...          CHTF18           HDAC7   \n",
       "16            AAK1         SLITRK5  ...             DCC          TNRC6C   \n",
       "17            DAB1            NEBL  ...           PTPRK             HRK   \n",
       "18          THSD7A           EGFL8  ...           SAMD5           SGPP2   \n",
       "19      AL138899.1          ATP8A1  ...           TTYH3         TMEM39A   \n",
       "\n",
       "   ENSG00000158023 ENSG00000186153 ENSG00000101966 ENSG00000073050  \\\n",
       "0            RPL10            RPF2            SYT4           RRAGB   \n",
       "1            RPS27          SPAG16           KITLG          KLHL29   \n",
       "2           ZNF397         UBE2QL1          PPFIA2         UBE2QL1   \n",
       "3            VIPR2           RRAGB           YIPF4           MTMR4   \n",
       "4           PIK3R3           MEF2C          DCBLD2             WLS   \n",
       "5             AFF3            NGFR            EBF3      AC021504.1   \n",
       "6            PTPRG           KCNC1            KSR1         ZNF385D   \n",
       "7          UBE2QL1         DPY19L2           CCNL1           PCLAF   \n",
       "8             STK4           ZGLP1             WLS          THSD7A   \n",
       "9          CCDC186           WDFY3          POU2F2           LRRN3   \n",
       "10         DNAJC22      AC137932.3            PUM2          PCSK1N   \n",
       "11            JUND           PNISR            GPHN         CARHSP1   \n",
       "12           LSAMP            HCN1           LRRC1             MDK   \n",
       "13             HRK            POLE         UBE2QL1           OLFM3   \n",
       "14            GBA2            AFF1         TMEM259            MADD   \n",
       "15        ATP6V0A1           LYPD1            CALY          ZRANB2   \n",
       "16            ZIC2            SUN1         TMEM39A           WDR70   \n",
       "17          ELAVL4            DBNL         RANGAP1           LYPD1   \n",
       "18         CARHSP1           PACS2      AL590705.1      AL133492.1   \n",
       "19          ZNF638          TRIM65           TIAM2           UBAP2   \n",
       "\n",
       "   ENSG00000134684 ENSG00000188994 ENSG00000166716 ENSG00000153896  \n",
       "0            LCORL            SYT1           MTMR4            IRX3  \n",
       "1             ESF1          CRABP2          TNRC6C            DMWD  \n",
       "2            TANC2          UNC13A            ECM1         CARHSP1  \n",
       "3            PTPRN             APP         DNAJC22           SRRM3  \n",
       "4          CCDC186            PAX2           WDR70         RAPGEF6  \n",
       "5               TH         DNAJC22        MTRNR2L8           XRRA1  \n",
       "6             LNPK           CEP78           CRNDE         ANKRD45  \n",
       "7           GTF2H2           LPAR2           PILRA           TTYH3  \n",
       "8            NEAT1           DHDDS      AL080317.2             CIC  \n",
       "9          FAM184A           CEBPG          CYB5R4           CCND2  \n",
       "10          VPS13A          MT-ND6           LCORL           PEAK1  \n",
       "11        TMEM191C           ITM2B            IRX3           RBCK1  \n",
       "12          ZNF397            SAT1            LAG3      STARD4-AS1  \n",
       "13      AC087190.1            RPS6            PAX2          ZNF816  \n",
       "14           FOXP1        LHX5-AS1          NHLRC3           MAST1  \n",
       "15         ANKRD45         CCDC186          LMBR1L          SEMA3D  \n",
       "16         S100A10            PXDN           SNX22           LPAR2  \n",
       "17         PPFIBP1           XRRA1       FEZF1-AS1            ULK4  \n",
       "18      AL023284.4            PUM2           ERBB4           ZMYM2  \n",
       "19            SNCG           PEAK1          CCDC78          SORBS2  \n",
       "\n",
       "[20 rows x 184 columns]"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Number of top genes to consider\n",
    "n_top_genes = 20\n",
    "\n",
    "# Extract the gene names for each group\n",
    "rank_genes = pd.DataFrame(adata.uns['rank_genes_groups']['names']).head(n_top_genes)\n",
    "rank_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENSG00000137845\n",
      "(181, 20)\n",
      "ENSG00000185736\n",
      "(137, 20)\n",
      "ENSG00000141385\n",
      "(181, 20)\n",
      "ENSG00000156709\n",
      "(145, 20)\n",
      "ENSG00000127914\n",
      "(165, 20)\n",
      "ENSG00000159461\n",
      "(120, 20)\n",
      "ENSG00000160746\n",
      "(226, 20)\n",
      "ENSG00000183020\n",
      "(239, 20)\n",
      "ENSG00000100823\n",
      "(161, 20)\n",
      "ENSG00000049618\n",
      "(160, 20)\n",
      "ENSG00000057663\n",
      "(119, 20)\n",
      "ENSG00000105409\n",
      "(205, 20)\n",
      "ENSG00000165629\n",
      "(111, 20)\n",
      "ENSG00000167863\n",
      "(152, 20)\n",
      "ENSG00000182220\n",
      "(204, 20)\n",
      "ENSG00000124788\n",
      "(188, 20)\n",
      "ENSG00000163635\n",
      "(153, 20)\n",
      "ENSG00000130638\n",
      "(140, 20)\n",
      "ENSG00000087088\n",
      "(178, 20)\n",
      "ENSG00000074582\n",
      "(149, 20)\n",
      "ENSG00000126581\n",
      "(234, 20)\n",
      "ENSG00000136717\n",
      "(145, 20)\n",
      "ENSG00000108578\n",
      "(109, 20)\n",
      "ENSG00000113734\n",
      "(168, 20)\n",
      "ENSG00000104327\n",
      "(152, 20)\n",
      "ENSG00000164305\n",
      "(244, 20)\n",
      "ENSG00000163171\n",
      "(143, 20)\n",
      "ENSG00000140937\n",
      "(125, 20)\n",
      "ENSG00000149187\n",
      "(149, 20)\n",
      "ENSG00000182923\n",
      "(232, 20)\n",
      "ENSG00000100075\n",
      "(197, 20)\n",
      "ENSG00000174469\n",
      "(137, 20)\n",
      "ENSG00000068120\n",
      "(180, 20)\n",
      "ENSG00000103051\n",
      "(169, 20)\n",
      "ENSG00000006695\n",
      "(203, 20)\n",
      "ENSG00000014919\n",
      "(154, 20)\n",
      "ENSG00000005339\n",
      "(324, 20)\n",
      "ENSG00000117984\n",
      "(120, 20)\n",
      "ENSG00000008283\n",
      "(279, 20)\n",
      "ENSG00000172115\n",
      "(99, 20)\n",
      "ENSG00000036530\n",
      "(133, 20)\n",
      "ENSG00000112977\n",
      "(162, 20)\n",
      "ENSG00000138231\n",
      "(138, 20)\n",
      "ENSG00000204843\n",
      "(144, 20)\n",
      "ENSG00000172869\n",
      "(248, 20)\n",
      "ENSG00000116675\n",
      "(362, 20)\n",
      "ENSG00000079805\n",
      "(203, 20)\n",
      "ENSG00000113360\n",
      "(139, 20)\n",
      "ENSG00000197102\n",
      "(171, 20)\n",
      "ENSG00000127884\n",
      "(181, 20)\n",
      "ENSG00000167658\n",
      "(194, 20)\n",
      "ENSG00000134001\n",
      "(111, 20)\n",
      "ENSG00000114867\n",
      "(138, 20)\n",
      "ENSG00000155849\n",
      "(195, 20)\n",
      "ENSG00000012660\n",
      "(195, 20)\n",
      "ENSG00000134014\n",
      "(148, 20)\n",
      "ENSG00000135999\n",
      "(155, 20)\n",
      "ENSG00000182944\n",
      "(141, 20)\n",
      "ENSG00000131558\n",
      "(139, 20)\n",
      "ENSG00000149926\n",
      "(175, 20)\n",
      "ENSG00000161682\n",
      "(152, 20)\n",
      "ENSG00000152767\n",
      "(211, 20)\n",
      "ENSG00000100225\n",
      "(193, 20)\n",
      "ENSG00000073712\n",
      "(169, 20)\n",
      "ENSG00000139132\n",
      "(117, 20)\n",
      "ENSG00000054598\n",
      "(222, 20)\n",
      "ENSG00000110074\n",
      "(195, 20)\n",
      "ENSG00000151474\n",
      "(241, 20)\n",
      "ENSG00000285130\n",
      "(216, 20)\n",
      "ENSG00000106105\n",
      "(207, 20)\n",
      "ENSG00000119392\n",
      "(171, 20)\n",
      "ENSG00000173905\n",
      "(182, 20)\n",
      "ENSG00000030582\n",
      "(205, 20)\n",
      "ENSG00000104687\n",
      "(177, 20)\n",
      "ENSG00000170899\n",
      "(114, 20)\n",
      "ENSG00000111358\n",
      "(217, 20)\n",
      "ENSG00000213614\n",
      "(229, 20)\n",
      "ENSG00000122566\n",
      "(137, 20)\n",
      "ENSG00000044574\n",
      "(81, 20)\n",
      "ENSG00000113013\n",
      "(167, 20)\n",
      "ENSG00000108576\n",
      "(277, 20)\n",
      "ENSG00000149428\n",
      "(187, 20)\n",
      "ENSG00000140443\n",
      "(178, 20)\n",
      "ENSG00000054523\n",
      "(73, 20)\n",
      "ENSG00000075945\n",
      "(152, 20)\n",
      "ENSG00000138095\n",
      "(129, 20)\n",
      "ENSG00000186868\n",
      "(248, 20)\n",
      "ENSG00000015479\n",
      "(235, 20)\n",
      "ENSG00000159111\n",
      "(210, 20)\n",
      "ENSG00000182979\n",
      "(149, 20)\n",
      "ENSG00000162736\n",
      "(108, 20)\n",
      "ENSG00000131495\n",
      "(107, 20)\n",
      "ENSG00000139180\n",
      "(195, 20)\n",
      "ENSG00000164182\n",
      "(159, 20)\n",
      "ENSG00000101247\n",
      "(124, 20)\n",
      "ENSG00000156170\n",
      "(160, 20)\n",
      "ENSG00000110717\n",
      "(152, 20)\n",
      "ENSG00000167792\n",
      "(256, 20)\n",
      "ENSG00000073969\n",
      "(180, 20)\n",
      "ENSG00000148053\n",
      "(178, 20)\n",
      "ENSG00000125741\n",
      "(130, 20)\n",
      "ENSG00000123240\n",
      "(203, 20)\n",
      "ENSG00000079156\n",
      "(126, 20)\n",
      "ENSG00000143799\n",
      "(214, 20)\n",
      "ENSG00000157212\n",
      "(305, 20)\n",
      "ENSG00000156453\n",
      "(137, 20)\n",
      "ENSG00000083642\n",
      "(148, 20)\n",
      "ENSG00000113068\n",
      "(131, 20)\n",
      "ENSG00000108518\n",
      "(142, 20)\n",
      "ENSG00000137338\n",
      "(167, 20)\n",
      "ENSG00000105223\n",
      "(133, 20)\n",
      "ENSG00000171680\n",
      "(273, 20)\n",
      "ENSG00000165688\n",
      "(197, 20)\n",
      "ENSG00000140521\n",
      "(159, 20)\n",
      "ENSG00000105854\n",
      "(219, 20)\n",
      "ENSG00000152192\n",
      "(167, 20)\n",
      "ENSG00000104866\n",
      "(167, 20)\n",
      "ENSG00000156475\n",
      "(264, 20)\n",
      "ENSG00000106617\n",
      "(119, 20)\n",
      "ENSG00000171867\n",
      "(137, 20)\n",
      "ENSG00000117523\n",
      "(161, 20)\n",
      "ENSG00000197746\n",
      "(206, 20)\n",
      "ENSG00000125818\n",
      "(175, 20)\n",
      "ENSG00000073008\n",
      "(138, 20)\n",
      "ENSG00000075785\n",
      "(266, 20)\n",
      "ENSG00000084733\n",
      "(128, 20)\n",
      "ENSG00000134453\n",
      "(178, 20)\n",
      "ENSG00000132563\n",
      "(72, 20)\n",
      "ENSG00000189056\n",
      "(166, 20)\n",
      "ENSG00000133119\n",
      "(143, 20)\n",
      "ENSG00000143344\n",
      "(173, 20)\n",
      "ENSG00000116741\n",
      "(182, 20)\n",
      "ENSG00000126858\n",
      "(175, 20)\n",
      "ENSG00000060709\n",
      "(128, 20)\n",
      "ENSG00000141622\n",
      "(116, 20)\n",
      "ENSG00000089009\n",
      "(223, 20)\n",
      "ENSG00000188846\n",
      "(134, 20)\n",
      "ENSG00000004139\n",
      "(134, 20)\n",
      "ENSG00000111961\n",
      "(233, 20)\n",
      "ENSG00000140386\n",
      "(159, 20)\n",
      "ENSG00000092108\n",
      "(142, 20)\n",
      "ENSG00000284194\n",
      "(152, 20)\n",
      "ENSG00000138802\n",
      "(152, 20)\n",
      "ENSG00000122335\n",
      "(190, 20)\n",
      "ENSG00000107290\n",
      "(127, 20)\n",
      "ENSG00000154447\n",
      "(85, 20)\n",
      "ENSG00000163798\n",
      "(171, 20)\n",
      "ENSG00000163681\n",
      "(180, 20)\n",
      "ENSG00000074317\n",
      "(297, 20)\n",
      "ENSG00000142168\n",
      "(149, 20)\n",
      "ENSG00000291237\n",
      "(161, 20)\n",
      "ENSG00000197912\n",
      "(184, 20)\n",
      "ENSG00000073849\n",
      "(139, 20)\n",
      "ENSG00000099365\n",
      "(192, 20)\n",
      "ENSG00000159082\n",
      "(230, 20)\n",
      "ENSG00000147133\n",
      "(375, 20)\n",
      "ENSG00000120948\n",
      "(45, 20)\n",
      "ENSG00000111802\n",
      "(183, 20)\n",
      "ENSG00000108064\n",
      "(182, 20)\n",
      "ENSG00000131931\n",
      "(143, 20)\n",
      "ENSG00000155099\n",
      "(177, 20)\n",
      "ENSG00000198900\n",
      "(253, 20)\n",
      "ENSG00000126602\n",
      "(175, 20)\n",
      "ENSG00000234127\n",
      "(164, 20)\n",
      "ENSG00000103671\n",
      "(147, 20)\n",
      "ENSG00000092439\n",
      "(161, 20)\n",
      "ENSG00000104833\n",
      "(142, 20)\n",
      "ENSG00000081307\n",
      "(304, 20)\n",
      "ENSG00000168246\n",
      "(192, 20)\n",
      "ENSG00000124164\n",
      "(157, 20)\n",
      "ENSG00000165280\n",
      "(128, 20)\n",
      "ENSG00000197969\n",
      "(160, 20)\n",
      "ENSG00000129003\n",
      "(187, 20)\n",
      "ENSG00000069329\n",
      "(135, 20)\n",
      "ENSG00000143952\n",
      "(214, 20)\n",
      "ENSG00000196998\n",
      "(208, 20)\n",
      "ENSG00000158023\n",
      "(147, 20)\n",
      "ENSG00000186153\n",
      "(93, 20)\n",
      "ENSG00000101966\n",
      "(158, 20)\n",
      "ENSG00000073050\n",
      "(200, 20)\n",
      "ENSG00000134684\n",
      "(160, 20)\n",
      "ENSG00000188994\n",
      "(127, 20)\n",
      "ENSG00000166716\n",
      "(148, 20)\n",
      "ENSG00000153896\n",
      "(210, 20)\n"
     ]
    }
   ],
   "source": [
    "perturbation_scores = {}\n",
    "\n",
    "for group in rank_genes.columns:\n",
    "    # Get the top marker genes for this group\n",
    "    marker_genes = rank_genes[group].dropna().values\n",
    "    marker_genes = [gene for gene in marker_genes if gene in adata.var_names]\n",
    "    print(group)\n",
    "    \n",
    "    if not marker_genes:\n",
    "        perturbation_scores[group] = np.nan\n",
    "        continue\n",
    "\n",
    "    # Get perturbed and control cells\n",
    "    perturbed_cells = adata.obs['gene_list'] == group\n",
    "    control_cells = adata.obs['gene_list'] == 'control'\n",
    "\n",
    "    # Extract expression\n",
    "    perturbed_expr = adata[perturbed_cells, marker_genes].X\n",
    "    control_expr = adata[control_cells, marker_genes].X\n",
    "\n",
    "    print(perturbed_expr.shape)\n",
    "\n",
    "    # Convert if sparse\n",
    "    if not isinstance(perturbed_expr, np.ndarray):\n",
    "        perturbed_expr = perturbed_expr.toarray()\n",
    "    if not isinstance(control_expr, np.ndarray):\n",
    "        control_expr = control_expr.toarray()\n",
    "\n",
    "    # Mean expression across marker genes and cells\n",
    "    perturbed_mean = np.mean(perturbed_expr, axis=0)\n",
    "    control_mean = np.mean(control_expr, axis=0)\n",
    "\n",
    "    # Calculate perturbation score: e.g., difference or log2 fold change\n",
    "    #score = float(np.abs(perturbed_mean - control_mean).mean())\n",
    "\n",
    "    #score = np.mean([wasserstein_distance(perturbed_expr[:, i], control_expr[:, i]) for i in range(perturbed_expr.shape[1])])\n",
    "\n",
    "    gene_std = np.std(control_expr, axis=0) + 1e-6  # avoid div by zero\n",
    "    z_scores = (perturbed_mean - control_mean) / gene_std\n",
    "    score = np.abs(z_scores).mean()\n",
    "\n",
    "    # score = np.log2((perturbed_mean + 1e-6) / (control_mean + 1e-6))  # alternative\n",
    "    perturbation_scores[group] = score\n",
    "    perturbation_scores = pd.Series(perturbation_scores)\n",
    "    perturbation_scores.name = 'perturbation_scores'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([1., 0., 1., 2., 1., 1., 1., 0., 2., 4., 4., 4., 1., 6., 7., 9., 5.,\n",
       "        2., 4., 8., 8., 9., 6., 2., 8., 2., 2., 4., 6., 9., 4., 5., 3., 5.,\n",
       "        4., 4., 4., 3., 2., 1., 1., 3., 0., 2., 4., 1., 1., 0., 0., 1., 1.,\n",
       "        2., 2., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 2., 0., 1., 0.,\n",
       "        0., 0., 1., 0., 0., 0., 0., 0., 2., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1.]),\n",
       " array([0.21883003, 0.22102354, 0.22321707, 0.22541058, 0.22760409,\n",
       "        0.22979762, 0.23199113, 0.23418464, 0.23637816, 0.23857167,\n",
       "        0.24076518, 0.24295869, 0.24515222, 0.24734573, 0.24953924,\n",
       "        0.25173277, 0.25392628, 0.25611979, 0.2583133 , 0.26050681,\n",
       "        0.26270035, 0.26489386, 0.26708737, 0.26928088, 0.27147439,\n",
       "        0.2736679 , 0.27586144, 0.27805495, 0.28024846, 0.28244197,\n",
       "        0.28463548, 0.28682899, 0.28902251, 0.29121605, 0.29340956,\n",
       "        0.29560307, 0.29779658, 0.29999009, 0.30218363, 0.30437714,\n",
       "        0.30657065, 0.30876416, 0.31095767, 0.31315118, 0.31534469,\n",
       "        0.3175382 , 0.31973174, 0.32192525, 0.32411876, 0.32631227,\n",
       "        0.32850578, 0.33069932, 0.33289284, 0.33508635, 0.33727986,\n",
       "        0.33947337, 0.34166688, 0.34386039, 0.3460539 , 0.34824741,\n",
       "        0.35044095, 0.35263446, 0.35482797, 0.35702151, 0.35921502,\n",
       "        0.36140853, 0.36360204, 0.36579555, 0.36798906, 0.37018257,\n",
       "        0.37237608, 0.37456959, 0.37676311, 0.37895665, 0.38115016,\n",
       "        0.38334367, 0.38553721, 0.38773072, 0.38992423, 0.39211774,\n",
       "        0.39431125, 0.39650476, 0.39869827, 0.40089178, 0.40308529,\n",
       "        0.40527883, 0.40747234, 0.40966585, 0.41185939, 0.4140529 ,\n",
       "        0.41624641, 0.41843992, 0.42063344, 0.42282695, 0.42502046,\n",
       "        0.42721397, 0.42940748, 0.43160099, 0.43379453, 0.43598804,\n",
       "        0.43818155]),\n",
       " <BarContainer object of 100 artists>)"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAhYAAAGdCAYAAABO2DpVAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjEsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvc2/+5QAAAAlwSFlzAAAPYQAAD2EBqD+naQAAFRVJREFUeJzt3XuQVnX9wPEPsvBwGRaTxGBAwClFwBKBkrDsYlhBU/2Rl9SIdCYmNIiZxmW0ABtYmCljvFEwhZSBjGOMzFAq04w3kJJbWTpiFLKjEtPFXdBpG+H8/mjYX8tF91k+z/Ls+nrNnD+es+fyffjugbfn2fV0K4qiCACABKed6gEAAF2HsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0tR09AkPHz4cr7zySvTr1y+6devW0acHANqhKIo4cOBADB48OE477cT3JTo8LF555ZUYOnRoR58WAEjQ0NAQQ4YMOeHXOzws+vXrFxH/HVhtbW1Hnx4AaIempqYYOnRoy7/jJ9LhYXHk44/a2lphAQCdzNv9GIMf3gQA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACBNhz82nY4zvG5Dq9d7Fk/pEudqi2obD8A7hTsWAEAaYQEApBEWAEAaYQEApBEWAEAaYQEApBEWAEAaYQEApBEWAEAaYQEApBEWAEAaYQEApBEWAEAaYQEApBEWAEAaYQEApBEWAEAaYQEApBEWAEAaYQEApBEWAEAaYQEApBEWAEAaYQEApBEWAEAaYQEApBEWAEAaYQEApBEWAEAaYQEApBEWAEAaYQEApBEWAEAaYQEApCkrLN5888249dZbY8SIEdG7d+8455xz4rbbbovDhw9XanwAQCdSU87GS5YsiR/96EexatWqGD16dGzdujWmT58e/fv3j1mzZlVqjABAJ1FWWDz99NPx+c9/PqZMmRIREcOHD481a9bE1q1bKzI4AKBzKeujkEsuuSR+85vfxK5duyIi4ve//3089dRT8dnPfvaE+zQ3N0dTU1OrBQDomsq6Y3HzzTdHY2NjjBw5Mrp37x6HDh2KhQsXxtVXX33Cferr62PBggUnPVBOneF1GzrsuHsWT6nIuQDoGGXdsVi7dm3cd999sXr16ti+fXusWrUqvv/978eqVatOuM/cuXOjsbGxZWloaDjpQQMA1amsOxbf/va3o66uLq666qqIiLjgggvipZdeivr6+pg2bdpx9ymVSlEqlU5+pABA1SvrjsUbb7wRp53Wepfu3bv7dVMAICLKvGPxuc99LhYuXBhnn312jB49Onbs2BG33357fO1rX6vU+ACATqSssLjzzjvjO9/5TnzjG9+I/fv3x+DBg+PrX/96fPe7363U+ACATqSssOjXr18sXbo0li5dWqHhAACdmWeFAABphAUAkEZYAABphAUAkEZYAABphAUAkEZYAABphAUAkEZYAABphAUAkEZYAABphAUAkEZYAABphAUAkEZYAABphAUAkEZYAABphAUAkEZYAABphAUAkEZYAABphAUAkEZYAABphAUAkEZYAABphAUAkEZYAABphAUAkEZYAABphAUAkEZYAABphAUAkEZYAABpak71AMgxvG7DqR4CALhjAQDkERYAQBphAQCkERYAQBphAQCkERYAQBphAQCkERYAQBphAQCkERYAQBphAQCkERYAQBphAQCkERYAQBphAQCkERYAQBphAQCkERYAQBphAQCkERYAQBphAQCkERYAQBphAQCkERYAQBphAQCkERYAQBphAQCkERYAQBphAQCkERYAQBphAQCkERYAQBphAQCkKTssXn755bj22mtjwIAB0adPn7jwwgtj27ZtlRgbANDJ1JSz8b/+9a+YNGlSfPzjH49f//rXMXDgwNi9e3ecfvrpFRoeANCZlBUWS5YsiaFDh8bKlStb1g0fPjx7TABAJ1XWRyHr16+P8ePHx5e+9KUYOHBgjB07NlasWFGpsQEAnUxZYfGXv/wlli1bFu973/vikUceiRkzZsQ3v/nN+NnPfnbCfZqbm6OpqanVAgB0TWV9FHL48OEYP358LFq0KCIixo4dG3/6059i2bJl8ZWvfOW4+9TX18eCBQtOfqS0Mrxuw6kewls63vj2LJ7SYecC4NQo647FoEGDYtSoUa3WnX/++bF3794T7jN37txobGxsWRoaGto3UgCg6pV1x2LSpEnxwgsvtFq3a9euGDZs2An3KZVKUSqV2jc6AKBTKeuOxbe+9a3YsmVLLFq0KP785z/H6tWrY/ny5TFz5sxKjQ8A6ETKCosJEybEunXrYs2aNTFmzJj43ve+F0uXLo1rrrmmUuMDADqRsj4KiYiYOnVqTJ06tRJjAQA6Oc8KAQDSCAsAII2wAADSCAsAII2wAADSCAsAII2wAADSCAsAII2wAADSCAsAII2wAADSCAsAII2wAADSCAsAII2wAADSCAsAII2wAADSCAsAII2wAADSCAsAII2wAADSCAsAII2wAADSCAsAII2wAADSCAsAII2wAADSCAsAII2wAADSCAsAII2wAADSCAsAII2wAADS1JzqAdBxhtdtOGbdnsVTTsFIqsPRfx7H+7NoyzYA/D93LACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEgjLACANMICAEhzUmFRX18f3bp1i9mzZycNBwDozNodFs8880wsX7483v/+92eOBwDoxNoVFgcPHoxrrrkmVqxYEe9617uyxwQAdFLtCouZM2fGlClT4rLLLnvbbZubm6OpqanVAgB0TTXl7nD//ffH9u3b45lnnmnT9vX19bFgwYKyB9ZVDa/bUPY+exZPqcBIjq894+to7Rlj1vtq73E6cg6P5+hxn+rxAF1XWXcsGhoaYtasWXHfffdFr1692rTP3Llzo7GxsWVpaGho10ABgOpX1h2Lbdu2xf79+2PcuHEt6w4dOhRPPPFE3HXXXdHc3Bzdu3dvtU+pVIpSqZQzWgCgqpUVFp/85Cfj2WefbbVu+vTpMXLkyLj55puPiQoA4J2lrLDo169fjBkzptW6vn37xoABA45ZDwC88/g/bwIAacr+rZCjPfbYYwnDAAC6AncsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASFNzqgcA1WJ43YYOO/aexVNSjnMyxwKoBHcsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASCMsAIA0wgIASFNWWNTX18eECROiX79+MXDgwPjCF74QL7zwQqXGBgB0MmWFxeOPPx4zZ86MLVu2xMaNG+PNN9+MyZMnx+uvv16p8QEAnUhNORs//PDDrV6vXLkyBg4cGNu2bYuPfvSjqQMDADqfssLiaI2NjRERccYZZ5xwm+bm5mhubm553dTUdDKnBACqWLvDoiiKmDNnTlxyySUxZsyYE25XX18fCxYsaO9pOsTwug2tXu9ZPKUix61GHTnGtpyrM/yZZTje++yq33dtGU/Wez/VKvV3yanWVd8XldHu3wq58cYb4w9/+EOsWbPmLbebO3duNDY2tiwNDQ3tPSUAUOXadcfipptuivXr18cTTzwRQ4YMecttS6VSlEqldg0OAOhcygqLoijipptuinXr1sVjjz0WI0aMqNS4AIBOqKywmDlzZqxevToeeuih6NevX+zbty8iIvr37x+9e/euyAABgM6jrJ+xWLZsWTQ2NsbHPvaxGDRoUMuydu3aSo0PAOhEyv4oBADgRDwrBABIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDQ1p3oAmYbXbThm3Z7FU952m7Ycpy2OPleW9o6HzqUj57kt52rL9/OpHnOlrrlKOvp9tPc9ZB2Hzq/avhfcsQAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACCNsAAA0ggLACBNu8LinnvuiREjRkSvXr1i3Lhx8eSTT2aPCwDohMoOi7Vr18bs2bPjlltuiR07dsRHPvKR+MxnPhN79+6txPgAgE6k7LC4/fbb4/rrr48bbrghzj///Fi6dGkMHTo0li1bVonxAQCdSE05G//nP/+Jbdu2RV1dXav1kydPjs2bNx93n+bm5mhubm553djYGBERTU1N5Y71bR1ufuOYdUef53jbZOnIc8ERlfq+a8s1eqrPVYm/R07G0WM83vjask3WubJ05LkoX0fNz5HjFkXx1hsWZXj55ZeLiCg2bdrUav3ChQuLc88997j7zJs3r4gIi8VisVgsXWBpaGh4y1Yo647FEd26dWv1uiiKY9YdMXfu3JgzZ07L68OHD8c///nPGDBgwAn3oXxNTU0xdOjQaGhoiNra2lM9HP6Huale5qZ6mZvqUxRFHDhwIAYPHvyW25UVFu9+97uje/fusW/fvlbr9+/fH2edddZx9ymVSlEqlVqtO/3008s5LWWora11EVYpc1O9zE31MjfVpX///m+7TVk/vNmzZ88YN25cbNy4sdX6jRs3xoc//OHyRgcAdDllfxQyZ86cuO6662L8+PExceLEWL58eezduzdmzJhRifEBAJ1I2WFx5ZVXxj/+8Y+47bbb4tVXX40xY8bEr371qxg2bFglxkcblUqlmDdv3jEfO3HqmZvqZW6ql7npvLoVb/t7IwAAbeNZIQBAGmEBAKQRFgBAGmEBAKQRFlWsnMfT//KXv4xPfepTceaZZ0ZtbW1MnDgxHnnkkVbb3HvvvdGtW7djln//+9+VfitdSjnz8tRTT8WkSZNiwIAB0bt37xg5cmT88Ic/PGa7Bx98MEaNGhWlUilGjRoV69atq+Rb6LKy58Y1k6ecuflfmzZtipqamrjwwguP+ZrrpkqV86wQOs79999f9OjRo1ixYkXx3HPPFbNmzSr69u1bvPTSS8fdftasWcWSJUuK3/3ud8WuXbuKuXPnFj169Ci2b9/ess3KlSuL2tra4tVXX2210Hblzsv27duL1atXF3/84x+Lv/71r8XPf/7zok+fPsWPf/zjlm02b95cdO/evVi0aFHx/PPPF4sWLSpqamqKLVu2dNTb6hIqMTeumRzlzs0Rr732WnHOOecUkydPLj7wgQ+0+prrpnoJiyr1wQ9+sJgxY0ardSNHjizq6urafIxRo0YVCxYsaHm9cuXKon///llDfEfKmJcvfvGLxbXXXtvy+oorrig+/elPt9rm8ssvL6666qqTG+w7TCXmxjWTo71zc+WVVxa33nprMW/evGPCwnVTvXwUUoWOPJ5+8uTJrda/1ePpj3b48OE4cOBAnHHGGa3WHzx4MIYNGxZDhgyJqVOnxo4dO9LG3dVlzMuOHTti8+bNcemll7ase/rpp4855uWXX97mY1K5uYlwzZys9s7NypUrY/fu3TFv3rzjft11U72ERRX6+9//HocOHTrmwW5nnXXWMQ+AO5Ef/OAH8frrr8cVV1zRsm7kyJFx7733xvr162PNmjXRq1evmDRpUrz44oup4++qTmZehgwZEqVSKcaPHx8zZ86MG264oeVr+/btO6m5pnJz45o5ee2ZmxdffDHq6uriF7/4RdTUHP9/EO26qV7temw6HaOcx9P/rzVr1sT8+fPjoYceioEDB7asv/jii+Piiy9ueT1p0qS46KKL4s4774w77rgjb+BdXHvm5cknn4yDBw/Gli1boq6uLt773vfG1VdffVLH5FjZc+OaydPWuTl06FB8+ctfjgULFsS5556bckw6lrCoQu15PP0Ra9eujeuvvz4eeOCBuOyyy95y29NOOy0mTJjgv77a6GTmZcSIERERccEFF8Tf/va3mD9/fss/Xu95z3vadUz+X6Xm5miumfKVOzcHDhyIrVu3xo4dO+LGG2+MiP9+tFsURdTU1MSjjz4an/jEJ1w3VcxHIVWovY+nX7NmTXz1q1+N1atXx5QpU972PEVRxM6dO2PQoEEnPeZ3gvbOy9GKoojm5uaW1xMnTjzmmI8++mhZx3ynq9TcHO/rrpnylDs3tbW18eyzz8bOnTtblhkzZsR5550XO3fujA996EMR4bqpaqfoh0Z5G0d+PesnP/lJ8dxzzxWzZ88u+vbtW+zZs6coiqKoq6srrrvuupbtV69eXdTU1BR33313q1+Le+2111q2mT9/fvHwww8Xu3fvLnbs2FFMnz69qKmpKX772992+PvrrMqdl7vuuqtYv359sWvXrmLXrl3FT3/606K2tra45ZZbWrbZtGlT0b1792Lx4sXF888/XyxevNivzbVDJebGNZOj3Lk52vF+K8R1U72ERRW7++67i2HDhhU9e/YsLrroouLxxx9v+dq0adOKSy+9tOX1pZdeWkTEMcu0adNatpk9e3Zx9tlnFz179izOPPPMYvLkycXmzZs78B11DeXMyx133FGMHj266NOnT1FbW1uMHTu2uOeee4pDhw61OuYDDzxQnHfeeUWPHj2KkSNHFg8++GBHvZ0uJXtuXDN5ypmbox0vLIrCdVOtPDYdAEjjZywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBIIywAgDTCAgBI83/exat0NY9KbgAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "perturbation_scores = pd.Series(perturbation_scores)\n",
    "plt.hist(perturbation_scores[perturbation_scores < 20], bins=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAigAAAHECAYAAAATY9HhAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjEsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvc2/+5QAAAAlwSFlzAAAPYQAAD2EBqD+naQAATg5JREFUeJzt3XlcFdX/P/DXZbvsVwHZFMFMTQU1xT0FXFBEzX1PTDNzS1MrsQVQE5cW/Wrap3Jf0nIv/Kgo4JJauCVqGSUoJoiSsu+c3x/+mM+93AtKXroDvJ6PxzwezJkzM+9zz0Xenpk5oxBCCBARERHJiJGhAyAiIiIqiwkKERERyQ4TFCIiIpIdJihEREQkO0xQiIiISHaYoBAREZHsMEEhIiIi2WGCQkRERLLDBIWIiIhkhwkKVRsKheKplpiYGEyYMAEeHh5VGk9iYiIUCgU+/vhjjfLi4mJMnDgRCoUCH330UZXGUFtlZGTgo48+gq+vL5ydnWFtbQ0vLy8sW7YMeXl5WvULCwsRFhYGDw8PKJVKvPDCC1i9erVWva+//hqDBg2Ch4cHLCws8Pzzz2Pq1KlITk7WqrtlyxaMGjUKzZo1g5GR0TN/30rb5O3tDVtbWyiVSnh4eGDixIm4ePGiVG/Tpk0a33cTExO4uLhg1KhRiI+P1zqur68vPD09Ncqys7OxbNkytG7dGra2trCxsUHjxo0xYsQInDhxQqoXExOjcS5jY2M4OTlh+PDh+PXXX6V6pb8LpYuRkRHq1q2Lnj174ujRo1oxhYaGQqFQPNPnRTWfiaEDIHpaZ8+e1VhftGgRoqOjERUVpVHeokULuLm5YdasWf9meACAgoICjB49Gvv378fatWsxderUfz2G2uD27dtYuXIlXnnlFcyZMwfW1tY4deoUQkNDERkZicjISI0/gNOmTcPWrVuxaNEitG/fHkeOHMGsWbOQmZmJBQsWSPVCQkLg5+eHJUuWoH79+rhx4wYWLVqEAwcO4NKlS3BycpLqbt26FSkpKejQoQNKSkpQWFj4j9vz559/wt/fH6mpqXjjjTcQFhYGa2trJCYm4ttvv0W7du3w6NEjqFQqaZ+NGzfihRdeQF5eHn788Ud89NFHiI6Oxm+//Ya6deuWe67i4mL4+/sjLi4Ob7/9Njp06AAAiI+Px/fff49Tp07Bx8dHY58lS5bAz88PBQUFOH/+PBYuXIjjx48jLi4O9evXl+rNnDkTY8aMQXFxMX777TeEhYWhX79+iIqKQvfu3aV6r732Gvr27fuPPy+qJQRRNRUUFCSsrKwMdv6EhAQBQKxYsUIIIURWVpbo1auXMDU1Fd98843B4qoNsrKyRFZWllb5ihUrBABx6tQpqezq1atCoVCIJUuWaNSdPHmysLCwEGlpaVLZvXv3tI4ZGxsrAIhFixZplBcXF0s/BwYGCnd393/UlqKiIuHl5SVsbW1FXFyczjqHDh0S2dnZQgghNm7cKACI2NhYjTphYWECgNiwYYNGuY+Pj2jZsqW0HhUVpbNeKfV2RUdHCwDiu+++06izfv16AUAsXrxYCKH9u1DqxIkTAoAYP358RR8BkU68xEM1kq5LPJ9//jm6d+8OR0dHWFlZwcvLC8uXL9f6n2/pkHhsbCy6desGS0tLPPfcc1i6dClKSkp0nu/hw4fo1asXfvzxR+zfvx+jRo3S2H7+/HmMGjVKunTg4eGB0aNH49atWxr1Sofvo6KiMHnyZNjb28PW1hbjx49HdnY2UlJSMGLECNSpUwcuLi6YN2+eRvylQ+3Lly/HRx99hIYNG8Lc3Bze3t44fvy4xrlKh9mvXLmC4cOHQ6VSwc7ODnPmzEFRURFu3LiBvn37wsbGBh4eHli+fLnG/nl5eZg7dy7atGkj7du5c2ccOHBA6/NRKBSYMWMGtm7diubNm8PS0hKtW7fGDz/8oDOma9euYfTo0VCpVHBycsLEiRORnp4u1bOysoKVlZXWeUpHA5KSkqSy/fv3QwiBV199VaPuq6++itzcXBw+fFgqc3R01Dpmu3btYGxsrHFMADAy0s8/n/v370dcXByCg4O1LsWUCggIgKWlZYXH8fb2BgDcu3evwnppaWkAABcXF53bn6ZdnTp1AgCt7+/TxsRLPPQ0mKBQrfHnn39izJgx2Lp1K3744QdMmjQJK1aswJQpU7TqpqSkYOzYsRg3bhwOHjyIgIAABAcHY9u2bVp1k5OT0b17d/z66684evQo+vXrp1UnMTERzZo1w8qVK3HkyBEsW7YMycnJaN++PR48eKBV/7XXXoNKpcLOnTvx/vvvY8eOHZg8eTICAwPRunVr7N69G0FBQfjkk0903kuxZs0aHD58GCtXrsS2bdtgZGSEgIAArctkADBixAi0bt0ae/bsweTJk/HZZ5/hrbfewqBBgxAYGIh9+/ahR48eePfdd7F3715pv/z8fPz999+YN28e9u/fj2+++QYvvfQShgwZgi1btmidJyIiAmvWrMHChQuxZ88e2NnZYfDgwbh586ZW3aFDh6Jp06bYs2cP5s+fjx07duCtt97SqldW6eW+li1bSmVXr15FvXr14OzsrFG3VatW0vaKnDhxAsXFxRrH1KfSezQGDRr0TMdJSEgAADRt2rTCet7e3jA1NcWsWbOwfft2nffXPMkff/wBAKhXr55eYiLSydBDOET/VEWXeIKCgiocci8uLhaFhYViy5YtwtjYWPz999/SNh8fHwFA/PTTTxr7tGjRQvTp00daLx3WLl2OHj361LEXFRWJrKwsYWVlJVatWiWVlw7fz5w5U6P+oEGDBADx6aefapS3adNGtG3bVismV1dXkZubK5VnZGQIOzs70atXL6ksJCREABCffPKJ1jEBiL1790plhYWFol69emLIkCEVtqmwsFBMmjRJvPjiixrbAAgnJyeRkZEhlaWkpAgjIyMRHh6uFdPy5cs19p82bZowNzcXJSUl5Z7/l19+ERYWFmLw4MEa5b179xbNmjXTuY+ZmZl4/fXXyz1mRkaGaN68uXBzcxOZmZnl1nuWSzx9+/YVAEReXt5T1S/9jpw7d04UFhaKzMxMcfjwYeHs7Cy6d+8uCgsLNeqXvcQjxONLNNbW1tJ318XFRYwfP16cPHlSo17pJZ5du3aJwsJCkZOTI06ePCmef/55YWxsLH755RchxP++d8uWLROFhYUiLy9PXL58WXTu3Fm4uLiIhIQEjeOW9jNRRTiCQrXGpUuXMHDgQNjb28PY2BimpqYYP348iouL8fvvv2vUdXZ2li4XlGrVqpXOIe0+ffpAqVRizpw5uH//vs5zZ2Vl4d1338Xzzz8PExMTmJiYwNraGtnZ2RpPQ5Tq37+/xnrz5s0BAIGBgVrlumIaMmQIzM3NpXUbGxsMGDAAJ0+eRHFx8RPPpVAoEBAQIJWZmJjg+eef1zrXd999h65du8La2homJiYwNTXF+vXrdbbJz88PNjY20rqTkxMcHR11xj9w4ECN9VatWiEvLw+pqaladYHHI1T9+/eHm5sbvv76a63tFV1OKG9bXl4ehgwZglu3buG7776DtbV1uccwhE6dOsHU1BQ2Njbo27cv6tatiwMHDsDE5MnPPkycOBF37tzBjh078Oabb8LNzQ3btm2Dj48PVqxYoVV/5MiRMDU1haWlJbp3747i4mLs3r1bGoUq9e6778LU1BTm5uZo06YNrl69iu+//77Kn6ijmokJCtUKt2/fRrdu3fDXX39h1apVOHXqFGJjY/H5558DAHJzczXq29vbax1DqVRq1QOAXr16Yd++fYiPj4efn5/OP6JjxozBmjVr8Nprr+HIkSP4+eefERsbi3r16uk8pp2dnca6mZlZueW6HqstezmjtKygoABZWVlPPJelpaVGgqPrXHv37sWIESNQv359bNu2DWfPnkVsbCwmTpyoM6bKfKZl6yqVSgDa/QQ8vg/Cz88PJiYmOH78uFZ77O3tpfsu1GVnZ6OgoECrPvD48tXgwYNx+vRpHDx4EB07dtSqoy8NGzYE8L/LIU9ry5YtiI2NRVRUFKZMmYJff/0Vo0ePfur9VSoVRo8ejVWrVuGnn37ClStX4OTkhPfeew+PHj3SqLts2TLExsbi4sWLuH37Nm7evKnzktSsWbMQGxuL06dP4+OPP0ZhYSFefvllnZ8/0ZPwMWOqFfbv34/s7Gzs3bsX7u7uUvnly5f1cvyAgAAcOHAAgwYNgp+fH6KioqRHUtPT0/HDDz8gJCQE8+fPl/YpvYejKqSkpOgsMzMz09tIwLZt29CoUSPs2rVLYxQiPz9fL8d/Grdu3YKvry+EEIiJiUGDBg206nh5eWHnzp1ISUnRSNzi4uIAQOvG1Pz8fAwaNAjR0dE4cOAAevbsWaVt6NOnD7788kvs379f4/vxJM2bN5duQvXz80NxcTG+/vpr7N69G8OGDat0HC1btsSoUaOwcuVK/P777xojiM8995x0roo0aNBAqte1a1c4Oztj3LhxCAkJwZo1ayodE9VuHEGhWqH0D2jp/8QBQAiBr776Sm/n6NOnDw4cOICbN2/Cz89PShIUCgWEEBrnBh5PClb2cou+7N27V2MUIzMzE99//z26desGY2NjvZxDoVDAzMxMIzlJSUnR+RRPVbh9+zZ8fX1RXFyMqKgojcRT3csvvwyFQoHNmzdrlG/atAkWFhYa83GUjpxERUVhz5496NOnT5W2oTQ+Ly8vhIeHl3vD7pEjR5CTk1PhcZYvX466deviww8/LPdpM+DxUzwFBQU6t/32228AAFdX16eMvmJjx46Fr68vvvrqqyc+8UNUFkdQqFbo3bs3zMzMMHr0aLzzzjvIy8vDunXr8PDhQ72ex9/fHwcPHsTLL78sjaS4uLige/fuWLFiBRwcHODh4YETJ05g/fr1qFOnjl7PX8rY2Bi9e/fGnDlzUFJSgmXLliEjIwNhYWF6O0f//v2xd+9eTJs2DcOGDUNSUhIWLVoEFxcXnTOa6lNqair8/PyQnJyM9evXIzU1VePSWoMGDaTRlJYtW2LSpEkICQmBsbEx2rdvj6NHj+LLL7/E4sWLNS7xDBs2DP/973/x3nvvwd7eHufOnZO22draokWLFtL69evXcf36dQCPE7OcnBzs3r0bwOPJAtXrVsTY2Bj79u2Dv78/OnfujKlTp8LPzw9WVla4desWdu/eje+///6J39W6desiODgY77zzDnbs2IFx48bprBcdHY1Zs2Zh7Nix6NKlC+zt7ZGamopvvvkGhw8fxvjx43WORP1Ty5YtQ8eOHbFo0SKd9wcRlYcJCtUKL7zwAvbs2YP3338fQ4YMgb29PcaMGYM5c+Zo3AyqD71798b333+PAQMGSEnKjh07MGvWLLzzzjsoKipC165dERkZqXXTq77MmDEDeXl5ePPNN5GamoqWLVsiIiICXbt21ds5Xn31VaSmpuKLL77Ahg0b8Nxzz2H+/Pm4c+eOXhMhXa5fvy49nqzrD3FISAhCQ0Ol9bVr16J+/fpYvXo1UlJS4OHhgVWrVmHmzJka+5XOy/LRRx9pvabAx8cHMTEx0vq3336r1c7hw4frPP+TNG7cGBcvXsTq1auxb98+rFu3Dvn5+VJye/r0aY1ZZMszc+ZM6VHu0aNH6xwt69SpEyZOnIjo6Ghs3boVDx48gIWFBVq0aIHVq1frffbjDh06YPjw4di8eTOCg4PRuHFjvR6fai6FEEIYOggi0o/ExEQ0atQIK1aswLx58wwdDhHRP8Z7UIiIiEh2eImHiKiKFBUVVbjdyMhIb1PmE9U0vMRDRFQFSi+3VaSy96oQ1SYcQSEiqgKurq6IjY19Yh0i0o0jKERERCQ7vPhJREREssMEhaqVHTt2YOXKlVV2fF9fX62pz5/V2rVrsWnTJq3yxMREKBQKnduo+tLVr2fOnEFoaKjWO24AwMPDQ+uFjU8rJycHoaGhGvOzlAoNDYVCocCDBw/+0bH1rSri8fX1ha+v7xPr8XetemKCQtVKVScoVaG8BMXFxQVnz56tssnayDB09euZM2cQFhamM0F5Fjk5OQgLC9OZoBBVd7xJlqqFnJwcWFpaVtvj66JUKtGpU6d/9Zw1hSH662nVhH4VQiAvLw8WFhaGDoVqMY6g0DMpHba9dOkShgwZAltbW6hUKowbNw7379/Xqr9r1y507twZVlZWsLa2Rp8+fXDp0iWNOhMmTIC1tTXi4uLg7+8PGxsb9OzZE76+voiIiMCtW7egUCikBQBiYmKgUCi0/iepa2i3vOOrO3XqFDp16gQLCwvUr18fH3zwgdaL/cLCwtCxY0fY2dnB1tYWbdu2xfr166F+37mHhweuXbuGEydOSPF6eHiUGxsAnD59Gj179oSNjQ0sLS3RpUsXREREaNTZtGkTFAoFoqOjMXXqVDg4OMDe3h5DhgzB3bt3y+0v9fOWtzzJzZs3MWrUKLi6ukKpVMLJyQk9e/bUejP0jh070LlzZ1hbW8Pa2hpt2rTB+vXrNeps2LABrVu3hrm5Oezs7DB48GD8+uuvGnUq6q+CggIsXrwYL7zwApRKJerVq4dXX31V67sXFRUFX19f2Nvbw8LCAg0bNsTQoUMrfAHf22+/DZVKpdHvM2fOhEKhwIoVK6SytLQ0GBkZYfXq1Rqfb2m/hoaG4u233wYANGrUSPqcy35XDx8+jLZt28LCwgIvvPACNmzYUG5speepV68egMffxdLjTpgwQaPevXv3MHr0aKhUKjg5OWHixIlIT0/XqKNQKDBjxgx88cUXaN68OZRKpfRyxfj4eIwZMwaOjo5QKpVo3rw5Pv/8c439S0pKsHjxYjRr1gwWFhaoU6cOWrVqhVWrVmnF/TTx5OXlITg4GI0aNYKZmRnq16+P6dOnP9UI1N27dzFixAjY2NhApVJh5MiROt/uTfLHERTSi8GDB2PEiBF44403cO3aNXzwwQe4fv06fvrpJ5iamgIAlixZgvfffx+vvvoq3n//fRQUFGDFihXo1q0bfv75Z42XqxUUFGDgwIGYMmUK5s+fj6KiIjRo0ACvv/46/vzzT+zbt++Z4tV1/FIpKSkYNWoU5s+fj4ULFyIiIgKLFy/Gw4cPNV4Zn5iYiClTpqBhw4YAgHPnzmHmzJn466+/8OGHHwIA9u3bh2HDhkGlUmHt2rUAoPVWY3UnTpxA79690apVK6xfvx5KpRJr167FgAED8M0332DkyJEa9V977TUEBgZix44dSEpKwttvv41x48YhKiqq3HOUXoJQd//+fYwbNw7169d/4mfXr18/FBcXY/ny5WjYsCEePHiAM2fOaPzx+PDDD7Fo0SIMGTIEc+fOhUqlwtWrVzXeaBseHo4FCxZg9OjRCA8PR1paGkJDQ9G5c2fExsaiSZMmUl1d/VVSUoKXX34Zp06dwjvvvIMuXbrg1q1bCAkJga+vL86fPw8LCwskJiYiMDAQ3bp1w4YNG1CnTh389ddfOHz4MAoKCsodienVqxc+/vhj/Pzzz+jcuTMA4NixY7CwsEBkZKSUdBw/fhxCCPTq1UvncV577TX8/fffWL16Nfbu3QsXFxcA0Pi+//LLL5g7dy7mz58PJycnfP3115g0aRKef/55dO/eXedxXVxccPjwYfTt2xeTJk3Ca6+9BgBS0lJq6NChGDlyJCZNmoS4uDgEBwcDgFYCtH//fpw6dQoffvghnJ2d4ejoiOvXr6NLly5o2LAhPvnkEzg7O+PIkSN488038eDBA4SEhAB4/Cbl0NBQvP/+++jevTsKCwvx22+/6UwonhSPEAKDBg3C8ePHERwcjG7duuHKlSsICQnB2bNncfbs2XJ/h3Jzc9GrVy/cvXsX4eHhaNq0KSIiIrR+b6iaEETPICQkRAAQb731lkb59u3bBQCxbds2IYQQt2/fFiYmJmLmzJka9TIzM4Wzs7MYMWKEVBYUFCQAiA0bNmidLzAwULi7u2uVR0dHCwAiOjpaozwhIUEAEBs3bnyq4/v4+AgA4sCBAxrlkydPFkZGRuLWrVs6P4fi4mJRWFgoFi5cKOzt7UVJSYm0rWXLlsLHx0drH12xderUSTg6OorMzEyprKioSHh6eooGDRpIx924caMAIKZNm6ZxzOXLlwsAIjk5WWecumRnZ4sOHToIFxcXkZiYWGHdBw8eCABi5cqV5da5efOmMDY2FmPHji23zsOHD4WFhYXo16+fRvnt27eFUqkUY8aMkcrK669vvvlGABB79uzRKI+NjRUAxNq1a4UQQuzevVsAEJcvX66wbWVlZ2cLMzMzsXDhQiGEEHfu3BEAxLvvvissLCxEXl6eEOLxd8PV1VXaT1e/rlixQgAQCQkJWudxd3cX5ubmGt+t3NxcYWdnJ6ZMmVJhjPfv3xcAREhIiNa20t/N5cuXa5RPmzZNmJuba3xHAQiVSiX+/vtvjbp9+vQRDRo0EOnp6RrlM2bMEObm5lL9/v37izZt2lQY69PGc/jwYZ31du3aJQCIL7/8Uirz8fHR+N1at25dub+/ZfuE5I+XeEgvxo4dq7E+YsQImJiYIDo6GgBw5MgRFBUVYfz48SgqKpIWc3NzrbfElho6dGiVxlze8W1sbDBw4ECNsjFjxqCkpAQnT56UyqKiotCrVy+oVCoYGxvD1NQUH374IdLS0pCamlrpeLKzs/HTTz9h2LBhsLa2lsqNjY3xyiuv4M6dO7hx44bGPmXjbNWqFQBojFRUpLi4GCNHjsSvv/6KQ4cOwd3dHcDj/8Wq91PpCJOdnR0aN26MFStW4NNPP8WlS5dQUlKicczIyEgUFxdj+vTp5Z737NmzyM3N1boc4ebmhh49euD48eNa+5Ttrx9++AF16tTBgAEDNOJs06YNnJ2dpe9UmzZtYGZmhtdffx2bN2+W3oL8JJaWlujcuTOOHTsmtatOnTp4++23UVBQgNOnTwN4PKpS3ujJ02rTpo00EgcA5ubmaNq06VP3Y0V0fUfy8vK0vqM9evRA3bp1pfW8vDwcP34cgwcPhqWlpcZn3K9fP+Tl5eHcuXMAHr+x+JdffsG0adNw5MgRZGRk/ON4Skf/yn43hg8fDisrK53fjVLR0dHl/v5S9cMEhfTC2dlZY93ExAT29vZIS0sD8Pi6MwC0b98epqamGsuuXbu0Hj20tLSEra1tlcVb0fGdnJy0ykrbV9qen3/+Gf7+/gCAr776Cj/++CNiY2Px3nvvAXg81FxZDx8+hBBCugSgrnTG0dLzl7K3t9dYLx36ftrzv/HGGzh8+DB2796NNm3aSOWbN2/W6ifg8b0Kx48fR58+fbB8+XK0bdsW9erVw5tvvonMzEwAkO7/aNCgQbnnLW1HeW0t205d/XXv3j08evQIZmZmWrGmpKRI36nGjRvj2LFjcHR0xPTp09G4cWM0btxY5/0RZfXq1Qvnzp1DdnY2jh07hh49esDe3h7t2rXDsWPHkJCQgISEhGdOUMr2I/C4L//J9+hJxy7vO1K2L9LS0lBUVITVq1drfb79+vUDAOkzDg4Oxscff4xz584hICAA9vb26NmzJ86fP1/peNLS0mBiYqJ1qUqhUMDZ2Vnru1E25op+f6l64T0opBcpKSka9y8UFRUhLS1N+sfIwcEBALB7927pf+kVeZqbNdWZm5sDAPLz8zXKy5tzoaLjlyZT6kpvsittz86dO2FqaooffvhBOjfw+Dr+P1W3bl0YGRkhOTlZa1vpja+ln6M+hIaG4uuvv8bGjRulZKvUgAEDyp2m3d3dXbrZ9ffff8e3336L0NBQFBQU4IsvvpD+sNy5cwdubm46j1H6OZbX1rLt1NVfpTcGHz58WOc5bGxspJ+7deuGbt26obi4GOfPn8fq1asxe/ZsODk5YdSoUTr3B4CePXvigw8+wMmTJ3H8+HHpnouePXvi6NGj0rt2yt5kXR2V/Yzr1q0rjd6VNxpW2n4TExPMmTMHc+bMwaNHj3Ds2DEsWLAAffr0QVJSUqWeuLK3t0dRURHu37+vkaQIIZCSkoL27dtXuO/PP/+sVc6bZKsnjqCQXmzfvl1j/dtvv0VRUZE0iVKfPn1gYmKCP//8E97e3jqXp1He/ypLn4y5cuWKRvnBgwcr3ZbMzEyt/Xbs2AEjIyPphkWFQgETExMYGxtLdXJzc7F169anjrksKysrdOzYEXv37tWoX1JSgm3btqFBgwZo2rRppdujy/r16xEWFoaFCxdqDaUDj/+hf5r+adq0Kd5//314eXnh4sWLAAB/f38YGxtj3bp15Z6/c+fOsLCwwLZt2zTK79y5g6ioqKf6g9+/f3+kpaWhuLhY5/epWbNmWvsYGxujY8eO0lMopTGXp0OHDrC1tcXKlSuRkpKC3r17A3g8snLp0iV8++23aNGixRPfqVPZka2nVVXHBR6PWvn5+eHSpUto1aqVzs9Y18hPnTp1MGzYMEyfPh1///03EhMTK3Xe0r4v+93Ys2cPsrOzK/xu+Pn5lfv7S9UPR1BIL/bu3QsTExP07t1beoqndevWGDFiBIDHCcTChQvx3nvv4ebNm+jbty/q1q2Le/fu4eeff4aVlRXCwsKeeB4vLy/s3bsX69atQ7t27WBkZARvb284OzujV69eCA8PR926deHu7o7jx49j7969lW6Lvb09pk6ditu3b6Np06Y4dOgQvvrqK0ydOlW6TyAwMBCffvopxowZg9dffx1paWn4+OOPdT5d4OXlhZ07d2LXrl147rnnYG5uDi8vL53nDg8PR+/eveHn54d58+bBzMwMa9euxdWrV/HNN99UemRJl7Nnz+KNN95A165d0bt3b+k+glIVzeFx5coVzJgxA8OHD0eTJk1gZmaGqKgoXLlyBfPnzwfwuK8XLFiARYsWITc3V3qk9Pr163jw4AHCwsJQp04dfPDBB1iwYAHGjx+P0aNHIy0tDWFhYTA3N5dGKioyatQobN++Hf369cOsWbPQoUMHmJqa4s6dO4iOjsbLL7+MwYMH44svvkBUVBQCAwPRsGFD5OXlSU+MPOnSjLGxMXx8fPD999+jUaNGaNy4MQCga9euUCqVOH78ON58880nxlra36tWrUJQUBBMTU3RrFkzjVGef8LGxgbu7u44cOAAevbsCTs7Ozg4OEgJ+7NatWoVXnrpJXTr1g1Tp06Fh4cHMjMz8ccff+D777+X7hcZMGAAPD094e3tjXr16uHWrVtYuXIl3N3dNZ7Gehq9e/dGnz598O677yIjIwNdu3aVnuJ58cUX8corr5S77/jx4/HZZ59h/Pjx+Oijj9CkSRMcOnQIR44ceabPgQzEwDfpUjVXemf+hQsXxIABA4S1tbWwsbERo0ePFvfu3dOqv3//fuHn5ydsbW2FUqkU7u7uYtiwYeLYsWNSnaCgIGFlZaXzfH///bcYNmyYqFOnjlAoFEL9K5ycnCyGDRsm7OzshEqlEuPGjRPnz5/X+RRPecf38fERLVu2FDExMcLb21solUrh4uIiFixYIAoLCzXqbtiwQTRr1kwolUrx3HPPifDwcLF+/XqtpzUSExOFv7+/sLGxEQCkp5B0Pe0hhBCnTp0SPXr0EFZWVsLCwkJ06tRJfP/99xp1Sp/iiY2N1Sgv72kmXfuWt1Tk3r17YsKECeKFF14QVlZWwtraWrRq1Up89tlnoqioSKPuli1bRPv27YW5ubmwtrYWL774olZbv/76a9GqVSthZmYmVCqVePnll8W1a9c06lTUX4WFheLjjz8WrVu3ls7zwgsviClTpoj4+HghhBBnz54VgwcPFu7u7kKpVAp7e3vh4+MjDh48WGFbS61atUoAEJMnT9Yo7927twCgdZzy+jU4OFi4uroKIyMjjT5yd3cXgYGBWuct+4RKeY4dOyZefPFFoVQqBQARFBQkhPjf7+b9+/c16pf2v/p3FICYPn26zuMnJCSIiRMnivr16wtTU1NRr1490aVLF7F48WKpzieffCK6dOkiHBwchJmZmWjYsKGYNGmSxlNhlYknNzdXvPvuu8Ld3V2YmpoKFxcXMXXqVPHw4cMnfkZ37twRQ4cOlf4tGjp0qDhz5gyf4qmG+DZjeiahoaEICwvD/fv39Xp/BBER1W68B4WIiIhkhwkKERERyQ4v8RAREZHscASFiIiIZIcJChEREckOExQiIiKSnWo5UVtJSQnu3r0LGxsbvUxcRURERFVPCIHMzEy4urrCyKjiMZJqmaDcvXu33Hd8EBERkbwlJSVV+EJRoJomKKXTQyclJen3jbfZ2UDpOzXu3gWsrPR3bCIiolouIyMDbm5uT/Wah2qZoJRe1rG1tdVvgqL24jfY2jJBISIiqgJPc3sGb5IlIiIi2amWIyhVxsQECAr6389ERERkEPwrrE6pBDZtMnQUREREtR4v8RAREZHscARFnRBATs7jny0tAc6xQkREZBAcQVGXkwNYWz9eShMVIiIi+tcxQSEiIiLZYYJCREREssMEhYiIiGSHCQoRERHJDhMUIiIikh0mKERERCQ7nAdFnbExMGzY/34mIiIig2CCos7cHPjuO0NHQUREVOsxQdHBY36Exnri0kADRUJERFQ78R4UIiIikh0mKOqyswGFAonL+sOiIM/Q0RAREdVaTFCIiIhIdpigEBERkewwQSEiIiLZqXSCcvLkSQwYMACurq5QKBTYv3+/xnaFQqFzWbFihVTH19dXa/uoUaOeuTFERERUM1Q6QcnOzkbr1q2xZs0anduTk5M1lg0bNkChUGDo0KEa9SZPnqxR7z//+c8/awERERHVOJWeByUgIAABAQHlbnd2dtZYP3DgAPz8/PDcc89plFtaWmrVJSIiIgKq+B6Ue/fuISIiApMmTdLatn37djg4OKBly5aYN28eMjMzyz1Ofn4+MjIyNJYqYWwM9OuHqOe8UWLE23OIiIgMpUpnkt28eTNsbGwwZMgQjfKxY8eiUaNGcHZ2xtWrVxEcHIxffvkFkZGROo8THh6OsLCwqgz1MXNzICICE8vMJEtERET/ripNUDZs2ICxY8fC3Nxco3zy5MnSz56enmjSpAm8vb1x8eJFtG3bVus4wcHBmDNnjrSekZEBNze3qguciIiIDKrKEpRTp07hxo0b2LVr1xPrtm3bFqampoiPj9eZoCiVSiiVyqoIk4iIiGSoym60WL9+Pdq1a4fWrVs/se61a9dQWFgIFxeXqgrn6WRnA1ZWuP7pUE51T0REZECVHkHJysrCH3/8Ia0nJCTg8uXLsLOzQ8OGDQE8vgTz3Xff4ZNPPtHa/88//8T27dvRr18/ODg44Pr165g7dy5efPFFdO3a9Rmaoic5ObA0dAxERES1XKUTlPPnz8PPz09aL703JCgoCJs2bQIA7Ny5E0IIjB49Wmt/MzMzHD9+HKtWrUJWVhbc3NwQGBiIkJAQGBsb/8NmEBERUU2iEEIIQwdRWRkZGVCpVEhPT4etra3+DpydDVhbAwCav7UbuWaPb+5NXBqov3MQERHVUpX5+83JPoiIiEh2mKAQERGR7DBBISIiItmp0onaqh0jI8DHB+dupqFEoTB0NERERLUWExR1FhZATAxGcap7IiIig+IlHiIiIpIdJihEREQkO0xQ1GVnA/Xq4cL/jeFU90RERAbEe1DKevAA9oaOgYiIqJbjCAoRERHJDhMUIiIikh0mKERERCQ7TFCIiIhIdpigEBERkezwKR51RkaAtzd+uZPOqe6JiIgMiAmKOgsLIDYWL3OqeyIiIoPiJR4iIiKSHSYoREREJDtMUNTl5AAeHji9biLMCznVPRERkaHwHhR1QgC3bqEBAIUwdDBERES1F0dQiIiISHaYoBAREZHsMEEhIiIi2WGCQkRERLLDBIWIiIhkh0/xqFMogBYt8Pu9LAjOdE9ERGQwTFDUWVoC167Bn1PdExERGRQv8RAREZHsMEEhIiIi2WGCoi4nB2jZEke/nsap7omIiAyI96CoEwK4fh1NwanuiYiIDIkjKERERCQ7TFCIiIhIdpigEBERkewwQSEiIiLZYYJCREREssOneNQpFIC7O+48zOVU90RERAbEBEWdpSWQmIiXONU9ERGRQVX6Es/JkycxYMAAuLq6QqFQYP/+/RrbJ0yYAIVCobF06tRJo05+fj5mzpwJBwcHWFlZYeDAgbhz584zNYSIiIhqjkonKNnZ2WjdujXWrFlTbp2+ffsiOTlZWg4dOqSxffbs2di3bx927tyJ06dPIysrC/3790dxcXHlW0BEREQ1TqUv8QQEBCAgIKDCOkqlEs7Ozjq3paenY/369di6dSt69eoFANi2bRvc3Nxw7Ngx9OnTp7Ih6U9uLtC9Ow7cSceIMUuRb6o0XCxERES1WJU8xRMTEwNHR0c0bdoUkydPRmpqqrTtwoULKCwshL+/v1Tm6uoKT09PnDlzRufx8vPzkZGRobFUiZIS4Px5tE6Jh5HgXPdERESGovcEJSAgANu3b0dUVBQ++eQTxMbGokePHsjPzwcApKSkwMzMDHXr1tXYz8nJCSkpKTqPGR4eDpVKJS1ubm76DpuIiIhkRO9P8YwcOVL62dPTE97e3nB3d0dERASGDBlS7n5CCCgUup/tDQ4Oxpw5c6T1jIwMJilEREQ1WJVP1Obi4gJ3d3fEx8cDAJydnVFQUICHDx9q1EtNTYWTk5POYyiVStja2mosREREVHNVeYKSlpaGpKQkuLi4AADatWsHU1NTREZGSnWSk5Nx9epVdOnSparDISIiomqg0pd4srKy8Mcff0jrCQkJuHz5Muzs7GBnZ4fQ0FAMHToULi4uSExMxIIFC+Dg4IDBgwcDAFQqFSZNmoS5c+fC3t4ednZ2mDdvHry8vKSneoiIiKh2q3SCcv78efj5+UnrpfeGBAUFYd26dYiLi8OWLVvw6NEjuLi4wM/PD7t27YKNjY20z2effQYTExOMGDECubm56NmzJzZt2gRjY2M9NOkZOTggLbvA0FEQERHVagohqt/ztBkZGVCpVEhPT6+S+1E8ykx1n7g0UO/nICIiqm0q8/ebbzMmIiIi2WGCQkRERLLDBEVdbi7g64udO+ZDWZhv6GiIiIhqLb1P1FatlZQAJ06gE8Cp7omIiAyIIyhEREQkO0xQiIiISHaYoBAREZHsMEEhIiIi2WGCQkRERLLDp3jKsrRETmGxoaMgIiKq1ZigqLOyArKz0aLMVPdERET07+IlHiIiIpIdJihEREQkO0xQ1OXlAYGB2PBdKJRFBYaOhoiIqNbiPSjqiouBQ4fQA4BRSYmhoyEiIqq1OIJCREREssMEhYiIiGSHCQoRERHJDhMUIiIikh0mKERERCQ7TFCIiIhIdviYsTorK0AIeHCqeyIiIoPiCAoRERHJDhMUIiIikh0mKOry8oDhw/H5/nBOdU9ERGRATFDUFRcDu3cj8MaPnOqeiIjIgJigEBERkewwQSEiIiLZYYJCREREssMEhYiIiGSHCQoRERHJDhMUIiIikh0mKOosLYGsLDR/azdyTZWGjoaIiKjW4rt41CkUgJUVcs3MDR0JERFRrcYRFCIiIpIdJijq8vOBCRPwccRnMCsqNHQ0REREtRYTFHVFRcDmzRh29TiMS4oNHQ0REVGtxQSFiIiIZKfSCcrJkycxYMAAuLq6QqFQYP/+/dK2wsJCvPvuu/Dy8oKVlRVcXV0xfvx43L17V+MYvr6+UCgUGsuoUaOeuTFERERUM1Q6QcnOzkbr1q2xZs0arW05OTm4ePEiPvjgA1y8eBF79+7F77//joEDB2rVnTx5MpKTk6XlP//5zz9rAREREdU4lX7MOCAgAAEBATq3qVQqREZGapStXr0aHTp0wO3bt9GwYUOp3NLSEs7OzpU9PREREdUCVX4PSnp6OhQKBerUqaNRvn37djg4OKBly5aYN28eMjMzyz1Gfn4+MjIyNBYiIiKquap0ora8vDzMnz8fY8aMga2trVQ+duxYNGrUCM7Ozrh69SqCg4Pxyy+/aI2+lAoPD0dYWFhVhkpEREQyohBCiH+8s0KBffv2YdCgQVrbCgsLMXz4cNy+fRsxMTEaCUpZFy5cgLe3Ny5cuIC2bdtqbc/Pz0d+fr60npGRATc3N6Snp1d43EoTAnjwAG0XReJvC9vHM8sCSFwaqL9zEBER1VIZGRlQqVRP9fe7SkZQCgsLMWLECCQkJCAqKuqJQbRt2xampqaIj4/XmaAolUoolf/Cu3EUCqBePfxtqar6cxEREVG59J6glCYn8fHxiI6Ohr29/RP3uXbtGgoLC+Hi4qLvcIiIiKgaqnSCkpWVhT/++ENaT0hIwOXLl2FnZwdXV1cMGzYMFy9exA8//IDi4mKkpKQAAOzs7GBmZoY///wT27dvR79+/eDg4IDr169j7ty5ePHFF9G1a1f9teyfyM8H5szBwrO3sLjHaygwMTVsPERERLVUpe9BiYmJgZ+fn1Z5UFAQQkND0ahRI537RUdHw9fXF0lJSRg3bhyuXr2KrKwsuLm5ITAwECEhIbCzs3uqGCpzDatSsrMBa2sAQPO3dktvNeY9KERERM+uSu9B8fX1RUU5zZPyHTc3N5w4caKypyUiIqJahO/iISIiItlhgkJERESywwSFiIiIZIcJChEREckOExQiIiKSnSp9F0+1Y2EBJCTgpWVRyDM1M3Q0REREtRYTFHVGRoCHB+6onAwdCRERUa3GSzxEREQkO0xQ1BUUAG+/jeDoDTAtLjR0NERERLUWExR1hYXAxx9jys97YVJcbOhoiIiIai0mKERERCQ7TFCIiIhIdpigEBERkewwQSEiIiLZYYJCREREssMEhYiIiGSHM8mqs7AArl5F709PcKp7IiIiA2KCos7ICGjZEvH1Eg0dCRERUa3GSzxEREQkO0xQ1BUUAKGhmH16O6e6JyIiMiAmKOoKC4GwMMz+8RtOdU9ERGRATFCIiIhIdpigEBERkewwQSEiIiLZYYJCREREssMEhYiIiGSHCQoRERHJDmeSVWduDvz8MwauOY18E1NDR0NERFRrMUFRZ2wMtG+PKy6pho6EiIioVuMlHiIiIpIdJijqCgqAFSvw+k97ONU9ERGRATFBUVdYCLzzDhbEbORU90RERAbEBIWIiIhkhwkKERERyQ4TFCIiIpIdJihEREQkO0xQiIiISHYqnaCcPHkSAwYMgKurKxQKBfbv36+xXQiB0NBQuLq6wsLCAr6+vrh27ZpGnfz8fMycORMODg6wsrLCwIEDcefOnWdqCBEREdUclU5QsrOz0bp1a6xZs0bn9uXLl+PTTz/FmjVrEBsbC2dnZ/Tu3RuZmZlSndmzZ2Pfvn3YuXMnTp8+jaysLPTv3x/Fhn6019wciI7GqNFLONU9ERGRASmEEOIf76xQYN++fRg0aBCAx6Mnrq6umD17Nt59910Aj0dLnJycsGzZMkyZMgXp6emoV68etm7dipEjRwIA7t69Czc3Nxw6dAh9+vR54nkzMjKgUqmQnp4OW1vbfxp+uTzmR2isJy4N1Ps5iIiIapvK/P3W6z0oCQkJSElJgb+/v1SmVCrh4+ODM2fOAAAuXLiAwsJCjTqurq7w9PSU6hAREVHtptcEJSUlBQDg5OSkUe7k5CRtS0lJgZmZGerWrVtunbLy8/ORkZGhsVSJwkLg88/xysUfYFJcVDXnICIioieqkqd4FAqFxroQQqusrIrqhIeHQ6VSSYubm5veYtVQUADMmIFFkV/AlAkKERGRweg1QXF2dgYArZGQ1NRUaVTF2dkZBQUFePjwYbl1ygoODkZ6erq0JCUl6TNsIiIikhm9JiiNGjWCs7MzIiMjpbKCggKcOHECXbp0AQC0a9cOpqamGnWSk5Nx9epVqU5ZSqUStra2GgsRERHVXCaV3SErKwt//PGHtJ6QkIDLly/Dzs4ODRs2xOzZs7FkyRI0adIETZo0wZIlS2BpaYkxY8YAAFQqFSZNmoS5c+fC3t4ednZ2mDdvHry8vNCrVy/9tYyIiIiqrUonKOfPn4efn5+0PmfOHABAUFAQNm3ahHfeeQe5ubmYNm0aHj58iI4dO+Lo0aOwsbGR9vnss89gYmKCESNGIDc3Fz179sSmTZtgbGyshyYRERFRdfdM86AYSpXNg5KdDVhbAwCav7UbuWbmADgPChERkT4YbB4UIiIiIn2o9CWeGk2pBH74Aa9uikUBp7onIiIyGCYo6kxMgMBARJ8ydCBERES1Gy/xEBERkewwQVFXWAhs2oRhccc41T0REZEB8RKPuoIC4NVX8TGAiGYvociYHw8REZEhcASFiIiIZIcJChEREckOExQiIiKSHSYoREREJDtMUIiIiEh2mKAQERGR7PA5WnVKJfDtt5i2/SKnuiciIjIgJijqTEyA4cNx6IKloSMhIiKq1XiJh4iIiGSHCYq6oiLgu+/Q77fTMC4pNnQ0REREtRYTFHX5+cCIEVh7YCnMigoNHQ0REVGtxQSFiIiIZIcJChEREckOExQiIiKSHSYoREREJDtMUIiIiEh2mKAQERGR7HAmWXVmZsDGjZj33S8oNOZHQ0REZCj8K6zO1BSYMAG7f4swdCRERES1Gi/xEBERkewwQVFXVARERMDvz1hOdU9ERGRATFDU5ecD/ftj4+4wTnVPRERkQExQiIiISHaYoBAREZHsMEEhIiIi2WGCQkRERLLDBIWIiIhkhwkKERERyQ5nklVnZgasWYMPDlzlVPdEREQGxL/C6kxNgenTsTWJU90TEREZEi/xEBERkewwQVFXXAzExKDT7Ssw4lT3REREBqP3BMXDwwMKhUJrmT59OgBgwoQJWts6deqk7zD+mbw8wM8PO79ZACWnuiciIjIYvd+DEhsbi+Li/40+XL16Fb1798bw4cOlsr59+2Ljxo3SupmZmb7DICIiompM7wlKvXr1NNaXLl2Kxo0bw8fHRypTKpVwdnbW96mJiIiohqjSe1AKCgqwbds2TJw4EQqFQiqPiYmBo6MjmjZtismTJyM1NbUqwyAiIqJqpkofM96/fz8ePXqECRMmSGUBAQEYPnw43N3dkZCQgA8++AA9evTAhQsXoFQqdR4nPz8f+fn50npGRkZVhk1EREQGVqUJyvr16xEQEABXV1epbOTIkdLPnp6e8Pb2hru7OyIiIjBkyBCdxwkPD0dYWFhVhkpEREQyUmWXeG7duoVjx47htddeq7Cei4sL3N3dER8fX26d4OBgpKenS0tSUpK+wyUiIiIZqbIRlI0bN8LR0RGBgYEV1ktLS0NSUhJcXFzKraNUKsu9/KNXpqbA8uVYcuhXFBkbV/35iIiISKcqGUEpKSnBxo0bERQUBBOT/+VAWVlZmDdvHs6ePYvExETExMRgwIABcHBwwODBg6silMoxMwPefhtfdhyKQmNTQ0dDRERUa1XJCMqxY8dw+/ZtTJw4UaPc2NgYcXFx2LJlCx49egQXFxf4+flh165dsLGxqYpQiIiIqBqqkgTF398fQgitcgsLCxw5cqQqTqkfxcXAxYtolfw7rjo1RokRL/MQEREZAt/Foy4vD+jQAQe3zOFU90RERAbEBIWIiIhkhwkKERERyQ4TFCIiIpIdJihEREQkO0xQiIiISHaYoBAREZHsVOnLAqsdU1MgJAQrj/3Oqe6JiIgMiAmKOjMzIDQUK/MiDB0JERFRrcZLPERERCQ7TFDUlZQA166hyf1bUIgSQ0dDRERUazFBUZebC3h6InLDdJgXFhg6GiIiolqLCQoRERHJDhMUIiIikh0mKERERCQ7TFCIiIhIdpigEBERkewwQSEiIiLZ4Uyy6kxNgXnz8J+TNznVPRERkQFxBEWdmRmwYgXC/Sai0NjU0NEQERHVWkxQiIiISHaYoKgrKQESE9Eg/R6nuiciIjIgJijqcnOBRo1w+otJnOqeiIjIgJigEBERkewwQSEiIiLZYYJCREREssMEhYiIiGSHCQoRERHJDhMUIiIikh1Oda/OxASYNg1bzt5CsRGnuiciIjIUjqCoUyqBzz/Hh/5TUWDCqe6JiIgMhQkKERERyQ4TFHVCAPfvwy4n/fHPREREZBBMUNTl5ACOjri4eiwsCvMNHQ0REVGtxQSFiIiIZIcJChEREckOExQiIiKSHSYoREREJDt6T1BCQ0OhUCg0FmdnZ2m7EAKhoaFwdXWFhYUFfH19ce3aNX2HQURERNVYlYygtGzZEsnJydISFxcnbVu+fDk+/fRTrFmzBrGxsXB2dkbv3r2RmZlZFaEQERFRNVQlU92bmJhojJqUEkJg5cqVeO+99zBkyBAAwObNm+Hk5IQdO3ZgypQpVRHO0zMxAYKCsPvCHU51T0REZEBVMoISHx8PV1dXNGrUCKNGjcLNmzcBAAkJCUhJSYG/v79UV6lUwsfHB2fOnKmKUCpHqQQ2bcK8wLc41T0REZEB6X0EpWPHjtiyZQuaNm2Ke/fuYfHixejSpQuuXbuGlJQUAICTk5PGPk5OTrh161a5x8zPz0d+/v8mTsvIyNB32ERERCQjek9QAgICpJ+9vLzQuXNnNG7cGJs3b0anTp0AAAqFQmMfIYRWmbrw8HCEhYXpO1RtQgA5ObAoyEOuqRKoICYiIiKqOlX+mLGVlRW8vLwQHx8v3ZdSOpJSKjU1VWtURV1wcDDS09OlJSkpqWqCzckBrK3x62fDONU9ERGRAVV5gpKfn49ff/0VLi4uaNSoEZydnREZGSltLygowIkTJ9ClS5dyj6FUKmFra6uxEBERUc2l90s88+bNw4ABA9CwYUOkpqZi8eLFyMjIQFBQEBQKBWbPno0lS5agSZMmaNKkCZYsWQJLS0uMGTNG36EQERFRNaX3BOXOnTsYPXo0Hjx4gHr16qFTp044d+4c3N3dAQDvvPMOcnNzMW3aNDx8+BAdO3bE0aNHYWNjo+9QiIiIqJpSCCGEoYOorIyMDKhUKqSnp+v3ck92NmBtDQBo/tZu5JqZAwASlwbq7xxERES1VGX+fvNdPERERCQ7TFCIiIhIdqpkqvtqy9gYGDYMEXHJKDFi7kZERGQoTFDUmZsD332H6fMjDB0JERFRrcZhAiIiIpIdJihEREQkO0xQ1GVnAwoFEpf1h0VBnqGjISIiqrWYoBAREZHsMEEhIiIi2WGCQkRERLLDBIWIiIhkhwkKERERyQ4TFCIiIpIdziSrztgY6NcPUb+lcqp7IiIiA2KCos7cHIiIwEROdU9ERGRQHCYgIiIi2WGCQkRERLLDBEVddjZgZYXrnw7lVPdEREQGxHtQysrJgaWhYyAiIqrlOIJCREREssMEhYiIiGSHCQoRERHJDhMUIiIikh0mKERERCQ7fIpHnZER4OODczfTUKJQGDoaIiKiWosJijoLCyAmBqM41T0REZFB8RIPERERyQ4TFCIiIpIdJijqsrOBevVw4f/GcKp7IiIiA+I9KGU9eAB7Q8dARERUy3EEhYiIiGSHCQoRERHJDhMUIiIikh0mKERERCQ7TFCIiIhIdvgUjzojI8DbG7/cSedU90RERAbEBEWdhQUQG4uXOdU9ERGRQfESDxEREcmO3hOU8PBwtG/fHjY2NnB0dMSgQYNw48YNjToTJkyAQqHQWDp16qTvUIiIiKia0nuCcuLECUyfPh3nzp1DZGQkioqK4O/vj+zsbI16ffv2RXJysrQcOnRI36FUXk4O4OGB0+smwryQU90TEREZit7vQTl8+LDG+saNG+Ho6IgLFy6ge/fuUrlSqYSzs7O+T/9shABu3UIDAAph6GCIiIhqryq/ByU9PR0AYGdnp1EeExMDR0dHNG3aFJMnT0Zqamq5x8jPz0dGRobGQkRERDVXlSYoQgjMmTMHL730Ejw9PaXygIAAbN++HVFRUfjkk08QGxuLHj16ID8/X+dxwsPDoVKppMXNza0qwyYiIiIDUwghquxixvTp0xEREYHTp0+jQYMG5dZLTk6Gu7s7du7ciSFDhmhtz8/P10heMjIy4ObmhvT0dNja2uov4OxswNoaAND8rd3INTMHACQuDdTfOYiIiGqpjIwMqFSqp/r7XWXzoMycORMHDx7EyZMnK0xOAMDFxQXu7u6Ij4/XuV2pVEKpVFZFmERERCRDek9QhBCYOXMm9u3bh5iYGDRq1OiJ+6SlpSEpKQkuLi76DoeIiIiqIb3fgzJ9+nRs27YNO3bsgI2NDVJSUpCSkoLc3FwAQFZWFubNm4ezZ88iMTERMTExGDBgABwcHDB48GB9h1M5CgXQogV+t28IwZnuiYiIDEbvIyjr1q0DAPj6+mqUb9y4ERMmTICxsTHi4uKwZcsWPHr0CC4uLvDz88OuXbtgY2Oj73Aqx9ISuHYN/pzqnoiIyKCq5BJPRSwsLHDkyBF9n5aIiIhqEL6Lh4iIiGSHCYq6nBygZUsc/Xoap7onIiIyoCp7zLhaEgK4fh1NwanuiYiIDIkjKERERCQ7TFCIiIhIdpigEBERkewwQSEiIiLZYYJCREREssOneNQpFIC7O+48zOVU90RERAbEBEWdpSWQmIiXONU9ERGRQfESDxEREckOExQiIiKSHSYo6nJzgfbtcWDzW1AW5hs6GiIiolqL96CoKykBzp9HawBGT3grMxEREVUdjqAQERGR7DBBISIiItlhgkJERESywwSFiIiIZIcJChEREckOn+Ipy8EBadkFho6CiIioVmOCos7KCrh/H+041T0REZFB8RIPERERyQ4TFCIiIpIdJijqcnMBX1/s3DGfU90TEREZEO9BUVdSApw4gU7gVPdERESGxBEUIiIikh0mKERERCQ7TFCIiIhIdpigEBERkewwQSEiIiLZ4VM8ZVlaIqew2NBREBER1WpMUNRZWQHZ2WjBqe6JiIgMipd4iIiISHaYoBAREZHsMEFRl5cHBAZiw3ehUBYVGDoaIiKiWov3oKgrLgYOHUIPAEYlJYaOhoiIqNbiCAoRERHJDhMUIiIikh2DJihr165Fo0aNYG5ujnbt2uHUqVOGDIeIiIhkwmAJyq5duzB79my89957uHTpErp164aAgADcvn3bUCERERGRTCiEEMIQJ+7YsSPatm2LdevWSWXNmzfHoEGDEB4eXuG+GRkZUKlUSE9Ph62trf6Cys4GrK0fx/LWbuSamZdbNXFpoP7Oq8ZDxyRx/+Rc+joOERHVfP/W34zK/P02yFM8BQUFuHDhAubPn69R7u/vjzNnzmjVz8/PR35+vrSenp4O4HFD9So7W/qxOD8HJaL8J3n0fu7/ryQ/Ry/n0tdxiIio5vu3/maUHvOpxkaEAfz1118CgPjxxx81yj/66CPRtGlTrfohISECABcuXLhw4cKlBixJSUlPzBUMOg+KQqHQWBdCaJUBQHBwMObMmSOtl5SU4O+//4a9vb3O+s8iIyMDbm5uSEpK0u/lI5mo6e0Dan4b2b7qr6a3ke2r/qqqjUIIZGZmwtXV9Yl1DZKgODg4wNjYGCkpKRrlqampcHJy0qqvVCqhVCo1yurUqVOVIcLW1rbGfvGAmt8+oOa3ke2r/mp6G9m+6q8q2qhSqZ6qnkGe4jEzM0O7du0QGRmpUR4ZGYkuXboYIiQiIiKSEYNd4pkzZw5eeeUVeHt7o3Pnzvjyyy9x+/ZtvPHGG4YKiYiIiGTCYAnKyJEjkZaWhoULFyI5ORmenp44dOgQ3N3dDRUSgMeXk0JCQrQuKdUUNb19QM1vI9tX/dX0NrJ91Z8c2miweVCIiIiIysN38RAREZHsMEEhIiIi2WGCQkRERLLDBIWIiIhkhwmKmrVr16JRo0YwNzdHu3btcOrUKUOHpFNoaCgUCoXG4uzsLG0XQiA0NBSurq6wsLCAr68vrl27pnGM/Px8zJw5Ew4ODrCyssLAgQNx584djToPHz7EK6+8ApVKBZVKhVdeeQWPHj3Se3tOnjyJAQMGwNXVFQqFAvv379fY/m+25/bt2xgwYACsrKzg4OCAN998EwUFBVXavgkTJmj1Z6dOnapN+8LDw9G+fXvY2NjA0dERgwYNwo0bNzTqVOc+fJr2Vfc+XLduHVq1aiVNytW5c2f897//lbZX5/57mvZV9/4rKzw8HAqFArNnz5bKqmUfPut7dWqKnTt3ClNTU/HVV1+J69evi1mzZgkrKytx69YtQ4emJSQkRLRs2VIkJydLS2pqqrR96dKlwsbGRuzZs0fExcWJkSNHChcXF5GRkSHVeeONN0T9+vVFZGSkuHjxovDz8xOtW7cWRUVFUp2+ffsKT09PcebMGXHmzBnh6ekp+vfvr/f2HDp0SLz33ntiz549AoDYt2+fxvZ/qz1FRUXC09NT+Pn5iYsXL4rIyEjh6uoqZsyYUaXtCwoKEn379tXoz7S0NI06cm5fnz59xMaNG8XVq1fF5cuXRWBgoGjYsKHIysqS6lTnPnya9lX3Pjx48KCIiIgQN27cEDdu3BALFiwQpqam4urVq0KI6t1/T9O+6t5/6n7++Wfh4eEhWrVqJWbNmiWVV8c+ZILy/3Xo0EG88cYbGmUvvPCCmD9/voEiKl9ISIho3bq1zm0lJSXC2dlZLF26VCrLy8sTKpVKfPHFF0IIIR49eiRMTU3Fzp07pTp//fWXMDIyEocPHxZCCHH9+nUBQJw7d06qc/bsWQFA/Pbbb1XQqsfK/gH/N9tz6NAhYWRkJP766y+pzjfffCOUSqVIT0+vkvYJ8fgfx5dffrncfapT+4QQIjU1VQAQJ06cEELUvD4s2z4hal4fCiFE3bp1xddff13j+q9s+4SoOf2XmZkpmjRpIiIjI4WPj4+UoFTXPuQlHgAFBQW4cOEC/P39Ncr9/f1x5swZA0VVsfj4eLi6uqJRo0YYNWoUbt68CQBISEhASkqKRluUSiV8fHyktly4cAGFhYUadVxdXeHp6SnVOXv2LFQqFTp27CjV6dSpE1Qq1b/6mfyb7Tl79iw8PT01XmLVp08f5Ofn48KFC1XazpiYGDg6OqJp06aYPHkyUlNTpW3VrX3p6ekAADs7OwA1rw/Ltq9UTenD4uJi7Ny5E9nZ2ejcuXON67+y7StVE/pv+vTpCAwMRK9evTTKq2sfGvRtxnLx4MEDFBcXa72o0MnJSeuFhnLQsWNHbNmyBU2bNsW9e/ewePFidOnSBdeuXZPi1dWWW7duAQBSUlJgZmaGunXratUp3T8lJQWOjo5a53Z0dPxXP5N/sz0pKSla56lbty7MzMyqtM0BAQEYPnw43N3dkZCQgA8++AA9evTAhQsXoFQqq1X7hBCYM2cOXnrpJXh6ekrnLY23bPzVrQ91tQ+oGX0YFxeHzp07Iy8vD9bW1ti3bx9atGgh/eGp7v1XXvuAmtF/O3fuxMWLFxEbG6u1rbr+DjJBUaNQKDTWhRBaZXIQEBAg/ezl5YXOnTujcePG2Lx5s3Rj1z9pS9k6uuob6jP5t9pjiDaPHDlS+tnT0xPe3t5wd3dHREQEhgwZUu5+cmzfjBkzcOXKFZw+fVprW03ow/LaVxP6sFmzZrh8+TIePXqEPXv2ICgoCCdOnCj3vNWt/8prX4sWLap9/yUlJWHWrFk4evQozM3Ny61X3fqQl3gAODg4wNjYWCu7S01N1coE5cjKygpeXl6Ij4+XnuapqC3Ozs4oKCjAw4cPK6xz7949rXPdv3//X/1M/s32ODs7a53n4cOHKCws/Ffb7OLiAnd3d8THx0txVYf2zZw5EwcPHkR0dDQaNGggldeUPiyvfbpUxz40MzPD888/D29vb4SHh6N169ZYtWpVjem/8tqnS3XrvwsXLiA1NRXt2rWDiYkJTExMcOLECfzf//0fTExMpGNXuz6s1B0rNViHDh3E1KlTNcqaN28uy5tky8rLyxP169cXYWFh0s1Qy5Ytk7bn5+frvBlq165dUp27d+/qvBnqp59+kuqcO3fOYDfJ/hvtKb256+7du1KdnTt3VvlNsmU9ePBAKJVKsXnz5mrRvpKSEjF9+nTh6uoqfv/9d53bq3MfPql9ulS3PtSlR48eIigoqNr335Pap0t167+MjAwRFxensXh7e4tx48aJuLi4atuHTFD+v9LHjNevXy+uX78uZs+eLaysrERiYqKhQ9Myd+5cERMTI27evCnOnTsn+vfvL2xsbKRYly5dKlQqldi7d6+Ii4sTo0eP1vk4WYMGDcSxY8fExYsXRY8ePXQ+TtaqVStx9uxZcfbsWeHl5VUljxlnZmaKS5cuiUuXLgkA4tNPPxWXLl2SHvH+t9pT+nhcz549xcWLF8WxY8dEgwYNnvkRwIral5mZKebOnSvOnDkjEhISRHR0tOjcubOoX79+tWnf1KlThUqlEjExMRqPaebk5Eh1qnMfPql9NaEPg4ODxcmTJ0VCQoK4cuWKWLBggTAyMhJHjx4VQlTv/ntS+2pC/+mi/hSPENWzD5mgqPn888+Fu7u7MDMzE23bttV4jFBOSp9fNzU1Fa6urmLIkCHi2rVr0vaSkhIREhIinJ2dhVKpFN27dxdxcXEax8jNzRUzZswQdnZ2wsLCQvTv31/cvn1bo05aWpoYO3assLGxETY2NmLs2LHi4cOHem9PdHS0AKC1lP7v5t9sz61bt0RgYKCwsLAQdnZ2YsaMGSIvL6/K2peTkyP8/f1FvXr1hKmpqWjYsKEICgrSil3O7dPVNgBi48aNUp3q3IdPal9N6MOJEydK//bVq1dP9OzZU0pOhKje/fek9tWE/tOlbIJSHftQIYQQlbsoRERERFS1eJMsERERyQ4TFCIiIpIdJihEREQkO0xQiIiISHaYoBAREZHsMEEhIiIi2WGCQkRERLLDBIWIiIhkhwkKERERyQ4TFCIiIpIdJihEREQkO0xQiIiISHb+H5ljOV4B9+d2AAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cutoff = 0.25\n",
    "\n",
    "plt.hist(perturbation_scores, bins=100)\n",
    "plt.axvline(cutoff, color='red', linestyle='--')\n",
    "plt.title(f'{perturbation_screen_name}\\nperturbation z-scores with threshold')\n",
    "plt.savefig(f'../figures/preprocessing/z_scores_{perturbation_screen_name}.png', dpi=400)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENSG00000137845     True\n",
      "ENSG00000185736     True\n",
      "ENSG00000141385     True\n",
      "ENSG00000156709     True\n",
      "ENSG00000127914     True\n",
      "                   ...  \n",
      "ENSG00000073050    False\n",
      "ENSG00000134684     True\n",
      "ENSG00000188994     True\n",
      "ENSG00000166716     True\n",
      "ENSG00000153896    False\n",
      "Name: perturbation_status, Length: 184, dtype: bool\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>guide_id</th>\n",
       "      <th>perturbation</th>\n",
       "      <th>tissue_type</th>\n",
       "      <th>celltype</th>\n",
       "      <th>cancer</th>\n",
       "      <th>disease</th>\n",
       "      <th>perturbation_type</th>\n",
       "      <th>organism</th>\n",
       "      <th>batch</th>\n",
       "      <th>nperts</th>\n",
       "      <th>ngenes</th>\n",
       "      <th>ncounts</th>\n",
       "      <th>percent_mito</th>\n",
       "      <th>percent_ribo</th>\n",
       "      <th>percent_hemo</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>gene_list</th>\n",
       "      <th>perturbation_scores</th>\n",
       "      <th>perturbation_status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAGACAACAGCGTTTA</th>\n",
       "      <td>SARM1_i2</td>\n",
       "      <td>SARM1</td>\n",
       "      <td>primary</td>\n",
       "      <td>iPSC-induced neuron</td>\n",
       "      <td>False</td>\n",
       "      <td>healthy</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>human</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>5167</td>\n",
       "      <td>17733.0</td>\n",
       "      <td>9.479501</td>\n",
       "      <td>6.609147</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5167</td>\n",
       "      <td>ENSG00000004139</td>\n",
       "      <td>0.282747</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAGCCATTCGAATGCT</th>\n",
       "      <td>SARM1_i2</td>\n",
       "      <td>SARM1</td>\n",
       "      <td>primary</td>\n",
       "      <td>iPSC-induced neuron</td>\n",
       "      <td>False</td>\n",
       "      <td>healthy</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>human</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>5311</td>\n",
       "      <td>16777.0</td>\n",
       "      <td>3.683614</td>\n",
       "      <td>6.967873</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5311</td>\n",
       "      <td>ENSG00000004139</td>\n",
       "      <td>0.282747</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AATCACGTCGCTAAAC</th>\n",
       "      <td>SARM1_i2</td>\n",
       "      <td>SARM1</td>\n",
       "      <td>primary</td>\n",
       "      <td>iPSC-induced neuron</td>\n",
       "      <td>False</td>\n",
       "      <td>healthy</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>human</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4029</td>\n",
       "      <td>11449.0</td>\n",
       "      <td>6.533322</td>\n",
       "      <td>6.716744</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4029</td>\n",
       "      <td>ENSG00000004139</td>\n",
       "      <td>0.282747</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AATTTCCCACGGCCAT</th>\n",
       "      <td>SARM1_i1</td>\n",
       "      <td>SARM1</td>\n",
       "      <td>primary</td>\n",
       "      <td>iPSC-induced neuron</td>\n",
       "      <td>False</td>\n",
       "      <td>healthy</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>human</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>5016</td>\n",
       "      <td>15883.0</td>\n",
       "      <td>5.546811</td>\n",
       "      <td>5.181641</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5016</td>\n",
       "      <td>ENSG00000004139</td>\n",
       "      <td>0.282747</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACAGGGAGTTGCATCA</th>\n",
       "      <td>SARM1_i1</td>\n",
       "      <td>SARM1</td>\n",
       "      <td>primary</td>\n",
       "      <td>iPSC-induced neuron</td>\n",
       "      <td>False</td>\n",
       "      <td>healthy</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>human</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3453</td>\n",
       "      <td>8636.0</td>\n",
       "      <td>5.280222</td>\n",
       "      <td>7.735063</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3453</td>\n",
       "      <td>ENSG00000004139</td>\n",
       "      <td>0.282747</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGGTTTCAGGACTAG</th>\n",
       "      <td>non-targeting_00655_i1</td>\n",
       "      <td>control</td>\n",
       "      <td>primary</td>\n",
       "      <td>iPSC-induced neuron</td>\n",
       "      <td>False</td>\n",
       "      <td>healthy</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>human</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>5319</td>\n",
       "      <td>18188.0</td>\n",
       "      <td>7.037607</td>\n",
       "      <td>9.528260</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5319</td>\n",
       "      <td>control</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGTTCAGTGTAAACA</th>\n",
       "      <td>non-targeting_00574_i1</td>\n",
       "      <td>control</td>\n",
       "      <td>primary</td>\n",
       "      <td>iPSC-induced neuron</td>\n",
       "      <td>False</td>\n",
       "      <td>healthy</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>human</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>4065</td>\n",
       "      <td>10626.0</td>\n",
       "      <td>6.446452</td>\n",
       "      <td>6.070017</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4065</td>\n",
       "      <td>control</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTACGTCACTGTCGG</th>\n",
       "      <td>non-targeting_00655_i1</td>\n",
       "      <td>control</td>\n",
       "      <td>primary</td>\n",
       "      <td>iPSC-induced neuron</td>\n",
       "      <td>False</td>\n",
       "      <td>healthy</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>human</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>3974</td>\n",
       "      <td>10514.0</td>\n",
       "      <td>4.441697</td>\n",
       "      <td>6.572189</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3974</td>\n",
       "      <td>control</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTCAGTTCCTAGCTC</th>\n",
       "      <td>non-targeting_00883_i1</td>\n",
       "      <td>control</td>\n",
       "      <td>primary</td>\n",
       "      <td>iPSC-induced neuron</td>\n",
       "      <td>False</td>\n",
       "      <td>healthy</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>human</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>5364</td>\n",
       "      <td>18802.0</td>\n",
       "      <td>4.286778</td>\n",
       "      <td>8.866078</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5364</td>\n",
       "      <td>control</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGATCCACGAAAGC</th>\n",
       "      <td>non-targeting_00110_i1</td>\n",
       "      <td>control</td>\n",
       "      <td>primary</td>\n",
       "      <td>iPSC-induced neuron</td>\n",
       "      <td>False</td>\n",
       "      <td>healthy</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>human</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>4295</td>\n",
       "      <td>12525.0</td>\n",
       "      <td>7.321357</td>\n",
       "      <td>5.996008</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4295</td>\n",
       "      <td>control</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>32300 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                guide_id perturbation tissue_type  \\\n",
       "AAGACAACAGCGTTTA                SARM1_i2        SARM1     primary   \n",
       "AAGCCATTCGAATGCT                SARM1_i2        SARM1     primary   \n",
       "AATCACGTCGCTAAAC                SARM1_i2        SARM1     primary   \n",
       "AATTTCCCACGGCCAT                SARM1_i1        SARM1     primary   \n",
       "ACAGGGAGTTGCATCA                SARM1_i1        SARM1     primary   \n",
       "...                                  ...          ...         ...   \n",
       "TTGGTTTCAGGACTAG  non-targeting_00655_i1      control     primary   \n",
       "TTGTTCAGTGTAAACA  non-targeting_00574_i1      control     primary   \n",
       "TTTACGTCACTGTCGG  non-targeting_00655_i1      control     primary   \n",
       "TTTCAGTTCCTAGCTC  non-targeting_00883_i1      control     primary   \n",
       "TTTGATCCACGAAAGC  non-targeting_00110_i1      control     primary   \n",
       "\n",
       "                             celltype  cancer  disease perturbation_type  \\\n",
       "AAGACAACAGCGTTTA  iPSC-induced neuron   False  healthy            CRISPR   \n",
       "AAGCCATTCGAATGCT  iPSC-induced neuron   False  healthy            CRISPR   \n",
       "AATCACGTCGCTAAAC  iPSC-induced neuron   False  healthy            CRISPR   \n",
       "AATTTCCCACGGCCAT  iPSC-induced neuron   False  healthy            CRISPR   \n",
       "ACAGGGAGTTGCATCA  iPSC-induced neuron   False  healthy            CRISPR   \n",
       "...                               ...     ...      ...               ...   \n",
       "TTGGTTTCAGGACTAG  iPSC-induced neuron   False  healthy            CRISPR   \n",
       "TTGTTCAGTGTAAACA  iPSC-induced neuron   False  healthy            CRISPR   \n",
       "TTTACGTCACTGTCGG  iPSC-induced neuron   False  healthy            CRISPR   \n",
       "TTTCAGTTCCTAGCTC  iPSC-induced neuron   False  healthy            CRISPR   \n",
       "TTTGATCCACGAAAGC  iPSC-induced neuron   False  healthy            CRISPR   \n",
       "\n",
       "                 organism batch  nperts  ngenes  ncounts  percent_mito  \\\n",
       "AAGACAACAGCGTTTA    human     1       1    5167  17733.0      9.479501   \n",
       "AAGCCATTCGAATGCT    human     1       1    5311  16777.0      3.683614   \n",
       "AATCACGTCGCTAAAC    human     1       1    4029  11449.0      6.533322   \n",
       "AATTTCCCACGGCCAT    human     1       1    5016  15883.0      5.546811   \n",
       "ACAGGGAGTTGCATCA    human     1       1    3453   8636.0      5.280222   \n",
       "...                   ...   ...     ...     ...      ...           ...   \n",
       "TTGGTTTCAGGACTAG    human     4       0    5319  18188.0      7.037607   \n",
       "TTGTTCAGTGTAAACA    human     4       0    4065  10626.0      6.446452   \n",
       "TTTACGTCACTGTCGG    human     4       0    3974  10514.0      4.441697   \n",
       "TTTCAGTTCCTAGCTC    human     4       0    5364  18802.0      4.286778   \n",
       "TTTGATCCACGAAAGC    human     4       0    4295  12525.0      7.321357   \n",
       "\n",
       "                  percent_ribo  percent_hemo  n_genes        gene_list  \\\n",
       "AAGACAACAGCGTTTA      6.609147           0.0     5167  ENSG00000004139   \n",
       "AAGCCATTCGAATGCT      6.967873           0.0     5311  ENSG00000004139   \n",
       "AATCACGTCGCTAAAC      6.716744           0.0     4029  ENSG00000004139   \n",
       "AATTTCCCACGGCCAT      5.181641           0.0     5016  ENSG00000004139   \n",
       "ACAGGGAGTTGCATCA      7.735063           0.0     3453  ENSG00000004139   \n",
       "...                        ...           ...      ...              ...   \n",
       "TTGGTTTCAGGACTAG      9.528260           0.0     5319          control   \n",
       "TTGTTCAGTGTAAACA      6.070017           0.0     4065          control   \n",
       "TTTACGTCACTGTCGG      6.572189           0.0     3974          control   \n",
       "TTTCAGTTCCTAGCTC      8.866078           0.0     5364          control   \n",
       "TTTGATCCACGAAAGC      5.996008           0.0     4295          control   \n",
       "\n",
       "                  perturbation_scores perturbation_status  \n",
       "AAGACAACAGCGTTTA             0.282747                True  \n",
       "AAGCCATTCGAATGCT             0.282747                True  \n",
       "AATCACGTCGCTAAAC             0.282747                True  \n",
       "AATTTCCCACGGCCAT             0.282747                True  \n",
       "ACAGGGAGTTGCATCA             0.282747                True  \n",
       "...                               ...                 ...  \n",
       "TTGGTTTCAGGACTAG                  NaN                 NaN  \n",
       "TTGTTCAGTGTAAACA                  NaN                 NaN  \n",
       "TTTACGTCACTGTCGG                  NaN                 NaN  \n",
       "TTTCAGTTCCTAGCTC                  NaN                 NaN  \n",
       "TTTGATCCACGAAAGC                  NaN                 NaN  \n",
       "\n",
       "[32300 rows x 19 columns]"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Classify cells\n",
    "perturbation_status = perturbation_scores > cutoff\n",
    "perturbation_status.name = 'perturbation_status'\n",
    "print(perturbation_status)\n",
    "adata.obs = pd.merge(left=adata.obs, right=perturbation_scores, left_on='gene_list', right_index=True, how='outer')\n",
    "adata.obs = pd.merge(left=adata.obs, right=perturbation_status, left_on='gene_list', right_index=True, how='outer')\n",
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['guide_id', 'perturbation', 'tissue_type', 'celltype', 'cancer',\n",
       "       'disease', 'perturbation_type', 'organism', 'batch', 'nperts', 'ngenes',\n",
       "       'ncounts', 'percent_mito', 'percent_ribo', 'percent_hemo', 'n_genes',\n",
       "       'gene_list', 'perturbation_scores', 'perturbation_status'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Target</th>\n",
       "      <th>Perturbed</th>\n",
       "      <th>Cell Line</th>\n",
       "      <th>ID</th>\n",
       "      <th>Experiment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAGACAACAGCGTTTA</th>\n",
       "      <td>ENSG00000004139</td>\n",
       "      <td>True</td>\n",
       "      <td>iPSC-induced neuron</td>\n",
       "      <td>AAGACAACAGCGTTTA</td>\n",
       "      <td>TianKampmann2021_CRISPRi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAGCCATTCGAATGCT</th>\n",
       "      <td>ENSG00000004139</td>\n",
       "      <td>True</td>\n",
       "      <td>iPSC-induced neuron</td>\n",
       "      <td>AAGCCATTCGAATGCT</td>\n",
       "      <td>TianKampmann2021_CRISPRi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AATCACGTCGCTAAAC</th>\n",
       "      <td>ENSG00000004139</td>\n",
       "      <td>True</td>\n",
       "      <td>iPSC-induced neuron</td>\n",
       "      <td>AATCACGTCGCTAAAC</td>\n",
       "      <td>TianKampmann2021_CRISPRi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AATTTCCCACGGCCAT</th>\n",
       "      <td>ENSG00000004139</td>\n",
       "      <td>True</td>\n",
       "      <td>iPSC-induced neuron</td>\n",
       "      <td>AATTTCCCACGGCCAT</td>\n",
       "      <td>TianKampmann2021_CRISPRi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACAGGGAGTTGCATCA</th>\n",
       "      <td>ENSG00000004139</td>\n",
       "      <td>True</td>\n",
       "      <td>iPSC-induced neuron</td>\n",
       "      <td>ACAGGGAGTTGCATCA</td>\n",
       "      <td>TianKampmann2021_CRISPRi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCATTCATCCGGGACT</th>\n",
       "      <td>ENSG00000291237</td>\n",
       "      <td>True</td>\n",
       "      <td>iPSC-induced neuron</td>\n",
       "      <td>TCATTCATCCGGGACT</td>\n",
       "      <td>TianKampmann2021_CRISPRi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCATTGTTCCCGAGGT</th>\n",
       "      <td>ENSG00000291237</td>\n",
       "      <td>True</td>\n",
       "      <td>iPSC-induced neuron</td>\n",
       "      <td>TCATTGTTCCCGAGGT</td>\n",
       "      <td>TianKampmann2021_CRISPRi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGTTCATGTCGACGCT</th>\n",
       "      <td>ENSG00000291237</td>\n",
       "      <td>True</td>\n",
       "      <td>iPSC-induced neuron</td>\n",
       "      <td>TGTTCATGTCGACGCT</td>\n",
       "      <td>TianKampmann2021_CRISPRi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTACGTTCAGGCTTGC</th>\n",
       "      <td>ENSG00000291237</td>\n",
       "      <td>True</td>\n",
       "      <td>iPSC-induced neuron</td>\n",
       "      <td>TTACGTTCAGGCTTGC</td>\n",
       "      <td>TianKampmann2021_CRISPRi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGCATTGTGCCTATA</th>\n",
       "      <td>ENSG00000291237</td>\n",
       "      <td>True</td>\n",
       "      <td>iPSC-induced neuron</td>\n",
       "      <td>TTGCATTGTGCCTATA</td>\n",
       "      <td>TianKampmann2021_CRISPRi</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>31863 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           Target Perturbed            Cell Line  \\\n",
       "AAGACAACAGCGTTTA  ENSG00000004139      True  iPSC-induced neuron   \n",
       "AAGCCATTCGAATGCT  ENSG00000004139      True  iPSC-induced neuron   \n",
       "AATCACGTCGCTAAAC  ENSG00000004139      True  iPSC-induced neuron   \n",
       "AATTTCCCACGGCCAT  ENSG00000004139      True  iPSC-induced neuron   \n",
       "ACAGGGAGTTGCATCA  ENSG00000004139      True  iPSC-induced neuron   \n",
       "...                           ...       ...                  ...   \n",
       "TCATTCATCCGGGACT  ENSG00000291237      True  iPSC-induced neuron   \n",
       "TCATTGTTCCCGAGGT  ENSG00000291237      True  iPSC-induced neuron   \n",
       "TGTTCATGTCGACGCT  ENSG00000291237      True  iPSC-induced neuron   \n",
       "TTACGTTCAGGCTTGC  ENSG00000291237      True  iPSC-induced neuron   \n",
       "TTGCATTGTGCCTATA  ENSG00000291237      True  iPSC-induced neuron   \n",
       "\n",
       "                                ID                Experiment  \n",
       "AAGACAACAGCGTTTA  AAGACAACAGCGTTTA  TianKampmann2021_CRISPRi  \n",
       "AAGCCATTCGAATGCT  AAGCCATTCGAATGCT  TianKampmann2021_CRISPRi  \n",
       "AATCACGTCGCTAAAC  AATCACGTCGCTAAAC  TianKampmann2021_CRISPRi  \n",
       "AATTTCCCACGGCCAT  AATTTCCCACGGCCAT  TianKampmann2021_CRISPRi  \n",
       "ACAGGGAGTTGCATCA  ACAGGGAGTTGCATCA  TianKampmann2021_CRISPRi  \n",
       "...                            ...                       ...  \n",
       "TCATTCATCCGGGACT  TCATTCATCCGGGACT  TianKampmann2021_CRISPRi  \n",
       "TCATTGTTCCCGAGGT  TCATTGTTCCCGAGGT  TianKampmann2021_CRISPRi  \n",
       "TGTTCATGTCGACGCT  TGTTCATGTCGACGCT  TianKampmann2021_CRISPRi  \n",
       "TTACGTTCAGGCTTGC  TTACGTTCAGGCTTGC  TianKampmann2021_CRISPRi  \n",
       "TTGCATTGTGCCTATA  TTGCATTGTGCCTATA  TianKampmann2021_CRISPRi  \n",
       "\n",
       "[31863 rows x 5 columns]"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "perturbation_df = pd.DataFrame(columns=['Target', 'Perturbed', 'Cell Line', 'ID', 'Experiment'])\n",
    "\n",
    "temp_df = adata.obs[(adata.obs['gene_list'] != 'control') & (adata.obs['gene_list'] != '')]\n",
    "\n",
    "perturbation_df['Target'] = temp_df['gene_list']\n",
    "perturbation_df['Perturbed'] = temp_df['perturbation_status']\n",
    "perturbation_df['Cell Line'] = temp_df['celltype']\n",
    "perturbation_df['ID'] = temp_df.index\n",
    "perturbation_df['Experiment'] = perturbation_screen_name\n",
    "perturbation_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "perturbation_df.to_csv(f'../data/perturbation_screens/{perturbation_screen_name}_processed.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_expr_ref_celline = pd.read_csv('../data/expression_reference/rna_celline.tsv', sep='\\t')\n",
    "df_expr_ref_cell_type = pd.read_csv('../data/expression_reference/rna_single_cell_type.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Cell type</th>\n",
       "      <th>nTPM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>Adipocytes</td>\n",
       "      <td>147.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>Alveolar cells type 1</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>Alveolar cells type 2</td>\n",
       "      <td>10.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>Astrocytes</td>\n",
       "      <td>13.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>B-cells</td>\n",
       "      <td>1.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1626637</th>\n",
       "      <td>ENSG00000288695</td>\n",
       "      <td>ENSG00000288695</td>\n",
       "      <td>Squamous epithelial cells</td>\n",
       "      <td>0.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1626638</th>\n",
       "      <td>ENSG00000288695</td>\n",
       "      <td>ENSG00000288695</td>\n",
       "      <td>Suprabasal keratinocytes</td>\n",
       "      <td>0.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1626639</th>\n",
       "      <td>ENSG00000288695</td>\n",
       "      <td>ENSG00000288695</td>\n",
       "      <td>Syncytiotrophoblasts</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1626640</th>\n",
       "      <td>ENSG00000288695</td>\n",
       "      <td>ENSG00000288695</td>\n",
       "      <td>T-cells</td>\n",
       "      <td>0.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1626641</th>\n",
       "      <td>ENSG00000288695</td>\n",
       "      <td>ENSG00000288695</td>\n",
       "      <td>Undifferentiated cells</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1626642 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Gene        Gene name                  Cell type   nTPM\n",
       "0        ENSG00000000003           TSPAN6                 Adipocytes  147.5\n",
       "1        ENSG00000000003           TSPAN6      Alveolar cells type 1    6.0\n",
       "2        ENSG00000000003           TSPAN6      Alveolar cells type 2   10.9\n",
       "3        ENSG00000000003           TSPAN6                 Astrocytes   13.6\n",
       "4        ENSG00000000003           TSPAN6                    B-cells    1.5\n",
       "...                  ...              ...                        ...    ...\n",
       "1626637  ENSG00000288695  ENSG00000288695  Squamous epithelial cells    0.1\n",
       "1626638  ENSG00000288695  ENSG00000288695   Suprabasal keratinocytes    0.3\n",
       "1626639  ENSG00000288695  ENSG00000288695       Syncytiotrophoblasts    0.0\n",
       "1626640  ENSG00000288695  ENSG00000288695                    T-cells    0.1\n",
       "1626641  ENSG00000288695  ENSG00000288695     Undifferentiated cells    0.0\n",
       "\n",
       "[1626642 rows x 4 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_expr_ref_cell_type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cell Line</th>\n",
       "      <th>Gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>K562</td>\n",
       "      <td>ENSG00000000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>K562</td>\n",
       "      <td>ENSG00000000419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>K562</td>\n",
       "      <td>ENSG00000000457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>K562</td>\n",
       "      <td>ENSG00000000460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>K562</td>\n",
       "      <td>ENSG00000000938</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16470</th>\n",
       "      <td>iPSC-induced neuron</td>\n",
       "      <td>ENSG00000286219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16471</th>\n",
       "      <td>iPSC-induced neuron</td>\n",
       "      <td>ENSG00000286237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16472</th>\n",
       "      <td>iPSC-induced neuron</td>\n",
       "      <td>ENSG00000287542</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16473</th>\n",
       "      <td>iPSC-induced neuron</td>\n",
       "      <td>ENSG00000288602</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16474</th>\n",
       "      <td>iPSC-induced neuron</td>\n",
       "      <td>ENSG00000288675</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>43821 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Cell Line             Gene\n",
       "0                     K562  ENSG00000000003\n",
       "1                     K562  ENSG00000000419\n",
       "2                     K562  ENSG00000000457\n",
       "3                     K562  ENSG00000000460\n",
       "4                     K562  ENSG00000000938\n",
       "...                    ...              ...\n",
       "16470  iPSC-induced neuron  ENSG00000286219\n",
       "16471  iPSC-induced neuron  ENSG00000286237\n",
       "16472  iPSC-induced neuron  ENSG00000287542\n",
       "16473  iPSC-induced neuron  ENSG00000288602\n",
       "16474  iPSC-induced neuron  ENSG00000288675\n",
       "\n",
       "[43821 rows x 2 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_expr_ref = pd.DataFrame(columns = ['Cell Line', 'Gene'])\n",
    "\n",
    "k562 = {'Cell Line': 'K562', 'Gene': list(df_expr_ref_celline[(df_expr_ref_celline['Cell line'] == 'K-562') & (df_expr_ref_celline['TPM'] > 0.0)]['Gene'].unique())}\n",
    "df_expr_ref = pd.concat([df_expr_ref, pd.DataFrame(k562)])\n",
    "melanocytes = {'Cell Line': 'melanocytes', 'Gene': list(df_expr_ref_cell_type[(df_expr_ref_cell_type['Cell type'] == 'Melanocytes') & (df_expr_ref_cell_type['nTPM'] > 0.0)]['Gene'].unique())}\n",
    "df_expr_ref = pd.concat([df_expr_ref, pd.DataFrame(melanocytes)])\n",
    "neurons = {'Cell Line': 'iPSC-induced neuron', 'Gene': list(df_expr_ref_cell_type[((df_expr_ref_cell_type['Cell type'] == 'Excitatory neurons') | (df_expr_ref_cell_type['Cell type'] == 'Inhibitory neurons')) & (df_expr_ref_cell_type['nTPM'] > 0.0)]['Gene'].unique())}\n",
    "df_expr_ref = pd.concat([df_expr_ref, pd.DataFrame(neurons)])\n",
    "\n",
    "df_expr_ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Gene\n",
       "ENSG00000285437    3\n",
       "ENSG00000285053    3\n",
       "ENSG00000003249    3\n",
       "ENSG00000003147    3\n",
       "ENSG00000003137    3\n",
       "                  ..\n",
       "ENSG00000282757    1\n",
       "ENSG00000281991    1\n",
       "ENSG00000280709    1\n",
       "ENSG00000280165    1\n",
       "ENSG00000279408    1\n",
       "Name: count, Length: 18097, dtype: int64"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_expr_ref['Gene'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Axes: xlabel='count', ylabel='Count'>"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAk0AAAGwCAYAAAC0HlECAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjEsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvc2/+5QAAAAlwSFlzAAAPYQAAD2EBqD+naQAALqpJREFUeJzt3X9UVWW+x/HPUX7qhRNggEyYVIQaVkZeRJtRQ9EKrdW6Y13rXCtTi9Sh9FqOU1HriumU2khj6phaZnbnqjPNrUgsxTF/o1SY0q2xxALRCQ/+IDDY94/GvTqC9nBEOYd5v9Y6a3X2/u59nq/bJz4+Z5+Dw7IsSwAAADindq09AAAAAH9AaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADAQ0NoDaEsaGhr0zTffKCwsTA6Ho7WHAwAADFiWpWPHjikuLk7t2p19PYnQ1IK++eYbxcfHt/YwAACAF8rKynTZZZeddT+hqQWFhYVJ+uEPPTw8vJVHAwAATFRXVys+Pt7+OX42hKYWdPotufDwcEITAAB+5qdureFGcAAAAAOEJgAAAAOEJgAAAAOEJgAAAAOEJgAAAAOEJgAAAAOEJgAAAAOEJgAAAAOEJgAAAAOEJgAAAAOEJgAAAAOEJgAAAAOEJgAAAAOEJgAAAAMBrT0AAADQ9h04cEBHjhw5r3N06tRJXbp0aaERNR+hCQAAXFAHDhxQt27dVVNz8rzOExraQfv27W214ERoAgAAF9SRI0dUU3NSqQ88rfDOXb06R3X5l9r2yjM6cuQIoQkAALRt4Z27KrJLUmsPw2vcCA4AAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCgVUPTxo0bNWzYMMXFxcnhcOhPf/qTx37LspSTk6O4uDiFhoZqwIAB2rNnj0dNbW2tJkyYoE6dOqljx44aPny4Dh486FFTVVUll8slp9Mpp9Mpl8ulo0ePetQcOHBAw4YNU8eOHdWpUydNnDhRdXV1F6JtAADgh1o1NJ04cULXXXed8vLymtw/a9YszZ49W3l5edqxY4diY2M1ePBgHTt2zK7Jzs7WmjVrtHLlSm3atEnHjx9XZmam6uvr7ZqRI0equLhY+fn5ys/PV3FxsVwul72/vr5et912m06cOKFNmzZp5cqVWrVqlSZNmnThmgcAAH4loDVf/JZbbtEtt9zS5D7LsjR37lxNmzZNd955pyRp2bJliomJ0YoVKzRu3Di53W4tXrxYr732mgYNGiRJWr58ueLj47Vu3ToNGTJEe/fuVX5+vrZu3arU1FRJ0qJFi5SWlqbS0lIlJSVp7dq1+vTTT1VWVqa4uDhJ0gsvvKD77rtP06dPV3h4+EX40wAAAL7MZ+9p2r9/vyoqKpSRkWFvCw4OVv/+/bV582ZJUlFRkU6dOuVRExcXp+TkZLtmy5YtcjqddmCSpD59+sjpdHrUJCcn24FJkoYMGaLa2loVFRWddYy1tbWqrq72eAAAgLbJZ0NTRUWFJCkmJsZje0xMjL2voqJCQUFBioiIOGdNdHR0o/NHR0d71Jz5OhEREQoKCrJrmjJjxgz7Pimn06n4+PhmdgkAAPyFz4am0xwOh8dzy7IabTvTmTVN1XtTc6apU6fK7Xbbj7KysnOOCwAA+C+fDU2xsbGS1Gilp7Ky0l4Vio2NVV1dnaqqqs5Zc+jQoUbnP3z4sEfNma9TVVWlU6dONVqB+rHg4GCFh4d7PAAAQNvks6EpISFBsbGxKigosLfV1dWpsLBQffv2lSSlpKQoMDDQo6a8vFwlJSV2TVpamtxut7Zv327XbNu2TW6326OmpKRE5eXlds3atWsVHByslJSUC9onAADwD6366bnjx4/r888/t5/v379fxcXFioyMVJcuXZSdna3c3FwlJiYqMTFRubm56tChg0aOHClJcjqdGj16tCZNmqSoqChFRkZq8uTJ6tmzp/1puu7du2vo0KEaM2aMFixYIEkaO3asMjMzlZSUJEnKyMhQjx495HK59Nvf/lbffvutJk+erDFjxrB6BAAAJLVyaNq5c6cGDhxoP3/sscckSaNGjdLSpUs1ZcoU1dTUKCsrS1VVVUpNTdXatWsVFhZmHzNnzhwFBARoxIgRqqmpUXp6upYuXar27dvbNa+//romTpxof8pu+PDhHt8N1b59e7399tvKyspSv379FBoaqpEjR+r555+/0H8EAADATzgsy7JaexBtRXV1tZxOp9xuNytUAAD8w65du5SSkqLB05YoskuSV+f49kCpCqbfr6KiIt1www0tOj7Tn98+e08TAACALyE0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGPDp0PT999/rN7/5jRISEhQaGqorrrhCzz77rBoaGuway7KUk5OjuLg4hYaGasCAAdqzZ4/HeWprazVhwgR16tRJHTt21PDhw3Xw4EGPmqqqKrlcLjmdTjmdTrlcLh09evRitAkAAPyAT4emmTNn6uWXX1ZeXp727t2rWbNm6be//a3mzZtn18yaNUuzZ89WXl6eduzYodjYWA0ePFjHjh2za7Kzs7VmzRqtXLlSmzZt0vHjx5WZman6+nq7ZuTIkSouLlZ+fr7y8/NVXFwsl8t1UfsFAAC+K6C1B3AuW7Zs0e23367bbrtNktS1a1e98cYb2rlzp6QfVpnmzp2radOm6c4775QkLVu2TDExMVqxYoXGjRsnt9utxYsX67XXXtOgQYMkScuXL1d8fLzWrVunIUOGaO/evcrPz9fWrVuVmpoqSVq0aJHS0tJUWlqqpKSkVugeAAD4Ep9eabrpppv0/vvv67PPPpMkffTRR9q0aZNuvfVWSdL+/ftVUVGhjIwM+5jg4GD1799fmzdvliQVFRXp1KlTHjVxcXFKTk62a7Zs2SKn02kHJknq06ePnE6nXdOU2tpaVVdXezwAAEDb5NMrTY8//rjcbre6deum9u3bq76+XtOnT9e///u/S5IqKiokSTExMR7HxcTE6KuvvrJrgoKCFBER0ajm9PEVFRWKjo5u9PrR0dF2TVNmzJihZ555xvsGAQCA3/DplaY333xTy5cv14oVK7Rr1y4tW7ZMzz//vJYtW+ZR53A4PJ5bltVo25nOrGmq/qfOM3XqVLndbvtRVlZm0hYAAPBDPr3S9J//+Z964okndPfdd0uSevbsqa+++kozZszQqFGjFBsbK+mHlaLOnTvbx1VWVtqrT7Gxsaqrq1NVVZXHalNlZaX69u1r1xw6dKjR6x8+fLjRKtaPBQcHKzg4+PwbBQAAPs+nV5pOnjypdu08h9i+fXv7KwcSEhIUGxurgoICe39dXZ0KCwvtQJSSkqLAwECPmvLycpWUlNg1aWlpcrvd2r59u12zbds2ud1uuwYAAPxz8+mVpmHDhmn69Onq0qWLrrnmGu3evVuzZ8/WAw88IOmHt9Sys7OVm5urxMREJSYmKjc3Vx06dNDIkSMlSU6nU6NHj9akSZMUFRWlyMhITZ48WT179rQ/Tde9e3cNHTpUY8aM0YIFCyRJY8eOVWZmJp+cAwAAknw8NM2bN09PPvmksrKyVFlZqbi4OI0bN05PPfWUXTNlyhTV1NQoKytLVVVVSk1N1dq1axUWFmbXzJkzRwEBARoxYoRqamqUnp6upUuXqn379nbN66+/rokTJ9qfshs+fLjy8vIuXrMAAMCnOSzLslp7EG1FdXW1nE6n3G63wsPDW3s4AAD4hF27diklJUWDpy1RZBfv3sH59kCpCqbfr6KiIt1www0tOj7Tn98+fU8TAACAryA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGPAqNF1xxRX6+9//3mj70aNHdcUVV5z3oAAAAHyNV6Hpyy+/VH19faPttbW1+vrrr897UAAAAL4moDnFb731lv3f7733npxOp/28vr5e77//vrp27dpigwMAAPAVzQpNd9xxhyTJ4XBo1KhRHvsCAwPVtWtXvfDCCy02OAAAAF/RrNDU0NAgSUpISNCOHTvUqVOnCzIoAAAAX+PVPU379++/aIHp66+/1r333quoqCh16NBB119/vYqKiuz9lmUpJydHcXFxCg0N1YABA7Rnzx6Pc9TW1mrChAnq1KmTOnbsqOHDh+vgwYMeNVVVVXK5XHI6nXI6nXK5XDp69OjFaBEAAPiBZq00/dj777+v999/X5WVlfYK1GmvvPLKeQ9M+iHI9OvXTwMHDtS7776r6OhoffHFF7rkkkvsmlmzZmn27NlaunSprr76av3Xf/2XBg8erNLSUoWFhUmSsrOz9Ze//EUrV65UVFSUJk2apMzMTBUVFal9+/aSpJEjR+rgwYPKz8+XJI0dO1Yul0t/+ctfWqQXAADg37wKTc8884yeffZZ3XjjjercubMcDkdLj0uSNHPmTMXHx2vJkiX2th/faG5ZlubOnatp06bpzjvvlCQtW7ZMMTExWrFihcaNGye3263Fixfrtdde06BBgyRJy5cvV3x8vNatW6chQ4Zo7969ys/P19atW5WamipJWrRokdLS0lRaWqqkpKQL0h8AAPAfXoWml19+WUuXLpXL5Wrp8Xh46623NGTIEP3yl79UYWGhfvaznykrK0tjxoyR9MPbhBUVFcrIyLCPCQ4OVv/+/bV582aNGzdORUVFOnXqlEdNXFyckpOTtXnzZg0ZMkRbtmyR0+m0A5Mk9enTR06nU5s3bz5raKqtrVVtba39vLq6uqX/CAAAgI/w6p6muro69e3bt6XH0sjf/vY3zZ8/X4mJiXrvvff00EMPaeLEiXr11VclSRUVFZKkmJgYj+NiYmLsfRUVFQoKClJERMQ5a6Kjoxu9fnR0tF3TlBkzZtj3QDmdTsXHx3vfLAAA8GlehaYHH3xQK1asaOmxNNLQ0KAbbrhBubm56tWrl8aNG6cxY8Zo/vz5HnVnvj1oWdZPvmV4Zk1T9T91nqlTp8rtdtuPsrIyk7YAAIAf8urtue+++04LFy7UunXrdO211yowMNBj/+zZs1tkcJ07d1aPHj08tnXv3l2rVq2SJMXGxkr6YaWoc+fOdk1lZaW9+hQbG6u6ujpVVVV5rDZVVlbaq2WxsbE6dOhQo9c/fPhwo1WsHwsODlZwcLCX3QEAAH/i1UrTxx9/rOuvv17t2rVTSUmJdu/ebT+Ki4tbbHD9+vVTaWmpx7bPPvtMl19+uaQfvi8qNjZWBQUF9v66ujoVFhbagSglJUWBgYEeNeXl5SopKbFr0tLS5Ha7tX37drtm27ZtcrvdF+VtSAAA4Pu8Wmlav359S4+jSY8++qj69u2r3NxcjRgxQtu3b9fChQu1cOFCST+8pZadna3c3FwlJiYqMTFRubm56tChg0aOHClJcjqdGj16tCZNmqSoqChFRkZq8uTJ6tmzp/1puu7du2vo0KEaM2aMFixYIOmHrxzIzMzkk3MAAEDSeXxP08XQu3dvrVmzRlOnTtWzzz6rhIQEzZ07V/fcc49dM2XKFNXU1CgrK0tVVVVKTU3V2rVr7e9okqQ5c+YoICBAI0aMUE1NjdLT07V06VL7O5ok6fXXX9fEiRPtT9kNHz5ceXl5F69ZAADg0xyWZVnNPWjgwIHnvEH6gw8+OK9B+avq6mo5nU653W6Fh4e39nAAAPAJu3btUkpKigZPW6LILt69g/PtgVIVTL9fRUVFuuGGG1p0fKY/v71aabr++us9np86dUrFxcUqKSlp9It8AQAA2gKvQtOcOXOa3J6Tk6Pjx4+f14AAAAB8kVefnjube++9t8V+7xwAAIAvadHQtGXLFoWEhLTkKQEAAHyCV2/Pnf7luKdZlqXy8nLt3LlTTz75ZIsMDAAAwJd4FZqcTqfH83bt2ikpKUnPPvusxy/GBQAAaCu8Ck1Llixp6XEAAAD4tPP6csuioiLt3btXDodDPXr0UK9evVpqXAAAAD7Fq9BUWVmpu+++Wxs2bNAll1wiy7Lkdrs1cOBArVy5UpdeemlLjxMAAKBVefXpuQkTJqi6ulp79uzRt99+q6qqKpWUlKi6uloTJ05s6TECAAC0Oq9WmvLz87Vu3Tp1797d3tajRw+99NJL3AgOAADaJK9WmhoaGhQYGNhoe2BgoBoaGs57UAAAAL7Gq9B0880361e/+pW++eYbe9vXX3+tRx99VOnp6S02OAAAAF/hVWjKy8vTsWPH1LVrV1155ZW66qqrlJCQoGPHjmnevHktPUYAAIBW59U9TfHx8dq1a5cKCgq0b98+WZalHj16aNCgQS09PgAAAJ/QrJWmDz74QD169FB1dbUkafDgwZowYYImTpyo3r1765prrtFf//rXCzJQAACA1tSs0DR37lyNGTNG4eHhjfY5nU6NGzdOs2fPbrHBAQAA+IpmhaaPPvpIQ4cOPev+jIwMFRUVnfegAAAAfE2zQtOhQ4ea/KqB0wICAnT48OHzHhQAAICvaVZo+tnPfqZPPvnkrPs//vhjde7c+bwHBQAA4GuaFZpuvfVWPfXUU/ruu+8a7aupqdHTTz+tzMzMFhscAACAr2jWVw785je/0erVq3X11Vdr/PjxSkpKksPh0N69e/XSSy+pvr5e06ZNu1BjBQAAaDXNCk0xMTHavHmzHn74YU2dOlWWZUmSHA6HhgwZot///veKiYm5IAMFAABoTc3+csvLL79c77zzjqqqqvT555/LsiwlJiYqIiLiQowPAADAJ3j1jeCSFBERod69e7fkWAAAAHyWV797DgAA4J8NoQkAAMAAoQkAAMAAoQkAAMAAoQkAAMAAoQkAAMAAoQkAAMAAoQkAAMAAoQkAAMAAoQkAAMAAoQkAAMAAoQkAAMAAoQkAAMAAoQkAAMAAoQkAAMAAoQkAAMAAoQkAAMAAoQkAAMAAoQkAAMAAoQkAAMAAoQkAAMAAoQkAAMAAoQkAAMAAoQkAAMAAoQkAAMAAoQkAAMAAoQkAAMAAoQkAAMAAoQkAAMCAX4WmGTNmyOFwKDs7295mWZZycnIUFxen0NBQDRgwQHv27PE4rra2VhMmTFCnTp3UsWNHDR8+XAcPHvSoqaqqksvlktPplNPplMvl0tGjRy9CVwAAwB/4TWjasWOHFi5cqGuvvdZj+6xZszR79mzl5eVpx44dio2N1eDBg3Xs2DG7Jjs7W2vWrNHKlSu1adMmHT9+XJmZmaqvr7drRo4cqeLiYuXn5ys/P1/FxcVyuVwXrT8AAODb/CI0HT9+XPfcc48WLVqkiIgIe7tlWZo7d66mTZumO++8U8nJyVq2bJlOnjypFStWSJLcbrcWL16sF154QYMGDVKvXr20fPlyffLJJ1q3bp0kae/evcrPz9cf/vAHpaWlKS0tTYsWLdL//u//qrS0tFV6BgAAvsUvQtMjjzyi2267TYMGDfLYvn//flVUVCgjI8PeFhwcrP79+2vz5s2SpKKiIp06dcqjJi4uTsnJyXbNli1b5HQ6lZqaatf06dNHTqfTrmlKbW2tqqurPR4AAKBtCmjtAfyUlStXateuXdqxY0ejfRUVFZKkmJgYj+0xMTH66quv7JqgoCCPFarTNaePr6ioUHR0dKPzR0dH2zVNmTFjhp555pnmNQQAAPyST680lZWV6Ve/+pWWL1+ukJCQs9Y5HA6P55ZlNdp2pjNrmqr/qfNMnTpVbrfbfpSVlZ3zNQEAgP/y6ZWmoqIiVVZWKiUlxd5WX1+vjRs3Ki8vz77fqKKiQp07d7ZrKisr7dWn2NhY1dXVqaqqymO1qbKyUn379rVrDh061Oj1Dx8+3GgV68eCg4MVHBx8fk0C8AkHDhzQkSNHzuscnTp1UpcuXVpoRAB8jU+HpvT0dH3yySce2+6//35169ZNjz/+uK644grFxsaqoKBAvXr1kiTV1dWpsLBQM2fOlCSlpKQoMDBQBQUFGjFihCSpvLxcJSUlmjVrliQpLS1Nbrdb27dv17/+679KkrZt2ya3220HKwBt14EDB9StW3fV1Jw8r/OEhnbQvn17CU5AG+XToSksLEzJycke2zp27KioqCh7e3Z2tnJzc5WYmKjExETl5uaqQ4cOGjlypCTJ6XRq9OjRmjRpkqKiohQZGanJkyerZ8+e9o3l3bt319ChQzVmzBgtWLBAkjR27FhlZmYqKSnpInYMoDUcOXJENTUnlfrA0wrv3NWrc1SXf6ltrzyjI0eOEJqANsqnQ5OJKVOmqKamRllZWaqqqlJqaqrWrl2rsLAwu2bOnDkKCAjQiBEjVFNTo/T0dC1dulTt27e3a15//XVNnDjR/pTd8OHDlZeXd9H7AdB6wjt3VWQX/qEEoGl+F5o2bNjg8dzhcCgnJ0c5OTlnPSYkJETz5s3TvHnzzloTGRmp5cuXt9AoAQBAW+PTn54DAADwFYQmAAAAA4QmAAAAA4QmAAAAA4QmAAAAA4QmAAAAA4QmAAAAA4QmAAAAA4QmAAAAA4QmAAAAA4QmAAAAA4QmAAAAA4QmAAAAA4QmAAAAA4QmAAAAAwGtPQCYOXDggI4cOXJe5+jUqZO6dOnSQiMCAOCfC6HJDxw4cEDdunVXTc3J8zpPaGgH7du3l+AEAIAXCE1+4MiRI6qpOanUB55WeOeuXp2juvxLbXvlGR05coTQBACAFwhNfiS8c1dFdklq7WEAAPBPiRvBAQAADBCaAAAADBCaAAAADBCaAAAADBCaAAAADBCaAAAADBCaAAAADBCaAAAADBCaAAAADBCaAAAADBCaAAAADBCaAAAADBCaAAAADBCaAAAADBCaAAAADBCaAAAADBCaAAAADBCaAAAADBCaAAAADBCaAAAADBCaAAAADBCaAAAADBCaAAAADBCaAAAADBCaAAAADBCaAAAADBCaAAAADBCaAAAADBCaAAAADBCaAAAADBCaAAAADBCaAAAADBCaAAAADBCaAAAADBCaAAAADPh0aJoxY4Z69+6tsLAwRUdH64477lBpaalHjWVZysnJUVxcnEJDQzVgwADt2bPHo6a2tlYTJkxQp06d1LFjRw0fPlwHDx70qKmqqpLL5ZLT6ZTT6ZTL5dLRo0cvdIsAAMBP+HRoKiws1COPPKKtW7eqoKBA33//vTIyMnTixAm7ZtasWZo9e7by8vK0Y8cOxcbGavDgwTp27Jhdk52drTVr1mjlypXatGmTjh8/rszMTNXX19s1I0eOVHFxsfLz85Wfn6/i4mK5XK6L2i8AAPBdAa09gHPJz8/3eL5kyRJFR0erqKhIv/jFL2RZlubOnatp06bpzjvvlCQtW7ZMMTExWrFihcaNGye3263Fixfrtdde06BBgyRJy5cvV3x8vNatW6chQ4Zo7969ys/P19atW5WamipJWrRokdLS0lRaWqqkpKQmx1dbW6va2lr7eXV19YX4YwAAAD7Ap1eazuR2uyVJkZGRkqT9+/eroqJCGRkZdk1wcLD69++vzZs3S5KKiop06tQpj5q4uDglJyfbNVu2bJHT6bQDkyT16dNHTqfTrmnKjBkz7LfznE6n4uPjW65ZAADgU/wmNFmWpccee0w33XSTkpOTJUkVFRWSpJiYGI/amJgYe19FRYWCgoIUERFxzpro6OhGrxkdHW3XNGXq1Klyu932o6yszPsGAQCAT/Ppt+d+bPz48fr444+1adOmRvscDofHc8uyGm0705k1TdX/1HmCg4MVHBz8U0MHAABtgF+sNE2YMEFvvfWW1q9fr8suu8zeHhsbK0mNVoMqKyvt1afY2FjV1dWpqqrqnDWHDh1q9LqHDx9utIoFAAD+Ofl0aLIsS+PHj9fq1av1wQcfKCEhwWN/QkKCYmNjVVBQYG+rq6tTYWGh+vbtK0lKSUlRYGCgR015eblKSkrsmrS0NLndbm3fvt2u2bZtm9xut10DAAD+ufn023OPPPKIVqxYoT//+c8KCwuzV5ScTqdCQ0PlcDiUnZ2t3NxcJSYmKjExUbm5uerQoYNGjhxp144ePVqTJk1SVFSUIiMjNXnyZPXs2dP+NF337t01dOhQjRkzRgsWLJAkjR07VpmZmWf95BwAAPjn4tOhaf78+ZKkAQMGeGxfsmSJ7rvvPknSlClTVFNTo6ysLFVVVSk1NVVr165VWFiYXT9nzhwFBARoxIgRqqmpUXp6upYuXar27dvbNa+//romTpxof8pu+PDhysvLu7ANAgAAv+HTocmyrJ+scTgcysnJUU5OzllrQkJCNG/ePM2bN++sNZGRkVq+fLk3wwQAAP8EfPqeJgAAAF9BaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaDrD73//eyUkJCgkJEQpKSn661//2tpDAgAAPoDQ9CNvvvmmsrOzNW3aNO3evVs///nPdcstt+jAgQOtPTQAANDKCE0/Mnv2bI0ePVoPPvigunfvrrlz5yo+Pl7z589v7aEBAIBWFtDaA/AVdXV1Kioq0hNPPOGxPSMjQ5s3b27ymNraWtXW1trP3W63JKm6urpFx3b8+HFJ0rdfler72hqvzlFd8cNqWVFRkX0+b7Rr104NDQ2tdryvnMMXxuAr5/CFMZzvOUpLSyUxx9raOXxhDC1xDl8Yw/meoyXn2PHjx1v85+zp81mWde5CC5ZlWdbXX39tSbI+/PBDj+3Tp0+3rr766iaPefrppy1JPHjw4MGDB4828CgrKztnVmCl6QwOh8PjuWVZjbadNnXqVD322GP284aGBn377beKioo66zHeqK6uVnx8vMrKyhQeHt5i5/Ulbb1H+vN/bb1H+vN/bb3HC9mfZVk6duyY4uLizllHaPqHTp06qX379qqoqPDYXllZqZiYmCaPCQ4OVnBwsMe2Sy655EINUeHh4W1yIvxYW++R/vxfW++R/vxfW+/xQvXndDp/soYbwf8hKChIKSkpKigo8NheUFCgvn37ttKoAACAr2Cl6Ucee+wxuVwu3XjjjUpLS9PChQt14MABPfTQQ609NAAA0MoITT9y11136e9//7ueffZZlZeXKzk5We+8844uv/zyVh1XcHCwnn766UZvBbYlbb1H+vN/bb1H+vN/bb1HX+jPYVk/9fk6AAAAcE8TAACAAUITAACAAUITAACAAUITAACAAULTRbZx40YNGzZMcXFxcjgc+tOf/vSTxxQWFiolJUUhISG64oor9PLLLzeqWbVqlXr06KHg4GD16NFDa9asuQCjN9PcHlevXq3Bgwfr0ksvVXh4uNLS0vTee+951CxdulQOh6PR47vvvruAnTStuf1t2LChybHv27fPo85XrmFz+7vvvvua7O+aa66xa3zp+s2YMUO9e/dWWFiYoqOjdccdd9i/F+tc/GkeetOjP81Db/rzp3noTX/+Ng/nz5+va6+91v6iyrS0NL377rvnPMYX5iCh6SI7ceKErrvuOuXl5RnV79+/X7feeqt+/vOfa/fu3fr1r3+tiRMnatWqVXbNli1bdNddd8nlcumjjz6Sy+XSiBEjtG3btgvVxjk1t8eNGzdq8ODBeuedd1RUVKSBAwdq2LBh2r17t0ddeHi4ysvLPR4hISEXooVzam5/p5WWlnqMPTEx0d7nS9ewuf29+OKLHn2VlZUpMjJSv/zlLz3qfOX6FRYW6pFHHtHWrVtVUFCg77//XhkZGTpx4sRZj/G3eehNj/40D73p7zR/mIfe9Odv8/Cyyy7Tc889p507d2rnzp26+eabdfvtt2vPnj1N1vvMHGyR33YLr0iy1qxZc86aKVOmWN26dfPYNm7cOKtPnz728xEjRlhDhw71qBkyZIh19913t9hYvWXSY1N69OhhPfPMM/bzJUuWWE6ns+UG1kJM+lu/fr0lyaqqqjprja9eQ2+u35o1ayyHw2F9+eWX9jZfvX6WZVmVlZWWJKuwsPCsNf4+D016bIq/zEOT/vx5Hnpz/fxtHlqWZUVERFh/+MMfmtznK3OQlSYft2XLFmVkZHhsGzJkiHbu3KlTp06ds2bz5s0XbZwtqaGhQceOHVNkZKTH9uPHj+vyyy/XZZddpszMzEb/AvZ1vXr1UufOnZWenq7169d77GtL13Dx4sUaNGhQoy+F9dXr53a7JanR37cf8/d5aNLjmfxpHjanP3+ch95cP3+ah/X19Vq5cqVOnDihtLS0Jmt8ZQ4SmnxcRUVFo18YHBMTo++//15Hjhw5Z82Zv3zYX7zwwgs6ceKERowYYW/r1q2bli5dqrfeektvvPGGQkJC1K9fP/3f//1fK47UTOfOnbVw4UKtWrVKq1evVlJSktLT07Vx40a7pq1cw/Lycr377rt68MEHPbb76vWzLEuPPfaYbrrpJiUnJ5+1zp/noWmPZ/KXeWjan7/OQ2+un7/Mw08++UT/8i//ouDgYD300ENas2aNevTo0WStr8xBfo2KH3A4HB7PrX98ifuPtzdVc+Y2f/DGG28oJydHf/7znxUdHW1v79Onj/r06WM/79evn2644QbNmzdPv/vd71pjqMaSkpKUlJRkP09LS1NZWZmef/55/eIXv7C3t4VruHTpUl1yySW64447PLb76vUbP368Pv74Y23atOkna/11Hjanx9P8aR6a9uev89Cb6+cv8zApKUnFxcU6evSoVq1apVGjRqmwsPCswckX5iArTT4uNja2UUqurKxUQECAoqKizllzZuL2dW+++aZGjx6t//7v/9agQYPOWduuXTv17t271VcqvNWnTx+PsbeFa2hZll555RW5XC4FBQWds9YXrt+ECRP01ltvaf369brsssvOWeuv87A5PZ7mT/PQm/5+zNfnoTf9+dM8DAoK0lVXXaUbb7xRM2bM0HXXXacXX3yxyVpfmYOEJh+XlpamgoICj21r167VjTfeqMDAwHPW9O3b96KN83y98cYbuu+++7RixQrddtttP1lvWZaKi4vVuXPnizC6lrd7926PsbeFa1hYWKjPP/9co0eP/sna1rx+lmVp/PjxWr16tT744AMlJCT85DH+Ng+96VHyn3nobX9n8tV5eD79+cs8bIplWaqtrW1yn8/MwRa7pRxGjh07Zu3evdvavXu3JcmaPXu2tXv3buurr76yLMuynnjiCcvlctn1f/vb36wOHTpYjz76qPXpp59aixcvtgIDA63/+Z//sWs+/PBDq3379tZzzz1n7d2713ruueesgIAAa+vWrRe9P8tqfo8rVqywAgICrJdeeskqLy+3H0ePHrVrcnJyrPz8fOuLL76wdu/ebd1///1WQECAtW3bNp/vb86cOdaaNWuszz77zCopKbGeeOIJS5K1atUqu8aXrmFz+zvt3nvvtVJTU5s8py9dv4cffthyOp3Whg0bPP6+nTx50q7x93noTY/+NA+96c+f5qE3/Z3mL/Nw6tSp1saNG639+/dbH3/8sfXrX//aateunbV27VrLsnx3DhKaLrLTH3s98zFq1CjLsixr1KhRVv/+/T2O2bBhg9WrVy8rKCjI6tq1qzV//vxG5/3jH/9oJSUlWYGBgVa3bt08/kdwsTW3x/79+5+z3rIsKzs72+rSpYsVFBRkXXrppVZGRoa1efPmi9vYPzS3v5kzZ1pXXnmlFRISYkVERFg33XST9fbbbzc6r69cQ2/+jh49etQKDQ21Fi5c2OQ5fen6NdWbJGvJkiV2jb/PQ2969Kd56E1//jQPvf076k/z8IEHHrAuv/xyeyzp6el2YLIs352DDsv6x51UAAAAOCvuaQIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAIAADBAaAKAC+zLL7+Uw+FQcXFxaw8FwHkgNAEAABggNAFo8xoaGjRz5kxdddVVCg4OVpcuXTR9+nRJ0ieffKKbb75ZoaGhioqK0tixY3X8+HH72AEDBig7O9vjfHfccYfuu+8++3nXrl2Vm5urBx54QGFhYerSpYsWLlxo709ISJAk9erVSw6HQwMGDLhgvQK4cAhNANq8qVOnaubMmXryySf16aefasWKFYqJidHJkyc1dOhQRUREaMeOHfrjH/+odevWafz48c1+jRdeeEE33nijdu/eraysLD388MPat2+fJGn79u2SpHXr1qm8vFyrV69u0f4AXBwBrT0AALiQjh07phdffFF5eXkaNWqUJOnKK6/UTTfdpEWLFqmmpkavvvqqOnbsKEnKy8vTsGHDNHPmTMXExBi/zq233qqsrCxJ0uOPP645c+Zow4YN6tatmy699FJJUlRUlGJjY1u4QwAXCytNANq0vXv3qra2Vunp6U3uu+666+zAJEn9+vVTQ0ODSktLm/U61157rf3fDodDsbGxqqys9H7gAHwOoQlAmxYaGnrWfZZlyeFwNLnv9PZ27drJsiyPfadOnWpUHxgY2Oj4hoaG5g4XgA8jNAFo0xITExUaGqr333+/0b4ePXqouLhYJ06csLd9+OGHateuna6++mpJ0qWXXqry8nJ7f319vUpKSpo1hqCgIPtYAP6L0ASgTQsJCdHjjz+uKVOm6NVXX9UXX3yhrVu3avHixbrnnnsUEhKiUaNGqaSkROvXr9eECRPkcrns+5luvvlmvf3223r77be1b98+ZWVl6ejRo80aQ3R0tEJDQ5Wfn69Dhw7J7XZfgE4BXGiEJgBt3pNPPqlJkybpqaeeUvfu3XXXXXepsrJSHTp00Hvvvadvv/1WvXv31r/9278pPT1deXl59rEPPPCARo0apf/4j/9Q//79lZCQoIEDBzbr9QMCAvS73/1OCxYsUFxcnG6//faWbhHAReCwznyzHgAAAI2w0gQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGCA0AQAAGDg/wGHvDmi5TeiHQAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.histplot(df_expr_ref['Gene'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_expr_ref.to_csv('../data/expression_reference/expression_reference.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
